<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>systemPipeR</title>
    <meta charset="utf-8" />
    <meta name="author" content=" Daniela Cassol" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/font-awesome/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/v4-shims.css" rel="stylesheet" />
    <link href="libs/tile-view/tile-view.css" rel="stylesheet" />
    <script src="libs/tile-view/tile-view.js"></script>
    <link href="libs/animate.css/animate.xaringan.css" rel="stylesheet" />
    <link href="libs/tachyons/tachyons.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
    <link rel="stylesheet" href="css/custom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <em>systemPipeR</em>
### <br/>Daniela Cassol
### <h3>
University of California Riverside
<h3/>
### <h3>
2020/10/20 (updated: 2021-04-25)
<h3/>

---


layout: true
background-image: url(https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png)
background-position: 99% 1%
background-size: 10%







---
class:  middle

## Outline

#### Introduction
#### Design
#### How to run a Workflow
#### Workflows Tutorial
#### Live Demo

&lt;!-- ################## --&gt;
&lt;!-- ## Introduction --&gt;
&lt;!-- ################## --&gt;
---
class: inverse, center, middle

# &lt;i class="fas fa-file-code"&gt;&lt;/i&gt; Introduction

&lt;!-- ################## Slide ################## --&gt;
---
##  Introduction

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
[_systemPipeR_](http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html) provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
_systemPipeR_ offers many utilities to build, control, and execute
workflows entirely from R

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
The environment takes advantage of central community S4 classes of the Bioconductor ecosystem

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Workflows are managed by generic workflow management containers supporting both analysis routines implemented in R code and/or command-line software

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Simple annotation system targets

&lt;!-- ################## Slide ################## --&gt;
---
### systemPipeR’s Core Functionalities 

.center[
&lt;img src="highLevel_color.svg" height="490px" class="center" /&gt;
]

&lt;!-- ################## Slide ################## --&gt;

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

--
.left-column[
#### WF infrastructure
]

.right-column[
*systemPipeR* offers many utilities to build, control, and execute
workflows entirely from R. The environment takes advantage of central community S4 classes
of the Bioconductor ecosystem. Workflows are managed by generic workflow management
containers supporting both analysis routines implemented in R code and/or command-line
software. A layered monitoring infrastructure is provided to design, control and debug each
step in a workflow. The run environment allows to execute workflows entirely or with a
intuitive to use step-wise execution syntax using R’s standard subsetting syntax
(runWF(sys[1:3])) or pipes (%&gt;%).
]

---

## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

.left-column[
#### WF infrastructure
#### Command-line support
]

.right-column[
An important feature of systemPipeR is support for
running command-line software by adopting the Common Workflow Language (CWL). The
latter is a widely adopted community standard for describing analysis workflows. This design
offers several advantages such as:

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
seamless integration of most command-line software

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
support to run systemPipeR workflows from R or many other popular computer languages

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
efficient sharing of workflows across different workflow environments.
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

.left-column[
#### WF infrastructure
#### Command-line support
#### Parallel evaluation 
]

.right-column[
The processing time of workflows can be greatly reduced by making use
of parallel evaluations across several CPU cores on single machines, or multiple nodes of
computer clusters and cloud-based systems. systemPipeR simplifies these parallelization tasks
without creating any limitations for users who do not have access to high-performance
computer resources
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

.left-column[
#### WF infrastructure
#### Command-line support
#### Parallel evaluation 
#### Reports infrastructure
]

.right-column[
*systemPipeR’s* reporting infrastructure
includes three types of interconnected reports each serving a different purpose:

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
a scientific report, based on R Markdown, contains all scientifically relevant results

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
a technical report captures all technical information important for each workflow step, including parameter
settings, software versions, and warning/error messages, etc. 

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
a visual report depicts the entire workflow topology including its run status in form of a workflow graph

]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

.left-column[
#### WF infrastructure
#### Command-line support
#### Parallel evaluation 
#### Reports infrastructure
#### Shiny Web Interface
]

.right-column[
Recently, the systemPipeShiny package has been added that allows
users to design workflows in an interactive graphical user interface (GUI). In addition to
designing workflows, this new interface allows users to run and to monitor workflows in an
intuitive manner without the need of knowing R.
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Structural Features

.left-column[
#### WF infrastructure
#### Command-line support
#### Parallel evaluation 
#### Reports infrastructure
#### Shiny Web Interface
#### Workflow Templates
]

.right-column[
A rich set of end-to-end workflow templates is provided by this
project for a wide range omics applications. In addition, users can contribute and share their
workflows with the community by submitting them to a central GitHub repository
]

&lt;!-- ################## Slide ################## --&gt;
---
##  Important Functions

.small[
| Function Name     | Description                                                                                                           | Category        |
|-------------------|-----------------------------------------------------------------------------------------------------------------------|-----------------|
| `genWorkenvir`    | Generates workflow templates provided by systemPipeRdata helper package / or from the individuals’ pipelines packages | Accessory       |
| `loadWorkflow`    | Constructs `SYSargs2` object from CWL param and targets files                                                         | ``SYSargs2``    |
| `renderWF`        | Populate all the command-line in an `SYSargs2` object                                                                 | ``SYSargs2``    |
| `subsetWF`        | Subsetting `SYSargs2` class slots                                                                                     | ``SYSargs2``    |
| `runCommandline`  | Executes command-line software on samples and parameters specified in `SYSargs2` object                               | ``SYSargs2``    |
| `clusterRun`      | Runs command-line software in parallel mode on a computer cluster                                                     | ``SYSargs2``    |
| `writeTargetsout` | Write updated targets out to file/Generate targets file with reference                                                | ``SYSargs2``    |
| `output_update`   | Updates the output files paths in the `SYSargs2` object                                                               | ``SYSargs2``    |
| `singleYML`       | Create automatically the `param.yml`                                                                                  | ``SYSargs2``    |
| `createWF`        | Create automatically `param.cwl` and the `param.yml` based on the command line                                        | ``SYSargs2``    |
| `config.param`    | Custom configuration of the CWL param files from R                                                                    | ``SYSargs2``    |
]

---
##  Important Functions

.small[
| Function Name     | Description                                                                                                           | Category        |
|-------------------|-----------------------------------------------------------------------------------------------------------------------|-----------------|
| `initWF`          | Constructs `SYSargsList` workflow control module (S4 object) from script file                                         | ``SYSargsList`` |
| `configWF`        | Control of which step of the workflow will be run and the generation of the new RMarkdown                             | ``SYSargsList`` |
| `runWF`           | Runs all the R chunk define in the RMarkdown file or a subset, e.g. `runWF[1:3]`                                      | ``SYSargsList`` |
| `renderReport`    | Render Scientific Report based on RMarkdown                                                                           | ``SYSargsList`` |
| `subsetRmd`       | Write updated subset Rmarkdown of R chunk with text associate in the step selected                                    | ``SYSargsList`` |
| `plotWF`          | Plot visual workflow designs and topologies with different graphical layouts                                          | ``SYSargsList`` |
| `statusWF`        | Return the overview of the workflow steps computational status                                                        | ``SYSargsList`` |
| `evalCode`        | Turn eval option `TRUE` or `FALSE` on RMarkdown file                                                                  | Accessory       |
| `tryCL`           | Checks if third-party software or utility is installed and set in the PATH                                            | Accessory       |
]



&lt;!-- ################## --&gt;
&lt;!-- ## Design --&gt;
&lt;!-- ################## --&gt;
---
class: inverse, center, middle

# &lt;i class="fas fa-drafting-compass"&gt;&lt;/i&gt; Design

---
### Workflow Management Solutions

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*systemPipeR* central concept for designing workflows is **workflow
management containers** (S4 class)

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*SYSargs2* controls workflow steps with input/output file operations 

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*SYSargs2* requires a targets and a set of workflow definition files (here param.cwl and param.yml)

&lt;!-- &lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt;  --&gt;
&lt;!-- Any number of predefined or custom workflow steps is supported --&gt;

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*SYSargsList* objects organize one or many *SYSargs2*
containers in a single compound object capturing all information required to run, control and
monitor complex workflows from start to finish

.center[
&lt;img src="SPRclass.png" class="center" /&gt;
]

---
## &lt;i class="far fa-folder-open"&gt;&lt;/i&gt; Directory Structure

The workflow templates generated by _`genWorkenvir`_ contain the following preconfigured directory structure:

&lt;img src="directory.png" height="425px" class="center" /&gt;

---
## &lt;i class="fas fa-project-diagram"&gt;&lt;/i&gt; Workflows Collection

Browse pipelines that are currently available as part of the *systemPipeR* toolkit

.small[
| WorkFlow |  Description  | Version | GitHub | CI Testing | 
|:-----------------------:|:----------------:|:----------------:|:--------------------:|:-------------------------------:|
| [systemPipeChIPseq](https://systempiper.github.io/systemPipeChIPseq/articles/systemPipeChIPseq.html) | ChIP-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeChIPseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeChIPseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeRIBOseq](https://systempiper.github.io/systemPipeRIBOseq/articles/systemPipeRIBOseq.html) | RIBO-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeRIBOseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeRIBOseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeRNAseq](https://systempiper.github.io/systemPipeRNAseq/articles/systemPipeRNAseq.html) | RNA-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeRNAseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeRNAseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeVARseq](https://systempiper.github.io/systemPipeVARseq/articles/systemPipeVARseq.html) | VAR-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeVARseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeVARseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeMethylseq](https://github.com/systemPipeR/systemPipeMethylseq) | Methyl-Seq Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeMethylseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeMethylseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeDeNovo](https://github.com/systemPipeR/systemPipeDeNovo) | De novo transcriptome assembly Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeDeNovo"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeDeNovo/workflows/R-CMD-check/badge.svg) |
| [systemPipeCLIPseq](https://github.com/systemPipeR/systemPipeCLIPseq) | CLIP-Seq Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeCLIPseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeCLIPseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeMetaTrans](https://github.com/systemPipeR/systemPipeMetaTrans) | Metatranscriptomic Sequencing Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeMetaTrans"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeMetaTrans/workflows/R-CMD-check/badge.svg) |
]

---
class: inverse, center, middle

# CWL

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; CWL

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt; **TODO:**
Add section with CWL details

.center[
&lt;img src="cwl.png" height="490px" class="center" /&gt;
]

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; CWL and SPR

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt; **TODO:**
How to use CWL definition with *systemPipeR*

- *SYSargs2* instances are constructed from a *`targets`* file and two *`param`* file
  - *`hisat2-mapping-se.cwl`* file contains the settings for running command-line software
  - *`hisat2-mapping-se.yml`* file define all the variables to be input in the specific command-line step &lt;/h4&gt;
    

```r
targets &lt;- system.file("extdata", "targets.txt", package="systemPipeR")
dir_path &lt;- system.file("extdata/cwl/hisat2/hisat2-se", package="systemPipeR")
align &lt;- loadWF(targets=targets, wf_file="hisat2-mapping-se.cwl",
                   input_file="hisat2-mapping-se.yml", dir_path=dir_path)
align &lt;- renderWF(align, inputvars=c(FileName="_FASTQ_PATH_", SampleName="_SampleName_"))

## Instance of 'SYSargs2':
##    Slot names/accessors: 
##       targets: 18 (M1A...V12B), targetsheader: 4 (lines)
##       modules: 2
##       wf: 0, clt: 1, yamlinput: 7 (components)
##       input: 18, output: 18
##       cmdlist: 18
##    WF Steps:
##       1. hisat2-mapping-se.cwl (rendered: TRUE)
```

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; CWL and SPR

## *`SYSargs2`* instance 

- Slots and accessor functions have the same names

```r
names(align)
#  [1] "targets"       "targetsheader" "modules"       "wf"            "clt"          
#  [6] "yamlinput"     "cmdlist"       "input"         "output"        "cwlfiles"     
# [11] "inputvars" 
```

- *`cmdlist`* return command-line arguments for the specific software, here *`HISAT2`* for the first sample

```r
cmdlist(align)[1]
# $M1A
# $M1A$`hisat2-mapping-se.cwl`
# [1] "hisat2 -S results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"
```

- The output components of *`SYSargs2`* define all the expected output files for each step in the workflow; some of which are the input for the next workflow step


```r
output(align)[1]
# $M1A
# $M1A$`hisat2-mapping-se.cwl`
# [1] "results/M1A.sam"
```

---
class: inverse, center, middle

# &lt;i class="fas fa-table"&gt;&lt;/i&gt; Metadata

---
### &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; *Targets* file organizes samples

- Structure of *`targets`* file for single-end (SE) library

```r
targetspath &lt;- system.file("extdata", "targets.txt", package="systemPipeR")
read.delim(targetspath, comment.char = "#")[1:3,1:4]
```

```
##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
```

- Structure of *`targets`* file for paired-end (PE) library

```r
targetspath &lt;- system.file("extdata", "targetsPE.txt", package="systemPipeR")
read.delim(targetspath, comment.char = "#")[1:3,1:5]
```

```
##                     FileName1                   FileName2 SampleName Factor
## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz        M1A     M1
## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz        M1B     M1
## 3 ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz        A1A     A1
##   SampleLong
## 1  Mock.1h.A
## 2  Mock.1h.B
## 3   Avr.1h.A
```

---
### &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Integration with *SummarizedExperiment*

- Integrates *targets* files and count table from *systemPipeR* to a *SummarizedExperiment* object


```r
## Create an object with targets file and comparison and count table
sprSE &lt;- SPRdata(targetspath = targetspath, cmp=TRUE)
metadata(sprSE)
# $version
# [1] ‘1.23.9’
# 
# $comparison
# $comparison$CMPset1
#       [,1]  [,2] 
#  [1,] "M1"  "A1" 
#  [2,] "M1"  "V1" 
#  [3,] "A1"  "V1" 
#  [4,] "M6"  "A6" 

colData(sprSE)
# DataFrame with 18 rows and 6 columns
#                    FileName  SampleName      Factor  SampleLong 
#                 &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; 
# M1A  ./data/SRR446027_1.f..         M1A          M1   Mock.1h.A 
# M1B  ./data/SRR446028_1.f..         M1B          M1   Mock.1h.B 
# ...                     ...         ...         ...         ... 
# M12B ./data/SRR446040_1.f..        M12B         M12  Mock.12h.B
```

---
class: inverse, center, middle

# &lt;i class="fas fa-code"&gt;&lt;/i&gt; Live Demo 

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Install Package

Install the **systemPipeRdata** package from [Bioconductor](http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html):


```r
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("systemPipeR")
```

### &lt;i class="fas fa-book"&gt;&lt;/i&gt; Load Package and Documentation

&lt;i class="fas fa-question" style="color:#00758a;"&gt;&lt;/i&gt; 
Load package:


```r
library("systemPipeR")
```

&lt;i class="fas fa-question" style="color:#00758a;"&gt;&lt;/i&gt; 
Access help:


```r
library(help="systemPipeR")
vignette("systemPipeR")
```

---
## &lt;i class="fas fa-code"&gt;&lt;/i&gt; Quick Start

### Load Sample Workflow

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*systemPipeRdata*

- Helper package to generate with a single command workflow templates for *systemPipeR*
- Includes sample data for testing
- User can create new workflows or change and extend existing ones
- Template Workflows:
    - Sample workflows can be loaded with the *`genWorkenvir`* function from **systemPipeRdata**

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Generate workflow template:


```r
library(systemPipeRdata)
genWorkenvir(workflow="rnaseq")
setwd("rnaseq")
```

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
More details about *systemPipeRdata* package [here](https://systempipe.org/pages/SPRdata_intro/SPRdata_intro#1)

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Install Workflow

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Check the workflow template availability 

```r
availableWF(github = TRUE)

# $systemPipeRdata
# [1] "chipseq" "new"     "riboseq" "rnaseq"  "varseq" 
# 
# $github
#                          workflow                       branches version                                               html       description
# 1   systemPipeR/systemPipeChIPseq                         master release   https://github.com/systemPipeR/systemPipeChIPseq Workflow Template
# 2   systemPipeR/systemPipeRIBOseq                         master release   https://github.com/systemPipeR/systemPipeRIBOseq Workflow Template
# 3    systemPipeR/systemPipeRNAseq cluster, master, singleMachine release    https://github.com/systemPipeR/systemPipeRNAseq Workflow Template
# 4    systemPipeR/systemPipeVARseq                         master release    https://github.com/systemPipeR/systemPipeVARseq Workflow Template
# 5   systemPipeR/systemPipeCLIPseq                         master   devel   https://github.com/systemPipeR/systemPipeCLIPseq Workflow Template
# 6    systemPipeR/systemPipeDeNovo                         master   devel    https://github.com/systemPipeR/systemPipeDeNovo Workflow Template
# 7 systemPipeR/systemPipeMetaTrans                         master   devel https://github.com/systemPipeR/systemPipeMetaTrans Workflow Template
# 8 systemPipeR/systemPipeMethylseq                         master   devel https://github.com/systemPipeR/systemPipeMethylseq Workflow Template
```

---
## &lt;i class="fas fa-mouse-pointer"&gt;&lt;/i&gt; Dynamic Workflow Template

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Create dynamic Workflow Templates with *RStudio*

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
`File -&gt; New File -&gt; R Markdown -&gt; From Template`
.center[
&lt;img src="rstudio.png" height="425px" class="center" /&gt;
]


---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Run a Workflow

.left-column[
### Setup
]

.right-column[

```r
library(systemPipeR)
targetspath &lt;- system.file("extdata", "targets.txt", package="systemPipeR") 
read.delim(targetspath, comment.char = "#")[1:4,1:4]
```

```
##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
## 4 ./data/SRR446030_1.fastq.gz        A1B     A1   Avr.1h.B
```

```r
script &lt;- system.file("extdata/workflows/rnaseq", "systemPipeRNAseq.Rmd", package="systemPipeRdata")
```
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Run a Workflow

.left-column[
### Setup
### `initWF`
]

.right-column[

```r
sysargslist &lt;- initWF(script = script, targets = targetspath, overwrite = TRUE)
# Project started with success: ./SYSproject and SYSconfig.yml were created.
```
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Run a Workflow

.left-column[
### Setup
### `initWF`
### `configWF`
]

.right-column[

```r
sysargslist &lt;- configWF(sysargslist, input_steps = "1:3")
sysargslist
# Instance of 'SYSargsList':
#    WF Steps:
# 1. Rmarkdown/HTML setting
# 2. Introduction
# 3. Samples and environment settings
#     3.1. Environment settings and input data
#     3.2. Required packages and resources
#     3.3. Experiment definition provided by `targets` file
```
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Run a Workflow

.left-column[
### Setup
### `initWF`
### `configWF`
### `runWF`
]

.right-column[

```r
sysargslist &lt;- runWF(sysargslist, steps = "1:2")
# Step: 1: Introduction --&gt; DONE 
# Step: 2: Samples and environment settings --&gt; DONE 
# Step: 2.1: Environment settings and input data --&gt; DONE 
# Step: 2.2: Required packages and resources --&gt; DONE 
# Step: 2.3: Experiment definition provided by `targets` file --&gt; DONE 
```


```r
sysargslist &lt;- runWF(sysargslist, steps = "ALL")
```
]

---
## &lt;i class="fas fa-toolbox"&gt;&lt;/i&gt; Run a Workflow

.left-column[
### Setup
### `initWF`
### `configWF`
### `runWF`
### `renderReport`
]

.right-column[

```r
sysargslist &lt;- renderReport(sysargslist = sysargslist)
```
]

---
## How to Use Pipes `%&gt;%`

Consider the following example, in which the steps are the initialization, configuration and running the entire workflow.


```r
library(systemPipeR)
sysargslist &lt;- initWF(script ="systemPipeRNAseq.Rmd", overwrite = TRUE) %&gt;%
    configWF(input_steps = "1:6") %&gt;%
    runWF(steps = "1:2")
```

---
class: inverse, center, middle

# &lt;i class="fas fa-tasks"&gt;&lt;/i&gt; Project Updates

---
### *targets* x *SummarizedExperiment*
&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Extension "SummarizedExperiment" methods:


```r
sprSE &lt;- addAssay(sprSE, assay(countMatrix), xName="countMatrix")
sprSE &lt;- addMetadata(sprSE, list(targets), xName="metadata")
```

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
New Function:


```r
## Create empty SummarizedExperiment
sprSE &lt;- SPRdata()

## Create an object with targets file and comparison and count table
sprSE &lt;- SPRdata(counts = countMatrix, cmp=TRUE, targetspath = targetspath)
metadata(sprSE)
colData(sprSE)
assays(sprSE)
```

---
## SPR Paper

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
[Link to draft](https://www.overleaf.com/project/5e94948a23918a000116311c)

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Added the main points to discuss in the draft

&lt;i class="fas fa-arrow-circle-right" style="color:#00758a;"&gt;&lt;/i&gt;
Writing: Results and introduction

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Improve Graphical Abstract 

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Show case?

---
## *SYSargsList*

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Explain how *SYSargsList* is implemented - Vignette

.small[
| Function Name  | Description                                                                               |
|----------------|-------------------------------------------------------------------------------------------|
| `initWF`       | Constructs `SYSargsList` workflow control module (S4 object) from script file             |
| `configWF`     | Control of which step of the workflow will be run and the generation of the new RMarkdown |
| `runWF`        | Runs all the R chunk define in the RMarkdown file or a subset, e.g. `runWF[1:3]`          |
| `renderReport` | Render Scientific Report based on RMarkdown                                               |
| `renderLog`    | Render logs Report based on RMarkdown                                                     |
| `updateWF`     | Recover the `SYSargsList` workflow previous ran and restarts the WF                       |
| `plotWF`       | Plot visual workflow designs and topologies with different graphical layouts              |
| `statusWF`     | Return the overview of the workflow steps computational status                            |
| `evalCode`     | Turn eval option `TRUE` or `FALSE` on RMarkdown file                                      |
| `tryCL`        | Checks if third-party software or utility is installed and set in the PATH                |
]

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Improve `statusWF()`

---
## Visualization in *systemPipeR*

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Add to vignette (SPR or SPS)
 - `exploreDDS`, `exploreDDSplot`, `GLMplot`, `MAplot`, `MDSplot`, `PCAplot`, `hclustplot`, `heatMaplot`, `tSNEplot`, `volcanoplot` 

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Enrichment analysis and visualization tool for SPR
- Integration with `FGSEA` 


---
## WebSite

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Updated the vignette

&lt;i class="fas fa-check-circle" style="color:#00758a;"&gt;&lt;/i&gt; 
Added *systemPipeRdata* vignette and presentation: [link](https://systempipe.org/docs/systemPipeRdata/)

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Redirect http://girke.bioinformatics.ucr.edu/systemPipeR/ to new page

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Add content to FAQ section

&lt;i class="fas fa-circle" style="color:#00758a;"&gt;&lt;/i&gt;
Add tutorials videos

---
class: middle

# Thanks!

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Browse source code at &lt;a href="https://github.com/tgirke/systemPipeR/"&gt; &lt;i class="fab fa-github fa-2x"&gt;&lt;/i&gt;&lt;/a&gt;

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Ask a question about systemPipeR at Bioconductor Support Page
&lt;a href="https://support.bioconductor.org/t/systempiper/"&gt;&lt;i class="far fa-question-circle fa-2x"&gt;&lt;/i&gt;&lt;/a&gt;

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*systemPipeRdata* at [Bioconductor](http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html)

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
[https://systempipe.org/](https://systempipe.org/)

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
