<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../../favicons/favicon.ico><link rel=apple-touch-icon href=../../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../../favicons/android-192x192.png sizes=192x192><title>Workflow templates and sample data | sysPipe</title><meta property="og:title" content="Workflow templates and sample data"><meta property="og:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: the most recent version of this vignette can be found here.
Note: if you use systemPipeR and systemPipeRdata in published research, please cite:
Backman, T.W.H and Girke, T. (2016)."><meta property="og:type" content="article"><meta property="og:url" content="/sp/sprdata/systempiperdata/"><meta property="article:modified_time" content="2021-06-15T12:04:48-07:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="Workflow templates and sample data"><meta itemprop=description content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: the most recent version of this vignette can be found here.
Note: if you use systemPipeR and systemPipeRdata in published research, please cite:
Backman, T.W.H and Girke, T. (2016)."><meta itemprop=dateModified content="2021-06-15T12:04:48-07:00"><meta itemprop=wordCount content="1160"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="Workflow templates and sample data"><meta name=twitter:description content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: the most recent version of this vignette can be found here.
Note: if you use systemPipeR and systemPipeRdata in published research, please cite:
Backman, T.W.H and Girke, T. (2016)."><link rel=preload href=../../../scss/main.min.7822b7c057b4a810809dabf336eddb9515f800a3275417dc6ae2da2d35a0d1d1.css as=style><link href=../../../scss/main.min.7822b7c057b4a810809dabf336eddb9515f800a3275417dc6ae2da2d35a0d1d1.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><script src=https://unpkg.com/lunr@2.3.8/lunr.min.js integrity=sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY crossorigin=anonymous></script><link rel=stylesheet href=../../../css/prism.css><script src=../../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../../css/driver0.9.8.min.css><link rel=stylesheet href=../../../css/site.css><script src=../../../js/site.js></script></head><body class="td-page line-numbers"><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../../><span class=font-weight-bold>sysPipe</span></a>
<row class=header-switch><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=sidebarSwitch checked>
<label class=custom-control-label for=sidebarSwitch>sidebar</label></div><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=tocSwitch checked>
<label class=custom-control-label for=tocSwitch>toc</label></div></row><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../about/overview/ style=padding-left:1rem;font-weight:calc(3/3*500)>Overview</a>
<a class=dropdown-item href=../../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sp/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sp/spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../../sp/spr/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../../sp/spr/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../../sp/spr/spr_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../../sp/spr/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../../sp/spr/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../../sp/spr/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../../sp/spr/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRTools</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:2rem;font-weight:calc(3/4*500)>Data Visualization</a>
<a class=dropdown-item href=../../../sp/sprdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../../sp/sprdata/systempiperdata/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../../sp/sprdata/newwf/ style=padding-left:2rem;font-weight:calc(3/4*500)>Rules</a>
<a class=dropdown-item href=../../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeTools style=padding-left:2rem;font-weight:calc(3/4*500)>SPT GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=../../../spr_wf/installwf/ style=padding-left:1rem;font-weight:calc(3/3*500)>How to install Workflow</a>
<a class=dropdown-item href=../../../spr_wf/newwf style=padding-left:1rem;font-weight:calc(3/3*500)>Rules to create Workflow</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../training/ id=navbarDropdown role=button>Training</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../training/conferences/ style=padding-left:1rem;font-weight:calc(3/3*500)>Conferences</a>
<a class=dropdown-item href=../../../training/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../../training/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../../offline-search-index.df39731a8502f32c043c477188267439.json data-offline-search-base-href=../../../ data-offline-search-max-results=10></div></nav></header><div class="container-fluid td-outer"><div class=td-main><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none"><div id=td-sidebar-menu class=td-sidebar__inner><form class="td-sidebar__search d-flex align-items-center"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../../offline-search-index.df39731a8502f32c043c477188267439.json data-offline-search-base-href=../../../ data-offline-search-max-results=10>
<button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type=button data-toggle=collapse data-target=#td-section-nav aria-controls=td-docs-nav aria-expanded=false aria-label="Toggle section navigation"></button></form><nav class="collapse td-sidebar-nav" id=td-section-nav><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/ class="align-left pl-0 pr-2 td-sidebar-link td-sidebar-link__section">Documentation</a></li><ul><li class="collapse show" id=sp><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/spr/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeR</a></li><ul><li class=collapse id=spspr><a class="td-sidebar-link td-sidebar-link__page" id=m-spsprintroduction href=../../../sp/spr/introduction/>Introduction</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprgettingstarted href=../../../sp/spr/gettingstarted/>Getting Started</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprspr_run href=../../../sp/spr/spr_run/>How to run a Workflow</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsteps href=../../../sp/spr/steps/>Workflow steps overview</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprtemplates href=../../../sp/spr/templates/>Workflow Templates</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprspr_funcs href=../../../sp/spr/spr_funcs/>Help Manual</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsprinstall href=../../../sp/spr/sprinstall/>SPR detailed installation instructions</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsp_docker href=../../../sp/spr/sp_docker/>SPR Docker container</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spspreditor href=../../../sp/spr/editor/>Workflow Plot Editor</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprviz_editorbower_componentssvg-pan-zoomissue_template href=../../../sp/spr/viz_editor/bower_components/svg-pan-zoom/issue_template/></a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/spt/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeTools</a></li><ul><li class=collapse id=spspt><a class="td-sidebar-link td-sidebar-link__page" id=m-spsptsystempipetools href=../../../sp/spt/systempipetools/>Data Visualizations</a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/sprdata/ class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">systemPipeRdata</a></li><ul><li class="collapse show" id=spsprdata><a class="td-sidebar-link td-sidebar-link__page active" id=m-spsprdatasystempiperdata href=../../../sp/sprdata/systempiperdata/>Workflow templates and sample data</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprdatanewwf href=../../../sp/sprdata/newwf/>Rules to create a new Workflow Template</a></li></ul></ul><a class="td-sidebar-link td-sidebar-link__page" id=m-spsp_docker href=../../../sp/sp_docker/>SPR Docker container</a></li></ul></ul></nav></div></div><div class="d-none d-xl-block col-xl-2 td-toc d-print-none"><div class="td-page-meta ml-2 pb-1 pt-2 mb-0"><a href=https://raw.githubusercontent.com/systemPipeR/systemPipeR.github.io/main/content/en/sp/sprdata/systemPipeRdata.Rmd target=_blank class=source-link><i class="fa fa-code fa-fw"></i>View Rmd source</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/blob/main/content/en/sp/sprdata/systemPipeRdata.md target=_blank><i class="fab fa-github fa-fw"></i>View on Github</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/edit/main/content/en/sp/sprdata/systemPipeRdata.md target=_blank><i class="fa fa-edit fa-fw"></i>Edit this page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/new/main/content/en/sp/sprdata/systemPipeRdata.md?filename=change-me.md&value=---%0Atitle%3A+%22Long+Page+Title%22%0AlinkTitle%3A+%22Short+Nav+Title%22%0Aweight%3A+100%0Adescription%3A+%3E-%0A+++++Page+description+for+heading+and+indexes.%0A---%0A%0A%23%23+Heading%0A%0AEdit+this+template+to+create+your+new+page.%0A%0A%2A+Give+it+a+good+name%2C+ending+in+%60.md%60+-+e.g.+%60getting-started.md%60%0A%2A+Edit+the+%22front+matter%22+section+at+the+top+of+the+page+%28weight+controls+how+its+ordered+amongst+other+pages+in+the+same+directory%3B+lowest+number+first%29.%0A%2A+Add+a+good+commit+message+at+the+bottom+of+the+page+%28%3C80+characters%3B+use+the+extended+description+field+for+more+detail%29.%0A%2A+Create+a+new+branch+so+you+can+preview+your+new+file+and+request+a+review+via+Pull+Request.%0A" target=_blank><i class="fa fa-edit fa-fw"></i>Create child page</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/issues/new target=_blank><i class="fas fa-tasks fa-fw"></i>Create project issue</a></div><nav id=TableOfContents><ul><li><a href=#introduction>Introduction</a></li><li><a href=#getting-started>Getting started</a><ul><li><a href=#installation>Installation</a></li><li><a href=#loading-package-and-documentation>Loading package and documentation</a></li><li><a href=#starting-with-pre-configured-workflow-templates>Starting with pre-configured workflow templates</a></li></ul></li><li><a href=#workflow-templates-collection>Workflow templates collection</a><ul><li><a href=#download-a-specific-r-markdown-file>Download a specific R Markdown file</a></li></ul></li><li><a href=#dynamic-generation-of-workflow-template>Dynamic generation of workflow template</a></li><li><a href=#directory-structure>Directory Structure</a></li><li><a href=#run-workflows>Run workflows</a><ul><li><a href=#return-paths-to-sample-data>Return paths to sample data</a></li></ul></li><li><a href=#version-information>Version information</a></li><li><a href=#funding>Funding</a></li><li><a href=#references>References</a></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 pl-md-5" role=main><nav aria-label=breadcrumb class="d-none d-md-block d-print-none"><ol class="breadcrumb spb-1"><li class=breadcrumb-item><a href=../../../sp/>Documentation</a></li><li class=breadcrumb-item><a href=../../../sp/sprdata/>systemPipeRdata</a></li><li class="breadcrumb-item active" aria-current=page><a href=../../../sp/sprdata/systempiperdata/>Workflow templates and sample data</a></li></ol></nav><div class=td-content><h1>Workflow templates and sample data</h1><style type=text/css>pre code{white-space:pre!important;overflow-x:scroll!important;word-break:keep-all!important;word-wrap:initial!important}</style><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){document.querySelector("h1").className="title";});</script><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){var links=document.links;for(var i=0,linksLength=links.length;i<linksLength;i++)
if(links[i].hostname!=window.location.hostname)
links[i].target='_blank';});</script><p><strong>Note:</strong> the most recent version of this vignette can be found <a href=https://github.com/tgirke/systemPipeRdata/blob/master/vignettes/systemPipeRdata.Rmd>here</a>.</p><p><strong>Note:</strong> if you use <em><code>systemPipeR</code></em> and <em><code>systemPipeRdata</code></em> in published research, please cite:</p><p>Backman, T.W.H and Girke, T. (2016). <em>systemPipeR</em>: Workflow and Report Generation Environment. <em>BMC Bioinformatics</em>, 17: 388. <a href=https://doi.org/10.1186/s12859-016-1241-0>10.1186/s12859-016-1241-0</a>.</p><h1 id=introduction>Introduction</h1><p><a href=https://github.com/tgirke/systemPipeRdata><em><code>systemPipeRdata</code></em></a> is a helper package
to generate with a single command workflow templates that are intended to be
used by its parent package <a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html><em><code>systemPipeR</code></em></a> (H Backman and Girke 2016).
The <em>systemPipeR</em> project provides a suite of R/Bioconductor packages for designing,
building and running end-to-end analysis workflows on local machines, HPC clusters
and cloud systems, while generating at the same time publication quality analysis reports.</p><p>To test workflows quickly or design new ones from existing templates, users can
generate with a single command workflow instances fully populated with sample data
and parameter files required for running a chosen workflow.
Pre-configured directory structure of the workflow environment and the sample data
used by <em><code>systemPipeRdata</code></em> are described <a href=http://bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#load-sample-data-and-workflow-templates>here</a>.</p><h1 id=getting-started>Getting started</h1><h2 id=installation>Installation</h2><p>The <em><code>systemPipeRdata</code></em> package is available at <a href=http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html>Bioconductor</a> and can be installed from within R as follows:</p><pre><code class=language-r>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;systemPipeRdata&quot;)
</code></pre><p>Also, it is possible to install the development version from <a href=http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html>Bioconductor</a>.</p><pre><code class=language-r>BiocManager::install(&quot;systemPipeRdata&quot;, version = &quot;devel&quot;, build_vignettes = TRUE, 
    dependencies = TRUE)  # Installs Devel version from Bioconductor
</code></pre><h2 id=loading-package-and-documentation>Loading package and documentation</h2><pre><code class=language-r>library(&quot;systemPipeRdata&quot;)  # Loads the package
</code></pre><pre><code class=language-r>library(help = &quot;systemPipeRdata&quot;)  # Lists package info
vignette(&quot;systemPipeRdata&quot;)  # Opens vignette
</code></pre><h2 id=starting-with-pre-configured-workflow-templates>Starting with pre-configured workflow templates</h2><p>Load one of the available workflows into your current working directory.
The following does this for the <em><code>varseq</code></em> workflow template. The name of the resulting
workflow directory can be specified under the <em><code>mydirname</code></em> argument. The default <em><code>NULL</code></em>
uses the name of the chosen workflow. An error is issued if a directory of the same
name and path exists already.</p><pre><code class=language-r>genWorkenvir(workflow = &quot;systemPipeR/SPvarseq&quot;, mydirname = &quot;varseq&quot;)
setwd(&quot;varseq&quot;)
</code></pre><p>On Linux and OS X systems the same can be achieved from the command-line of a terminal with the following commands.</p><pre><code class=language-bash>$ Rscript -e &quot;systemPipeRdata::genWorkenvir(workflow='systemPipeR/SPvarseq', mydirname='varseq')&quot;
</code></pre><h1 id=workflow-templates-collection>Workflow templates collection</h1><p>A collection of workflow templates are available, and it is possible to browse the
current availability, as follows:</p><pre><code class=language-r>availableWF(github = TRUE)
</code></pre><p>This function returns the list of workflow templates available within the package
and <a href=https://github.com/systemPipeR>systemPipeR Organization</a> on GitHub. Each one
listed template can be created as described above.</p><p>The workflow template choose from Github will be installed as an R package, and
also it creates the environment with all the settings and files to run the demo
analysis.</p><pre><code class=language-r>genWorkenvir(workflow=&quot;systemPipeR/SPrnaseq&quot;, mydirname=&quot;NULL&quot;)
setwd(&quot;SPrnaseq&quot;)
</code></pre><p>Besides, it is possible to choose different versions of the workflow template,
defined through other branches on the GitHub Repository. By default, the <em><code>master</code></em>
branch is selected, however, it is possible to define a different branch with the <em><code>ref</code></em> argument.</p><pre><code class=language-r>genWorkenvir(workflow=&quot;systemPipeR/SPrnaseq&quot;, ref = &quot;singleMachine&quot;)
setwd(&quot;SPrnaseq&quot;)
</code></pre><h2 id=download-a-specific-r-markdown-file>Download a specific R Markdown file</h2><p>Also, it is possible to download a specific workflow script for your analysis.
The URL can be specified under <em><code>url</code></em> argument and the R Markdown file name in
the <em><code>urlname</code></em> argument. The default <em><code>NULL</code></em> copies the current version available in the chose template.</p><pre><code class=language-r>genWorkenvir(workflow=&quot;systemPipeR/SPrnaseq&quot;, url = &quot;https://raw.githubusercontent.com/systemPipeR/systemPipeRNAseq/cluster/vignettes/systemPipeRNAseq.Rmd&quot;, 
             urlname = &quot;rnaseq_V-cluster.Rmd&quot;)
setwd(&quot;rnaseq&quot;)
</code></pre><h1 id=dynamic-generation-of-workflow-template>Dynamic generation of workflow template</h1><p>It is possible to create a new workflow structure from RStudio
menu <code>File -> New File -> R Markdown -> From Template -> systemPipeR New WorkFlow</code>.
This interactive option creates the same environment as demonstrated above.</p><p><img src=results/rstudio.png alt>
<strong>Figure 1:</strong> Selecting workflow template within RStudio.</p><h1 id=directory-structure>Directory Structure</h1><p>The workflow templates generated by <em><code>genWorkenvir</code></em> contain the following preconfigured directory structure:</p><ul><li><span style=color:green><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>rnaseq/</em>)<ul><li>This is the root directory of the R session running the workflow.</li><li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li><li>Note, this directory can have any name (<em>e.g.</em> <span style=color:green><em><strong>rnaseq</strong></em></span>, <span style=color:green><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li><li><strong>Important subdirectories</strong>:<ul><li><span style=color:green><em><strong>param/</strong></em></span><ul><li>Stores non-CWL parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*.run.sh</em>. These files are only required for backwards compatibility to run old workflows using the previous custom command-line interface.</li><li><span style=color:green><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the CWL parameter files. To organize workflows, each can have its own subdirectory, where all <code>CWL param</code> and <code>input.yml</code> files need to be in the same subdirectory.</li></ul></li><li><span style=color:green><em><strong>data/</strong></em></span><ul><li>FASTQ files</li><li>FASTA file of reference (<em>e.g.</em> reference genome)</li><li>Annotation files</li><li>etc.</li></ul></li><li><span style=color:green><em><strong>results/</strong></em></span><ul><li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li><li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li></ul></li></ul></li></ul></li></ul><p><strong>Note</strong>: Directory names are indicated in <span style=color:grey><em><strong>green</strong></em></span>.
Users can change this structure as needed, but need to adjust the code in their workflows
accordingly.</p><center><img src=results/directory.png></center><p><strong>Figure 2:</strong> <em>systemPipeR’s</em> preconfigured directory structure.</p><h1 id=run-workflows>Run workflows</h1><p>Next, run from within R the chosen sample workflow by executing the code provided
in the corresponding <em><code>*.Rmd</code>* template file.
Much more detailed information on running and customizing </em><code>systemPipeR</code>*<a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html></a>
workflows is available in its overview vignette <a href=http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html>here</a>.
This vignette can also be opened from R with the following command.</p><pre><code class=language-r>library(&quot;systemPipeR&quot;)  # Loads systemPipeR which needs to be installed via BiocManager::install() from Bioconductor
</code></pre><pre><code class=language-r>vignette(&quot;systemPipeR&quot;, package = &quot;systemPipeR&quot;)
</code></pre><h2 id=return-paths-to-sample-data>Return paths to sample data</h2><p>The location of the sample data provided by <em><code>systemPipeRdata</code></em> can be returned as a <em><code>list</code></em>.</p><pre><code class=language-r>pathList()
</code></pre><pre><code>## $targets
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/param/targets.txt&quot;
## 
## $targetsPE
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/param/targetsPE.txt&quot;
## 
## $annotationdir
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/annotation/&quot;
## 
## $fastqdir
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/fastq/&quot;
## 
## $bamdir
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/bam/&quot;
## 
## $paramdir
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/param/&quot;
## 
## $workflows
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/&quot;
## 
## $chipseq
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/chipseq/&quot;
## 
## $rnaseq
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/rnaseq/&quot;
## 
## $riboseq
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/riboseq/&quot;
## 
## $varseq
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/varseq/&quot;
## 
## $new
## [1] &quot;/home/dcassol/src/R-devel/library/systemPipeRdata/extdata/workflows/new/&quot;
</code></pre><h1 id=version-information>Version information</h1><pre><code class=language-r>sessionInfo()
</code></pre><pre><code>## R Under development (unstable) (2021-02-04 r79940)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets 
## [6] methods   base     
## 
## other attached packages:
## [1] systemPipeRdata_1.19.2 BiocStyle_2.19.2      
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.33          magrittr_2.0.1     
##  [3] BiocGenerics_0.37.4 R6_2.5.0           
##  [5] rlang_0.4.11        stringr_1.4.0      
##  [7] tools_4.1.0         parallel_4.1.0     
##  [9] xfun_0.22           jquerylib_0.1.4    
## [11] htmltools_0.5.1.1   remotes_2.3.0      
## [13] yaml_2.2.1          digest_0.6.27      
## [15] bookdown_0.22       formatR_1.9        
## [17] BiocManager_1.30.12 sass_0.3.1         
## [19] codetools_0.2-18    evaluate_0.14      
## [21] rmarkdown_2.7.12    blogdown_1.3       
## [23] stringi_1.5.3       compiler_4.1.0     
## [25] bslib_0.2.4         jsonlite_1.7.2
</code></pre><h1 id=funding>Funding</h1><p>This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).</p><h1 id=references>References</h1><div id=refs class="references csl-bib-body hanging-indent"><div id=ref-H_Backman2016-bt class=csl-entry><p>H Backman, Tyler W, and Thomas Girke. 2016. “<span class=nocase>systemPipeR: NGS workflow and report generation environment</span>.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href=https://doi.org/10.1186/s12859-016-1241-0>https://doi.org/10.1186/s12859-016-1241-0</a>.</p></div></div><div class="text-muted mt-5 pt-3 border-top">Last modified 2021-06-15: <a href=https://github.com/systemPipeR/systemPipeR.github.io/commit/79ca2e8e74de301d5b5ce699143e6bd4a5ed965f>Merge pull request #18 from systemPipeR/imgbot (79ca2e8)</a></div></div></main></div></div><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2021 The systemPipeR Project All Rights Reserved</small><p class=mt-2><a href=../../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script id=MathJax-script async src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js integrity="sha256-+CC/xQfD6+/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin=anonymous></script><script src=../../../js/prism.js></script></body></html>