<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../../favicons/favicon.ico><link rel=apple-touch-icon href=../../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../../favicons/android-192x192.png sizes=192x192><title>Introduction | sysPipe</title><meta property="og:title" content="Introduction"><meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: if you use systemPipeR in published research, please cite: Backman, T.W.H and Girke, T. (2016). systemPipeR: NGS Workflow and Report Generation Environment. BMC Bioinformatics, 17: 388. 10.1186/s12859-016-1241-0.
Introduction systemPipeR provides flexible utilities for building and running automated end-to-end analysis workflows for a wide range of research applications, including next-generation sequencing (NGS) experiments, such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq (H Backman and Girke 2016)."><meta property="og:type" content="article"><meta property="og:url" content="/spr/systempiper/introduction/"><meta property="article:modified_time" content="2021-03-04T13:17:32-08:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="Introduction"><meta itemprop=description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: if you use systemPipeR in published research, please cite: Backman, T.W.H and Girke, T. (2016). systemPipeR: NGS Workflow and Report Generation Environment. BMC Bioinformatics, 17: 388. 10.1186/s12859-016-1241-0.
Introduction systemPipeR provides flexible utilities for building and running automated end-to-end analysis workflows for a wide range of research applications, including next-generation sequencing (NGS) experiments, such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq (H Backman and Girke 2016)."><meta itemprop=dateModified content="2021-03-04T13:17:32-08:00"><meta itemprop=wordCount content="1445"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="Introduction"><meta name=twitter:description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Note: if you use systemPipeR in published research, please cite: Backman, T.W.H and Girke, T. (2016). systemPipeR: NGS Workflow and Report Generation Environment. BMC Bioinformatics, 17: 388. 10.1186/s12859-016-1241-0.
Introduction systemPipeR provides flexible utilities for building and running automated end-to-end analysis workflows for a wide range of research applications, including next-generation sequencing (NGS) experiments, such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq (H Backman and Girke 2016)."><link rel=preload href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css as=style><link href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><link rel=stylesheet href=../../../css/prism.css><script src=../../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../../css/driver0.9.8.min.css><link rel=stylesheet href=../../../css/site.css><script src=../../../js/site.js></script></head><body class="td-page line-numbers"><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../../><span class=font-weight-bold>sysPipe</span></a>
<row class=header-switch><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=sidebarSwitch checked>
<label class=custom-control-label for=sidebarSwitch>sidebar</label></div><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=tocSwitch checked>
<label class=custom-control-label for=tocSwitch>toc</label></div></row><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../about/overview/ style=padding-left:1rem;font-weight:calc(3/3*500)>Overview</a>
<a class=dropdown-item href=../../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr/systempiper/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../../spr/systempiper/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../../spr/systempiper/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../../spr/systempiper/spr_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../../spr/systempiper/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../../spr/systempiper/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../../spr/systempiper/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../../spr/systempiper/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../../spr/systempiperdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../../spr/systempiperdata/vignette/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeRNAseq/articles/systemPipeRNAseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>RNA-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeChIPseq/articles/systemPipeChIPseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>ChIP-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeRIBOseq/articles/systemPipeRIBOseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>RIBO-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeVARseq/articles/systemPipeVARseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>VAR-Seq</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../training/ id=navbarDropdown role=button>Training</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../training/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../../training/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002 Search this site…" aria-label="Search this site…" autocomplete=off></div></nav></header><div class="container-fluid td-outer"><div class=td-main><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none"><div id=td-sidebar-menu class=td-sidebar__inner><form class="td-sidebar__search d-flex align-items-center"><input type=search class="form-control td-search-input" placeholder="&#xf002 Search this site…" aria-label="Search this site…" autocomplete=off>
<button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type=button data-toggle=collapse data-target=#td-section-nav aria-controls=td-docs-nav aria-expanded=false aria-label="Toggle section navigation"></button></form><nav class="collapse td-sidebar-nav" id=td-section-nav><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/ class="align-left pl-0 pr-2 td-sidebar-link td-sidebar-link__section">Documentation</a></li><ul><li class="collapse show" id=spr><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/systempiper/ class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">systemPipeR</a></li><ul><li class="collapse show" id=sprsystempiper><a class="td-sidebar-link td-sidebar-link__page active" id=m-sprsystempiperintroduction href=../../../spr/systempiper/introduction/>Introduction</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipergettingstarted href=../../../spr/systempiper/gettingstarted/>Getting Started</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperspr_run href=../../../spr/systempiper/spr_run/>How to run a Workflow</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipersteps href=../../../spr/systempiper/steps/>Workflow steps overview</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipertemplates href=../../../spr/systempiper/templates/>Workflow Templates</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperspr_funcs href=../../../spr/systempiper/spr_funcs/>systemPipeR Help Manual</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipersprinstall href=../../../spr/systempiper/sprinstall/>SPR detailed installation instructions</a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/systempiperdata/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeRdata</a></li><ul><li class=collapse id=sprsystempiperdata><a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperdatavignette href=../../../spr/systempiperdata/vignette/>SRPdata Vignette</a></li></ul></ul></li></ul></ul></nav></div></div><div class="d-none d-xl-block col-xl-2 td-toc d-print-none"><div class="td-page-meta ml-2 pb-1 pt-2 mb-0"><a href=https://github.com/systemPipeR/systemPipeR.github.io/blob/main/content/en/spr/systempiper/introduction.markdown target=_blank><i class="fa fa-code fa-fw"></i>View source code</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/edit/main/content/en/spr/systempiper/introduction.markdown target=_blank><i class="fa fa-edit fa-fw"></i>Edit this page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/new/main/content/en/spr/systempiper/introduction.markdown?filename=change-me.md&value=---%0Atitle%3A+%22Long+Page+Title%22%0AlinkTitle%3A+%22Short+Nav+Title%22%0Aweight%3A+100%0Adescription%3A+%3E-%0A+++++Page+description+for+heading+and+indexes.%0A---%0A%0A%23%23+Heading%0A%0AEdit+this+template+to+create+your+new+page.%0A%0A%2A+Give+it+a+good+name%2C+ending+in+%60.md%60+-+e.g.+%60getting-started.md%60%0A%2A+Edit+the+%22front+matter%22+section+at+the+top+of+the+page+%28weight+controls+how+its+ordered+amongst+other+pages+in+the+same+directory%3B+lowest+number+first%29.%0A%2A+Add+a+good+commit+message+at+the+bottom+of+the+page+%28%3C80+characters%3B+use+the+extended+description+field+for+more+detail%29.%0A%2A+Create+a+new+branch+so+you+can+preview+your+new+file+and+request+a+review+via+Pull+Request.%0A" target=_blank><i class="fa fa-edit fa-fw"></i>Create child page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/issues/new?title=Introduction" target=_blank><i class="fab fa-github fa-fw"></i>Create documentation issue</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/issues/new target=_blank><i class="fas fa-tasks fa-fw"></i>Create project issue</a></div><nav id=TableOfContents><ul><li><ul><li><a href=#introduction>Introduction</a><ul><li><a href=#workflow-design-structure-using-sysargs2>Workflow design structure using <em><code>SYSargs2</code></em></a></li><li><a href=#workflow-management-using-sysargslist>Workflow Management using <em><code>SYSargsList</code></em></a></li><li><a href=#workflow-design-structure-using-sysargs-previous-version>Workflow design structure using <em><code>SYSargs</code></em>: Previous version</a></li><li><a href=#reference>Reference</a></li></ul></li></ul></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 pl-md-5" role=main><nav aria-label=breadcrumb class="d-none d-md-block d-print-none"><ol class="breadcrumb spb-1"><li class=breadcrumb-item><a href=../../../spr/>Documentation</a></li><li class=breadcrumb-item><a href=../../../spr/systempiper/>systemPipeR</a></li><li class="breadcrumb-item active" aria-current=page><a href=../../../spr/systempiper/introduction/>Introduction</a></li></ol></nav><div class=td-content><h1>Introduction</h1><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){document.querySelector("h1").className="title";});</script><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){var links=document.links;for(var i=0,linksLength=links.length;i<linksLength;i++)
if(links[i].hostname!=window.location.hostname)
links[i].target='_blank';});</script><p><strong>Note:</strong> if you use <em><code>systemPipeR</code></em> in published research, please cite:
Backman, T.W.H and Girke, T. (2016). <em><code>systemPipeR</code></em>: NGS Workflow and Report Generation Environment. <em>BMC Bioinformatics</em>, 17: 388. <a href=https://doi.org/10.1186/s12859-016-1241-0>10.1186/s12859-016-1241-0</a>.</p><h2 id=introduction>Introduction</h2><p><a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html><em><code>systemPipeR</code></em></a> provides flexible utilities for building and running automated end-to-end analysis workflows for a wide range of research applications, including next-generation sequencing (NGS) experiments, such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq (H Backman and Girke 2016). Important features include a uniform workflow interface across different data analysis applications, automated report generation, and support for running both R and command-line software, such as NGS aligners or peak/variant callers, on local computers or compute clusters (Figure 1). The latter supports interactive job submissions and batch submissions to queuing systems of clusters. For instance, <em><code>systemPipeR</code>* can be used with most command-line aligners such as <code>BWA</code> (Heng Li 2013; H. Li and Durbin 2009), <code>HISAT2</code> (Kim, Langmead, and Salzberg 2015), <code>TopHat2</code> (Kim et al. 2013) and <code>Bowtie2</code> (Langmead and Salzberg 2012), as well as the R-based NGS aligners </em><code>Rsubread</code>*<a href=http://www.bioconductor.org/packages/devel/bioc/html/Rsubread.html></a> (Liao, Smyth, and Shi 2013) and <a href=http://www.bioconductor.org/packages/devel/bioc/html/gmapR.html><em><code>gsnap (gmapR)</code></em></a> (Wu and Nacu 2010). Efficient handling of complex sample sets (<em>e.g.</em> FASTQ/BAM files) and experimental designs are facilitated by a well-defined sample annotation infrastructure which improves reproducibility and user-friendliness of many typical analysis workflows in the NGS area (Lawrence et al. 2013).</p><p>The main motivation and advantages of using <em><code>systemPipeR</code></em> for complex data analysis tasks are:</p><ol><li>Facilitates the design of complex NGS workflows involving multiple R/Bioconductor packages</li><li>Common workflow interface for different NGS applications</li><li>Makes NGS analysis with Bioconductor utilities more accessible to new users</li><li>Simplifies usage of command-line software from within R</li><li>Reduces the complexity of using compute clusters for R and command-line software</li><li>Accelerates runtime of workflows via parallelization on computer systems with multiple CPU cores and/or multiple compute nodes</li><li>Improves reproducibility by automating analyses and generation of analysis reports</li></ol><center><img src=utilities.png></center><p><strong>Figure 1:</strong> Relevant features in <em><code>systemPipeR</code></em>.
Workflow design concepts are illustrated under (A & B). Examples of
<em>systemPipeR’s</em> visualization functionalities are given under (C).</br></p><p>A central concept for designing workflows within the <em><code>systemPipeR</code></em> environment
is the use of workflow management containers. In previous versions, <em><code>systemPipeR</code></em>
used a custom command-line interface called <em><code>SYSargs</code></em> (see Figure 3) and for
this purpose will continue to be supported for some time. With the latest <a href=http://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html>Bioconductor Release 3.9</a>,
we are adopting for this functionality the widely used community standard
<a href=https://www.commonwl.org/>Common Workflow Language</a> (CWL) for describing
analysis workflows in a generic and reproducible manner, introducing <em><code>SYSargs2</code></em>
workflow control class (see Figure 2). Using this community standard in <em><code>systemPipeR</code></em>
has many advantages. For instance, the integration of CWL allows running <em><code>systemPipeR</code></em>
workflows from a single specification instance either entirely from within R, from various command-line
wrappers (e.g., <em>cwl-runner</em>) or from other languages (<em>, e.g.,</em> Bash or Python).
<em><code>systemPipeR</code></em> includes support for both command-line and R/Bioconductor software
as well as resources for containerization, parallel evaluations on computer clusters
along with the automated generation of interactive analysis reports.</p><p>An important feature of <em><code>systemPipeR's</code></em> CWL interface is that it provides two
options to run command-line tools and workflows based on CWL. First, one can
run CWL in its native way via an R-based wrapper utility for <em>cwl-runner</em> or
<em>cwl-tools</em> (CWL-based approach). Second, one can run workflows using CWL’s
command-line and workflow instructions from within R (R-based approach). In the
latter case the same CWL workflow definition files (<em>e.g.</em> <code>*.cwl</code> and <code>*.yml</code>)
are used but rendered and executed entirely with R functions defined by
<em><code>systemPipeR</code></em>, and thus use CWL mainly as a command-line and workflow
definition format rather than software to run workflows. In this regard
<em><code>systemPipeR</code></em> also provides several convenience functions that are useful for
designing and debugging workflows, such as a command-line rendering function to
retrieve the exact command-line strings for each data set and processing step
prior to running a command-line.</p><p>This overview introduces the design of a new CWL S4 class in <em><code>systemPipeR</code></em>,
as well as the custom command-line interface, combined with the overview of all
the common analysis steps of NGS experiments.</p><h3 id=workflow-design-structure-using-sysargs2>Workflow design structure using <em><code>SYSargs2</code></em></h3><p>The flexibility of <em><code>systemPipeR's</code></em> new interface workflow control class is the driving factor behind
the use of as many steps necessary for the analysis, as well as the connection
between command-line- or R-based software. The connectivity among all
workflow steps is achieved by the <em><code>SYSargs2</code></em> workflow control class (see Figure 3).
This S4 class is a list-like container where each instance stores all the
input/output paths and parameter components required for a particular data
analysis step. <em><code>SYSargs2</code></em> instances are generated by two constructor
functions, <em>loadWorkflow</em> and <em>renderWF</em>, using as data input <em>targets</em> or
<em>yaml</em> files as well as two <em>cwl</em> parameter files (for details see below). When
running preconfigured workflows, the only input the user needs to provide is
the initial <em>targets</em> file containing the paths to the input files (<em>e.g.</em>
FASTQ) along with unique sample labels. Subsequent targets instances are
created automatically. The parameters required for running command-line
software is provided by the parameter (<em>.cwl</em>) files described below.</p><p>We also introduce the <em><code>SYSargs2Pipe</code></em> class that organizes one or many
SYSargs2 containers in a single compound object capturing all information
required to run, control and monitor complex workflows from start to finish. This
design enhances the <em><code>systemPipeR</code></em> workflow framework with a generalized,
flexible, and robust design.</p><center><img src=SYS_WF.png></center><p><strong>Figure 2:</strong> Workflow steps with input/output file operations are controlled by
<em><code>SYSargs2</code></em> objects. Each <em><code>SYSargs2</code></em> instance is constructed from one <em>targets</em>
and two <em>param</em> files. The only input provided by the user is the initial <em>targets</em>
file. Subsequent <em>targets</em> instances are created automatically, from the previous
output files. Any number of predefined or custom workflow steps are supported. One
or many <em><code>SYSargs2</code></em> objects are organized in an <em><code>SYSargs2Pipe</code></em> container.</p><h3 id=workflow-management-using-sysargslist>Workflow Management using <em><code>SYSargsList</code></em></h3><p><strong>systemPipeR</strong> allows creation (multi-step analyses) and execution of workflow entirely for R, with control, flexibility, and scalability of the all process. The execution of the workflow can be sent to a HPC, can be parallelizes, accelerating results acquisition. A workflow management system provides an infrastructure for the set-up, performance and monitoring of a defined sequence of tasks, arranged as a workflow application.</p><center><img src=sysargslist.png></center><p><strong>Figure 3:</strong> Workflow Management using <em><code>SYSargsList</code></em>.</p><h3 id=workflow-design-structure-using-sysargs-previous-version>Workflow design structure using <em><code>SYSargs</code></em>: Previous version</h3><p>Instances of this S4 object class are constructed by the <em><code>systemArgs</code></em> function
from two simple tabular files: a <em><code>targets</code></em> file and a <em><code>param</code></em> file. The latter
is optional for workflow steps lacking command-line software. Typically, a
<em><code>SYSargs</code></em> instance stores all sample-level inputs as well as the paths to the
corresponding outputs generated by command-line- or R-based software generating
sample-level output files, such as read preprocessors (trimmed/filtered FASTQ
files), aligners (SAM/BAM files), variant callers (VCF/BCF files) or peak callers
(BED/WIG files). Each sample level input/output operation uses its own <em><code>SYSargs</code></em>
instance. The outpaths of <em><code>SYSargs</code></em> usually define the sample inputs for the
next <em><code>SYSargs</code></em> instance. This connectivity is established by writing the
outpaths with the <em><code>writeTargetsout</code></em> function to a new <em><code>targets</code></em> file that
serves as input to the next <em><code>systemArgs</code></em> call. Typically, the user has to
provide only the initial <em><code>targets</code></em> file. All downstream <em><code>targets</code></em> files are
generated automatically. By chaining several <em><code>SYSargs</code></em> steps together one can
construct complex workflows involving many sample-level input/output file
operations with any combination of command-line or R-based software.</p><center><img src=SystemPipeR_Workflow.png></center><p><strong>Figure 4:</strong> Workflow design structure of <em><code>systemPipeR</code></em> using <em><code>SYSargs</code></em>.</p><h3 id=reference>Reference</h3><div id=refs class="references csl-bib-body hanging-indent"><div id=ref-H_Backman2016-bt class=csl-entry><p>H Backman, Tyler W, and Thomas Girke. 2016. “<span class=nocase>systemPipeR: NGS workflow and report generation environment</span>.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href=https://doi.org/10.1186/s12859-016-1241-0>https://doi.org/10.1186/s12859-016-1241-0</a>.</p></div><div id=ref-Kim2015-ve class=csl-entry><p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4): 357–60.</p></div><div id=ref-Kim2013-vg class=csl-entry><p>Kim, Daehwan, Geo Pertea, Cole Trapnell, Harold Pimentel, Ryan Kelley, and Steven L Salzberg. 2013. “TopHat2: Accurate Alignment of Transcriptomes in the Presence of Insertions, Deletions and Gene Fusions.” <em>Genome Biol.</em> 14 (4): R36. <a href=https://doi.org/10.1186/gb-2013-14-4-r36>https://doi.org/10.1186/gb-2013-14-4-r36</a>.</p></div><div id=ref-Langmead2012-bs class=csl-entry><p>Langmead, Ben, and Steven L Salzberg. 2012. “Fast Gapped-Read Alignment with Bowtie 2.” <em>Nat. Methods</em> 9 (4): 357–59. <a href=https://doi.org/10.1038/nmeth.1923>https://doi.org/10.1038/nmeth.1923</a>.</p></div><div id=ref-Lawrence2013-kt class=csl-entry><p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href=https://doi.org/10.1371/journal.pcbi.1003118>https://doi.org/10.1371/journal.pcbi.1003118</a>.</p></div><div id=ref-Li2009-oc class=csl-entry><p>Li, H, and R Durbin. 2009. “Fast and Accurate Short Read Alignment with Burrows-Wheeler Transform.” <em>Bioinformatics</em> 25 (14): 1754–60. <a href=https://doi.org/10.1093/bioinformatics/btp324>https://doi.org/10.1093/bioinformatics/btp324</a>.</p></div><div id=ref-Li2013-oy class=csl-entry><p>Li, Heng. 2013. “Aligning Sequence Reads, Clone Sequences and Assembly Contigs with BWA-MEM.” <em>arXiv [q-Bio.GN]</em>, March. <a href=http://arxiv.org/abs/1303.3997>http://arxiv.org/abs/1303.3997</a>.</p></div><div id=ref-Liao2013-bn class=csl-entry><p>Liao, Yang, Gordon K Smyth, and Wei Shi. 2013. “The Subread Aligner: Fast, Accurate and Scalable Read Mapping by Seed-and-Vote.” <em>Nucleic Acids Res.</em> 41 (10): e108. <a href=https://doi.org/10.1093/nar/gkt214>https://doi.org/10.1093/nar/gkt214</a>.</p></div><div id=ref-Wu2010-iq class=csl-entry><p>Wu, T D, and S Nacu. 2010. “Fast and <span class=nocase>SNP-tolerant</span> Detection of Complex Variants and Splicing in Short Reads.” <em>Bioinformatics</em> 26 (7): 873–81. <a href=https://doi.org/10.1093/bioinformatics/btq057>https://doi.org/10.1093/bioinformatics/btq057</a>.</p></div></div><div class="text-muted mt-5 pt-3 border-top">Last modified 2021-03-04: <a href=https://github.com/systemPipeR/systemPipeR.github.io/commit/cee9f4108731fd68bd9162da8d8c2cb81b878b21>Merge pull request #3 from systemPipeR/dependabot/npm_and_yarn/postcss-8.2.7 (cee9f41)</a></div></div></main></div></div><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2021 The systemPipeR Project All Rights Reserved</small><p class=mt-2><a href=../../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src=../../../js/mathjax3.1.2.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../../js/main.min.b5fc1b29d2465835844254bd4d804738eaf24c0cec53009b4c6899989be55a47.js integrity="sha256-tfwbKdJGWDWEQlS9TYBHOOryTAzsUwCbTGiZmJvlWkc=" crossorigin=anonymous></script><script src=../../../js/prism.js></script></body></html>