<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../../favicons/favicon.ico><link rel=apple-touch-icon href=../../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../../favicons/android-192x192.png sizes=192x192><title>Getting Started | sysPipe</title><meta property="og:title" content="Getting Started"><meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Getting Started Installation systemPipeR environment can be installed from the R console using the BiocManager::install command. The associated data package systemPipeRdata can be installed the same way."><meta property="og:type" content="article"><meta property="og:url" content="/sp/spr/gettingstarted/"><meta property="article:modified_time" content="2021-10-30T02:23:57+00:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="Getting Started"><meta itemprop=description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Getting Started Installation systemPipeR environment can be installed from the R console using the BiocManager::install command. The associated data package systemPipeRdata can be installed the same way."><meta itemprop=dateModified content="2021-10-30T02:23:57+00:00"><meta itemprop=wordCount content="1429"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="Getting Started"><meta name=twitter:description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Getting Started Installation systemPipeR environment can be installed from the R console using the BiocManager::install command. The associated data package systemPipeRdata can be installed the same way."><link rel=preload href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css as=style><link href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><script src=https://unpkg.com/lunr@2.3.8/lunr.min.js integrity=sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY crossorigin=anonymous></script><link rel=stylesheet href=../../../css/prism.css><script src=../../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../../css/driver0.9.8.min.css><link rel=stylesheet href=../../../css/site.css><script src=../../../js/site.js></script></head><body class="td-page line-numbers"><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../../><span class=font-weight-bold>sysPipe</span></a>
<row class=header-switch><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=sidebarSwitch checked>
<label class=custom-control-label for=sidebarSwitch>sidebar</label></div><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=tocSwitch checked>
<label class=custom-control-label for=tocSwitch>toc</label></div></row><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../about/overview/ style=padding-left:1rem;font-weight:calc(3/3*500)>Overview</a>
<a class=dropdown-item href=../../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sp/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sp/spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../../sp/spr/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../../sp/spr/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../../sp/spr/spr_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../../sp/spr/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../../sp/spr/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../../sp/spr/cwl_syntax/ style=padding-left:2rem;font-weight:calc(3/4*500)>CWL syntax</a>
<a class=dropdown-item href=../../../sp/spr/cwl_and_spr/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR and CWL</a>
<a class=dropdown-item href=../../../sp/spr/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRTools</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:2rem;font-weight:calc(3/4*500)>Data Visualization</a>
<a class=dropdown-item href=../../../sp/spr/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../../sp/sprdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../../sp/sprdata/systempiperdata/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../../sp/sprdata/newwf/ style=padding-left:2rem;font-weight:calc(3/4*500)>Rules</a>
<a class=dropdown-item href=../../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeTools style=padding-left:2rem;font-weight:calc(3/4*500)>SPT GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=../../../spr_wf/installwf/ style=padding-left:1rem;font-weight:calc(3/3*500)>How to install Workflow</a>
<a class=dropdown-item href=../../../spr_wf/newwf style=padding-left:1rem;font-weight:calc(3/3*500)>Rules to create Workflow</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../outreach/ id=navbarDropdown role=button>Outreach</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../training/conferences/ style=padding-left:1rem;font-weight:calc(3/3*500)>Conferences</a>
<a class=dropdown-item href=../../../training/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../../training/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../news/release_3_13/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.13</a>
<a class=dropdown-item href=../../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../../offline-search-index.a27377dce846ec069f6d13ce96c94d0d.json data-offline-search-base-href=../../../ data-offline-search-max-results=10></div></nav></header><div class="container-fluid td-outer"><div class=td-main><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none"><div id=td-sidebar-menu class=td-sidebar__inner><form class="td-sidebar__search d-flex align-items-center"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../../offline-search-index.a27377dce846ec069f6d13ce96c94d0d.json data-offline-search-base-href=../../../ data-offline-search-max-results=10>
<button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type=button data-toggle=collapse data-target=#td-section-nav aria-controls=td-docs-nav aria-expanded=false aria-label="Toggle section navigation"></button></form><nav class="collapse td-sidebar-nav" id=td-section-nav><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/ class="align-left pl-0 pr-2 td-sidebar-link td-sidebar-link__section">Documentation</a></li><ul><li class="collapse show" id=sp><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/spr/ class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">systemPipeR</a></li><ul><li class="collapse show" id=spspr><a class="td-sidebar-link td-sidebar-link__page" id=m-spsprintroduction href=../../../sp/spr/introduction/>Introduction</a>
<a class="td-sidebar-link td-sidebar-link__page active" id=m-spsprgettingstarted href=../../../sp/spr/gettingstarted/>Getting Started</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprspr_run href=../../../sp/spr/spr_run/>How to run a Workflow</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsteps href=../../../sp/spr/steps/>Workflow steps overview</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprtemplates href=../../../sp/spr/templates/>Workflow Templates</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprcwl_syntax href=../../../sp/spr/cwl_syntax/>CWL syntax</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprcwl_and_spr href=../../../sp/spr/cwl_and_spr/>SPR and CWL</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsprinstall href=../../../sp/spr/sprinstall/>SPR detailed installation instructions</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprspr_funcs href=../../../sp/spr/spr_funcs/>Help Manual</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprsp_docker href=../../../sp/spr/sp_docker/>SPR Docker container</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spspreditor href=../../../sp/spr/editor/>Workflow Plot Editor</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprviz_editorbower_componentssvg-pan-zoomissue_template href=../../../sp/spr/viz_editor/bower_components/svg-pan-zoom/issue_template/></a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/spt/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeTools</a></li><ul><li class=collapse id=spspt><a class="td-sidebar-link td-sidebar-link__page" id=m-spsptsystempipetools href=../../../sp/spt/systempipetools/>Data Visualizations</a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../sp/sprdata/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeRdata</a></li><ul><li class=collapse id=spsprdata><a class="td-sidebar-link td-sidebar-link__page" id=m-spsprdatasystempiperdata href=../../../sp/sprdata/systempiperdata/>Workflow templates and sample data</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-spsprdatanewwf href=../../../sp/sprdata/newwf/>Rules to create a new Workflow Template</a></li></ul></ul><a class="td-sidebar-link td-sidebar-link__page" id=m-spsp_docker href=../../../sp/sp_docker/>SPR Docker container</a></li></ul></ul></nav></div></div><div class="d-none d-xl-block col-xl-2 td-toc d-print-none"><div class="td-page-meta ml-2 pb-1 pt-2 mb-0"><a href=https://raw.githubusercontent.com/systemPipeR/systemPipeR.github.io/main/content/en/sp/spr/gettingstarted.Rmd target=_blank class=source-link><i class="fa fa-code fa-fw"></i>View Rmd source</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/blob/main/content/en/sp/spr/gettingstarted.md target=_blank><i class="fab fa-github fa-fw"></i>View on Github</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/edit/main/content/en/sp/spr/gettingstarted.md target=_blank><i class="fa fa-edit fa-fw"></i>Edit this page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/new/main/content/en/sp/spr/gettingstarted.md?filename=change-me.md&value=---%0Atitle%3A+%22Long+Page+Title%22%0AlinkTitle%3A+%22Short+Nav+Title%22%0Aweight%3A+100%0Adescription%3A+%3E-%0A+++++Page+description+for+heading+and+indexes.%0A---%0A%0A%23%23+Heading%0A%0AEdit+this+template+to+create+your+new+page.%0A%0A%2A+Give+it+a+good+name%2C+ending+in+%60.md%60+-+e.g.+%60getting-started.md%60%0A%2A+Edit+the+%22front+matter%22+section+at+the+top+of+the+page+%28weight+controls+how+its+ordered+amongst+other+pages+in+the+same+directory%3B+lowest+number+first%29.%0A%2A+Add+a+good+commit+message+at+the+bottom+of+the+page+%28%3C80+characters%3B+use+the+extended+description+field+for+more+detail%29.%0A%2A+Create+a+new+branch+so+you+can+preview+your+new+file+and+request+a+review+via+Pull+Request.%0A" target=_blank><i class="fa fa-edit fa-fw"></i>Create child page</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/issues/new target=_blank><i class="fas fa-tasks fa-fw"></i>Create project issue</a></div><nav id=TableOfContents><ul><li><ul><li><a href=#getting-started>Getting Started</a><ul><li><a href=#installation>Installation</a></li><li><a href=#loading-package-and-documentation>Loading package and documentation</a></li><li><a href=#how-to-get-help-for-systempiper>How to get help for systemPipeR</a></li></ul></li><li><a href=#project-structure>Project structure</a><ul><li><a href=#directory-structure>Directory Structure</a></li><li><a href=#structure-of-initial-targets-file>Structure of initial <em><code>targets</code></em> file</a><ul><li><a href=#structure-of-targets-file-for-single-end-se-samples>Structure of <em><code>targets</code></em> file for single-end (SE) samples</a></li><li><a href=#structure-of-targets-file-for-paired-end-pe-samples>Structure of <em><code>targets</code></em> file for paired-end (PE) samples</a></li><li><a href=#structure-of-targets-file-for-hello-world-example>Structure of <em><code>targets</code></em> file for “Hello World” example</a></li><li><a href=#sample-comparisons>Sample comparisons</a></li></ul></li><li><a href=#downstream-targets-files-description>Downstream targets files description</a></li></ul></li><li><a href=#structure-of-the-new-parameters-files>Structure of the new parameters files</a></li><li><a href=#references>References</a></li></ul></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 pl-md-5" role=main><nav aria-label=breadcrumb class="d-none d-md-block d-print-none"><ol class="breadcrumb spb-1"><li class=breadcrumb-item><a href=../../../sp/>Documentation</a></li><li class=breadcrumb-item><a href=../../../sp/spr/>systemPipeR</a></li><li class="breadcrumb-item active" aria-current=page><a href=../../../sp/spr/gettingstarted/>Getting Started</a></li></ol></nav><div class=td-content><h1>Getting Started</h1><script src=../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js></script><link href=../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/datatables-binding/datatables.js></script><script src=../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js></script><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css rel=stylesheet><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js></script><link href=../../../rmarkdown-libs/dt-ext-fixedcolumns/css/fixedColumns.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-fixedcolumns/js/dataTables.fixedColumns.min.js></script><link href=../../../rmarkdown-libs/dt-ext-scroller/css/scroller.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-scroller/js/dataTables.scroller.min.js></script><link href=../../../rmarkdown-libs/crosstalk/css/crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js></script><script src=../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js></script><link href=../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/datatables-binding/datatables.js></script><script src=../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js></script><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css rel=stylesheet><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js></script><link href=../../../rmarkdown-libs/dt-ext-fixedcolumns/css/fixedColumns.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-fixedcolumns/js/dataTables.fixedColumns.min.js></script><link href=../../../rmarkdown-libs/dt-ext-scroller/css/scroller.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-scroller/js/dataTables.scroller.min.js></script><link href=../../../rmarkdown-libs/crosstalk/css/crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js></script><script src=../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js></script><link href=../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/datatables-binding/datatables.js></script><script src=../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js></script><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css rel=stylesheet><link href=../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js></script><link href=../../../rmarkdown-libs/dt-ext-fixedcolumns/css/fixedColumns.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-fixedcolumns/js/dataTables.fixedColumns.min.js></script><link href=../../../rmarkdown-libs/dt-ext-scroller/css/scroller.dataTables.min.css rel=stylesheet><script src=../../../rmarkdown-libs/dt-ext-scroller/js/dataTables.scroller.min.js></script><link href=../../../rmarkdown-libs/crosstalk/css/crosstalk.css rel=stylesheet><script src=../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js></script><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){document.querySelector("h1").className="title";});</script><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){var links=document.links;for(var i=0,linksLength=links.length;i<linksLength;i++)
if(links[i].hostname!=window.location.hostname)
links[i].target='_blank';});</script><h2 id=getting-started>Getting Started</h2><h3 id=installation>Installation</h3><p><a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html><em><code>systemPipeR</code></em></a>
environment can be installed from the R console using the <a href=https://cran.r-project.org/web/packages/BiocManager/index.html><em><code>BiocManager::install</code></em></a>
command. The associated data package <a href=http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html><em><code>systemPipeRdata</code></em></a>
can be installed the same way. The latter is a helper package for generating <em><code>systemPipeR</code></em>
workflow environments with a single command containing all parameter files and
sample data required to quickly test and run workflows.</p><pre><code class=language-r>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;systemPipeR&quot;)
BiocManager::install(&quot;systemPipeRdata&quot;)
</code></pre><p>Please note that if you desire to use a third-party command-line tool, the particular
tool and dependencies need to be installed and exported in your PATH.
See <a href=https://systempipe.org/sp/spr/sprinstall/>details</a>.</p><h3 id=loading-package-and-documentation>Loading package and documentation</h3><pre><code class=language-r>library(&quot;systemPipeR&quot;)  # Loads the package
library(help = &quot;systemPipeR&quot;)  # Lists package info
vignette(&quot;systemPipeR&quot;)  # Opens vignette
</code></pre><h3 id=how-to-get-help-for-systempiper>How to get help for systemPipeR</h3><p>All questions about the package or any particular function should be posted to
the Bioconductor support site <a href=https://support.bioconductor.org>https://support.bioconductor.org</a>.</p><p>Please add the “<em><code>systemPipeR</code></em>” tag to your question, and the package authors will
automatically receive an alert.</p><p>We appreciate receiving reports of bugs in the functions or documentation and
suggestions for improvement. For that, please consider opening an issue at
<a href=https://github.com/tgirke/systemPipeR/issues/new>GitHub</a>.</p><h2 id=project-structure>Project structure</h2><p><em><code>systemPipeR</code></em> expects a project directory structure that consists of a directory
where users may store all the raw data, the results directory that will be reserved
for all the outfiles files or new output folders, and the parameters directory.</p><p>This structure allows reproducibility and collaboration across the data science
team since internally relative paths are used. Users could transfer this project
to a different location and still be able to run the entire workflow. Also, it
increases efficiency and data management once the raw data is kept in a separate
folder and avoids duplication.</p><h3 id=directory-structure>Directory Structure</h3><p><a href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html><em><code>systemPipeRdata</code></em></a>,
helper package, provides pre-configured workflows, reporting
templates, and sample data loaded as demonstrated below. With a single command,
the package allows creating the workflow environment containing the structure
described here (see Figure <a href=#fig:dir>1</a>).</p><p>Directory names are indicated in <span style=color:grey><em><strong>green</strong></em></span>.
Users can change this structure as needed, but need to adjust the code in their
workflows accordingly.</p><ul><li><span style=color:green><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>myproject/</em>)<ul><li>This is the root directory of the R session running the workflow.</li><li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li><li>Note, this directory can have any name (<em>e.g.</em> <span style=color:green><em><strong>myproject</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li><li><strong>Important subdirectories</strong>:<ul><li><span style=color:green><em><strong>param/</strong></em></span><ul><li><span style=color:green><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the parameter and configuration files. To organize workflows, each can have its own subdirectory, where all <code>*.cwl</code> and <code>*input.yml</code> files need to be in the same subdirectory.</li></ul></li><li><span style=color:green><em><strong>data/</strong></em></span><ul><li>Raw data (<em>e.g.</em> FASTQ files)</li><li>FASTA file of reference (<em>e.g.</em> reference genome)</li><li>Annotation files</li><li>Metadata</li><li>etc.</li></ul></li><li><span style=color:green><em><strong>results/</strong></em></span><ul><li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li><li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li></ul></li></ul></li></ul></li></ul><div class=figure style=text-align:center><p><img src=spr_project.png alt="systemPipeR's preconfigured directory structure." width=100%></p><p class=caption>Figure 1: systemPipeR’s preconfigured directory structure.</p></div><p>The following parameter files are included in each workflow template:</p><ol><li><em><code>targets.txt</code></em>: initial one provided by user; downstream <em><code>targets_*.txt</code></em> files are generated automatically</li><li><em><code>*.param/cwl</code></em>: defines parameter for input/output file operations, <em>e.g.</em>:<ul><li><em><code>hisat2/hisat2-mapping-se.cwl</code></em></li><li><em><code>hisat2/hisat2-mapping-se.yml</code></em></li></ul></li><li><em><code>*_run.sh</code></em>: optional bash scripts</li><li>Configuration files for computer cluster environments (skip on single machines):<ul><li><em><code>.batchtools.conf.R</code></em>: defines the type of scheduler for <em><code>batchtools</code></em> pointing to template file of cluster, and located in user’s home directory</li><li><em><code>batchtools.*.tmpl</code></em>: specifies parameters of scheduler used by a system, <em>e.g.</em> Torque, SGE, Slurm, etc.</li></ul></li></ol><h3 id=structure-of-initial-targets-file>Structure of initial <em><code>targets</code></em> file</h3><p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample
comparisons of an analysis workflow. The following shows the format of a sample
<em><code>targets</code></em> file included in the package. It also can be viewed and downloaded
from <em><code>systemPipeR</code></em>’s GitHub repository <a href=https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt>here</a>.
In a target file with a single type of input files, here FASTQ files of
single-end (SE) reads, the first column describe the path and the second column
represents a unique id name for each sample. The third column called <code>Factor</code>
represents the biological replicates. All subsequent columns are additional
information, and any number of extra columns can be added as needed.</p><p>Users should note here, the usage of targets files is optional when using
<em><code>systemPipeR's</code></em> new workflow management interface. They can be replaced by a standard YAML
input file used by CWL. Since for organizing experimental variables targets
files are extremely useful and user-friendly. Thus, we encourage users to keep using
them.</p><h4 id=structure-of-targets-file-for-single-end-se-samples>Structure of <em><code>targets</code></em> file for single-end (SE) samples</h4><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targets.txt&quot;, package = &quot;systemPipeR&quot;)
showDF(read.delim(targetspath, comment.char = &quot;#&quot;))
</code></pre><pre><code>## Loading required namespace: DT
</code></pre><div id=htmlwidget-1 style=width:100%;height:auto class="datatables html-widget"></div><script type=application/json data-for=htmlwidget-1>{"x":{"filter":"none","vertical":false,"extensions":["FixedColumns","Scroller"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["./data/SRR446027_1.fastq.gz","./data/SRR446028_1.fastq.gz","./data/SRR446029_1.fastq.gz","./data/SRR446030_1.fastq.gz","./data/SRR446031_1.fastq.gz","./data/SRR446032_1.fastq.gz","./data/SRR446033_1.fastq.gz","./data/SRR446034_1.fastq.gz","./data/SRR446035_1.fastq.gz","./data/SRR446036_1.fastq.gz","./data/SRR446037_1.fastq.gz","./data/SRR446038_1.fastq.gz","./data/SRR446039_1.fastq.gz","./data/SRR446040_1.fastq.gz","./data/SRR446041_1.fastq.gz","./data/SRR446042_1.fastq.gz","./data/SRR446043_1.fastq.gz","./data/SRR446044_1.fastq.gz"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],["M1","M1","A1","A1","V1","V1","M6","M6","A6","A6","V6","V6","M12","M12","A12","A12","V12","V12"],["Mock.1h.A","Mock.1h.B","Avr.1h.A","Avr.1h.B","Vir.1h.A","Vir.1h.B","Mock.6h.A","Mock.6h.B","Avr.6h.A","Avr.6h.B","Vir.6h.A","Vir.6h.B","Mock.12h.A","Mock.12h.B","Avr.12h.A","Avr.12h.B","Vir.12h.A","Vir.12h.B"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName<\/th>\n      <th>SampleName<\/th>\n      <th>Factor<\/th>\n      <th>SampleLong<\/th>\n      <th>Experiment<\/th>\n      <th>Date<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing
the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the
path to the corresponding <em><code>targets</code></em> file.</p><h4 id=structure-of-targets-file-for-paired-end-pe-samples>Structure of <em><code>targets</code></em> file for paired-end (PE) samples</h4><p>For paired-end (PE) samples, the structure of the targets file is similar, where
users need to provide two FASTQ path columns: <em><code>FileName1</code></em> and <em><code>FileName2</code></em>
with the paths to the PE FASTQ files.</p><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE.txt&quot;, package = &quot;systemPipeR&quot;)
showDF(read.delim(targetspath, comment.char = &quot;#&quot;))
</code></pre><div id=htmlwidget-2 style=width:100%;height:auto class="datatables html-widget"></div><script type=application/json data-for=htmlwidget-2>{"x":{"filter":"none","vertical":false,"extensions":["FixedColumns","Scroller"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["./data/SRR446027_1.fastq.gz","./data/SRR446028_1.fastq.gz","./data/SRR446029_1.fastq.gz","./data/SRR446030_1.fastq.gz","./data/SRR446031_1.fastq.gz","./data/SRR446032_1.fastq.gz","./data/SRR446033_1.fastq.gz","./data/SRR446034_1.fastq.gz","./data/SRR446035_1.fastq.gz","./data/SRR446036_1.fastq.gz","./data/SRR446037_1.fastq.gz","./data/SRR446038_1.fastq.gz","./data/SRR446039_1.fastq.gz","./data/SRR446040_1.fastq.gz","./data/SRR446041_1.fastq.gz","./data/SRR446042_1.fastq.gz","./data/SRR446043_1.fastq.gz","./data/SRR446044_1.fastq.gz"],["./data/SRR446027_2.fastq.gz","./data/SRR446028_2.fastq.gz","./data/SRR446029_2.fastq.gz","./data/SRR446030_2.fastq.gz","./data/SRR446031_2.fastq.gz","./data/SRR446032_2.fastq.gz","./data/SRR446033_2.fastq.gz","./data/SRR446034_2.fastq.gz","./data/SRR446035_2.fastq.gz","./data/SRR446036_2.fastq.gz","./data/SRR446037_2.fastq.gz","./data/SRR446038_2.fastq.gz","./data/SRR446039_2.fastq.gz","./data/SRR446040_2.fastq.gz","./data/SRR446041_2.fastq.gz","./data/SRR446042_2.fastq.gz","./data/SRR446043_2.fastq.gz","./data/SRR446044_2.fastq.gz"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],["M1","M1","A1","A1","V1","V1","M6","M6","A6","A6","V6","V6","M12","M12","A12","A12","V12","V12"],["Mock.1h.A","Mock.1h.B","Avr.1h.A","Avr.1h.B","Vir.1h.A","Vir.1h.B","Mock.6h.A","Mock.6h.B","Avr.6h.A","Avr.6h.B","Vir.6h.A","Vir.6h.B","Mock.12h.A","Mock.12h.B","Avr.12h.A","Avr.12h.B","Vir.12h.A","Vir.12h.B"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName1<\/th>\n      <th>FileName2<\/th>\n      <th>SampleName<\/th>\n      <th>Factor<\/th>\n      <th>SampleLong<\/th>\n      <th>Experiment<\/th>\n      <th>Date<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"columnDefs":[{"className":"dt-right","targets":6},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><h4 id=structure-of-targets-file-for-hello-world-example>Structure of <em><code>targets</code></em> file for “Hello World” example</h4><p>In this example, <em><code>targets</code></em> file presents only two columns, which the first one
are the different phrases used by the <code>echo</code> command-line and the second column
it is the sample <code>id</code>. The <code>id</code> column is required, and each sample id should be unique.</p><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata/cwl/example/targets_example.txt&quot;, package = &quot;systemPipeR&quot;)
showDF(read.delim(targetspath, comment.char = &quot;#&quot;))
</code></pre><div id=htmlwidget-3 style=width:100%;height:auto class="datatables html-widget"></div><script type=application/json data-for=htmlwidget-3>{"x":{"filter":"none","vertical":false,"extensions":["FixedColumns","Scroller"],"data":[["1","2","3"],["Hello World!","Hello USA!","Hello Bioconductor!"],["M1","M2","M3"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Message<\/th>\n      <th>SampleName<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script><h4 id=sample-comparisons>Sample comparisons</h4><p>Sample comparisons are defined in the header lines of the <em><code>targets</code></em> file
starting with ‘<code># &lt;CMP></code>.’</p><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE.txt&quot;, package = &quot;systemPipeR&quot;)
readLines(targetspath)[1:4]
</code></pre><pre><code>## [1] &quot;# Project ID: Arabidopsis - Pseudomonas alternative splicing study (SRA: SRP010938; PMID: 24098335)&quot;                                                                              
## [2] &quot;# The following line(s) allow to specify the contrasts needed for comparative analyses, such as DEG identification. All possible comparisons can be specified with 'CMPset: ALL'.&quot;
## [3] &quot;# &lt;CMP&gt; CMPset1: M1-A1, M1-V1, A1-V1, M6-A6, M6-V6, A6-V6, M12-A12, M12-V12, A12-V12&quot;                                                                                             
## [4] &quot;# &lt;CMP&gt; CMPset2: ALL&quot;
</code></pre><p>The function <em><code>readComp</code></em> imports the comparison information and stores it in a
<em><code>list</code></em>. Alternatively, <em><code>readComp</code></em> can obtain the comparison information from
the corresponding <em><code>SYSargsList</code></em> step (see below). Note, these header lines are
optional. They are mainly useful for controlling comparative analyses according
to certain biological expectations, such as identifying differentially expressed
genes in RNA-Seq experiments based on simple pair-wise comparisons.</p><pre><code class=language-r>readComp(file = targetspath, format = &quot;vector&quot;, delim = &quot;-&quot;)
</code></pre><pre><code>## $CMPset1
## [1] &quot;M1-A1&quot;   &quot;M1-V1&quot;   &quot;A1-V1&quot;   &quot;M6-A6&quot;   &quot;M6-V6&quot;   &quot;A6-V6&quot;   &quot;M12-A12&quot;
## [8] &quot;M12-V12&quot; &quot;A12-V12&quot;
## 
## $CMPset2
##  [1] &quot;M1-A1&quot;   &quot;M1-V1&quot;   &quot;M1-M6&quot;   &quot;M1-A6&quot;   &quot;M1-V6&quot;   &quot;M1-M12&quot;  &quot;M1-A12&quot; 
##  [8] &quot;M1-V12&quot;  &quot;A1-V1&quot;   &quot;A1-M6&quot;   &quot;A1-A6&quot;   &quot;A1-V6&quot;   &quot;A1-M12&quot;  &quot;A1-A12&quot; 
## [15] &quot;A1-V12&quot;  &quot;V1-M6&quot;   &quot;V1-A6&quot;   &quot;V1-V6&quot;   &quot;V1-M12&quot;  &quot;V1-A12&quot;  &quot;V1-V12&quot; 
## [22] &quot;M6-A6&quot;   &quot;M6-V6&quot;   &quot;M6-M12&quot;  &quot;M6-A12&quot;  &quot;M6-V12&quot;  &quot;A6-V6&quot;   &quot;A6-M12&quot; 
## [29] &quot;A6-A12&quot;  &quot;A6-V12&quot;  &quot;V6-M12&quot;  &quot;V6-A12&quot;  &quot;V6-V12&quot;  &quot;M12-A12&quot; &quot;M12-V12&quot;
## [36] &quot;A12-V12&quot;
</code></pre><h3 id=downstream-targets-files-description>Downstream targets files description</h3><p>After the step which required the initial targets file information, the downstream
targets files are created automatically (see Figure <a href=#fig:targetsFig>2</a>).
Each step that uses the previous step outfiles as an input, the new targets input
will be managed internally by the workflow instances, establishing connectivity
among the steps in the workflow.
<em><code>systemPipeR</code></em> provides features to automatically and systematically build this
connection, providing security that all the samples will be managed efficiently
and reproducibly.</p><div class=figure style=text-align:center><p><img src=targets_con.png alt="_systemPipeR_ automatically creates the downstream `targets` files based on the previous steps outfiles. A) Usually, users provide the initial `targets` files, and this step will generate some outfiles, as demonstrated on B. Then, those files are used to build the new `targets` files as inputs in the next step. _`systemPipeR`_ (C) manages this connectivity among the steps automatically for the users." width=100%></p><p class=caption>Figure 2: *systemPipeR* automatically creates the downstream `targets` files based on the previous steps outfiles. A) Usually, users provide the initial `targets` files, and this step will generate some outfiles, as demonstrated on B. Then, those files are used to build the new `targets` files as inputs in the next step. *`systemPipeR`* (C) manages this connectivity among the steps automatically for the users.</p></div><h2 id=structure-of-the-new-parameters-files>Structure of the new parameters files</h2><p>The parameters and configuration required for running command-line software are
provided by the widely used community standard <a href=https://www.commonwl.org/>Common Workflow Language</a> (CWL) (Amstutz et al. 2016), which describes
parameters analysis workflows in a generic and reproducible manner.
For R-based workflow steps, param files are not required.
For a complete overview of the CWL syntax, please see this <a href=https://systempipe.org/sp/spr/cwl_syntax/>section</a>.
Also, we have a dedicated section explain how to <em><code>systemPipeR</code></em> establish the
connection between the CWL parameters files and the targets files. Please see <a href=https://systempipe.org/sp/spr/cwl_and_spr>here</a>.</p><h2 id=references>References</h2><div id=refs class="references csl-bib-body hanging-indent"><div id=ref-Amstutz2016-ka class=csl-entry><p>Amstutz, Peter, Michael R Crusoe, Nebojša Tijanić, Brad Chapman, John Chilton, Michael Heuer, Andrey Kartashov, et al. 2016. “Common Workflow Language, V1.0,” July. <a href=https://doi.org/10.6084/m9.figshare.3115156.v2>https://doi.org/10.6084/m9.figshare.3115156.v2</a>.</p></div></div><div class="text-muted mt-5 pt-3 border-top">Last modified 2021-10-30: <a href=https://github.com/systemPipeR/systemPipeR.github.io/commit/196d9eded237b176972e85ca00a224195dc55254>Bot build site (196d9ed)</a></div></div></main></div></div><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2021 The systemPipe Project All Rights Reserved</small><p class=mt-2><a href=../../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script id=MathJax-script async src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js integrity="sha256-+CC/xQfD6+/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin=anonymous></script><script src=../../../js/prism.js></script></body></html>