<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>systemPipeRdata</title>
    <meta charset="utf-8" />
    <meta name="author" content=" Daniela Cassol" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/font-awesome/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/v4-shims.css" rel="stylesheet" />
    <link href="libs/tile-view/tile-view.css" rel="stylesheet" />
    <script src="libs/tile-view/tile-view.js"></script>
    <link href="libs/animate.css/animate.xaringan.css" rel="stylesheet" />
    <link href="libs/tachyons/tachyons.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
    <link rel="stylesheet" href="css/custom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <em>systemPipeRdata</em>
### <br/>Daniela Cassol
### <h3>
University of California Riverside
<h3/>
### <h3>
2020/08/28 (updated: 2020-10-26)
<h3/>

---


layout: true
background-image: url(https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png)
background-position: 99% 1%
background-size: 10%







---
class:  middle

## Outline

#### Introduction
#### Live Demo
    - Install Package
    - Quick Start
    - Directory Structure
    - Workflow Collection
    - Install Workflow Template

---
class: inverse, center, middle

# Introduction

---
## Introduction

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
[_systemPipeRdata_](http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html) is a helper package to generate with a single command workflow templates 
that are intended to be used by its parent package [_systemPipeR_](http://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html).

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Includes sample data for testing.

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Users can create *new workflows* or change and extend existing ones. 

---
class: inverse, center, middle

# &lt;i class="fas fa-code"&gt;&lt;/i&gt; Live Demo 

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Install Package

Install the **systemPipeRdata** package from [Bioconductor](http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html):


```r
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("systemPipeRdata")
```

### &lt;i class="fas fa-book"&gt;&lt;/i&gt; Load Package and Documentation

&lt;i class="fas fa-question" style="color:#00758a;"&gt;&lt;/i&gt; 
Load package:


```r
library("systemPipeRdata")
```

&lt;i class="fas fa-question" style="color:#00758a;"&gt;&lt;/i&gt; 
Access help:


```r
library(help="systemPipeRdata")
vignette("systemPipeRdata")
```

---
## &lt;i class="fas fa-code"&gt;&lt;/i&gt; Quick Start

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Generate workflow template:


```r
genWorkenvir(workflow="varseq")
setwd("varseq")
```

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Command-line alternative for generating workflow environments:


```bash
$ Rscript -e "systemPipeRdata::genWorkenvir(workflow='varseq', mydirname=NULL, url=NULL, urlname=NULL)"
```

---
## &lt;i class="far fa-folder-open"&gt;&lt;/i&gt; Directory Structure

The workflow templates generated by _`genWorkenvir`_ contain the following preconfigured directory structure:

&lt;img src="directory.png" height="425px" class="center" /&gt;

---
## &lt;i class="fas fa-project-diagram"&gt;&lt;/i&gt; Workflows Collection

Browse pipelines that are currently available as part of the *systemPipeR* toolkit

.small[
| WorkFlow |  Description  | Version | GitHub | CI Testing | 
|:-----------------------:|:----------------:|:----------------:|:--------------------:|:-------------------------------:|
| [systemPipeChIPseq](https://systempiper.github.io/systemPipeChIPseq/articles/systemPipeChIPseq.html) | ChIP-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeChIPseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeChIPseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeRIBOseq](https://systempiper.github.io/systemPipeRIBOseq/articles/systemPipeRIBOseq.html) | RIBO-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeRIBOseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeRIBOseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeRNAseq](https://systempiper.github.io/systemPipeRNAseq/articles/systemPipeRNAseq.html) | RNA-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeRNAseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeRNAseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeVARseq](https://systempiper.github.io/systemPipeVARseq/articles/systemPipeVARseq.html) | VAR-Seq Workflow Template | v1.0 &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png" align="center" style="width:20px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeVARseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeVARseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeMethylseq](https://github.com/systemPipeR/systemPipeMethylseq) | Methyl-Seq Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeMethylseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeMethylseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeDeNovo](https://github.com/systemPipeR/systemPipeDeNovo) | De novo transcriptome assembly Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeDeNovo"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeDeNovo/workflows/R-CMD-check/badge.svg) |
| [systemPipeCLIPseq](https://github.com/systemPipeR/systemPipeCLIPseq) | CLIP-Seq Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeCLIPseq"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeCLIPseq/workflows/R-CMD-check/badge.svg) |
| [systemPipeMetaTrans](https://github.com/systemPipeR/systemPipeMetaTrans) | Metatranscriptomic Sequencing Workflow Template | devel &lt;img src="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png" align="center" style="width:25px;" /&gt; | &lt;a href="https://github.com/systemPipeR/systemPipeMetaTrans"&gt; &lt;img src="https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png" align="center" style="width:20px;" /&gt; | ![R-CMD-check](https://github.com/systemPipeR/systemPipeMetaTrans/workflows/R-CMD-check/badge.svg) |
]

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Install Workflow

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Check the workflow template availability 

```r
availableWF(github = TRUE)

# $systemPipeRdata
# [1] "chipseq" "new"     "riboseq" "rnaseq"  "varseq" 
# 
# $github
#                          workflow                       branches version                                               html       description
# 1   systemPipeR/systemPipeChIPseq                         master release   https://github.com/systemPipeR/systemPipeChIPseq Workflow Template
# 2   systemPipeR/systemPipeRIBOseq                         master release   https://github.com/systemPipeR/systemPipeRIBOseq Workflow Template
# 3    systemPipeR/systemPipeRNAseq cluster, master, singleMachine release    https://github.com/systemPipeR/systemPipeRNAseq Workflow Template
# 4    systemPipeR/systemPipeVARseq                         master release    https://github.com/systemPipeR/systemPipeVARseq Workflow Template
# 5   systemPipeR/systemPipeCLIPseq                         master   devel   https://github.com/systemPipeR/systemPipeCLIPseq Workflow Template
# 6    systemPipeR/systemPipeDeNovo                         master   devel    https://github.com/systemPipeR/systemPipeDeNovo Workflow Template
# 7 systemPipeR/systemPipeMetaTrans                         master   devel https://github.com/systemPipeR/systemPipeMetaTrans Workflow Template
# 8 systemPipeR/systemPipeMethylseq                         master   devel https://github.com/systemPipeR/systemPipeMethylseq Workflow Template
```

---
## &lt;i class="fas fa-box-open"&gt;&lt;/i&gt; Install Workflow

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Install Workflow Template from [*systemPipeR* Organization](https://github.com/systemPipeR):


```r
genWorkenvir(workflow="systemPipeR/systemPipeVARseq", mydirname=NULL)

# [1] "Package systemPipeVARseq is successfully installed!"
# [1] "The 'demo data'was successfully copied to your project."
# [1] "The 'param' directory was successfully copied to your project."
# [1] "Generated systemPipeVARseq directory. Next run in systemPipeR/systemPipeVARseq directory, the R code from *.Rmd template interactively. Alternatively, workflows can be exectued with a single command as instructed in the vignette."
```

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Install Workflow Template with a specific *R Markdown* file:
 

```r
genWorkenvir(workflow="systemPipeR/systemPipeRNAseq", mydirname="rnaseq",
             url = "https://raw.githubusercontent.com/systemPipeR/systemPipeRNAseq/cluster/vignettes/systemPipeRNAseq.Rmd",
             urlname = "rnaseq_V-cluster.Rmd")
```

---
## &lt;i class="fas fa-mouse-pointer"&gt;&lt;/i&gt; Dynamic Workflow Template

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Create dynamic Workflow Templates with *RStudio*

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
`File -&gt; New File -&gt; R Markdown -&gt; From Template`
.center[
&lt;img src="rstudio.png" height="425px" class="center" /&gt;
]


---
class: middle

# Thanks!

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Browse source code at &lt;a href="https://github.com/tgirke/systemPipeRdata/"&gt; &lt;i class="fab fa-github fa-2x"&gt;&lt;/i&gt;&lt;/a&gt;

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
Ask a question about systemPipeRdata at Bioconductor Support Page
&lt;a href="https://support.bioconductor.org/t/systempiper/"&gt;&lt;i class="far fa-question-circle fa-2x"&gt;&lt;/i&gt;&lt;/a&gt;

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
*systemPipeRdata* at [Bioconductor](http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html)

&lt;i class="fas fa-hand-point-right" style="color:#00758a;"&gt;&lt;/i&gt; 
[https://systempipe.org/](https://systempipe.org/)

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
