<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>GO term enrichment analysis for large numbers of gene sets — GOHyperGAll • systemPipeR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="GO term enrichment analysis for large numbers of gene sets — GOHyperGAll"><meta property="og:description" content="To test a sample population of genes for over-representation of GO terms, the
core function GOHyperGAll computes for all nodes in the three GO networks
(BP, CC and MF) an enrichment test based on the hypergeometric distribution and 
returns the corresponding raw and Bonferroni corrected p-values. 
Subsequently, a filter function supports GO Slim analyses using default or 
custom GO Slim categories. Several convenience functions are provided to process
large numbers of gene sets (e.g. clusters from partitioning results) and to 
visualize the results.
Note: GOHyperGAll provides similar utilities as the GOHyperG
function in the GOstats package. The main difference is that
GOHyperGAll simplifies processing of large numbers of gene sets, as well
as the usage of custom array-to-gene and gene-to-GO mappings."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tgirke/systemPipeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GO term enrichment analysis for large numbers of gene sets</h1>
    
    <div class="hidden name"><code>GOHyperGAll.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>To test a sample population of genes for over-representation of GO terms, the
core function <code>GOHyperGAll</code> computes for all nodes in the three GO networks
(BP, CC and MF) an enrichment test based on the hypergeometric distribution and 
returns the corresponding raw and Bonferroni corrected p-values. 
Subsequently, a filter function supports GO Slim analyses using default or 
custom GO Slim categories. Several convenience functions are provided to process
large numbers of gene sets (e.g. clusters from partitioning results) and to 
visualize the results.</p>
<p>Note: <code>GOHyperGAll</code> provides similar utilities as the <code>GOHyperG</code>
function in the <code>GOstats</code> package. The main difference is that
<code>GOHyperGAll</code> simplifies processing of large numbers of gene sets, as well
as the usage of custom array-to-gene and gene-to-GO mappings.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co">## Generate gene-to-GO mappings and store as catDB object </span>
<span class="fu">makeCATdb</span><span class="op">(</span><span class="va">myfile</span>, lib <span class="op">=</span> <span class="cn">NULL</span>, org <span class="op">=</span> <span class="st">""</span>, colno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, idconv <span class="op">=</span> <span class="cn">NULL</span>, 
            rootUK<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co">## Enrichment function</span>
<span class="fu">GOHyperGAll</span><span class="op">(</span><span class="va">catdb</span>, gocat <span class="op">=</span> <span class="st">"MF"</span>, <span class="va">sample</span>, Nannot <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="co">## GO slim analysis</span>
<span class="fu">GOHyperGAll_Subset</span><span class="op">(</span><span class="va">catdb</span>, <span class="va">GOHyperGAll_result</span>, sample <span class="op">=</span> <span class="va">test_sample</span>, 
            type <span class="op">=</span> <span class="st">"goSlim"</span>, <span class="va">myslimv</span><span class="op">)</span>

<span class="co">## Reduce GO term redundancy </span>
<span class="fu">GOHyperGAll_Simplify</span><span class="op">(</span><span class="va">GOHyperGAll_result</span>, gocat <span class="op">=</span> <span class="st">"MF"</span>, cutoff <span class="op">=</span> <span class="fl">0.001</span>, correct <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Batch analysis of many gene sets</span>
<span class="fu">GOCluster_Report</span><span class="op">(</span><span class="va">catdb</span>, <span class="va">setlist</span>, id_type <span class="op">=</span> <span class="st">"affy"</span>, method <span class="op">=</span> <span class="st">"all"</span>, CLSZ <span class="op">=</span> <span class="fl">10</span>, 
            cutoff <span class="op">=</span> <span class="fl">0.001</span>, gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, myslimv <span class="op">=</span> <span class="st">"default"</span>,
            correct <span class="op">=</span> <span class="cn">TRUE</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span>

<span class="co">## Bar plot of GOCluster_Report results</span>
<span class="fu">goBarplot</span><span class="op">(</span><span class="va">GOBatchResult</span>, <span class="va">gocat</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>myfile</dt>
<dd><p>File with gene-to-GO mappings. Sample files can be downloaded from geneontology.org (http://geneontology.org/GO.downloads.annotations.shtml) or from BioMart as shown in example below.</p></dd>
<dt>colno</dt>
<dd><p>Column numbers referencing in <code>myfile</code> the three target columns containing GOID, GeneID and GOCAT, in that order.</p></dd>
<dt>org</dt>
<dd><p>Optional argument. Currently, the only valid option is <code>org="Arabidopsis"</code> to get rid of transcript duplications in this particular annotation.</p></dd>
<dt>lib</dt>
<dd><p>If the gene-to-GO mappings are obtained from a <code>*.db</code> package from Bioconductor then the package name can be specified under the <code>lib</code> argument of the <code>sampleDFgene2GO</code> function.</p></dd>
<dt>idconv</dt>
<dd><p>Optional id conversion <code>data.frame</code></p></dd>
<dt>catdb</dt>
<dd><p><code>catdb</code> object storing mappings of genes to annotation categories. For details, see ?"SYSargs-class".</p></dd>
<dt>rootUK</dt>
<dd><p>If the argument <code>rootUK</code> is set to <code>TRUE</code> then the root nodes are treated as terminal nodes to account for the new unknown terms.</p></dd>
<dt>sample</dt>
<dd><p><code>character vector</code> containing the test set of gene identifiers</p></dd>
<dt>Nannot</dt>
<dd><p>Defines the minimum number of direct annotations per GO node from the sample 
        set to determine the number of tested hypotheses for the p-value adjustment.</p></dd>
<dt>gocat</dt>
<dd><p>Specifies the GO type, can be assigned one of the following character values: "MF", "BP" and "CC".</p></dd>
<dt>GOHyperGAll_result</dt>
<dd><p><code>data.frame</code> generated by <code>GOHyperGAll</code></p></dd>
<dt>type</dt>
<dd><p>The function <code>GOHyperGAll_Subset</code> subsets the <code>GOHyperGAll</code>
	results by directly assigned GO nodes or custom <code>goSlim</code> categories. 
	The argument <code>type</code> can be assigned the values <code>goSlim</code> or <code>assigned</code>.</p></dd>
<dt>myslimv</dt>
<dd><p>optional argument to provide custom <code>goSlim</code> vector</p></dd>
<dt>cutoff</dt>
<dd><p>p-value cutoff for GO terms to show in result <code>data.frame</code></p></dd>
<dt>correct</dt>
<dd><p>If <code>TRUE</code> the function will favor the selection of terminal (informationich) 
        GO terms that have at the same time a large number of sample matches.</p></dd>
<dt>setlist</dt>
<dd><p><code>list</code> of <code>character vectors</code> containing gene IDs (or array feature 
	IDs). The names of the <code>list</code> components correspond to the set labels, 
	e.g. DEG comparisons or cluster IDs.</p></dd>
<dt>id_type</dt>
<dd><p>specifies type of IDs in input, can be assigned <code>gene</code> or <code>affy</code></p></dd>
<dt>method</dt>
<dd><p>Specifies analysis type. Current options are <code>all</code> for <code>GOHyperGAll</code>, 
	<code>slim</code> for <code>GOHyperGAll_Subset</code> or <code>simplify</code> for 
	<code>GOHyperGAll_Simplify</code>.</p></dd>
<dt>CLSZ</dt>
<dd><p>minimum gene set (cluster) size to consider. Gene sets below this cutoff 
	will be ignored.</p></dd>
<dt>gocats</dt>
<dd><p>Specifies GO type, can be assigned the values "MF", "BP" and "CC".</p></dd>
<dt>recordSpecGO</dt>
<dd><p>argument to report in the result <code>data.frame</code> specific GO IDs for any 
	of the 3 ontologies disregarding whether they meet the specified p-value <code>cutoff</code>, e.g: <code>recordSpecGO=c("GO:0003674", "GO:0008150", "GO:0005575")</code></p></dd>
<dt>GOBatchResult</dt>
<dd><p><code>data.frame</code> generated by <code>GOCluster_Report</code></p></dd>
<dt>...</dt>
<dd><p>additional arguments to pass on</p></dd>
</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p><code>GOHyperGAll_Simplify</code>: The result data frame from <code>GOHyperGAll</code>
	will often contain several connected GO terms with significant scores which 
	can complicate the interpretation of large sample sets. To reduce this redundancy,
	the function <code>GOHyperGAll_Simplify</code> subsets the data frame
	by a user specified p-value cutoff and removes from it all GO nodes with 
        overlapping children sets (OFFSPRING), while the best scoring nodes are 
	retained in the result <code>data.frame</code>.</p>
<p><code>GOCluster_Report</code>: performs the three types of GO term enrichment 
	analyses in batch mode: <code>GOHyperGAll</code>, <code>GOHyperGAll_Subset</code> or 
	<code>GOHyperGAll_Simplify</code>. It processes many gene sets (e.g. gene expression 
        clusters) and returns the results conveniently organized in a single result data frame.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    <p><code>makeCATdb</code> generates <code>catDB</code> object from file.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>This workflow has been published in Plant Physiol (2008) 147, 41-57.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thomas Girke</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>GOHyperGAll_Subset</code>, <code>GOHyperGAll_Simplify</code>, <code>GOCluster_Report</code>, <code>goBarplot</code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

