<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot read distribution across genomic features — plotfeaturetypeCounts • systemPipeR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot read distribution across genomic features — plotfeaturetypeCounts"><meta property="og:description" content="Function to visualize the distribution of reads across different feature types
for many alignment files in parallel. The plots are stacked bar plots
representing the raw or normalized read counts for the sense and antisense
strand of each feature. The graphics results are generated with
ggplot2. Typically, the expected input is generated with the
affiliated featuretypeCounts function."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.22</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tgirke/systemPipeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot read distribution across genomic features</h1>
    
    <div class="hidden name"><code>plotfeaturetypeCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function to visualize the distribution of reads across different feature types
for many alignment files in parallel. The plots are stacked bar plots
representing the raw or normalized read counts for the sense and antisense
strand of each feature. The graphics results are generated with
<code>ggplot2</code>. Typically, the expected input is generated with the
affiliated <code>featuretypeCounts</code> function.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotfeaturetypeCounts</span><span class="op">(</span><span class="va">x</span>, <span class="va">graphicsfile</span>, graphicsformat <span class="op">=</span> <span class="st">"pdf"</span>, scales <span class="op">=</span> <span class="st">"fixed"</span>, anyreadlength <span class="op">=</span> <span class="cn">FALSE</span>, 
                      drop_N_total_aligned <span class="op">=</span> <span class="cn">TRUE</span>, scale_count_val <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span>, scale_length_val <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p><code>data.frame</code> with feature counts generated by the <code>featuretypeCounts</code>
function.</p></dd>
<dt>graphicsfile</dt>
<dd><p>Path to file where to write the output graphics. Note, the function returns
the graphics instructions from <code>ggplot2</code> for interactive plotting in R.
However, due to the complexity of the graphics generated here, the finished results
are written to a file directly.</p></dd>
<dt>graphicsformat</dt>
<dd><p>Graphics file format. Currently, supported formats are: pdf, png or jpeg.
Argument accepts one of them as character string.</p></dd>
<dt>scales</dt>
<dd><p>Scales setting passed on to the <code>facet_wrap</code> function of <code>ggplot2</code>. 
For details see <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">ggplot2::facet_wrap</a></code>. The default <code>fixed</code> assures a
constant scale across all bar plot panels, while <code>free</code> uses the optimum
scale within each bar plot panel. To evaluate plots in all their details, it
may be necessary to generate two graphics files one for each scaling option.</p></dd>
<dt>anyreadlength</dt>
<dd><p>If set to <code>TRUE</code> read length specific read counts will be summed up to 
a single count value to plot read counts for any read length. Otherwise 
the bar plots will show the counts for each read length value.</p></dd>
<dt>drop_N_total_aligned</dt>
<dd><p>If set to <code>TRUE</code> the special feature count <code>N_total_aligned</code> 
will not be included as a separate feature in the plots. However, the 
information will still be used internally for scaling the read counts
to a fixed value if this option is requested under the 
<code>scale_count_val</code> argument.</p></dd>
<dt>scale_count_val</dt>
<dd><p>Scales (normalizes) the read counts to a fixed value of aligned reads
in each sample such as counts per million aligned reads (default is 10^6). 
For this calculation the <code>N_total_aligned</code> values are used that are 
reported in the input <code>data.frame</code> generated by the upstream 
<code>featuretypeCounds</code> function. Assign <code>NULL</code> to turn off
scaling by aligned reads.</p></dd>
<dt>scale_length_val</dt>
<dd><p>Allows to adjust the raw or scaled read counts to a constant length interval
(e.g. <code>scale_length_val=10^3</code> in bps) considering the total genomic length 
of the corresponding feature type. The required genomic length information for 
each feature type is obtained from the <code>Featuretypelength</code> column of the 
input <code>data.frame</code> generated by the <code>featuretypeCount</code> function. To 
turn off feature length adjustment, assign <code>NULL</code> (default).</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function returns bar plot graphics for aligned read counts with read
length resolution if the input contains this information and argument
<code>anyreadlength</code> is set to <code>FALSE</code>. If the input contains counts for
any read length and/or <code>anyreadlength=TRUE</code> then there will be only one
bar per feature and sample. Due to the complexity of the plots, the results
are directly written to file in the chosen graphics format. However, the
function also returns the plotting instructions returned by <code>ggplot2</code> to
display the result components using R's plotting device.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thomas Girke</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>featuretypeCounts</code>, <code>genFeatures</code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Construct SYSargs2 object from param and targets files </span></span>
<span class="r-in"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets<span class="op">=</span><span class="va">targets</span>, wf_file<span class="op">=</span><span class="st">"hisat2/hisat2-mapping-se.cwl"</span>, </span>
<span class="r-in">                  input_file<span class="op">=</span><span class="st">"hisat2/hisat2-mapping-se.yml"</span>, dir_path<span class="op">=</span><span class="va">dir_path</span><span class="op">)</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="loadWorkflow.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>FileName<span class="op">=</span><span class="st">"_FASTQ_PATH1_"</span>, SampleName<span class="op">=</span><span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">args</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">Instance of 'SYSargs2':</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Slot names/accessors: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       targets: 18 (M1A...V12B), targetsheader: 4 (lines)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       modules: 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       wf: 0, clt: 1, yamlinput: 7 (inputs)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       input: 18, output: 18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       cmdlist: 18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Sub Steps:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       1. hisat2-mapping-se (rendered: TRUE)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="co">## Run alignments</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"><span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Features from sample data of systemPipeRdata package</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span class="r-in"><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/annotation"</span>, <span class="st">"tair10.gff"</span>, package<span class="op">=</span><span class="st">"systemPipeRdata"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">file</span>, format<span class="op">=</span><span class="st">"gff3"</span>, organism<span class="op">=</span><span class="st">"Arabidopsis"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="genFeatures.html">genFeatures</a></span><span class="op">(</span><span class="va">txdb</span>, featuretype<span class="op">=</span><span class="st">"all"</span>, reduce_ranges<span class="op">=</span><span class="cn">TRUE</span>, upstream<span class="op">=</span><span class="fl">1000</span>, downstream<span class="op">=</span><span class="fl">0</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Generate and plot feature counts for specific read lengths</span></span>
<span class="r-in"><span class="va">fc</span> <span class="op">&lt;-</span> <span class="fu"><a href="featuretypeCounts.html">featuretypeCounts</a></span><span class="op">(</span>bfl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="va">outpaths</span>, yieldSize<span class="op">=</span><span class="fl">50000</span><span class="op">)</span>, grl<span class="op">=</span><span class="va">feat</span>, singleEnd<span class="op">=</span><span class="cn">TRUE</span>, readlength<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">74</span><span class="op">:</span><span class="fl">76</span>,<span class="fl">99</span><span class="op">:</span><span class="fl">102</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"data.frame"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plotfeaturetypeCounts</span><span class="op">(</span>x<span class="op">=</span><span class="va">fc</span>, graphicsfile<span class="op">=</span><span class="st">"featureCounts.pdf"</span>, graphicsformat<span class="op">=</span><span class="st">"pdf"</span>, scales<span class="op">=</span><span class="st">"fixed"</span>, anyreadlength<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Generate and plot feature counts for any read length  </span></span>
<span class="r-in"><span class="va">fc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="featuretypeCounts.html">featuretypeCounts</a></span><span class="op">(</span>bfl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="va">outpaths</span>, yieldSize<span class="op">=</span><span class="fl">50000</span><span class="op">)</span>, grl<span class="op">=</span><span class="va">feat</span>, singleEnd<span class="op">=</span><span class="cn">TRUE</span>, readlength<span class="op">=</span><span class="cn">NULL</span>, type<span class="op">=</span><span class="st">"data.frame"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plotfeaturetypeCounts</span><span class="op">(</span>x<span class="op">=</span><span class="va">fc2</span>, graphicsfile<span class="op">=</span><span class="st">"featureCounts2.pdf"</span>, graphicsformat<span class="op">=</span><span class="st">"pdf"</span>, scales<span class="op">=</span><span class="st">"fixed"</span>, anyreadlength<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

