<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Scale spliced ranges to genome coordinates — scaleRanges • systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Scale spliced ranges to genome coordinates — scaleRanges" />
<meta property="og:description" content="Function to scale mappings of spliced features (query ranges) to their
corresponding genome coordinates (subject ranges). The method accounts for
introns in the subject ranges that are absent in the query ranges. A use case
example are uORFs predicted in the 5' UTRs sequences using predORF.
These query ranges are given relative to the 5' UTR sequence. The
scaleRanges function will scale them to the corresponding genome
coordinates. This way they can be used in RNA-Seq expression experiments
like other gene ranges." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.27.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Scale spliced ranges to genome coordinates</h1>
    
    <div class="hidden name"><code>scaleRanges.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function to scale mappings of spliced features (query ranges) to their
corresponding genome coordinates (subject ranges). The method accounts for
introns in the subject ranges that are absent in the query ranges. A use case
example are uORFs predicted in the 5' UTRs sequences using <code>predORF</code>.
These query ranges are given relative to the 5' UTR sequence. The
<code>scaleRanges</code> function will scale them to the corresponding genome
coordinates. This way they can be used in RNA-Seq expression experiments
like other gene ranges.</p>
    </div>

    <pre class="usage"><span class='fu'>scaleRanges</span><span class='op'>(</span><span class='va'>subject</span>, <span class='va'>query</span>, type <span class='op'>=</span> <span class='st'>"custom"</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>subject</th>
      <td><p>Genomic ranges provided as <code>GRangesList</code> object. Their name and length
requirements are described under <code>query</code>.</p></td>
    </tr>
    <tr>
      <th>query</th>
      <td><p>Feature level ranges provided as <code>GRangesList</code> object. The names of the 
query ranges need to match the names of the <code>GRangesList</code> object 
assigened to the <code>subject</code> argument. In addition, the length of 
each <code>query</code> range cannot exceed the total length of the corresponding 
<code>subject</code> range set.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>Feature name to use in <code>type</code> column of <code>GRangesList</code> result.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>The setting <code>verbose=FALSE</code> suppresses all print messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Object of class <code>GRangesList</code></p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thomas Girke</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>predORF</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Usage for simple example</span>
<span class='va'>subject</span> <span class='op'>&lt;-</span> <span class='fu'>GRanges</span><span class='op'>(</span>seqnames<span class='op'>=</span><span class='st'>"Chr1"</span>, <span class='fu'>IRanges</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>5</span>,<span class='fl'>15</span>,<span class='fl'>30</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>10</span>,<span class='fl'>25</span>,<span class='fl'>40</span><span class='op'>)</span><span class='op'>)</span>, strand<span class='op'>=</span><span class='st'>"+"</span><span class='op'>)</span> 
<span class='va'>query</span> <span class='op'>&lt;-</span> <span class='fu'>GRanges</span><span class='op'>(</span>seqnames<span class='op'>=</span><span class='st'>"myseq"</span>, <span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>9</span><span class='op'>)</span>, strand<span class='op'>=</span><span class='st'>"+"</span><span class='op'>)</span>
<span class='fu'>scaleRanges</span><span class='op'>(</span><span class='fu'>GRangesList</span><span class='op'>(</span>myid1<span class='op'>=</span><span class='va'>subject</span><span class='op'>)</span>, <span class='fu'>GRangesList</span><span class='op'>(</span>myid1<span class='op'>=</span><span class='va'>query</span><span class='op'>)</span>, type<span class='op'>=</span><span class='st'>"test"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Scaled range 1 of 1 ranges: myid1:1_9 </div><div class='output co'>#&gt; GRangesList object of length 1:
#&gt; $`myid1:1_9`
#&gt; GRanges object with 2 ranges and 3 metadata columns:
#&gt;       seqnames    ranges strand |      feature_by featuretype_id featuretype
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;    &lt;character&gt; &lt;character&gt;
#&gt;   [1]     Chr1      5-10      + |       myid1:1_9    myid1:1_9.1        test
#&gt;   [2]     Chr1     15-17      + |       myid1:1_9    myid1:1_9.2        test
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths
#&gt; </div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'>## Usage for more complex example</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicFeatures'>GenomicFeatures</a></span><span class='op'>)</span>; <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/tgirke/systemPipeRdata'>systemPipeRdata</a></span><span class='op'>)</span>
<span class='va'>gff</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/annotation"</span>, <span class='st'>"tair10.gff"</span>, package<span class='op'>=</span><span class='st'>"systemPipeRdata"</span><span class='op'>)</span>
<span class='va'>txdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span>file<span class='op'>=</span><span class='va'>gff</span>, format<span class='op'>=</span><span class='st'>"gff3"</span>, organism<span class='op'>=</span><span class='st'>"Arabidopsis"</span><span class='op'>)</span>
<span class='va'>futr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html'>fiveUTRsByTranscript</a></span><span class='op'>(</span><span class='va'>txdb</span>, use.names<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>genome</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/annotation"</span>, <span class='st'>"tair10.fasta"</span>, package<span class='op'>=</span><span class='st'>"systemPipeRdata"</span><span class='op'>)</span>
<span class='va'>dna</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html'>extractTranscriptSeqs</a></span><span class='op'>(</span><span class='fu'>FaFile</span><span class='op'>(</span><span class='va'>genome</span><span class='op'>)</span>, <span class='va'>futr</span><span class='op'>)</span>
<span class='va'>uorf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='predORF.html'>predORF</a></span><span class='op'>(</span><span class='va'>dna</span>, n<span class='op'>=</span><span class='st'>"all"</span>, mode<span class='op'>=</span><span class='st'>"orf"</span>, longest_disjoint<span class='op'>=</span><span class='cn'>TRUE</span>, strand<span class='op'>=</span><span class='st'>"sense"</span><span class='op'>)</span>
<span class='va'>grl_scaled</span> <span class='op'>&lt;-</span> <span class='fu'>scaleRanges</span><span class='op'>(</span>subject<span class='op'>=</span><span class='va'>futr</span>, query<span class='op'>=</span><span class='va'>uorf</span>, type<span class='op'>=</span><span class='st'>"uORF"</span>, verbose<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'>export.gff3</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='va'>grl_scaled</span><span class='op'>)</span>, <span class='st'>"uorf.gff"</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


