<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate Variant Report — variantReport • systemPipeR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate Variant Report — variantReport"><meta property="og:description" content="Functions for generating tabular variant reports including genomic context 
annotations and confidence statistics of variants. The annotations are 
obtained with utilities provided by the VariantAnnotation package and 
the variant statistics are retrieved from the input VCF files."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.31</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tgirke/systemPipeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate Variant Report</h1>
    
    <div class="hidden name"><code>variantReport.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Functions for generating tabular variant reports including genomic context 
annotations and confidence statistics of variants. The annotations are 
obtained with utilities provided by the <code>VariantAnnotation</code> package and 
the variant statistics are retrieved from the input VCF files.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co">## Variant report</span>
<span class="fu">variantReport</span><span class="op">(</span><span class="va">files</span>, <span class="va">txdb</span>, <span class="va">fa</span>, <span class="va">organism</span>, out_dir <span class="op">=</span> <span class="st">"results"</span><span class="op">)</span>

<span class="co">## Combine variant reports</span>
<span class="fu">combineVarReports</span><span class="op">(</span><span class="va">files</span>, <span class="va">filtercol</span>, ncol <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>

<span class="co">## Create summary statistics of variants</span>
<span class="fu">varSummary</span><span class="op">(</span><span class="va">files</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>files</dt>
<dd><p><code>named character vector</code> with paths of the input VCF files.</p></dd>
<dt>txdb</dt>
<dd><p>Annotation data stored as <code>TranscriptDb</code> object, which can be obtained 
	from GFF/GTF files, BioMart, Bioc Annotation packages, UCSC, etc. 
	For details see the vignette of the <code>GenomicFeatures</code> package.
	It is important to use here matching versions of the <code>txdb</code> and 
	<code>fa</code> objects. The latter is the genome sequence used for read mapping 
	and variant calling.</p></dd>
<dt>fa</dt>
<dd><p><code>FaFile</code> object pointing to the sequence file of the corresponding r
	eference genome stored in FASTA format or a <code>BSgenome</code> instance.</p></dd>
<dt>organism</dt>
<dd><p>Character vector specifying the organism name of the reference genome.</p></dd>
<dt>filtercol</dt>
<dd><p>Named character vector containing in the name field the column titles to 
	filter on, and in the data field the corresponding values to include in 
	the report. For instance, the setting <code>filtercol=c(Consequence="nonsynonymous")</code> 
	will include only nonsysynonymous variances listed in the <code>Consequence</code> 
	column. To omit the filtering step, one can use the setting 
	<code>filtercol="All"</code>.</p></dd>
<dt>ncol</dt>
<dd><p>Integer specifying the number of columns in the tabular input files. 
	Default is set to 15.</p></dd>
<dt>out_dir</dt>
<dd><p>Character vector of a <code>results</code> directory name. 
	Default is <code>results</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Tabular output files.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thomas Girke</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>filterVars</code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Alignment with BWA (sequentially on single machine)</span></span>
<span class="r-in"><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bwa.param"</span>, package<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="va">param</span>, mytargets<span class="op">=</span><span class="va">targets</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[1]="./data/SRR446027_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[2]="./data/SRR446028_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[3]="./data/SRR446029_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[4]="./data/SRR446030_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[5]="./data/SRR446031_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[6]="./data/SRR446032_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[7]="./data/SRR446033_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[8]="./data/SRR446034_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[9]="./data/SRR446035_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[10]="./data/SRR446036_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[11]="./data/SRR446037_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[12]="./data/SRR446038_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[13]="./data/SRR446039_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[14]="./data/SRR446040_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[15]="./data/SRR446041_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[16]="./data/SRR446042_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[17]="./data/SRR446043_1.fastq.gz": No such file or directory</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>path[18]="./data/SRR446044_1.fastq.gz": No such file or directory</span>
<span class="r-in"><span class="fu"><a href="sysargs.html">sysargs</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                                                                                                                                                                                                        M1A </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "bwa mem -t 4 -M -R '@RG\\tID:group1\\tSM:sample1\\tPL:illumina\\tLB:lib1\\tPU:unit1' /home/runner/work/systemPipeR/systemPipeR/docs/reference/data/tair10.fasta ./data/SRR446027_1.fastq.gz  &gt; /home/runner/work/systemPipeR/systemPipeR/docs/reference/results/SRR446027_1.fastq.gz.sam" </span>
<span class="r-in"></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"bwa index -a bwtsw ./data/tair10.fasta"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="runCommandline.html">runCommandline</a></span><span class="op">(</span>args<span class="op">=</span><span class="va">args</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Alignment with BWA (parallelized on compute cluster)</span></span>
<span class="r-in"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>walltime<span class="op">=</span><span class="fl">120</span>, ntasks<span class="op">=</span><span class="fl">1</span>, ncpus<span class="op">=</span><span class="fu"><a href="sysargs.html">cores</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, memory<span class="op">=</span><span class="fl">1024</span><span class="op">)</span> </span>
<span class="r-in"><span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>, template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>, Njobs<span class="op">=</span><span class="fl">18</span>, runid<span class="op">=</span><span class="st">"01"</span>, resourceList<span class="op">=</span><span class="va">resources</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Variant calling with GATK</span></span>
<span class="r-in"><span class="co">## The following creates in the inital step a new targets file</span></span>
<span class="r-in"><span class="co">## (targets_bam.txt). The first column of this file gives the paths to</span></span>
<span class="r-in"><span class="co">## the BAM files created in the alignment step. The new targets file and the</span></span>
<span class="r-in"><span class="co">## parameter file gatk.param are used to create a new SYSargs</span></span>
<span class="r-in"><span class="co">## instance for running GATK. Since GATK involves many processing steps, it is</span></span>
<span class="r-in"><span class="co">## executed by a bash script gatk_run.sh where the user can specify the</span></span>
<span class="r-in"><span class="co">## detailed run parameters. All three files are expected to be located in the</span></span>
<span class="r-in"><span class="co">## current working directory. Samples files for gatk.param and</span></span>
<span class="r-in"><span class="co">## gatk_run.sh are available in the subdirectory ./inst/extdata/ of the </span></span>
<span class="r-in"><span class="co">## source file of the systemPipeR package. </span></span>
<span class="r-in"><span class="fu"><a href="writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">args</span>, file<span class="op">=</span><span class="st">"targets_bam.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"java -jar CreateSequenceDictionary.jar R=./data/tair10.fasta O=./data/tair10.dict"</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># system("java -jar /opt/picard/1.81/CreateSequenceDictionary.jar R=./data/tair10.fasta O=./data/tair10.dict")</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"gatk.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_bam.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>walltime<span class="op">=</span><span class="fl">120</span>, ntasks<span class="op">=</span><span class="fl">1</span>, ncpus<span class="op">=</span><span class="fu"><a href="sysargs.html">cores</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, memory<span class="op">=</span><span class="fl">1024</span><span class="op">)</span> </span>
<span class="r-in"><span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>, template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>, Njobs<span class="op">=</span><span class="fl">18</span>, runid<span class="op">=</span><span class="st">"01"</span>, resourceList<span class="op">=</span><span class="va">resources</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">args</span>, file<span class="op">=</span><span class="st">"targets_gatk.txt"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Variant calling with BCFtools</span></span>
<span class="r-in"><span class="co">## The following runs the variant calling with BCFtools. This step requires in</span></span>
<span class="r-in"><span class="co">## the current working directory the parameter file sambcf.param and the </span></span>
<span class="r-in"><span class="co">## bash script sambcf_run.sh.</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"sambcf.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_bam.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>walltime<span class="op">=</span><span class="fl">120</span>, ntasks<span class="op">=</span><span class="fl">1</span>, ncpus<span class="op">=</span><span class="fu"><a href="sysargs.html">cores</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, memory<span class="op">=</span><span class="fl">1024</span><span class="op">)</span> </span>
<span class="r-in"><span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>, template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>, Njobs<span class="op">=</span><span class="fl">18</span>, runid<span class="op">=</span><span class="st">"01"</span>, resourceList<span class="op">=</span><span class="va">resources</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">args</span>, file<span class="op">=</span><span class="st">"targets_sambcf.txt"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Filtering of VCF files generated by GATK</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"filter_gatk.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_gatk.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">filter</span> <span class="op">&lt;-</span> <span class="st">"totalDepth(vr) &gt;= 2 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8) &amp; rowSums(softFilterMatrix(vr))==4"</span></span>
<span class="r-in"><span class="co"># filter &lt;- "totalDepth(vr) &gt;= 20 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8) &amp; rowSums(softFilterMatrix(vr))==6"</span></span>
<span class="r-in"><span class="fu"><a href="filterVars.html">filterVars</a></span><span class="op">(</span><span class="va">args</span>, <span class="va">filter</span>, varcaller<span class="op">=</span><span class="st">"gatk"</span>, organism<span class="op">=</span><span class="st">"A. thaliana"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">args</span>, file<span class="op">=</span><span class="st">"targets_gatk_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Filtering of VCF files generated by BCFtools</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"filter_sambcf.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_sambcf.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">filter</span> <span class="op">&lt;-</span> <span class="st">"rowSums(vr) &gt;= 2 &amp; (rowSums(vr[,3:4])/rowSums(vr[,1:4]) &gt;= 0.8)"</span></span>
<span class="r-in"><span class="co"># filter &lt;- "rowSums(vr) &gt;= 20 &amp; (rowSums(vr[,3:4])/rowSums(vr[,1:4]) &gt;= 0.8)"</span></span>
<span class="r-in"><span class="fu"><a href="filterVars.html">filterVars</a></span><span class="op">(</span><span class="va">args</span>, <span class="va">filter</span>, varcaller<span class="op">=</span><span class="st">"bcftools"</span>, organism<span class="op">=</span><span class="st">"A. thaliana"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">args</span>, file<span class="op">=</span><span class="st">"targets_sambcf_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Annotate filtered variants from GATK</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_gatk_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">FaFile</a></span><span class="op">(</span><span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="sysargs.html">reference</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">variantReport</span><span class="op">(</span>args<span class="op">=</span><span class="va">args</span>, txdb<span class="op">=</span><span class="va">txdb</span>, fa<span class="op">=</span><span class="va">fa</span>, organism<span class="op">=</span><span class="st">"A. thaliana"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Annotate filtered variants from BCFtools</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_sambcf_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">FaFile</a></span><span class="op">(</span><span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="sysargs.html">reference</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">variantReport</span><span class="op">(</span>args<span class="op">=</span><span class="va">args</span>, txdb<span class="op">=</span><span class="va">txdb</span>, fa<span class="op">=</span><span class="va">fa</span>, organism<span class="op">=</span><span class="st">"A. thaliana"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Combine results from GATK</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_gatk_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">combineDF</span> <span class="op">&lt;-</span> <span class="fu">combineVarReports</span><span class="op">(</span><span class="va">args</span>, filtercol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>Consequence<span class="op">=</span><span class="st">"nonsynonymous"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">combineDF</span>, <span class="st">"./results/combineDF_nonsyn_gatk.xls"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Combine results from BCFtools</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_sambcf_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">combineDF</span> <span class="op">&lt;-</span> <span class="fu">combineVarReports</span><span class="op">(</span><span class="va">args</span>, filtercol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>Consequence<span class="op">=</span><span class="st">"nonsynonymous"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">combineDF</span>, <span class="st">"./results/combineDF_nonsyn_sambcf.xls"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Summary for GATK</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_gatk_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu">varSummary</span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, <span class="st">"./results/variantStats_gatk.xls"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Summary for BCFtools</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_sambcf_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu">varSummary</span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, <span class="st">"./results/variantStats_sambcf.xls"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Venn diagram of variants</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_gatk_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">varlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="sysargs.html">outpaths</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="sysargs.html">outpaths</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">VARID</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">vennset_gatk</span> <span class="op">&lt;-</span> <span class="fu"><a href="overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">varlist</span>, type<span class="op">=</span><span class="st">"vennsets"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma<span class="op">=</span><span class="st">"annotate_vars.param"</span>, mytargets<span class="op">=</span><span class="st">"targets_sambcf_filtered.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">varlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="sysargs.html">outpaths</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="sysargs.html">outpaths</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">VARID</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">vennset_bcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">varlist</span>, type<span class="op">=</span><span class="st">"vennsets"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="vennPlot.html">vennPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">vennset_gatk</span>, <span class="va">vennset_bcf</span><span class="op">)</span>, mymain<span class="op">=</span><span class="st">""</span>, mysub<span class="op">=</span><span class="st">"GATK: red; BCFtools: blue"</span>, colmode<span class="op">=</span><span class="fl">2</span>, ccol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

