<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipeR: Workflows collection • systemPipeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipeR: Workflows collection">
<meta property="og:description" content="systemPipeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeR_workflows_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipeR: Workflows collection</h1>
                        <h4 class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 class="date">Last update: 21 February, 2021</h4>
      
      
      <div class="hidden name"><code>systemPipeR_workflows.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeR_workflow.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeR_workflow.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R systemPipeR.knit.md 2; Rscript -e "rmarkdown::render('systemPipeR_workflow.Rmd', c('BiocStyle::pdf_document'))"
--><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Note:</strong> the most recent version of this tutorial can be found <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a>.</p>
<p><strong>Note:</strong> if you use <em><code>systemPipeR</code></em> in published research, please cite: Backman, T.W.H and Girke, T. (2016). <em><code>systemPipeR</code></em>: NGS Workflow and Report Generation Environment. <em>BMC Bioinformatics</em>, 17: 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">10.1186/s12859-016-1241-0</a>.</p>
<div id="workflow-templates" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-templates" class="anchor"></a>Workflow templates</h1>
<p>The intended way of running <em><code>sytemPipeR</code></em> workflows is via <em><code>*.Rmd</code></em> files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line. This way comprehensive and reproducible analysis reports can be generated in PDF or HTML format in a fully automated manner by making use of the highly functional reporting utilities available for R. The following shows how to execute a workflow (<em>e.g.</em>, systemPipeRNAseq.Rmd) from the command-line.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> -e <span class="st">"rmarkdown::render('systemPipeRNAseq.Rmd')"</span></span></code></pre></div>
<p>Templates for setting up custom project reports are provided as <em><code>*.Rmd</code></em> files by the helper package <em><code>systemPipeRdata</code></em> and in the vignettes subdirectory of <em><code>systemPipeR</code></em>. The corresponding HTML of these report templates are available here: <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.html"><em><code>systemPipeRNAseq</code></em></a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.html"><em><code>systemPipeRIBOseq</code></em></a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.html"><em><code>systemPipeChIPseq</code></em></a> and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.html"><em><code>systemPipeVARseq</code></em></a>. To work with <em><code>*.Rmd</code></em> files efficiently, basic knowledge of <a href="http://yihui.name/knitr/"><em><code>knitr</code></em></a> and <a href="http://www.latex-project.org/"><em><code>Latex</code></em></a> or <a href="http://rmarkdown.rstudio.com/"><em><code>R Markdown v2</code></em></a> is required.</p>
<div id="directory-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#directory-structure" class="anchor"></a>Directory Structure</h2>
<p>The working environment of the sample data loaded in the previous step contains the following pre-configured directory structure. Directory names are indicated in <span style="color:grey"><strong><em>green</em></strong></span>. Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>
<ul>
<li>
<span style="color:green"><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>rnaseq/</em>)
<ul>
<li>This is the root directory of the R session running the workflow.</li>
<li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
<li>Note, this directory can have any name (<em>e.g.</em> <span style="color:green"><em><strong>rnaseq</strong></em></span>, <span style="color:green"><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
<li>
<strong>Important subdirectories</strong>:
<ul>
<li>
<span style="color:green"><em><strong>param/</strong></em></span>
<ul>
<li>Stores non-CWL parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*.run.sh</em>. These files are only required for backwards compatibility to run old workflows using the previous custom command-line interface.</li>
<li>
<span style="color:green"><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the CWL parameter files. To organize workflows, each can have its own subdirectory, where all <code>CWL param</code> and <code>input.yml</code> files need to be in the same subdirectory.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>data/</strong></em> </span>
<ul>
<li>FASTQ files</li>
<li>FASTA file of reference (<em>e.g.</em> reference genome)</li>
<li>Annotation files</li>
<li>etc.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>results/</strong></em></span>
<ul>
<li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li>
<li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>The following parameter files are included in each workflow template:</p>
<ol style="list-style-type: decimal">
<li>
<em><code>targets.txt</code></em>: initial one provided by user; downstream <em><code>targets_*.txt</code></em> files are generated automatically</li>
<li>
<em><code>*.param/cwl</code></em>: defines parameter for input/output file operations, <em>e.g.</em>:
<ul>
<li><em><code>hisat2-se/hisat2-mapping-se.cwl</code></em></li>
<li><em><code>hisat2-se/hisat2-mapping-se.yml</code></em></li>
</ul>
</li>
<li>
<em><code>*_run.sh</code></em>: optional bash scripts</li>
<li>Configuration files for computer cluster environments (skip on single machines):
<ul>
<li>
<em><code>.batchtools.conf.R</code></em>: defines the type of scheduler for <em><code>batchtools</code></em> pointing to template file of cluster, and located in user’s home directory</li>
<li>
<em><code>*.tmpl</code></em>: specifies parameters of scheduler used by a system, <em>e.g.</em> Torque, SGE, Slurm, etc.</li>
</ul>
</li>
</ol>
</div>
</div>
<div id="rna-seq-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#rna-seq-workflow" class="anchor"></a>RNA-Seq Workflow</h1>
<p>This workflow demonstrates how to use various utilities for building and running automated end-to-end analysis workflows for <em><code>RNA-Seq</code></em> data.</p>
<p><strong>The full workflow can be found here</strong>: <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.html">HTML</a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.Rmd">.Rmd</a>, and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.R">.R</a>.</p>
<div id="loading-package-and-workflow-template" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-workflow-template" class="anchor"></a>Loading package and workflow template</h2>
<p>Load the <em><code>RNA-Seq</code></em> sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#run-workflow" class="anchor"></a>Run workflow</h2>
<p>Next, run the chosen sample workflow <em><code>systemPipeRNAseq</code></em> (<a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.Rmd">.Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>rnaseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p><strong>Workflow includes following steps:</strong></p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>HISAT2</code></em> (or any other RNA-Seq aligner)</li>
<li>Alignment stats</li>
<li>Read counting</li>
<li>Sample-wise correlation analysis</li>
<li>Analysis of differentially expressed genes (DEGs)</li>
<li>GO term enrichment analysis</li>
<li>Gene-wise clustering</li>
</ol>
</div>
</div>
<div id="chip-seq-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#chip-seq-workflow" class="anchor"></a>ChIP-Seq Workflow</h1>
<p>This workflow demonstrates how to use various utilities for building and running automated end-to-end analysis workflows for <em><code>ChIP-Seq</code></em> data.</p>
<p><strong>The full workflow can be found here</strong>: <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.html">HTML</a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.Rmd">.Rmd</a>, and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.R">.R</a>.</p>
<div id="loading-package-and-workflow-template-1" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-workflow-template-1" class="anchor"></a>Loading package and workflow template</h2>
<p>Load the <em><code>ChIP-Seq</code></em> sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"chipseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"chipseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-workflow-1" class="section level2">
<h2 class="hasAnchor">
<a href="#run-workflow-1" class="anchor"></a>Run workflow</h2>
<p>Next, run the chosen sample workflow <em><code>systemPipeChIPseq</code></em> (<a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.Rmd">.Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>chipseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p><strong>Workflow includes following steps:</strong></p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>Bowtie2</code></em> or <em><code>rsubread</code></em>
</li>
<li>Alignment stats</li>
<li>Peak calling: <em><code>MACS2</code></em>, <em><code>BayesPeak</code></em>
</li>
<li>Peak annotation with genomic context</li>
<li>Differential binding analysis</li>
<li>GO term enrichment analysis</li>
<li>Motif analysis</li>
</ol>
</div>
</div>
<div id="var-seq-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#var-seq-workflow" class="anchor"></a>VAR-Seq Workflow</h1>
<p>This workflow demonstrates how to use various utilities for building and running automated end-to-end analysis workflows for <em><code>VAR-Seq</code></em> data.</p>
<p><strong>The full workflow can be found here:</strong> <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.html">HTML</a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.Rmd">.Rmd</a>, and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.R">.R</a>.</p>
<div id="loading-package-and-workflow-template-2" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-workflow-template-2" class="anchor"></a>Loading package and workflow template</h2>
<p>Load the <em><code>VAR-Seq</code></em> sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"varseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"varseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-workflow-2" class="section level2">
<h2 class="hasAnchor">
<a href="#run-workflow-2" class="anchor"></a>Run workflow</h2>
<p>Next, run the chosen sample workflow <em><code>systemPipeVARseq</code></em> (<a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.Rmd">.Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>varseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p><strong>Workflow includes following steps:</strong></p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>gsnap</code></em>, <em><code>bwa</code></em>
</li>
<li>Variant calling: <em><code>VariantTools</code></em>, <em><code>GATK</code></em>, <em><code>BCFtools</code></em>
</li>
<li>Variant filtering: <em><code>VariantTools</code></em> and <em><code>VariantAnnotation</code></em>
</li>
<li>Variant annotation: <em><code>VariantAnnotation</code></em>
</li>
<li>Combine results from many samples</li>
<li>Summary statistics of samples</li>
</ol>
</div>
</div>
<div id="ribo-seq-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#ribo-seq-workflow" class="anchor"></a>Ribo-Seq Workflow</h1>
<p>This workflow demonstrates how to use various utilities for building and running automated end-to-end analysis workflows for <em><code>RIBO-Seq</code></em> data.</p>
<p><strong>The full workflow can be found here:</strong> <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.html">HTML</a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.Rmd">.Rmd</a>, and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.R">.R</a>.</p>
<div id="loading-package-and-workflow-template-3" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-workflow-template-3" class="anchor"></a>Loading package and workflow template</h2>
<p>Load the <em><code>RIBO-Seq</code></em> sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"riboseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"riboseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-workflow-3" class="section level2">
<h2 class="hasAnchor">
<a href="#run-workflow-3" class="anchor"></a>Run workflow</h2>
<p>Next, run the chosen sample workflow <em><code>systemPipeRIBOseq</code></em> (<a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.Rmd">.Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>ribseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p><strong>Workflow includes following steps:</strong></p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Adaptor trimming and quality filtering</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>HISAT2</code></em> (or any other RNA-Seq aligner)</li>
<li>Alignment stats</li>
<li>Compute read distribution across genomic features</li>
<li>Adding custom features to workflow (e.g. uORFs)</li>
<li>Genomic read coverage along transcripts</li>
<li>Read counting</li>
<li>Sample-wise correlation analysis</li>
<li>Analysis of differentially expressed genes (DEGs)</li>
<li>GO term enrichment analysis</li>
<li>Gene-wise clustering</li>
<li>Differential ribosome binding (translational efficiency)</li>
</ol>
</div>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version information</h1>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-4.0.3/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] systemPipeR_1.25.6          ShortRead_1.48.0           
##  [3] GenomicAlignments_1.26.0    SummarizedExperiment_1.20.0
##  [5] Biobase_2.50.0              MatrixGenerics_1.2.1       
##  [7] matrixStats_0.58.0          BiocParallel_1.24.1        
##  [9] Rsamtools_2.6.0             Biostrings_2.58.0          
## [11] XVector_0.30.0              GenomicRanges_1.42.0       
## [13] GenomeInfoDb_1.26.2         IRanges_2.24.1             
## [15] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [17] BiocStyle_2.18.1           
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-0         rjson_0.2.20             hwriter_1.3.2           
##   [4] ellipsis_0.3.1           rprojroot_2.0.2          fs_1.5.0                
##   [7] bit64_4.0.5              AnnotationDbi_1.52.0     xml2_1.3.2              
##  [10] codetools_0.2-18         splines_4.0.3            cachem_1.0.3            
##  [13] knitr_1.31               jsonlite_1.7.2           annotate_1.68.0         
##  [16] GO.db_3.12.1             dbplyr_2.1.0             png_0.1-7               
##  [19] pheatmap_1.0.12          graph_1.68.0             BiocManager_1.30.10     
##  [22] compiler_4.0.3           httr_1.4.2               backports_1.2.1         
##  [25] GOstats_2.56.0           assertthat_0.2.1         Matrix_1.3-2            
##  [28] fastmap_1.1.0            limma_3.46.0             formatR_1.7             
##  [31] htmltools_0.5.1.1        prettyunits_1.1.1        tools_4.0.3             
##  [34] gtable_0.3.0             glue_1.4.2               GenomeInfoDbData_1.2.4  
##  [37] Category_2.56.0          dplyr_1.0.4              rsvg_2.1                
##  [40] batchtools_0.9.15        rappdirs_0.3.3           V8_3.4.0                
##  [43] Rcpp_1.0.6               pkgdown_1.6.1            vctrs_0.3.6             
##  [46] rtracklayer_1.50.0       xfun_0.21                stringr_1.4.0           
##  [49] lifecycle_1.0.0.9000     XML_3.99-0.5             edgeR_3.32.1            
##  [52] zlibbioc_1.36.0          scales_1.1.1             BSgenome_1.58.0         
##  [55] VariantAnnotation_1.36.0 ragg_0.4.1               hms_1.0.0               
##  [58] RBGL_1.66.0              RColorBrewer_1.1-2       yaml_2.2.1              
##  [61] curl_4.3                 memoise_2.0.0            ggplot2_3.3.3           
##  [64] biomaRt_2.46.3           latticeExtra_0.6-29      stringi_1.5.3           
##  [67] RSQLite_2.2.3            genefilter_1.72.1        desc_1.2.0              
##  [70] checkmate_2.0.0          GenomicFeatures_1.42.1   DOT_0.1                 
##  [73] rlang_0.4.10             pkgconfig_2.0.3          systemfonts_1.0.1       
##  [76] bitops_1.0-6             evaluate_0.14            lattice_0.20-41         
##  [79] purrr_0.3.4              bit_4.0.4                tidyselect_1.1.0        
##  [82] GSEABase_1.52.1          AnnotationForge_1.32.0   magrittr_2.0.1          
##  [85] bookdown_0.21            R6_2.5.0                 generics_0.1.0          
##  [88] base64url_1.4            DelayedArray_0.16.1      DBI_1.1.1               
##  [91] withr_2.4.1              pillar_1.4.7             survival_3.2-7          
##  [94] RCurl_1.98-1.2           tibble_3.0.6             crayon_1.4.1            
##  [97] BiocFileCache_1.14.0     rmarkdown_2.6            jpeg_0.1-8.1            
## [100] progress_1.2.2           locfit_1.5-9.4           grid_4.0.3              
## [103] data.table_1.13.6        blob_1.2.1               Rgraphviz_2.34.0        
## [106] digest_0.6.27            xtable_1.8-4             brew_1.0-6              
## [109] textshaping_0.3.0        openssl_1.4.3            munsell_0.5.0           
## [112] askpass_1.1</code></pre>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This project is funded by NSF award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152">ABI-1661152</a>.</p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
