<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipeR: Workflow design and reporting generation environment • systemPipeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipeR: Workflow design and reporting generation environment">
<meta property="og:description" content="systemPipeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeR_files/header-attrs-2.6/header-attrs.js"></script><script src="systemPipeR_files/kePrint-0.0.1/kePrint.js"></script><link href="systemPipeR_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipeR: Workflow design and reporting generation environment</h1>
                        <h4 class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 class="date">Last update: 21 February, 2021</h4>
      
      
      <div class="hidden name"><code>systemPipeR.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeR.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeR.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R systemPipeR.knit.md 2; Rscript -e "rmarkdown::render('systemPipeR.Rmd', c('BiocStyle::pdf_document'))"
--><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Note:</strong> if you use <em><code>systemPipeR</code></em> in published research, please cite: Backman, T.W.H and Girke, T. (2016). <em><code>systemPipeR</code></em>: NGS Workflow and Report Generation Environment. <em>BMC Bioinformatics</em>, 17: 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">10.1186/s12859-016-1241-0</a>.</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code>systemPipeR</code></em></a> provides flexible utilities for building and running automated end-to-end analysis workflows for a wide range of research applications, including next-generation sequencing (NGS) experiments, such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq <span class="citation">(H Backman and Girke 2016)</span>. Important features include a uniform workflow interface across different data analysis applications, automated report generation, and support for running both R and command-line software, such as NGS aligners or peak/variant callers, on local computers or compute clusters (Figure 1). The latter supports interactive job submissions and batch submissions to queuing systems of clusters. For instance, <em><code>systemPipeR</code></em> can be used with most command-line aligners such as <code>BWA</code> <span class="citation">(Heng Li 2013; H. Li and Durbin 2009)</span>, <code>HISAT2</code> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span>, <code>TopHat2</code> <span class="citation">(Kim et al. 2013)</span> and <code>Bowtie2</code> <span class="citation">(Langmead and Salzberg 2012)</span>, as well as the R-based NGS aligners <a href="http://www.bioconductor.org/packages/devel/bioc/html/Rsubread.html"><em><code>Rsubread</code></em></a> <span class="citation">(Liao, Smyth, and Shi 2013)</span> and <a href="http://www.bioconductor.org/packages/devel/bioc/html/gmapR.html"><em><code>gsnap (gmapR)</code></em></a> <span class="citation">(Wu and Nacu 2010)</span>. Efficient handling of complex sample sets (<em>e.g.</em> FASTQ/BAM files) and experimental designs are facilitated by a well-defined sample annotation infrastructure which improves reproducibility and user-friendliness of many typical analysis workflows in the NGS area <span class="citation">(Lawrence et al. 2013)</span>.</p>
<p>The main motivation and advantages of using <em><code>systemPipeR</code></em> for complex data analysis tasks are:</p>
<ol style="list-style-type: decimal">
<li>Facilitates the design of complex NGS workflows involving multiple R/Bioconductor packages</li>
<li>Common workflow interface for different NGS applications</li>
<li>Makes NGS analysis with Bioconductor utilities more accessible to new users</li>
<li>Simplifies usage of command-line software from within R</li>
<li>Reduces the complexity of using compute clusters for R and command-line software</li>
<li>Accelerates runtime of workflows via parallelization on computer systems with multiple CPU cores and/or multiple compute nodes</li>
<li>Improves reproducibility by automating analyses and generation of analysis reports</li>
</ol>
<center>
<img src="utilities.png">
</center>
<p><strong>Figure 1:</strong> Relevant features in <em><code>systemPipeR</code></em>. Workflow design concepts are illustrated under (A &amp; B). Examples of <em>systemPipeR’s</em> visualization functionalities are given under (C). </p>
<p>A central concept for designing workflows within the <em><code>systemPipeR</code></em> environment is the use of workflow management containers. In previous versions, <em><code>systemPipeR</code></em> used a custom command-line interface called <em><code>SYSargs</code></em> (see Figure 3) and for this purpose will continue to be supported for some time. With the latest <a href="http://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html">Bioconductor Release 3.9</a>, we are adopting for this functionality the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) for describing analysis workflows in a generic and reproducible manner, introducing <em><code>SYSargs2</code></em> workflow control class (see Figure 2). Using this community standard in <em><code>systemPipeR</code></em> has many advantages. For instance, the integration of CWL allows running <em><code>systemPipeR</code></em> workflows from a single specification instance either entirely from within R, from various command-line wrappers (e.g., <em>cwl-runner</em>) or from other languages (<em>, e.g.,</em> Bash or Python). <em><code>systemPipeR</code></em> includes support for both command-line and R/Bioconductor software as well as resources for containerization, parallel evaluations on computer clusters along with the automated generation of interactive analysis reports.</p>
<p>An important feature of <em><code>systemPipeR's</code></em> CWL interface is that it provides two options to run command-line tools and workflows based on CWL. First, one can run CWL in its native way via an R-based wrapper utility for <em>cwl-runner</em> or <em>cwl-tools</em> (CWL-based approach). Second, one can run workflows using CWL’s command-line and workflow instructions from within R (R-based approach). In the latter case the same CWL workflow definition files (<em>e.g.</em> <code>*.cwl</code> and <code>*.yml</code>) are used but rendered and executed entirely with R functions defined by <em><code>systemPipeR</code></em>, and thus use CWL mainly as a command-line and workflow definition format rather than software to run workflows. In this regard <em><code>systemPipeR</code></em> also provides several convenience functions that are useful for designing and debugging workflows, such as a command-line rendering function to retrieve the exact command-line strings for each data set and processing step prior to running a command-line.</p>
<p>This overview introduces the design of a new CWL S4 class in <em><code>systemPipeR</code></em>, as well as the custom command-line interface, combined with the overview of all the common analysis steps of NGS experiments.</p>
<div id="workflow-design-structure-using-sysargs2" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-design-structure-using-sysargs2" class="anchor"></a>Workflow design structure using <em><code>SYSargs2</code></em>
</h2>
<p>The flexibility of <em><code>systemPipeR's</code></em> new interface workflow control class is the driving factor behind the use of as many steps necessary for the analysis, as well as the connection between command-line- or R-based software. The connectivity among all workflow steps is achieved by the <em><code>SYSargs2</code></em> workflow control class (see Figure 3). This S4 class is a list-like container where each instance stores all the input/output paths and parameter components required for a particular data analysis step. <em><code>SYSargs2</code></em> instances are generated by two constructor functions, <em>loadWorkflow</em> and <em>renderWF</em>, using as data input <em>targets</em> or <em>yaml</em> files as well as two <em>cwl</em> parameter files (for details see below). When running preconfigured workflows, the only input the user needs to provide is the initial <em>targets</em> file containing the paths to the input files (<em>e.g.</em> FASTQ) along with unique sample labels. Subsequent targets instances are created automatically. The parameters required for running command-line software is provided by the parameter (<em>.cwl</em>) files described below.</p>
<p>We also introduce the <em><code>SYSargs2Pipe</code></em> class that organizes one or many SYSargs2 containers in a single compound object capturing all information required to run, control and monitor complex workflows from start to finish. This design enhances the <em><code>systemPipeR</code></em> workflow framework with a generalized, flexible, and robust design.</p>
<center>
<img src="SYS_WF.png">
</center>
<p><strong>Figure 2:</strong> Workflow steps with input/output file operations are controlled by <em><code>SYSargs2</code></em> objects. Each <em><code>SYSargs2</code></em> instance is constructed from one <em>targets</em> and two <em>param</em> files. The only input provided by the user is the initial <em>targets</em> file. Subsequent <em>targets</em> instances are created automatically, from the previous output files. Any number of predefined or custom workflow steps are supported. One or many <em><code>SYSargs2</code></em> objects are organized in an <em><code>SYSargs2Pipe</code></em> container.</p>
</div>
<div id="workflow-management-using-sysargslist" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-management-using-sysargslist" class="anchor"></a>Workflow Management using <em><code>SYSargsList</code></em>
</h2>
<p><strong>systemPipeR</strong> allows creation (multi-step analyses) and execution of workflow entirely for R, with control, flexibility, and scalability of the all process. The execution of the workflow can be sent to a HPC, can be parallelizes, accelerating results acquisition. A workflow management system provides an infrastructure for the set-up, performance and monitoring of a defined sequence of tasks, arranged as a workflow application.</p>
<center>
<img src="sysargslist.png">
</center>
<p><strong>Figure 3:</strong> Workflow Management using <em><code>SYSargsList</code></em>.</p>
</div>
<div id="workflow-design-structure-using-sysargs-previous-version" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-design-structure-using-sysargs-previous-version" class="anchor"></a>Workflow design structure using <em><code>SYSargs</code></em>: Previous version</h2>
<p>Instances of this S4 object class are constructed by the <em><code>systemArgs</code></em> function from two simple tabular files: a <em><code>targets</code></em> file and a <em><code>param</code></em> file. The latter is optional for workflow steps lacking command-line software. Typically, a <em><code>SYSargs</code></em> instance stores all sample-level inputs as well as the paths to the corresponding outputs generated by command-line- or R-based software generating sample-level output files, such as read preprocessors (trimmed/filtered FASTQ files), aligners (SAM/BAM files), variant callers (VCF/BCF files) or peak callers (BED/WIG files). Each sample level input/output operation uses its own <em><code>SYSargs</code></em> instance. The outpaths of <em><code>SYSargs</code></em> usually define the sample inputs for the next <em><code>SYSargs</code></em> instance. This connectivity is established by writing the outpaths with the <em><code>writeTargetsout</code></em> function to a new <em><code>targets</code></em> file that serves as input to the next <em><code>systemArgs</code></em> call. Typically, the user has to provide only the initial <em><code>targets</code></em> file. All downstream <em><code>targets</code></em> files are generated automatically. By chaining several <em><code>SYSargs</code></em> steps together one can construct complex workflows involving many sample-level input/output file operations with any combination of command-line or R-based software.</p>
<center>
<img src="SystemPipeR_Workflow.png">
</center>
<p><strong>Figure 4:</strong> Workflow design structure of <em><code>systemPipeR</code></em> using <em><code>SYSargs</code></em>.</p>
</div>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h1>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The R software for running <a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code>systemPipeR</code></em></a> can be downloaded from <a href="http://cran.at.r-project.org/"><em>CRAN</em></a>. The <em><code>systemPipeR</code></em> environment can be installed from the R console using the <a href="https://cran.r-project.org/web/packages/BiocManager/index.html"><em><code>BiocManager::install</code></em></a> command. The associated data package <a href="http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a> can be installed the same way. The latter is a helper package for generating <em><code>systemPipeR</code></em> workflow environments with a single command containing all parameter files and sample data required to quickly test and run workflows.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeRdata"</span><span class="op">)</span></code></pre></div>
<p>Please note that if you desire to use a third-party command line tool, the particular tool and dependencies need to be installed and exported in your PATH. See <a href="#tools">details</a>.</p>
</div>
<div id="loading-package-and-documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-documentation" class="anchor"></a>Loading package and documentation</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">"systemPipeR"</a></span><span class="op">)</span>  <span class="co"># Loads the package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span>help <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>  <span class="co"># Lists package info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/vignette.html">vignette</a></span><span class="op">(</span><span class="st">"systemPipeR"</span><span class="op">)</span>  <span class="co"># Opens vignette</span></code></pre></div>
</div>
<div id="load-sample-data-and-workflow-templates" class="section level2">
<h2 class="hasAnchor">
<a href="#load-sample-data-and-workflow-templates" class="anchor"></a>Load sample data and workflow templates</h2>
<p>The mini sample FASTQ files used by this overview vignette as well as the associated workflow reporting vignettes can be loaded via the <em><code>systemPipeRdata</code></em> package as shown below. The chosen data set <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938"><code>SRP010938</code></a> obtains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em> <span class="citation">(Howard et al. 2013)</span>. To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the <em>A. thalina</em> genome. The corresponding reference genome sequence (FASTA) and its GFF annotation files (provided in the same download) have been truncated accordingly. This way the entire test sample data set requires less than 200MB disk storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single-end) reads or PE reads.</p>
<p>The following generates a fully populated <em><code>systemPipeR</code></em> workflow environment (here for RNA-Seq) in the current working directory of an R session. At this time the package includes workflow templates for RNA-Seq, ChIP-Seq, VAR-Seq, and Ribo-Seq. Templates for additional NGS applications will be provided in the future.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
<p>If you desire run this tutorial with your data set, please follow the instruction here:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"new"</span>, mydirname <span class="op">=</span> <span class="st">"FEB_project"</span><span class="op">)</span></code></pre></div>
<div id="workflow-template-from-an-individuals-package" class="section level3">
<h3 class="hasAnchor">
<a href="#workflow-template-from-an-individuals-package" class="anchor"></a>Workflow template from an individual’s package</h3>
<p>The package provides pre-configured workflows and reporting templates for a wide range of NGS applications that are listed <a href="https://github.com/tgirke/systemPipeR/tree/devel#workflow">here</a>. Additional workflow templates will be provided in the future. If you desire to use an individual package and version, follow the instruction below:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="cn">NULL</span>, package_repo <span class="op">=</span> <span class="st">"systemPipeR/systemPipeRIBOseq"</span>, ref <span class="op">=</span> <span class="st">"master"</span>, 
    subdir <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="cn">NULL</span>, package_repo <span class="op">=</span> <span class="st">"systemPipeR/systemPipeRNAseq"</span>, ref <span class="op">=</span> <span class="st">"singleMachine"</span>, 
    subdir <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="directory-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#directory-structure" class="anchor"></a>Directory Structure</h2>
<p>The working environment of the sample data loaded in the previous step contains the following pre-configured directory structure (Figure 4). Directory names are indicated in <span style="color:grey"><strong><em>green</em></strong></span>. Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>
<ul>
<li>
<span style="color:green"><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>rnaseq/</em>)
<ul>
<li>This is the root directory of the R session running the workflow.</li>
<li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
<li>Note, this directory can have any name (<em>e.g.</em> <span style="color:green"><em><strong>rnaseq</strong></em></span>, <span style="color:green"><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
<li>
<strong>Important subdirectories</strong>:
<ul>
<li>
<span style="color:green"><em><strong>param/</strong></em></span>
<ul>
<li>Stores non-CWL parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*.run.sh</em>. These files are only required for backwards compatibility to run old workflows using the previous custom command-line interface.</li>
<li>
<span style="color:green"><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the CWL parameter files. To organize workflows, each can have its own subdirectory, where all <code>CWL param</code> and <code>input.yml</code> files need to be in the same subdirectory.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>data/</strong></em> </span>
<ul>
<li>FASTQ files</li>
<li>FASTA file of reference (<em>e.g.</em> reference genome)</li>
<li>Annotation files</li>
<li>etc.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>results/</strong></em></span>
<ul>
<li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li>
<li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<center>
<img src="SYSdir.png">
</center>
<p><strong>Figure 5:</strong> <em>systemPipeR’s</em> preconfigured directory structure.</p>
<p>The following parameter files are included in each workflow template:</p>
<ol style="list-style-type: decimal">
<li>
<em><code>targets.txt</code></em>: initial one provided by user; downstream <em><code>targets_*.txt</code></em> files are generated automatically</li>
<li>
<em><code>*.param/cwl</code></em>: defines parameter for input/output file operations, <em>e.g.</em>:
<ul>
<li><em><code>hisat2-se/hisat2-mapping-se.cwl</code></em></li>
<li><em><code>hisat2-se/hisat2-mapping-se.yml</code></em></li>
</ul>
</li>
<li>
<em><code>*_run.sh</code></em>: optional bash scripts</li>
<li>Configuration files for computer cluster environments (skip on single machines):
<ul>
<li>
<em><code>.batchtools.conf.R</code></em>: defines the type of scheduler for <em><code>batchtools</code></em> pointing to template file of cluster, and located in user’s home directory</li>
<li>
<em><code>*.tmpl</code></em>: specifies parameters of scheduler used by a system, <em>e.g.</em> Torque, SGE, Slurm, etc.</li>
</ul>
</li>
</ol>
</div>
<div id="structure-of-targets-file" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-targets-file" class="anchor"></a>Structure of <em><code>targets</code></em> file</h2>
<p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code>targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code>systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single-end (SE) reads, the first three columns are mandatory including their column names, while it is four mandatory columns for FASTQ files of PE reads. All subsequent columns are optional and any number of additional columns can be added as needed.</p>
<p>Users should note here, the usage of targets files is optional when using <em>systemPipeR’s</em> new CWL interface. They can be replaced by a standard YAML input file used by CWL. Since for organizing experimental variables targets files are extremely useful and user-friendly. Thus, we encourage users to keep using them.</p>
<div id="structure-of-targets-file-for-single-end-se-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#structure-of-targets-file-for-single-end-se-samples" class="anchor"></a>Structure of <em><code>targets</code></em> file for single-end (SE) samples</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
<span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span></code></pre></div>
<pre><code>##                      FileName SampleName Factor SampleLong Experiment
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A          1
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B          1
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A          1
## 4 ./data/SRR446030_1.fastq.gz        A1B     A1   Avr.1h.B          1
##          Date
## 1 23-Mar-2012
## 2 23-Mar-2012
## 3 23-Mar-2012
## 4 23-Mar-2012</code></pre>
<p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the path to the corresponding <em><code>targets</code></em> file.</p>
</div>
<div id="structure-of-targets-file-for-paired-end-pe-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#structure-of-targets-file-for-paired-end-pe-samples" class="anchor"></a>Structure of <em><code>targets</code></em> file for paired-end (PE) samples</h3>
<p>For paired-end (PE) samples, the structure of the targets file is similar, where users need to provide two FASTQ path columns: <em><code>FileName1</code></em> and <em><code>FileName2</code></em> with the paths to the PE FASTQ files.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></code></pre></div>
<pre><code>##                     FileName1                   FileName2 SampleName Factor
## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz        M1A     M1
## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz        M1B     M1
##   SampleLong Experiment
## 1  Mock.1h.A          1
## 2  Mock.1h.B          1</code></pre>
</div>
<div id="sample-comparisons" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-comparisons" class="anchor"></a>Sample comparisons</h3>
<p>Sample comparisons are defined in the header lines of the <em><code>targets</code></em> file starting with ‘<code># &lt;CMP&gt;</code>.’</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="va">targetspath</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "# Project ID: Arabidopsis - Pseudomonas alternative splicing study (SRA: SRP010938; PMID: 24098335)"                                                                              
## [2] "# The following line(s) allow to specify the contrasts needed for comparative analyses, such as DEG identification. All possible comparisons can be specified with 'CMPset: ALL'."
## [3] "# &lt;CMP&gt; CMPset1: M1-A1, M1-V1, A1-V1, M6-A6, M6-V6, A6-V6, M12-A12, M12-V12, A12-V12"                                                                                             
## [4] "# &lt;CMP&gt; CMPset2: ALL"</code></pre>
<p>The function <em><code>readComp</code></em> imports the comparison information and stores it in a <em><code>list</code></em>. Alternatively, <em><code>readComp</code></em> can obtain the comparison information from the corresponding <em><code>SYSargs</code></em> object (see below). Note, these header lines are optional. They are mainly useful for controlling comparative analyses according to certain biological expectations, such as identifying differentially expressed genes in RNA-Seq experiments based on simple pair-wise comparisons.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/readComp.html">readComp</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">targetspath</span>, format <span class="op">=</span> <span class="st">"vector"</span>, delim <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></code></pre></div>
<pre><code>## $CMPset1
## [1] "M1-A1"   "M1-V1"   "A1-V1"   "M6-A6"   "M6-V6"   "A6-V6"   "M12-A12"
## [8] "M12-V12" "A12-V12"
## 
## $CMPset2
##  [1] "M1-A1"   "M1-V1"   "M1-M6"   "M1-A6"   "M1-V6"   "M1-M12"  "M1-A12" 
##  [8] "M1-V12"  "A1-V1"   "A1-M6"   "A1-A6"   "A1-V6"   "A1-M12"  "A1-A12" 
## [15] "A1-V12"  "V1-M6"   "V1-A6"   "V1-V6"   "V1-M12"  "V1-A12"  "V1-V12" 
## [22] "M6-A6"   "M6-V6"   "M6-M12"  "M6-A12"  "M6-V12"  "A6-V6"   "A6-M12" 
## [29] "A6-A12"  "A6-V12"  "V6-M12"  "V6-A12"  "V6-V12"  "M12-A12" "M12-V12"
## [36] "A12-V12"</code></pre>
</div>
</div>
<div id="structure-of-the-new-param-files-and-construct-sysargs2-container" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-the-new-param-files-and-construct-sysargs2-container" class="anchor"></a>Structure of the new <em><code>param</code></em> files and construct <em><code>SYSargs2</code></em> container</h2>
<p><em><code>SYSargs2</code></em> stores all the information and instructions needed for processing a set of input files with a single or many command-line steps within a workflow (<em>i.e.</em> several components of the software or several independent software tools). The <em><code>SYSargs2</code></em> object is created and fully populated with the <em>loadWF</em> and <em>renderWF</em> functions, respectively.</p>
<p>In CWL, files with the extension <em><code>.cwl</code></em> define the parameters of a chosen command-line step or workflow, while files with the extension <em><code>.yml</code></em> define the input variables of command-line steps. Note, input variables provided by a <em>targets</em> file can be passed on to a <em><code>SYSargs2</code></em> instance via the <em>inputvars</em> argument of the <em>renderWF</em> function.</p>
<p>The following imports a <em><code>.cwl</code></em> file (here <em><code>hisat2-mapping-se.cwl</code></em>) for running the short read aligner HISAT2 <span class="citation">(Kim, Langmead, and Salzberg 2015)</span>. The <em>loadWorkflow</em> and <em>renderWF</em> functions render the proper command-line strings for each sample and software tool.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/hisat2/hisat2-se"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">WF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.cwl"</span>, input_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>

<span class="va">WF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">WF</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Several accessor methods are available that are named after the slot names of the <em><code>SYSargs2</code></em> object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "targets"       "targetsheader" "modules"       "wf"           
##  [5] "clt"           "yamlinput"     "cmdlist"       "input"        
##  [9] "output"        "cwlfiles"      "inputvars"</code></pre>
<p>Of particular interest is the <em><code><a href="../reference/sysargs2.html">cmdlist()</a></code></em> method. It constructs the system commands for running command-line software as specified by a given <em><code>.cwl</code></em> file combined with the paths to the input samples (<em>e.g.</em> FASTQ files) provided by a <em><code>targets</code></em> file. The example below shows the <em><code><a href="../reference/sysargs2.html">cmdlist()</a></code></em> output for running HISAT2 on the first SE read sample. Evaluating the output of <em><code><a href="../reference/sysargs2.html">cmdlist()</a></code></em> can be very helpful for designing and debugging <em><code>.cwl</code></em> files of new command-line software or changing the parameter settings of existing ones.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$`hisat2-mapping-se`
## [1] "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"</code></pre>
<p>The output components of <em><code>SYSargs2</code></em> define the expected output files for each step in the workflow; some of which are the input for the next workflow step, here next <em><code>SYSargs2</code></em> instance (see Figure 2).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$`hisat2-mapping-se`
## [1] "./results/M1A.sam"</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs.html">modules</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span></code></pre></div>
<pre><code>##        module1 
## "hisat2/2.1.0"</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">targets</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$FileName
## [1] "./data/SRR446027_1.fastq.gz"
## 
## $M1A$SampleName
## [1] "M1A"
## 
## $M1A$Factor
## [1] "M1"
## 
## $M1A$SampleLong
## [1] "Mock.1h.A"
## 
## $M1A$Experiment
## [1] 1
## 
## $M1A$Date
## [1] "23-Mar-2012"</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">targets</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></code></pre></div>
<pre><code>##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
## 4 ./data/SRR446030_1.fastq.gz        A1B     A1   Avr.1h.B</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$`hisat2-mapping-se`
## [1] "./results/M1A.sam"</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">cwlfiles</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span></code></pre></div>
<pre><code>## $cwl
## [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e3ecaa520/systemPipeR/extdata/cwl/hisat2/hisat2-se/hisat2-mapping-se.cwl"
## 
## $yml
## [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e3ecaa520/systemPipeR/extdata/cwl/hisat2/hisat2-se/hisat2-mapping-se.yml"
## 
## $steps
## [1] "hisat2-mapping-se"
## 
## $targets
## [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e3ecaa520/systemPipeR/extdata/targets.txt"</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">inputvars</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span></code></pre></div>
<pre><code>## $FileName
## [1] "_FASTQ_PATH1_"
## 
## $SampleName
## [1] "_SampleName_"</code></pre>
<p>In an ‘R-centric’ rather than a ‘CWL-centric’ workflow design the connectivity among workflow steps is established by writing all relevant output with the <em>writeTargetsout</em> function to a new targets file that serves as input to the next <em>loadWorkflow</em> and <em>renderWF</em> call. By chaining several <em><code>SYSargs2</code></em> steps together one can construct complex workflows involving many sample-level input/output file operations with any combination of command-line or R-based software. Alternatively, a CWL-centric workflow design can be used that defines all/most workflow steps with CWL workflow and parameter files. Due to time and space restrictions, the CWL-centric approach is not covered by this tutorial.</p>
<div id="tools" class="section level3">
<h3 class="hasAnchor">
<a href="#tools" class="anchor"></a>Third-party software tools</h3>
<p>Current, <em>systemPipeR</em> provides the <em><code>param</code></em> file templates for third-party software tools. Please check the listed software tools.</p>
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:500px; overflow-x: scroll; width:80%; ">
<table class="table table table-striped table-hover table-condensed">
<thead><tr>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Tool Name
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Step
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
<a href="http://bio-bwa.sourceforge.net/bwa.shtml">bwa</a>
</td>
<td style="text-align:center;">
BWA is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. 
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie2</a>
</td>
<td style="text-align:center;">
Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://hannonlab.cshl.edu/fastx_toolkit/commandline.html">FASTX-Toolkit</a>
</td>
<td style="text-align:center;">
FASTX-Toolkit is a collection of command line tools for Short-Reads FASTA/FASTQ files preprocessing.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #EC7770 !important;">Read Preprocessing</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://hibberdlab.com/transrate/">TransRate</a>
</td>
<td style="text-align:center;">
Transrate is software for de-novo transcriptome assembly quality analysis.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #D98576 !important;">Quality</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://research-pub.gene.com/gmap/">Gsnap</a>
</td>
<td style="text-align:center;">
GSNAP is a genomic short-read nucleotide alignment program.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://www.htslib.org/doc/samtools-1.2.html">Samtools</a>
</td>
<td style="text-align:center;">
Samtools is a suite of programs for interacting with high-throughput sequencing data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #D08C79 !important;">Post-processing</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a>
</td>
<td style="text-align:center;">
Trimmomatic is a flexible read trimming tool for Illumina NGS data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #EC7770 !important;">Read Preprocessing</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://bioconductor.org/packages/release/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf">Rsubread</a>
</td>
<td style="text-align:center;">
Rsubread is a Bioconductor software package that provides high-performance alignment and read counting functions for RNA-seq reads.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://broadinstitute.github.io/picard/">Picard</a>
</td>
<td style="text-align:center;">
Picard is a set of command line tools for manipulating high-throughput sequencing (HTS) data and formats such as SAM/BAM/CRAM and VCF.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #B4A082 !important;">Manipulating HTS data</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://busco.ezlab.org/">Busco</a>
</td>
<td style="text-align:center;">
BUSCO assesses genome assembly and annotation completeness with Benchmarking Universal Single-Copy Orthologs.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #D98576 !important;">Quality</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://ccb.jhu.edu/software/hisat2/manual.shtml">Hisat2</a>
</td>
<td style="text-align:center;">
HISAT2 is a fast and sensitive alignment program for mapping NGS reads (both DNA and RNA) to reference genomes.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://ccb.jhu.edu/software/tophat/manual.shtml">Tophat2</a>
</td>
<td style="text-align:center;">
TopHat is a fast splice junction mapper for RNA-Seq reads.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://gatk.broadinstitute.org/hc/en-us">GATK</a>
</td>
<td style="text-align:center;">
Variant Discovery in High-Throughput Sequencing Data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #FF6A6A !important;">Variant Discovery</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://github.com/alexdobin/STAR">STAR</a>
</td>
<td style="text-align:center;">
STAR is an ultrafast universal RNA-seq aligner.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8FBC8F !important;">Alignment</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://github.com/FelixKrueger/TrimGalore">Trim_galore</a>
</td>
<td style="text-align:center;">
Trim Galore is a wrapper around Cutadapt and FastQC to consistently apply adapter and quality trimming to FastQ files.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #EC7770 !important;">Read Preprocessing</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://github.com/TransDecoder/TransDecoder/wiki">TransDecoder</a>
</td>
<td style="text-align:center;">
TransDecoder identifies candidate coding regions within transcript sequences.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #ABA785 !important;">Find Coding Regions</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity</a>
</td>
<td style="text-align:center;">
Trinity assembles transcript sequences from Illumina RNA-Seq data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #A1AE88 !important;">denovo Transcriptome Assembly</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://github.com/Trinotate/Trinotate.github.io/wiki">Trinotate</a>
</td>
<td style="text-align:center;">
Trinotate is a comprehensive annotation suite designed for automatic functional annotation of transcriptomes.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #F5706D !important;">Transcriptome Functional Annotation</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://macs3-project.github.io/MACS/">MACS2</a>
</td>
<td style="text-align:center;">
MACS2 identifies transcription factor binding sites in ChIP-seq data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #C7937C !important;">Peak calling</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://pachterlab.github.io/kallisto/manual">Kallisto</a>
</td>
<td style="text-align:center;">
kallisto is a program for quantifying abundances of transcripts from RNA-Seq data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #E37E73 !important;">Read counting</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://samtools.github.io/bcftools/howtos/index.html">BCFtools</a>
</td>
<td style="text-align:center;">
BCFtools is a program for variant calling and manipulating files in the Variant Call Format (VCF) and its binary counterpart BCF.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #FF6A6A !important;">Variant Discovery</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a>
</td>
<td style="text-align:center;">
Bismark is a program to map bisulfite treated sequencing reads to a genome of interest and perform methylation calls in a single step.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #98B58B !important;">Bisulfite mapping</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">Fastqc</a>
</td>
<td style="text-align:center;">
FastQC is a quality control tool for high throughput sequence data.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #D98576 !important;">Quality</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<a href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">Blast</a>
</td>
<td style="text-align:center;">
BLAST finds regions of similarity between biological sequences.
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #BD997F !important;">Blast</span>
</td>
</tr>
</tbody>
</table>
</div>
<p>Remember, if you desire to run any of these tools, make sure to have the respective software installed on your system and configure in the PATH. You can check as follows:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/tryCL.html">tryCL</a></span><span class="op">(</span>command <span class="op">=</span> <span class="st">"grep"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="structure-of-param-file-and-sysargs-container-previous-version" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-param-file-and-sysargs-container-previous-version" class="anchor"></a>Structure of <em><code>param</code></em> file and <em><code>SYSargs</code></em> container (Previous version)</h2>
<p>The <em><code>param</code></em> file defines the parameters of a chosen command-line software. The following shows the format of a sample <em><code>param</code></em> file provided by this package.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">parampath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tophat.param"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">parampath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></code></pre></div>
<pre><code>##      PairSet         Name                                  Value
## 1    modules         &lt;NA&gt;                          bowtie2/2.2.5
## 2    modules         &lt;NA&gt;                          tophat/2.0.14
## 3   software         &lt;NA&gt;                                 tophat
## 4      cores           -p                                      4
## 5      other         &lt;NA&gt; -g 1 --segment-length 25 -i 30 -I 3000
## 6   outfile1           -o                            &lt;FileName1&gt;
## 7   outfile1         path                             ./results/
## 8   outfile1       remove                                   &lt;NA&gt;
## 9   outfile1       append                                .tophat
## 10  outfile1 outextension              .tophat/accepted_hits.bam
## 11 reference         &lt;NA&gt;                    ./data/tair10.fasta
## 12   infile1         &lt;NA&gt;                            &lt;FileName1&gt;
## 13   infile1         path                                   &lt;NA&gt;
## 14   infile2         &lt;NA&gt;                            &lt;FileName2&gt;
## 15   infile2         path                                   &lt;NA&gt;</code></pre>
<p>The <em><code>systemArgs</code></em> function imports the definitions of both the <em><code>param</code></em> file and the <em><code>targets</code></em> file, and stores all relevant information in a <em><code>SYSargs</code></em> object (S4 class). To run the pipeline without command-line software, one can assign <em><code>NULL</code></em> to <em><code>sysma</code></em> instead of a <em><code>param</code></em> file. In addition, one can start <em><code>systemPipeR</code></em> workflows with pre-generated BAM files by providing a targets file where the <em><code>FileName</code></em> column provides the paths to the BAM files. Note, in the following example the usage of <em><code><a href="https://rdrr.io/r/base/warning.html">suppressWarnings()</a></code></em> is only relevant for building this vignette. In typical workflows it should be removed.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma <span class="op">=</span> <span class="va">parampath</span>, mytargets <span class="op">=</span> <span class="va">targetspath</span><span class="op">)</span><span class="op">)</span>
<span class="va">args</span></code></pre></div>
<pre><code>## An instance of 'SYSargs' for running 'tophat' on 18 samples</code></pre>
<p>Several accessor methods are available that are named after the slot names of the <em><code>SYSargs</code></em> object.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "targetsin"     "targetsout"    "targetsheader" "modules"      
##  [5] "software"      "cores"         "other"         "reference"    
##  [9] "results"       "infile1"       "infile2"       "outfile1"     
## [13] "sysargs"       "outpaths"</code></pre>
<p>Of particular interest is the <em><code><a href="../reference/sysargs.html">sysargs()</a></code></em> method. It constructs the system commands for running command-lined software as specified by a given <em><code>param</code></em> file combined with the paths to the input samples (<em>e.g.</em> FASTQ files) provided by a <em><code>targets</code></em> file. The example below shows the <em><code><a href="../reference/sysargs.html">sysargs()</a></code></em> output for running TopHat2 on the first PE read sample. Evaluating the output of <em><code><a href="../reference/sysargs.html">sysargs()</a></code></em> can be very helpful for designing and debugging <em><code>param</code></em> files of new command-line software or changing the parameter settings of existing ones.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs.html">sysargs</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>##                                                                                                                                                                                                                                                                                          M1A 
## "tophat -p 4 -g 1 --segment-length 25 -i 30 -I 3000 -o /home/dcassol/danielac@ucr.edu/projects/SPR_WF_org/systemPipeR/vignettes/results/SRR446027_1.fastq.gz.tophat /home/dcassol/danielac@ucr.edu/projects/SPR_WF_org/systemPipeR/vignettes/data/tair10.fasta ./data/SRR446027_1.fastq.gz "</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs.html">modules</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "bowtie2/2.2.5" "tophat/2.0.14"</code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs.html">cores</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs.html">outpaths</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>##                                                                                                                              M1A 
## "/home/dcassol/danielac@ucr.edu/projects/SPR_WF_org/systemPipeR/vignettes/results/SRR446027_1.fastq.gz.tophat/accepted_hits.bam"</code></pre>
<p>The content of the <em><code>param</code></em> file can also be returned as JSON object as follows (requires <em><code>rjson</code></em> package).</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma <span class="op">=</span> <span class="va">parampath</span>, mytargets <span class="op">=</span> <span class="va">targetspath</span>, type <span class="op">=</span> <span class="st">"json"</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "{\"modules\":{\"n1\":\"\",\"v2\":\"bowtie2/2.2.5\",\"n1\":\"\",\"v2\":\"tophat/2.0.14\"},\"software\":{\"n1\":\"\",\"v1\":\"tophat\"},\"cores\":{\"n1\":\"-p\",\"v1\":\"4\"},\"other\":{\"n1\":\"\",\"v1\":\"-g 1 --segment-length 25 -i 30 -I 3000\"},\"outfile1\":{\"n1\":\"-o\",\"v2\":\"&lt;FileName1&gt;\",\"n3\":\"path\",\"v4\":\"./results/\",\"n5\":\"remove\",\"v1\":\"\",\"n2\":\"append\",\"v3\":\".tophat\",\"n4\":\"outextension\",\"v5\":\".tophat/accepted_hits.bam\"},\"reference\":{\"n1\":\"\",\"v1\":\"./data/tair10.fasta\"},\"infile1\":{\"n1\":\"\",\"v2\":\"&lt;FileName1&gt;\",\"n1\":\"path\",\"v2\":\"\"},\"infile2\":{\"n1\":\"\",\"v2\":\"&lt;FileName2&gt;\",\"n1\":\"path\",\"v2\":\"\"}}"</code></pre>
</div>
</div>
<div id="how-to-run-a-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-run-a-workflow" class="anchor"></a>How to run a Workflow</h1>
<p>This tutorial introduces the basic ideas and tools needed to build a specific workflow from preconfigured templates.</p>
<div id="load-sample-data-and-workflow-templates-1" class="section level2">
<h2 class="hasAnchor">
<a href="#load-sample-data-and-workflow-templates-1" class="anchor"></a>Load sample data and workflow templates</h2>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="setup-and-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#setup-and-requirements" class="anchor"></a>Setup and Requirements</h2>
<p>To go through this tutorial, you need the following software installed:</p>
<ul>
<li>R/&gt;=3.6.2</li>
<li>systemPipeR R package (version 1.22)</li>
<li>Hisat2/2.1.0</li>
</ul>
<p>If you desire to build your pipeline with any different software, make sure to have the respective software installed and configured in your PATH. To make sure if the configuration is right, you always can test as follow:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/tryCL.html">tryCL</a></span><span class="op">(</span>command <span class="op">=</span> <span class="st">"hisat2"</span><span class="op">)</span>  <span class="co">## 'All set up, proceed!'</span></code></pre></div>
</div>
<div id="project-initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#project-initialization" class="anchor"></a>Project Initialization</h2>
<p>The Project management structure is essential, especially for reproducibility and efficiency in the analysis. Here we show how to construct an instance of this S4 object class by the <em><code>initWF</code></em> function. The object of class <em><code>SYSarsgsList</code></em> storing all the configuration information for the project and allows management and control at a high level.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">script</span> <span class="op">&lt;-</span> <span class="st">"systemPipeRNAseq.Rmd"</span>
<span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="st">"targets.txt"</span>
<span class="va">sysargslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initWF.html">initWF</a></span><span class="op">(</span>script <span class="op">=</span> <span class="va">script</span>, targets <span class="op">=</span> <span class="va">targetspath</span><span class="op">)</span></code></pre></div>
</div>
<div id="project-initialization-in-a-temporary-directory" class="section level2">
<h2 class="hasAnchor">
<a href="#project-initialization-in-a-temporary-directory" class="anchor"></a>Project Initialization in a Temporary Directory</h2>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="va">script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/workflows/rnaseq"</span>, <span class="st">"systemPipeRNAseq.Rmd"</span>, package <span class="op">=</span> <span class="st">"systemPipeRdata"</span><span class="op">)</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">SYSconfig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initWF.html">initProject</a></span><span class="op">(</span>projPath <span class="op">=</span> <span class="va">dir_path</span>, targets <span class="op">=</span> <span class="va">targets</span>, script <span class="op">=</span> <span class="va">script</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sysargslist_temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initWF.html">initWF</a></span><span class="op">(</span>sysconfig <span class="op">=</span> <span class="st">"SYSconfig.yml"</span><span class="op">)</span></code></pre></div>
</div>
<div id="configuration-and-run-of-the-project" class="section level2">
<h2 class="hasAnchor">
<a href="#configuration-and-run-of-the-project" class="anchor"></a>Configuration and run of the project</h2>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sysargslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/configWF.html">configWF</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sysargslist</span>, input_steps <span class="op">=</span> <span class="st">"1:3"</span><span class="op">)</span>
<span class="va">sysargslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runWF.html">runWF</a></span><span class="op">(</span>sysargslist <span class="op">=</span> <span class="va">sysargslist</span>, steps <span class="op">=</span> <span class="st">"ALL"</span><span class="op">)</span>
<span class="va">sysargslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runWF.html">runWF</a></span><span class="op">(</span>sysargslist <span class="op">=</span> <span class="va">sysargslist</span>, steps <span class="op">=</span> <span class="st">"1:2"</span><span class="op">)</span></code></pre></div>
</div>
<div id="how-to-use-pipes-with-systempiper" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-use-pipes-with-systempiper" class="anchor"></a>How to Use Pipes with <em>systemPipeR</em>
</h2>
<p>At first encounter, you may wonder whether an operator such as <em>%&gt;%</em> can really be all that beneficial; but as you may notice, it semantically changes your code in a way that makes it more intuitive to both read and write.</p>
<p>Consider the following example, in which the steps are the initialization, configuration and running the entire workflow.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
<span class="va">sysargslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initWF.html">initWF</a></span><span class="op">(</span>script <span class="op">=</span> <span class="st">"systemPipeRNAseq.Rmd"</span>, overwrite <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/configWF.html">configWF</a></span><span class="op">(</span>input_steps <span class="op">=</span> <span class="st">"1:3"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="../reference/runWF.html">runWF</a></span><span class="op">(</span>steps <span class="op">=</span> <span class="st">"1:2"</span><span class="op">)</span></code></pre></div>
</div>
<div id="how-to-run-the-workflow-on-a-cluster" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-run-the-workflow-on-a-cluster" class="anchor"></a>How to run the workflow on a cluster</h2>
<p>This section of the tutorial provides an introduction to the usage of the <em>systemPipeR</em> features on a cluster.</p>
<p>Now open the R markdown script <code>*.Rmd</code>in your R IDE (_e.g._vim-r or RStudio) and run the workflow as outlined below. If you work under Vim-R-Tmux, the following command sequence will connect the user in an interactive session with a node on the cluster. The code of the <code>Rmd</code> script can then be sent from Vim on the login (head) node to an open R session running on the corresponding computer node. This is important since Tmux sessions should not be run on the computer nodes.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/quit.html">q</a></span><span class="op">(</span><span class="st">"no"</span><span class="op">)</span>  <span class="co"># closes R session on head node</span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> --x11 --partition=short --mem=2gb --cpus-per-task 4 --ntasks 1 --time 2:00:00 --pty bash -l</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load R/3.4.2</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Now check whether your R session is running on a computer node of the cluster and not on a head node.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"hostname"</span><span class="op">)</span>  <span class="co"># should return name of a compute node starting with i or c </span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># checks current working directory of R session</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># returns content of current working directory</span></code></pre></div>
<div id="parallelization-on-clusters" class="section level3">
<h3 class="hasAnchor">
<a href="#parallelization-on-clusters" class="anchor"></a>Parallelization on clusters</h3>
<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. For this the <em><code>clusterRun</code></em> function submits the computing requests to the scheduler using the run specifications defined by <em><code>runCommandline</code></em>.</p>
<p>To avoid over-subscription of CPU cores on the compute nodes, the value from <em><code>yamlinput(args)['thread']</code></em> is passed on to the submission command, here <em><code>ncpus</code></em> in the <em><code>resources</code></em> list object. The number of independent parallel cluster processes is defined under the <em><code>Njobs</code></em> argument. The following example will run 18 processes in parallel using for each 4 CPU cores. If the resources available on a cluster allow running all 18 processes at the same time then the shown sample submission will utilize in total 72 CPU cores. Note, <em><code>clusterRun</code></em> can be used with most queueing systems as it is based on utilities from the <em><code>batchtools</code></em> package which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a conf file (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>) and a template file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>) for the queueing available on a system. The following example uses the sample conf and template files for the Slurm scheduler provided by this package.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span>, 
    dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/waitForJobs.html">waitForJobs</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
<div id="workflow-steps-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-steps-overview" class="anchor"></a>Workflow steps overview</h1>
<div id="define-environment-settings-and-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#define-environment-settings-and-samples" class="anchor"></a>Define environment settings and samples</h2>
<p>A typical workflow starts with generating the expected working environment containing the proper directory structure, input files, and parameter settings. To simplify this task, one can load one of the existing NGS workflows templates provided by <em><code>systemPipeRdata</code></em> into the current working directory. The following does this for the <em><code>rnaseq</code></em> template. The name of the resulting workflow directory can be specified under the <em><code>mydirname</code></em> argument. The default <em><code>NULL</code></em> uses the name of the chosen workflow. An error is issued if a directory of the same name and path exists already. On Linux and OS X systems one can also create new workflow instances from the command-line of a terminal as shown <a href="http://bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#generate-workflow-template">here</a>. To apply workflows to custom data, the user needs to modify the <em><code>targets</code></em> file and if necessary update the corresponding <em><code>.cwl</code></em> and <em><code>.yml</code></em> files. A collection of pre-generated <em><code>.cwl</code></em> and <em><code>.yml</code></em> files are provided in the <em><code>param/cwl</code></em> subdirectory of each workflow template. They are also viewable in the GitHub repository of <em><code>systemPipeRdata</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/tree/master/inst/extdata/param">see here</a>).</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span>, mydirname <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#read-preprocessing" class="anchor"></a>Read Preprocessing</h2>
<div id="preprocessing-with-preprocessreads-function" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-preprocessreads-function" class="anchor"></a>Preprocessing with <em><code>preprocessReads</code></em> function</h3>
<p>The function <em><code>preprocessReads</code></em> allows to apply predefined or custom read preprocessing functions to all FASTQ files referenced in a <em><code>SYSargs2</code></em> container, such as quality filtering or adaptor trimming routines. The paths to the resulting output FASTQ files are stored in the <em><code>output</code></em> slot of the <em><code>SYSargs2</code></em> object. Internally, <em><code>preprocessReads</code></em> uses the <em><code>FastqStreamer</code></em> function from the <em><code>ShortRead</code></em> package to stream through large FASTQ files in a memory-efficient manner. The following example performs adaptor trimming with the <em><code>trimLRPatterns</code></em> function from the <em><code>Biostrings</code></em> package. After the trimming step a new targets file is generated (here <em><code>targets_trimPE.txt</code></em>) containing the paths to the trimmed FASTQ files. The new targets file can be used for the next workflow step with an updated <em><code>SYSargs2</code></em> instance, <em>e.g.</em> running the NGS alignments with the trimmed FASTQ files.</p>
<p>Construct <em><code>SYSargs2</code></em> object from <em><code>cwl</code></em> and <em><code>yml</code></em> param and <em><code>targets</code></em> files.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/preprocessReads/trim-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trim-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trim-pe.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trim</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trim</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">trim</span>, Fct <span class="op">=</span> <span class="st">"trimLRPatterns(Rpattern='GCCCGGGTAA', 
                subject=fq)"</span>, 
    batchsize <span class="op">=</span> <span class="fl">1e+05</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span>, compress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>The following example shows how one can design a custom read preprocessing function using utilities provided by the <em><code>ShortRead</code></em> package, and then run it in batch mode with the <em>‘preprocessReads’</em> function (here on paired-end reads).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filterFct</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fq</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">20</span>, <span class="va">Nexceptions</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">qcount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="fu">quality</span><span class="op">(</span><span class="va">fq</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">cutoff</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="co"># Retains reads where Phred scores are &gt;= cutoff with N exceptions</span>
    <span class="va">fq</span><span class="op">[</span><span class="va">qcount</span> <span class="op">&lt;=</span> <span class="va">Nexceptions</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">trim</span>, Fct <span class="op">=</span> <span class="st">"filterFct(fq, cutoff=20, Nexceptions=0)"</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span><span class="op">)</span></code></pre></div>
</div>
<div id="preprocessing-with-trimgalore" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-trimgalore" class="anchor"></a>Preprocessing with TrimGalore!</h3>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">TrimGalore!</a> is a wrapper tool to consistently apply quality and adapter trimming to fastq files, with some extra functionality for removing Reduced Representation Bisulfite-Seq (RRBS) libraries.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/trim_galore/trim_galore-se"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"trim_galore-se.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trim_galore-se.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trimG</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trimG</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Run Single Machine Option</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="preprocessing-with-trimmomatic" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-trimmomatic" class="anchor"></a>Preprocessing with Trimmomatic</h3>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/trimmomatic/trimmomatic-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trimmomatic-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trimmomatic-pe.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trimM</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trimM</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Run Single Machine Option</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="fastq-quality-report" class="section level2">
<h2 class="hasAnchor">
<a href="#fastq-quality-report" class="anchor"></a>FASTQ quality report</h2>
<p>The following <em><code>seeFastq</code></em> and <em><code>seeFastqPlot</code></em> functions generate and plot a series of useful quality statistics for a set of FASTQ files including per cycle quality box plots, base proportions, base-level quality trends, relative k-mer diversity, length and occurrence distribution of reads, number of reads above quality cutoffs and mean quality distribution.<br>
The function <em><code>seeFastq</code></em> computes the quality statistics and stores the results in a relatively small list object that can be saved to disk with <em><code><a href="https://rdrr.io/r/base/save.html">save()</a></code></em> and reloaded with <em><code><a href="https://rdrr.io/r/base/load.html">load()</a></code></em> for later plotting. The argument <em><code>klength</code></em> specifies the k-mer length and <em><code>batchsize</code></em> the number of reads to a random sample from each FASTQ file.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="../reference/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="st">"./results/fastqReport.pdf"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<center>
<img src="fastqReport.png">
</center>
<div align="center">
<strong>Figure 5:</strong> FASTQ quality report
</div>
<p></p>
<p>Parallelization of FASTQ quality report on a single machine with multiple cores.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="../reference/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu">bplapply</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">trim</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">fqlist</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Parallelization of FASTQ quality report via scheduler (<em>e.g.</em> Slurm) across several compute nodes.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
    <span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
    <span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/preprocessReads/trim-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
    <span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trim-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trim-pe.yml"</span>, 
        dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
    <span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trim</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="../reference/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="../reference/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">trim</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span>
<span class="fu"><a href="../reference/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">fqlist</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="ngs-alignment-software" class="section level2">
<h2 class="hasAnchor">
<a href="#ngs-alignment-software" class="anchor"></a>NGS Alignment software</h2>
<p>After quality control, the sequence reads can be aligned to a reference genome or transcriptome database. The following sessions present some NGS sequence alignment software. Select the most accurate aligner and determining the optimal parameter for your custom data set project.</p>
<p>For all the following examples, it is necessary to install the respective software and export the <code>PATH</code> accordingly. If it is available <a href="http://modules.sourceforge.net/">Environment Module</a> in the system, you can load all the request software with <em><code><a href="../reference/moduleload.html">moduleload(args)</a></code></em> function.</p>
<div id="alignment-with-hisat2-using-sysargs2" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-hisat2-using-sysargs2" class="anchor"></a>Alignment with <code>HISAT2</code> using <em><code>SYSargs2</code></em>
</h3>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span> in both interactive job submissions and batch submissions to queuing systems of clusters using the <em><code>systemPipeR's</code></em> new CWL command-line interface.</p>
<p>The parameter settings of the aligner are defined in the <code>hisat2-mapping-se.cwl</code> and <code>hisat2-mapping-se.yml</code> files. The following shows how to construct the corresponding <em>SYSargs2</em> object, here <em>args</em>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/hisat2/hisat2-se"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.cwl"</span>, input_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">args</span></code></pre></div>
<pre><code>## Instance of 'SYSargs2':
##    Slot names/accessors: 
##       targets: 18 (M1A...V12B), targetsheader: 4 (lines)
##       modules: 1
##       wf: 0, clt: 1, yamlinput: 7 (components)
##       input: 18, output: 18
##       cmdlist: 18
##    WF Steps:
##       1. hisat2-mapping-se (rendered: TRUE)</code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$`hisat2-mapping-se`
## [1] "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"
## 
## 
## $M1B
## $M1B$`hisat2-mapping-se`
## [1] "hisat2 -S ./results/M1B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446028_1.fastq.gz --threads 4"</code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<pre><code>## $M1A
## $M1A$`hisat2-mapping-se`
## [1] "./results/M1A.sam"
## 
## 
## $M1B
## $M1B$`hisat2-mapping-se`
## [1] "./results/M1B.sam"</code></pre>
<p>Subsetting <em><code>SYSargs2</code></em> class slots for each workflow step.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"input"</span>, subset <span class="op">=</span> <span class="st">"FileName"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>  <span class="co">## Subsetting the input files for this particular workflow </span></code></pre></div>
<pre><code>##                           M1A                           M1B 
## "./data/SRR446027_1.fastq.gz" "./data/SRR446028_1.fastq.gz"</code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>  <span class="co">## Subsetting the output files for one particular step in the workflow </span></code></pre></div>
<pre><code>##                 M1A                 M1B 
## "./results/M1A.sam" "./results/M1B.sam"</code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"step"</span>, subset <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>  <span class="co">## Subsetting the command-lines for one particular step in the workflow </span></code></pre></div>
<pre><code>##                                                                                                                                               M1A 
## "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"</code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span>, delete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>  <span class="co">## DELETING specific output files</span></code></pre></div>
<pre><code>## The subset cannot be deleted: no such file</code></pre>
<pre><code>##                 M1A 
## "./results/M1A.sam"</code></pre>
<p>Build <code>Hisat2</code> index.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/hisat2/hisat2-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"hisat2-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>

<span class="co">## Run</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div id="interactive-job-submissions-in-a-single-machine" class="section level4">
<h4 class="hasAnchor">
<a href="#interactive-job-submissions-in-a-single-machine" class="anchor"></a>Interactive job submissions in a single machine</h4>
<p>To simplify the short read alignment execution for the user, the command-line can be run with the <em><code>runCommandline</code></em> function. The execution will be on a single machine without submitting to a queuing system of a computer cluster. This way, the input FASTQ files will be processed sequentially. By default <em><code>runCommandline</code></em> auto detects SAM file outputs and converts them to sorted and indexed BAM files, using internally the <code>Rsamtools</code> package <span class="citation">(Morgan et al. 2019)</span>. Besides, <em><code>runCommandline</code></em> allows the user to create a dedicated results folder for each workflow and a sub-folder for each sample defined in the <em>targets</em> file. This includes all the output and log files for each step. When these options are used, the output location will be updated by default and can be assigned to the same object.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co">## generates alignments and writes *.sam files to ./results folder </span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co">## same as above but writes files and converts *.sam files to sorted and indexed BAM files. Assigning the new extention of the output files to the object args.</span></code></pre></div>
<p>If available, multiple CPU cores can be used for processing each file. The number of CPU cores (here 4) to use for each process is defined in the <em><code>*.yml</code></em> file. With <em><code>yamlinput(args)['thread']</code></em> one can return this value from the <em><code>SYSargs2</code></em> object.</p>
</div>
<div id="parallelization-on-clusters-1" class="section level4">
<h4 class="hasAnchor">
<a href="#parallelization-on-clusters-1" class="anchor"></a>Parallelization on clusters</h4>
<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. For this the <em><code>clusterRun</code></em> function submits the computing requests to the scheduler using the run specifications defined by <em><code>runCommandline</code></em>.</p>
<p>To avoid over-subscription of CPU cores on the compute nodes, the value from <em><code>yamlinput(args)['thread']</code></em> is passed on to the submission command, here <em><code>ncpus</code></em> in the <em><code>resources</code></em> list object. The number of independent parallel cluster processes is defined under the <em><code>Njobs</code></em> argument. The following example will run 18 processes in parallel using for each 4 CPU cores. If the resources available on a cluster allow running all 18 processes at the same time then the shown sample submission will utilize in total 72 CPU cores. Note, <em><code>clusterRun</code></em> can be used with most queueing systems as it is based on utilities from the <em><code>batchtools</code></em> package which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a conf file (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>) and a template file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>) for the queueing available on a system. The following example uses the sample conf and template files for the Slurm scheduler provided by this package.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span>, 
    dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/waitForJobs.html">waitForJobs</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
<p>Check and update the output location if necessary.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/output_update.html">output_update</a></span><span class="op">(</span><span class="va">args</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, extension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">".sam"</span>, <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span>  <span class="co">## Updates the output(args) to the right location in the subfolders</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span></code></pre></div>
</div>
<div id="create-new-targets-file" class="section level4">
<h4 class="hasAnchor">
<a href="#create-new-targets-file" class="anchor"></a>Create new targets file</h4>
<p>To establish the connectivity to the next workflow step, one can write a new <em>targets</em> file with the <em><code>writeTargetsout</code></em> function. The new <em>targets</em> file serves as input to the next <em><code>loadWorkflow</code></em> and <em><code>renderWF</code></em> call.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="alignment-with-hisat2-and-samtools" class="section level4">
<h4 class="hasAnchor">
<a href="#alignment-with-hisat2-and-samtools" class="anchor"></a>Alignment with <code>HISAT2</code> and <code>SAMtools</code>
</h4>
<p>Alternatively, it possible to build an workflow with <code>HISAT2</code> and <code>SAMtools</code>.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/workflow-hisat2/workflow-hisat2-se"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">WF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"workflow_hisat2-se.cwl"</span>, input_file <span class="op">=</span> <span class="st">"workflow_hisat2-se.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">WF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">WF</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">WF</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">WF</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
</div>
</div>
<div id="alignment-with-tophat2" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-tophat2" class="anchor"></a>Alignment with <em><code>Tophat2</code></em>
</h3>
<p>The NGS reads of this project can also be aligned against the reference genome sequence using <code>Bowtie2/TopHat2</code> <span class="citation">(Kim et al. 2013; Langmead and Salzberg 2012)</span>.</p>
<p>Build <em><code>Bowtie2</code></em> index.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"bowtie2-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>

<span class="co">## Run in single machine</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>The parameter settings of the aligner are defined in the <code>tophat2-mapping-pe.cwl</code> and <code>tophat2-mapping-pe.yml</code> files. The following shows how to construct the corresponding <em>SYSargs2</em> object, here <em>tophat2PE</em>.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/tophat2/tophat2-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">tophat2PE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"tophat2-mapping-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"tophat2-mapping-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">tophat2PE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">tophat2PE</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">tophat2PE</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">tophat2PE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">tophat2PE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>

<span class="co">## Run in single machine</span>
<span class="va">tophat2PE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">tophat2PE</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Parallelization on clusters.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">tophat2PE</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">tophat2PE</span>, 
    make_bam <span class="op">=</span> <span class="cn">TRUE</span>, dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/waitForJobs.html">waitForJobs</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
<p>Create new targets file</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">tophat2PE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tophat2PE</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"tophat2PE"</span>, 
    new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="alignment-with-bowtie2-e-g--for-mirna-profiling" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-bowtie2-e-g--for-mirna-profiling" class="anchor"></a>Alignment with <em><code>Bowtie2</code></em> (<em>e.g.</em> for miRNA profiling)</h3>
<p>The following example runs <em><code>Bowtie2</code></em> as a single process without submitting it to a cluster.</p>
<p>Building the index:</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"bowtie2-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>

<span class="co">## Run in single machine</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Building all the command-line:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">bowtiePE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-mapping-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bowtie2-mapping-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">bowtiePE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">bowtiePE</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bowtiePE</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">bowtiePE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">bowtiePE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<p>Running all the jobs to computing nodes.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">bowtiePE</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bowtiePE</span>, 
    dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
<p>Alternatively, it possible to run all the jobs in a single machine.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bowtiePE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">bowtiePE</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">bowtiePE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bowtiePE</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"bowtiePE"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="alignment-with-bwa-mem-e-g--for-var-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-bwa-mem-e-g--for-var-seq" class="anchor"></a>Alignment with <em><code>BWA-MEM</code></em> (<em>e.g.</em> for VAR-Seq)</h3>
<p>The following example runs BWA-MEM as a single process without submitting it to a cluster. ##TODO: add reference</p>
<p>Build the index:</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bwa/bwa-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bwa-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"bwa-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>  <span class="co"># Indexes reference genome</span>

<span class="co">## Run</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Running the alignment:</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bwa/bwa-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"bwa-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"bwa-pe.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">bwaPE</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bwaPE</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Single Machine</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bwaPE</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">## Cluster</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">bwaPE</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bwaPE</span>, dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, 
    conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, Njobs <span class="op">=</span> <span class="fl">18</span>, 
    runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bwaPE</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"bwaPE"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="alignment-with-rsubread-e-g--for-rna-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-rsubread-e-g--for-rna-seq" class="anchor"></a>Alignment with <em><code>Rsubread</code></em> (<em>e.g.</em> for RNA-Seq)</h3>
<p>The following example shows how one can use within the  environment the R-based aligner , allowing running from R or command-line.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Build the index:</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/rsubread/rsubread-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"rsubread-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"rsubread-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">## Running the alignment:</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/rsubread/rsubread-se"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">rsubread</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"rsubread-mapping-se.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"rsubread-mapping-se.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">rsubread</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">rsubread</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">rsubread</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">rsubread</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>

<span class="co">## Single Machine</span>
<span class="va">rsubread</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">rsubread</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">rsubread</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rsubread</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"rsubread"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="alignment-with-gsnap-e-g--for-var-seq-and-rna-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-gsnap-e-g--for-var-seq-and-rna-seq" class="anchor"></a>Alignment with <em><code>gsnap</code></em> (<em>e.g.</em> for VAR-Seq and RNA-Seq)</h3>
<p>Another R-based short read aligner is <em><code>gsnap</code></em> from the <em><code>gmapR</code></em> package <span class="citation">(Wu and Nacu 2010)</span>. The code sample below introduces how to run this aligner on multiple nodes of a compute cluster.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Build the index:</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/gsnap/gsnap-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"gsnap-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gsnap-index.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="fu"><a href="../reference/runCommandline.html">runCommandline</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">## Running the alignment:</span>
<span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/gsnap/gsnap-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">gsnap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"gsnap-mapping-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gsnap-mapping-pe.yml"</span>, 
    dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">gsnap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">gsnap</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gsnap</span>
<span class="fu"><a href="../reference/sysargs2.html">cmdlist</a></span><span class="op">(</span><span class="va">gsnap</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="../reference/sysargs2.html">output</a></span><span class="op">(</span><span class="va">gsnap</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>

<span class="co">## Cluster</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">gsnap</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">gsnap</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, 
    conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, Njobs <span class="op">=</span> <span class="fl">18</span>, 
    runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="va">gsnap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/output_update.html">output_update</a></span><span class="op">(</span><span class="va">gsnap</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, extension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">".sam"</span>, 
    <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">clt</a></span><span class="op">(</span><span class="va">gsnap</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gsnap</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"gsnap"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="create-symbolic-links-for-viewing-bam-files-in-igv" class="section level2">
<h2 class="hasAnchor">
<a href="#create-symbolic-links-for-viewing-bam-files-in-igv" class="anchor"></a>Create symbolic links for viewing BAM files in IGV</h2>
<p>The genome browser IGV supports reading of indexed/sorted BAM files via web URLs. This way it can be avoided to create unnecessary copies of these large files. To enable this approach, an HTML directory with Http access needs to be available in the user account (<em>e.g.</em> <em><code>home/publichtml</code></em>) of a system. If this is not the case then the BAM files need to be moved or copied to the system where IGV runs. In the following, <em><code>htmldir</code></em> defines the path to the HTML directory with http access where the symbolic links to the BAM files will be stored. The corresponding URLs will be written to a text file specified under the <code>_urlfile</code>_ argument.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/symLink2bam.html">symLink2bam</a></span><span class="op">(</span>sysargs <span class="op">=</span> <span class="va">args</span>, htmldir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"~/.html/"</span>, <span class="st">"somedir/"</span><span class="op">)</span>, urlbase <span class="op">=</span> <span class="st">"http://myserver.edu/~username/"</span>, 
    urlfile <span class="op">=</span> <span class="st">"IGVurl.txt"</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-counting-for-mrna-profiling-experiments" class="section level2">
<h2 class="hasAnchor">
<a href="#read-counting-for-mrna-profiling-experiments" class="anchor"></a>Read counting for mRNA profiling experiments</h2>
<p>Create <em><code>txdb</code></em> (needs to be done only once).</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures">GenomicFeatures</a></span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>, format <span class="op">=</span> <span class="st">"gff"</span>, dataSource <span class="op">=</span> <span class="st">"TAIR"</span>, 
    organism <span class="op">=</span> <span class="st">"Arabidopsis thaliana"</span><span class="op">)</span>
<span class="fu">saveDb</span><span class="op">(</span><span class="va">txdb</span>, file <span class="op">=</span> <span class="st">"./data/tair10.sqlite"</span><span class="op">)</span></code></pre></div>
<p>The following performs read counting with <em><code>summarizeOverlaps</code></em> in parallel mode with multiple cores.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">loadDb</span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>
<span class="va">eByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">multicoreParam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="va">multicoreParam</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">registered</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">counteByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="va">bfl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">summarizeOverlaps</span><span class="op">(</span><span class="va">eByg</span>, <span class="va">x</span>, mode <span class="op">=</span> <span class="st">"Union"</span>, 
    ignore.strand <span class="op">=</span> <span class="cn">TRUE</span>, inter.feature <span class="op">=</span> <span class="cn">TRUE</span>, singleEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Note: for strand-specific RNA-Seq set 'ignore.strand=FALSE' and for PE data set</span>
<span class="co"># 'singleEnd=FALSE'</span>
<span class="va">countDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">counteByg</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">assays</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>
<span class="va">rpkmDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">countDFeByg</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/returnRPKM.html">returnRPKM</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">x</span>, ranges <span class="op">=</span> <span class="va">eByg</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">countDFeByg</span>, <span class="st">"results/countDFeByg.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">rpkmDFeByg</span>, <span class="st">"results/rpkmDFeByg.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
<p>Please note, in addition to read counts this step generates RPKM normalized expression values. For most statistical differential expression or abundance analysis methods, such as <em><code>edgeR</code></em> or <em><code>DESeq2</code></em>, the raw count values should be used as input. The usage of RPKM values should be restricted to specialty applications required by some users, <em>e.g.</em> manually comparing the expression levels of different genes or features.</p>
<p>Read counting with <em><code>summarizeOverlaps</code></em> using multiple nodes of a cluster.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures">GenomicFeatures</a></span><span class="op">)</span>
    <span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">loadDb</span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>
    <span class="va">eByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>
    <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/systemArgs.html">systemArgs</a></span><span class="op">(</span>sysma <span class="op">=</span> <span class="st">"param/tophat.param"</span>, mytargets <span class="op">=</span> <span class="st">"targets.txt"</span><span class="op">)</span>
    <span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
    <span class="fu">summarizeOverlaps</span><span class="op">(</span><span class="va">eByg</span>, <span class="va">bfl</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, mode <span class="op">=</span> <span class="st">"Union"</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span>, inter.feature <span class="op">=</span> <span class="cn">TRUE</span>, 
        singleEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">counteByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">args</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span>
<span class="va">countDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">counteByg</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">assays</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span></code></pre></div>
<p>Useful commands for monitoring the progress of submitted jobs</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/loadResult.html">loadResult</a></span><span class="op">(</span><span class="va">reg</span>, id <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span>  <span class="co"># Works after job completion</span></code></pre></div>
<div id="read-and-alignment-count-stats" class="section level4">
<h4 class="hasAnchor">
<a href="#read-and-alignment-count-stats" class="anchor"></a>Read and alignment count stats</h4>
<p>Generate a table of read and alignment counts for all samples.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/alignStats.html">alignStats</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats.xls"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
<p>The following shows the first four lines of the sample alignment stats file provided by the <em><code>systemPipeR</code></em> package. For simplicity the number of PE reads is multiplied here by 2 to approximate proper alignment frequencies where each read in a pair is counted.</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"alignStats.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, 
    <span class="op">]</span></code></pre></div>
<pre><code>##   FileName Nreads2x Nalign Perc_Aligned Nalign_Primary Perc_Aligned_Primary
## 1      M1A   192918 177961     92.24697         177961             92.24697
## 2      M1B   197484 159378     80.70426         159378             80.70426
## 3      A1A   189870 176055     92.72397         176055             92.72397
## 4      A1B   188854 147768     78.24457         147768             78.24457</code></pre>
<p>Parallelization of read/alignment stats on single machine with multiple cores.</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/alignStats.html">alignStats</a></span><span class="op">(</span><span class="va">args</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span>
<span class="va">read_statsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">args</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span>
<span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">read_statsList</span><span class="op">)</span></code></pre></div>
<p>Parallelization of read/alignment stats via scheduler (<em>e.g.</em> Slurm) across several compute nodes.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span>
    <span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
    <span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="st">"param/cwl/hisat2/hisat2-se"</span>  <span class="co">## TODO: replace path to system.file </span>
    <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.cwl"</span>, input_file <span class="op">=</span> <span class="st">"hisat2-mapping-se.yml"</span>, 
        dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
    <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/output_update.html">output_update</a></span><span class="op">(</span><span class="va">args</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, extension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">".sam"</span>, 
        <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="../reference/alignStats.html">alignStats</a></span><span class="op">(</span><span class="va">args</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">read_statsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">args</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span>
<span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">read_statsList</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="read-counting-for-mirna-profiling-experiments" class="section level2">
<h2 class="hasAnchor">
<a href="#read-counting-for-mirna-profiling-experiments" class="anchor"></a>Read counting for miRNA profiling experiments</h2>
<p>Download miRNA genes from miRBase.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"wget ftp://mirbase.org/pub/mirbase/19/genomes/My_species.gff3 -P ./data/"</span><span class="op">)</span>
<span class="va">gff</span> <span class="op">&lt;-</span> <span class="fu">import.gff</span><span class="op">(</span><span class="st">"./data/My_species.gff3"</span><span class="op">)</span>
<span class="va">gff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">gff</span>, <span class="fu">elementMetadata</span><span class="op">(</span><span class="va">gff</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span>
<span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bampaths</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bams</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">targets</span><span class="op">$</span><span class="va">SampleName</span>
<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">bams</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">countDFmiR</span> <span class="op">&lt;-</span> <span class="fu">summarizeOverlaps</span><span class="op">(</span><span class="va">gff</span>, <span class="va">bfl</span>, mode <span class="op">=</span> <span class="st">"Union"</span>, ignore.strand <span class="op">=</span> <span class="cn">FALSE</span>, 
    inter.feature <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># Note: inter.feature=FALSE important since pre and mature miRNA ranges overlap</span>
<span class="va">rpkmDFmiR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">countDFmiR</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/returnRPKM.html">returnRPKM</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">x</span>, gffsub <span class="op">=</span> <span class="va">gff</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">countDFmiR</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span>, <span class="st">"results/countDFmiR.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, 
    quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">rpkmDFmiR</span>, <span class="st">"results/rpkmDFmiR.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
</div>
<div id="correlation-analysis-of-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#correlation-analysis-of-samples" class="anchor"></a>Correlation analysis of samples</h2>
<p>The following computes the sample-wise Spearman correlation coefficients from the <em><code>rlog</code></em> (regularized-logarithm) transformed expression values generated with the <em><code>DESeq2</code></em> package. After transformation to a distance matrix, hierarchical clustering is performed with the <em><code>hclust</code></em> function and the result is plotted as a dendrogram (<a href="./results/sample_tree.pdf">sample_tree.pdf</a>).</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span>, warn.conflicts <span class="op">=</span> <span class="cn">FALSE</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span>, warn.conflicts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">countDFpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"countDFeByg.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">countDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="va">countDFpath</span><span class="op">)</span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="../reference/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">SampleName</span>, condition <span class="op">=</span> <span class="fu"><a href="../reference/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="../reference/sysargs2.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">Factor</span><span class="op">)</span>
<span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">countDF</span>, colData <span class="op">=</span> <span class="va">colData</span>, design <span class="op">=</span> <span class="op">~</span><span class="va">condition</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
## design formula are characters, converting to factors</code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>
<span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">d</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html">plot.phylo</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"p"</span>, edge.col <span class="op">=</span> <span class="fl">4</span>, edge.width <span class="op">=</span> <span class="fl">3</span>, show.node.label <span class="op">=</span> <span class="cn">TRUE</span>, 
    no.margin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="../docs/articles/systemPipeR_files/figure-html/sample_tree_rlog-1.png" width="700"></p>
<div align="center">
<strong>Figure 6:</strong> Correlation dendrogram of samples for <em><code>rlog</code></em> values.
</div>
<p></p>
<p>Alternatively, the clustering can be performed with <em><code>RPKM</code></em> normalized expression values. In combination with Spearman correlation the results of the two clustering methods are often relatively similar.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rpkmDFeBygpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rpkmDFeByg.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">rpkmDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="va">rpkmDFeBygpath</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">rpkmDFeByg</span> <span class="op">&lt;-</span> <span class="va">rpkmDFeByg</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">rpkmDFeByg</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">50</span>, <span class="op">]</span>
<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">rpkmDFeByg</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>
<span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">d</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html">plot.phylo</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"p"</span>, edge.col <span class="op">=</span> <span class="st">"blue"</span>, edge.width <span class="op">=</span> <span class="fl">2</span>, show.node.label <span class="op">=</span> <span class="cn">TRUE</span>, 
    no.margin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="deg-analysis-with-edger" class="section level2">
<h2 class="hasAnchor">
<a href="#deg-analysis-with-edger" class="anchor"></a>DEG analysis with <em><code>edgeR</code></em>
</h2>
<p>The following <em><code>run_edgeR</code></em> function is a convenience wrapper for identifying differentially expressed genes (DEGs) in batch mode with <em><code>edgeR</code></em>’s GML method <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span> for any number of pairwise sample comparisons specified under the <em><code>cmp</code></em> argument. Users are strongly encouraged to consult the <a href="%5Chref%7Bhttp://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf"><em><code>edgeR</code></em></a> vignette for more detailed information on this topic and how to properly run <em><code>edgeR</code></em> on data sets with more complex experimental designs.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
<span class="va">cmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readComp.html">readComp</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">targetspath</span>, format <span class="op">=</span> <span class="st">"matrix"</span>, delim <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span>
<span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code>##       [,1]  [,2] 
##  [1,] "M1"  "A1" 
##  [2,] "M1"  "V1" 
##  [3,] "A1"  "V1" 
##  [4,] "M6"  "A6" 
##  [5,] "M6"  "V6" 
##  [6,] "A6"  "V6" 
##  [7,] "M12" "A12"
##  [8,] "M12" "V12"
##  [9,] "A12" "V12"</code></pre>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">countDFeBygpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"countDFeByg.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">countDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">countDFeBygpath</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">edgeDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_edgeR.html">run_edgeR</a></span><span class="op">(</span>countDF <span class="op">=</span> <span class="va">countDFeByg</span>, targets <span class="op">=</span> <span class="va">targets</span>, cmp <span class="op">=</span> <span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, independent <span class="op">=</span> <span class="cn">FALSE</span>, 
    mdsplot <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<pre><code>## Disp = 0.21829 , BCV = 0.4672</code></pre>
<p>Filter and plot DEG results for up and down-regulated genes. Because of the small size of the toy data set used by this vignette, the <em>FDR</em> value has been set to a relatively high threshold (here 10%). More commonly used <em>FDR</em> cutoffs are 1% or 5%. The definition of ‘<em>up</em>’ and ‘<em>down</em>’ is given in the corresponding help file. To open it, type <em><code><a href="../reference/filterDEGs.html">?filterDEGs</a></code></em> in the R console.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEG_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterDEGs.html">filterDEGs</a></span><span class="op">(</span>degDF <span class="op">=</span> <span class="va">edgeDF</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, FDR <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="../docs/articles/systemPipeR_files/figure-html/edger_deg_counts-1.png" width="700"></p>
<div align="center">
<strong>Figure 7:</strong> Up and down regulated DEGs identified by <em><code>edgeR</code></em>.
</div>
<p></p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "UporDown" "Up"       "Down"     "Summary"</code></pre>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEG_list</span><span class="op">$</span><span class="va">Summary</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span></code></pre></div>
<pre><code>##       Comparisons Counts_Up_or_Down Counts_Up Counts_Down
## M1-A1       M1-A1                 0         0           0
## M1-V1       M1-V1                 1         1           0
## A1-V1       A1-V1                 1         1           0
## M6-A6       M6-A6                 0         0           0</code></pre>
</div>
<div id="deg-analysis-with-deseq2" class="section level2">
<h2 class="hasAnchor">
<a href="#deg-analysis-with-deseq2" class="anchor"></a>DEG analysis with <em><code>DESeq2</code></em>
</h2>
<p>The following <em><code>run_DESeq2</code></em> function is a convenience wrapper for identifying DEGs in batch mode with <em><code>DESeq2</code></em> <span class="citation">(Love, Huber, and Anders 2014)</span> for any number of pairwise sample comparisons specified under the <em><code>cmp</code></em> argument. Users are strongly encouraged to consult the <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf"><em><code>DESeq2</code></em></a> vignette for more detailed information on this topic and how to properly run <em><code>DESeq2</code></em> on data sets with more complex experimental designs.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degseqDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_DESeq2.html">run_DESeq2</a></span><span class="op">(</span>countDF <span class="op">=</span> <span class="va">countDFeByg</span>, targets <span class="op">=</span> <span class="va">targets</span>, cmp <span class="op">=</span> <span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, 
    independent <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
## design formula are characters, converting to factors</code></pre>
<p>Filter and plot DEG results for up and down-regulated genes.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEG_list2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterDEGs.html">filterDEGs</a></span><span class="op">(</span>degDF <span class="op">=</span> <span class="va">degseqDF</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, FDR <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="../docs/articles/systemPipeR_files/figure-html/deseq2_deg_counts-1.png" width="700"></p>
<div align="center">
<strong>Figure 8:</strong> Up and down regulated DEGs identified by <em><code>DESeq2</code></em>.
</div>
<p></p>
</div>
<div id="venn-diagrams" class="section level2">
<h2 class="hasAnchor">
<a href="#venn-diagrams" class="anchor"></a>Venn Diagrams</h2>
<p>The function <em><code>overLapper</code></em> can compute Venn intersects for large numbers of sample sets (up to 20 or more) and <em><code>vennPlot</code></em> can plot 2-5 way Venn diagrams. A useful feature is the possibility to combine the counts from several Venn comparisons with the same number of sample sets in a single Venn diagram (here for 4 up and down DEG sets).</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vennsetup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">$</span><span class="va">Up</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"vennsets"</span><span class="op">)</span>
<span class="va">vennsetdown</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">$</span><span class="va">Down</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"vennsets"</span><span class="op">)</span>
<span class="fu"><a href="../reference/vennPlot.html">vennPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">vennsetup</span>, <span class="va">vennsetdown</span><span class="op">)</span>, mymain <span class="op">=</span> <span class="st">""</span>, mysub <span class="op">=</span> <span class="st">""</span>, colmode <span class="op">=</span> <span class="fl">2</span>, ccol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, 
    <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="../docs/articles/systemPipeR_files/figure-html/vennplot-1.png" width="700"></p>
<div align="center">
<strong>Figure 9:</strong> Venn Diagram for 4 Up and Down DEG Sets.
</div>
<p></p>
</div>
<div id="go-term-enrichment-analysis-of-degs" class="section level2">
<h2 class="hasAnchor">
<a href="#go-term-enrichment-analysis-of-degs" class="anchor"></a>GO term enrichment analysis of DEGs</h2>
<div id="obtain-gene-to-go-mappings" class="section level3">
<h3 class="hasAnchor">
<a href="#obtain-gene-to-go-mappings" class="anchor"></a>Obtain gene-to-GO mappings</h3>
<p>The following shows how to obtain gene-to-GO mappings from <em><code>biomaRt</code></em> (here for <em>A. thaliana</em>) and how to organize them for the downstream GO term enrichment analysis. Alternatively, the gene-to-GO mappings can be obtained for many organisms from Bioconductor’s <em><code>*.db</code></em> genome annotation packages or GO annotation files provided by various genome databases. For each annotation, this relatively slow preprocessing step needs to be performed only once. Subsequently, the preprocessed data can be loaded with the <em><code>load</code></em> function as shown in the next subsection.</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listMarts.html">listMarts</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># To choose BioMart database</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listMarts.html">listMarts</a></span><span class="op">(</span>host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">listDatasets</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>  <span class="co"># Choose data types you want to download</span>
<span class="va">go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"go_id"</span>, <span class="st">"tair_locus"</span>, <span class="st">"namespace_1003"</span><span class="op">)</span>, mart <span class="op">=</span> <span class="va">m</span><span class="op">)</span>
<span class="va">go</span> <span class="op">&lt;-</span> <span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">!=</span> <span class="st">""</span>, <span class="op">]</span>
<span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"molecular_function"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"F"</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"biological_process"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P"</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"cellular_component"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"C"</span>
<span class="va">go</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="st">"./data/GO"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">go</span>, <span class="st">"data/GO/GOannotationsBiomart_mod.txt"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, 
    col.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="va">catdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GOHyperGAll.html">makeCATdb</a></span><span class="op">(</span>myfile <span class="op">=</span> <span class="st">"data/GO/GOannotationsBiomart_mod.txt"</span>, lib <span class="op">=</span> <span class="cn">NULL</span>, org <span class="op">=</span> <span class="st">""</span>, 
    colno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, idconv <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">catdb</span>, file <span class="op">=</span> <span class="st">"data/GO/catdb.RData"</span><span class="op">)</span></code></pre></div>
</div>
<div id="batch-go-term-enrichment-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#batch-go-term-enrichment-analysis" class="anchor"></a>Batch GO term enrichment analysis</h3>
<p>Apply the enrichment analysis to the DEG sets obtained in the above differential expression analysis. Note, in the following example the <em>FDR</em> filter is set here to an unreasonably high value, simply because of the small size of the toy data set used in this vignette. Batch enrichment analysis of many gene sets is performed with the <em><code>GOCluster_Report</code></em> function. When <em><code>method="all"</code></em>, it returns all GO terms passing the p-value cutoff specified under the <em><code>cutoff</code></em> arguments. When <em><code>method="slim"</code></em>, it returns only the GO terms specified under the <em><code>myslimv</code></em> argument. The given example shows how one can obtain such a GO slim vector from BioMart for a specific organism.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/GO/catdb.RData"</span><span class="op">)</span>
<span class="va">DEG_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterDEGs.html">filterDEGs</a></span><span class="op">(</span>degDF <span class="op">=</span> <span class="va">edgeDF</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, FDR <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">up_down</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">UporDown</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up_down</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up_down</span><span class="op">)</span>, <span class="st">"_up_down"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">up</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">Up</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up</span><span class="op">)</span>, <span class="st">"_up"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">down</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">Down</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">down</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">down</span><span class="op">)</span>, <span class="st">"_down"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">DEGlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">up_down</span>, <span class="va">up</span>, <span class="va">down</span><span class="op">)</span>
<span class="va">DEGlist</span> <span class="op">&lt;-</span> <span class="va">DEGlist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">DEGlist</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>
<span class="va">BatchResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GOHyperGAll.html">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">DEGlist</span>, method <span class="op">=</span> <span class="st">"all"</span>, 
    id_type <span class="op">=</span> <span class="st">"gene"</span>, CLSZ <span class="op">=</span> <span class="fl">2</span>, cutoff <span class="op">=</span> <span class="fl">0.9</span>, gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="va">goslimvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"goslim_goa_accession"</span><span class="op">)</span>, mart <span class="op">=</span> <span class="va">m</span><span class="op">)</span><span class="op">[</span>, 
    <span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">BatchResultslim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GOHyperGAll.html">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">DEGlist</span>, method <span class="op">=</span> <span class="st">"slim"</span>, 
    id_type <span class="op">=</span> <span class="st">"gene"</span>, myslimv <span class="op">=</span> <span class="va">goslimvec</span>, CLSZ <span class="op">=</span> <span class="fl">10</span>, cutoff <span class="op">=</span> <span class="fl">0.01</span>, gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, 
        <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
</div>
<div id="plot-batch-go-term-results" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-batch-go-term-results" class="anchor"></a>Plot batch GO term results</h3>
<p>The <em><code>data.frame</code></em> generated by <em><code>GOCluster_Report</code></em> can be plotted with the <em><code>goBarplot</code></em> function. Because of the variable size of the sample sets, it may not always be desirable to show the results from different DEG sets in the same bar plot. Plotting single sample sets is achieved by subsetting the input data frame as shown in the first line of the following example.</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gos</span> <span class="op">&lt;-</span> <span class="va">BatchResultslim</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"M6-V6_up_down"</span>, <span class="va">BatchResultslim</span><span class="op">$</span><span class="va">CLID</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">gos</span> <span class="op">&lt;-</span> <span class="va">BatchResultslim</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="st">"GOslimbarplotMF.pdf"</span>, height <span class="op">=</span> <span class="fl">8</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="../reference/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"MF"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"BP"</span><span class="op">)</span>
<span class="fu"><a href="../reference/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"CC"</span><span class="op">)</span></code></pre></div>
<img src="GOslimbarplotMF.png"><div align="center">
<strong>Figure 10:</strong> GO Slim Barplot for MF Ontology.
</div>
<p></p>
</div>
</div>
<div id="clustering-and-heat-maps" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-and-heat-maps" class="anchor"></a>Clustering and heat maps</h2>
<p>The following example performs hierarchical clustering on the <em><code>rlog</code></em> transformed expression matrix subsetted by the DEGs identified in the above differential expression analysis. It uses a Pearson correlation-based distance measure and complete linkage for cluster join.</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="va">geneids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">geneids</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="st">"heatmap1.pdf"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">y</span>, scale <span class="op">=</span> <span class="st">"row"</span>, clustering_distance_rows <span class="op">=</span> <span class="st">"correlation"</span>, clustering_distance_cols <span class="op">=</span> <span class="st">"correlation"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<center>
<img src="heatmap1.png">
</center>
<div align="center">
<strong>Figure 11:</strong> Heat map with hierarchical clustering dendrograms of DEGs.
</div>
<p></p>
</div>
</div>
<div id="workflow-templates" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-templates" class="anchor"></a>Workflow templates</h1>
<p>The intended way of running <em><code>systemPipeR</code></em> workflows is via <em><code>*.Rmd</code></em> files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line. This way comprehensive and reproducible analysis reports can be generated in PDF or HTML format in a fully automated manner by making use of the highly functional reporting utilities available for R. The following shows how to execute a workflow (<em>e.g.</em>, systemPipeRNAseq.Rmd) from the command-line.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> -e <span class="st">"rmarkdown::render('systemPipeRNAseq.Rmd')"</span></span></code></pre></div>
<p>Templates for setting up custom project reports are provided as <em><code>*.Rmd</code></em> files by the helper package <em><code>systemPipeRdata</code></em> and in the vignettes subdirectory of <em><code>systemPipeR</code></em>. The corresponding HTML of these report templates are available here: <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.html"><em><code>systemPipeRNAseq</code></em></a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.html"><em><code>systemPipeRIBOseq</code></em></a>, <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.html"><em><code>systemPipeChIPseq</code></em></a> and <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.html"><em><code>systemPipeVARseq</code></em></a>. To work with <em><code>*.Rnw</code></em> or <em><code>*.Rmd</code></em> files efficiently, basic knowledge of <a href="https://www.stat.uni-muenchen.de/~leisch/Sweave/"><em><code>Sweave</code></em></a> or <a href="http://yihui.name/knitr/"><em><code>knitr</code></em></a> and <a href="http://www.latex-project.org/"><em><code>Latex</code></em></a> or <a href="http://rmarkdown.rstudio.com/"><em><code>R Markdown v2</code></em></a> is required.</p>
<div id="rna-seq-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#rna-seq-sample" class="anchor"></a>RNA-Seq sample</h2>
<p>Load the RNA-Seq sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
<div id="run-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#run-workflow" class="anchor"></a>Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code>systemPipeRNAseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq.Rmd">Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>rnaseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p>The workflow includes following steps:</p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>Tophat2</code></em> (or any other RNA-Seq aligner)</li>
<li>Alignment stats</li>
<li>Read counting</li>
<li>Sample-wise correlation analysis</li>
<li>Analysis of differentially expressed genes (DEGs)</li>
<li>GO term enrichment analysis</li>
<li>Gene-wise clustering</li>
</ol>
</div>
</div>
<div id="chip-seq-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#chip-seq-sample" class="anchor"></a>ChIP-Seq sample</h2>
<p>Load the ChIP-Seq sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"chipseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"chipseq"</span><span class="op">)</span></code></pre></div>
<div id="run-workflow-1" class="section level3">
<h3 class="hasAnchor">
<a href="#run-workflow-1" class="anchor"></a>Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code>systemPipeChIPseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq.Rmd">Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>chipseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p>The workflow includes the following steps:</p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>Bowtie2</code></em> or <em><code>rsubread</code></em>
</li>
<li>Alignment stats</li>
<li>Peak calling: <em><code>MACS2</code></em>, <em><code>BayesPeak</code></em>
</li>
<li>Peak annotation with genomic context</li>
<li>Differential binding analysis</li>
<li>GO term enrichment analysis</li>
<li>Motif analysis</li>
</ol>
</div>
</div>
<div id="var-seq-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#var-seq-sample" class="anchor"></a>VAR-Seq sample</h2>
<div id="var-seq-workflow-for-the-single-machine" class="section level3">
<h3 class="hasAnchor">
<a href="#var-seq-workflow-for-the-single-machine" class="anchor"></a>VAR-Seq workflow for the single machine</h3>
<p>Load the VAR-Seq sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"varseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"varseq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-workflow-2" class="section level3">
<h3 class="hasAnchor">
<a href="#run-workflow-2" class="anchor"></a>Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code>systemPipeVARseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.Rmd">Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>varseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p>The workflow includes following steps:</p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Quality filtering (trimming)</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>gsnap</code></em>, <em><code>bwa</code></em>
</li>
<li>Variant calling: <em><code>VariantTools</code></em>, <em><code>GATK</code></em>, <em><code>BCFtools</code></em>
</li>
<li>Variant filtering: <em><code>VariantTools</code></em> and <em><code>VariantAnnotation</code></em>
</li>
<li>Variant annotation: <em><code>VariantAnnotation</code></em>
</li>
<li>Combine results from many samples</li>
<li>Summary statistics of samples</li>
</ol>
</div>
<div id="var-seq-workflow-for-computer-cluster" class="section level3">
<h3 class="hasAnchor">
<a href="#var-seq-workflow-for-computer-cluster" class="anchor"></a>VAR-Seq workflow for computer cluster</h3>
<p>The workflow template provided for this step is called <em><code>systemPipeVARseq.Rmd</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.Rmd">Rmd</a>). It runs the above VAR-Seq workflow in parallel on multiple compute nodes of an HPC system using Slurm as the scheduler.</p>
</div>
</div>
<div id="ribo-seq-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#ribo-seq-sample" class="anchor"></a>Ribo-Seq sample</h2>
<p>Load the Ribo-Seq sample workflow into your current working directory.</p>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"riboseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"riboseq"</span><span class="op">)</span></code></pre></div>
<div id="run-workflow-3" class="section level3">
<h3 class="hasAnchor">
<a href="#run-workflow-3" class="anchor"></a>Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code>systemPipeRIBOseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/riboseq/systemPipeRIBOseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/ribseq/systemPipeRIBOseq.Rmd">Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>ribseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p>
<p>The workflow includes following steps:</p>
<ol style="list-style-type: decimal">
<li>Read preprocessing
<ul>
<li>Adaptor trimming and quality filtering</li>
<li>FASTQ quality report</li>
</ul>
</li>
<li>Alignments: <em><code>Tophat2</code></em> (or any other RNA-Seq aligner)</li>
<li>Alignment stats</li>
<li>Compute read distribution across genomic features</li>
<li>Adding custom features to the workflow (e.g. uORFs)</li>
<li>Genomic read coverage along with transcripts</li>
<li>Read counting</li>
<li>Sample-wise correlation analysis</li>
<li>Analysis of differentially expressed genes (DEGs)</li>
<li>GO term enrichment analysis</li>
<li>Gene-wise clustering</li>
<li>Differential ribosome binding (translational efficiency)</li>
</ol>
</div>
</div>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version information</h1>
<p><strong>Note:</strong> the most recent version of this tutorial can be found <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a>.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-4.0.3/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] DESeq2_1.30.0               kableExtra_1.3.1           
##  [3] dplyr_1.0.4                 magrittr_2.0.1             
##  [5] batchtools_0.9.15           ape_5.4-1                  
##  [7] ggplot2_3.3.3               systemPipeR_1.25.6         
##  [9] ShortRead_1.48.0            GenomicAlignments_1.26.0   
## [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [13] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [15] BiocParallel_1.24.1         Rsamtools_2.6.0            
## [17] Biostrings_2.58.0           XVector_0.30.0             
## [19] GenomicRanges_1.42.0        GenomeInfoDb_1.26.2        
## [21] IRanges_2.24.1              S4Vectors_0.28.1           
## [23] BiocGenerics_0.36.0         BiocStyle_2.18.1           
## 
## loaded via a namespace (and not attached):
##   [1] GOstats_2.56.0           backports_1.2.1          BiocFileCache_1.14.0    
##   [4] systemfonts_1.0.1        GSEABase_1.52.1          splines_4.0.3           
##   [7] digest_0.6.27            htmltools_0.5.1.1        GO.db_3.12.1            
##  [10] checkmate_2.0.0          memoise_2.0.0            BSgenome_1.58.0         
##  [13] base64url_1.4            limma_3.46.0             annotate_1.68.0         
##  [16] askpass_1.1              pkgdown_1.6.1            prettyunits_1.1.1       
##  [19] jpeg_0.1-8.1             colorspace_2.0-0         blob_1.2.1              
##  [22] rvest_0.3.6              rappdirs_0.3.3           textshaping_0.3.0       
##  [25] xfun_0.21                crayon_1.4.1             RCurl_1.98-1.2          
##  [28] jsonlite_1.7.2           graph_1.68.0             genefilter_1.72.1       
##  [31] brew_1.0-6               survival_3.2-7           VariantAnnotation_1.36.0
##  [34] glue_1.4.2               gtable_0.3.0             zlibbioc_1.36.0         
##  [37] webshot_0.5.2            DelayedArray_0.16.1      V8_3.4.0                
##  [40] Rgraphviz_2.34.0         scales_1.1.1             pheatmap_1.0.12         
##  [43] DBI_1.1.1                edgeR_3.32.1             Rcpp_1.0.6              
##  [46] viridisLite_0.3.0        xtable_1.8-4             progress_1.2.2          
##  [49] bit_4.0.4                rsvg_2.1                 AnnotationForge_1.32.0  
##  [52] httr_1.4.2               RColorBrewer_1.1-2       ellipsis_0.3.1          
##  [55] farver_2.0.3             pkgconfig_2.0.3          XML_3.99-0.5            
##  [58] dbplyr_2.1.0             locfit_1.5-9.4           labeling_0.4.2          
##  [61] tidyselect_1.1.0         rlang_0.4.10             AnnotationDbi_1.52.0    
##  [64] munsell_0.5.0            tools_4.0.3              cachem_1.0.3            
##  [67] generics_0.1.0           RSQLite_2.2.3            evaluate_0.14           
##  [70] stringr_1.4.0            fastmap_1.1.0            yaml_2.2.1              
##  [73] ragg_0.4.1               knitr_1.31               bit64_4.0.5             
##  [76] fs_1.5.0                 purrr_0.3.4              RBGL_1.66.0             
##  [79] nlme_3.1-152             formatR_1.7              xml2_1.3.2              
##  [82] biomaRt_2.46.3           compiler_4.0.3           rstudioapi_0.13         
##  [85] curl_4.3                 png_0.1-7                geneplotter_1.68.0      
##  [88] tibble_3.0.6             stringi_1.5.3            highr_0.8               
##  [91] GenomicFeatures_1.42.1   desc_1.2.0               lattice_0.20-41         
##  [94] Matrix_1.3-2             vctrs_0.3.6              pillar_1.4.7            
##  [97] lifecycle_1.0.0.9000     BiocManager_1.30.10      data.table_1.13.6       
## [100] bitops_1.0-6             rtracklayer_1.50.0       R6_2.5.0                
## [103] latticeExtra_0.6-29      hwriter_1.3.2            bookdown_0.21           
## [106] codetools_0.2-18         assertthat_0.2.1         openssl_1.4.3           
## [109] Category_2.56.0          rprojroot_2.0.2          rjson_0.2.20            
## [112] withr_2.4.1              GenomeInfoDbData_1.2.4   hms_1.0.0               
## [115] grid_4.0.3               DOT_0.1                  rmarkdown_2.6</code></pre>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This project is funded by NSF award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152">ABI-1661152</a>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Howard2013-fq" class="csl-entry">
Howard, Brian E, Qiwen Hu, Ahmet Can Babaoglu, Manan Chandra, Monica Borghi, Xiaoping Tan, Luyan He, et al. 2013. <span>“High-Throughput <span>RNA</span> Sequencing of Pseudomonas-Infected Arabidopsis Reveals Hidden Transcriptome Complexity and Novel Splice Variants.”</span> <em>PLoS One</em> 8 (10): e74183. <a href="https://doi.org/10.1371/journal.pone.0074183">https://doi.org/10.1371/journal.pone.0074183</a>.
</div>
<div id="ref-Kim2015-ve" class="csl-entry">
Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. <span>“<span>HISAT</span>: A Fast Spliced Aligner with Low Memory Requirements.”</span> <em>Nat. Methods</em> 12 (4): 357–60.
</div>
<div id="ref-Kim2013-vg" class="csl-entry">
Kim, Daehwan, Geo Pertea, Cole Trapnell, Harold Pimentel, Ryan Kelley, and Steven L Salzberg. 2013. <span>“<span>TopHat2</span>: Accurate Alignment of Transcriptomes in the Presence of Insertions, Deletions and Gene Fusions.”</span> <em>Genome Biol.</em> 14 (4): R36. <a href="https://doi.org/10.1186/gb-2013-14-4-r36">https://doi.org/10.1186/gb-2013-14-4-r36</a>.
</div>
<div id="ref-Langmead2012-bs" class="csl-entry">
Langmead, Ben, and Steven L Salzberg. 2012. <span>“Fast Gapped-Read Alignment with Bowtie 2.”</span> <em>Nat. Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923">https://doi.org/10.1038/nmeth.1923</a>.
</div>
<div id="ref-Lawrence2013-kt" class="csl-entry">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-Li2009-oc" class="csl-entry">
Li, H, and R Durbin. 2009. <span>“Fast and Accurate Short Read Alignment with <span>Burrows-Wheeler</span> Transform.”</span> <em>Bioinformatics</em> 25 (14): 1754–60. <a href="https://doi.org/10.1093/bioinformatics/btp324">https://doi.org/10.1093/bioinformatics/btp324</a>.
</div>
<div id="ref-Li2013-oy" class="csl-entry">
Li, Heng. 2013. <span>“Aligning Sequence Reads, Clone Sequences and Assembly Contigs with <span>BWA-MEM</span>.”</span> <em>arXiv [q-Bio.GN]</em>, March. <a href="http://arxiv.org/abs/1303.3997">http://arxiv.org/abs/1303.3997</a>.
</div>
<div id="ref-Liao2013-bn" class="csl-entry">
Liao, Yang, Gordon K Smyth, and Wei Shi. 2013. <span>“The Subread Aligner: Fast, Accurate and Scalable Read Mapping by Seed-and-Vote.”</span> <em>Nucleic Acids Res.</em> 41 (10): e108. <a href="https://doi.org/10.1093/nar/gkt214">https://doi.org/10.1093/nar/gkt214</a>.
</div>
<div id="ref-Love2014-sh" class="csl-entry">
Love, Michael, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Genome Biol.</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Rsamtools" class="csl-entry">
Morgan, Martin, Hervé Pagès, Valerie Obenchain, and Nathaniel Hayden. 2019. <em>Rsamtools: Binary Alignment (BAM), FASTA, Variant Call (BCF), and Tabix File Import</em>. <a href="http://bioconductor.org/packages/Rsamtools">http://bioconductor.org/packages/Rsamtools</a>.
</div>
<div id="ref-Robinson2010-uk" class="csl-entry">
Robinson, M D, D J McCarthy, and G K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-Wu2010-iq" class="csl-entry">
Wu, T D, and S Nacu. 2010. <span>“Fast and <span class="nocase">SNP-tolerant</span> Detection of Complex Variants and Splicing in Short Reads.”</span> <em>Bioinformatics</em> 26 (7): 873–81. <a href="https://doi.org/10.1093/bioinformatics/btq057">https://doi.org/10.1093/bioinformatics/btq057</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
