<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../../favicons/favicon.ico><link rel=apple-touch-icon href=../../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../../favicons/android-192x192.png sizes=192x192><title>Workflow Templates | sysPipe</title><meta property="og:title" content="Workflow Templates"><meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Workflow templates The intended way of running systemPipeR workflows is via *.Rmd files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line. This way comprehensive and reproducible analysis reports can be generated in PDF or HTML format in a fully automated manner by making use of the highly functional reporting utilities available for R."><meta property="og:type" content="article"><meta property="og:url" content="/spr/systempiper/templates/"><meta property="article:modified_time" content="2021-03-04T13:17:32-08:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="Workflow Templates"><meta itemprop=description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Workflow templates The intended way of running systemPipeR workflows is via *.Rmd files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line. This way comprehensive and reproducible analysis reports can be generated in PDF or HTML format in a fully automated manner by making use of the highly functional reporting utilities available for R."><meta itemprop=dateModified content="2021-03-04T13:17:32-08:00"><meta itemprop=wordCount content="927"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="Workflow Templates"><meta name=twitter:description content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Workflow templates The intended way of running systemPipeR workflows is via *.Rmd files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line. This way comprehensive and reproducible analysis reports can be generated in PDF or HTML format in a fully automated manner by making use of the highly functional reporting utilities available for R."><link rel=preload href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css as=style><link href=../../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><link rel=stylesheet href=../../../css/prism.css><script src=../../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../../css/driver0.9.8.min.css><link rel=stylesheet href=../../../css/site.css><script src=../../../js/site.js></script></head><body class="td-page line-numbers"><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../../><span class=font-weight-bold>sysPipe</span></a>
<row class=header-switch><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=sidebarSwitch checked>
<label class=custom-control-label for=sidebarSwitch>sidebar</label></div><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=tocSwitch checked>
<label class=custom-control-label for=tocSwitch>toc</label></div></row><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../about/overview/ style=padding-left:1rem;font-weight:calc(3/3*500)>Overview</a>
<a class=dropdown-item href=../../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr/systempiper/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../../spr/systempiper/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../../spr/systempiper/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../../spr/systempiper/spr_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../../spr/systempiper/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../../spr/systempiper/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../../spr/systempiper/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../../spr/systempiper/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../../spr/systempiperdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../../spr/systempiperdata/vignette/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeRNAseq/articles/systemPipeRNAseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>RNA-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeChIPseq/articles/systemPipeChIPseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>ChIP-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeRIBOseq/articles/systemPipeRIBOseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>RIBO-Seq</a>
<a class=dropdown-item href=https://systempipe.org/systemPipeVARseq/articles/systemPipeVARseq.html style=padding-left:1rem;font-weight:calc(3/3*500)>VAR-Seq</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../training/ id=navbarDropdown role=button>Training</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../training/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../../training/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002 Search this site…" aria-label="Search this site…" autocomplete=off></div></nav></header><div class="container-fluid td-outer"><div class=td-main><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none"><div id=td-sidebar-menu class=td-sidebar__inner><form class="td-sidebar__search d-flex align-items-center"><input type=search class="form-control td-search-input" placeholder="&#xf002 Search this site…" aria-label="Search this site…" autocomplete=off>
<button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type=button data-toggle=collapse data-target=#td-section-nav aria-controls=td-docs-nav aria-expanded=false aria-label="Toggle section navigation"></button></form><nav class="collapse td-sidebar-nav" id=td-section-nav><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/ class="align-left pl-0 pr-2 td-sidebar-link td-sidebar-link__section">Documentation</a></li><ul><li class="collapse show" id=spr><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/systempiper/ class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">systemPipeR</a></li><ul><li class="collapse show" id=sprsystempiper><a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperintroduction href=../../../spr/systempiper/introduction/>Introduction</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipergettingstarted href=../../../spr/systempiper/gettingstarted/>Getting Started</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperspr_run href=../../../spr/systempiper/spr_run/>How to run a Workflow</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipersteps href=../../../spr/systempiper/steps/>Workflow steps overview</a>
<a class="td-sidebar-link td-sidebar-link__page active" id=m-sprsystempipertemplates href=../../../spr/systempiper/templates/>Workflow Templates</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperspr_funcs href=../../../spr/systempiper/spr_funcs/>systemPipeR Help Manual</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempipersprinstall href=../../../spr/systempiper/sprinstall/>SPR detailed installation instructions</a></li></ul></ul><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../../spr/systempiperdata/ class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">systemPipeRdata</a></li><ul><li class=collapse id=sprsystempiperdata><a class="td-sidebar-link td-sidebar-link__page" id=m-sprsystempiperdatavignette href=../../../spr/systempiperdata/vignette/>SRPdata Vignette</a></li></ul></ul></li></ul></ul></nav></div></div><div class="d-none d-xl-block col-xl-2 td-toc d-print-none"><div class="td-page-meta ml-2 pb-1 pt-2 mb-0"><a href=https://github.com/systemPipeR/systemPipeR.github.io/blob/main/content/en/spr/systempiper/templates.markdown target=_blank><i class="fa fa-code fa-fw"></i>View source code</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/edit/main/content/en/spr/systempiper/templates.markdown target=_blank><i class="fa fa-edit fa-fw"></i>Edit this page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/new/main/content/en/spr/systempiper/templates.markdown?filename=change-me.md&value=---%0Atitle%3A+%22Long+Page+Title%22%0AlinkTitle%3A+%22Short+Nav+Title%22%0Aweight%3A+100%0Adescription%3A+%3E-%0A+++++Page+description+for+heading+and+indexes.%0A---%0A%0A%23%23+Heading%0A%0AEdit+this+template+to+create+your+new+page.%0A%0A%2A+Give+it+a+good+name%2C+ending+in+%60.md%60+-+e.g.+%60getting-started.md%60%0A%2A+Edit+the+%22front+matter%22+section+at+the+top+of+the+page+%28weight+controls+how+its+ordered+amongst+other+pages+in+the+same+directory%3B+lowest+number+first%29.%0A%2A+Add+a+good+commit+message+at+the+bottom+of+the+page+%28%3C80+characters%3B+use+the+extended+description+field+for+more+detail%29.%0A%2A+Create+a+new+branch+so+you+can+preview+your+new+file+and+request+a+review+via+Pull+Request.%0A" target=_blank><i class="fa fa-edit fa-fw"></i>Create child page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/issues/new?title=Workflow%20Templates" target=_blank><i class="fab fa-github fa-fw"></i>Create documentation issue</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/issues/new target=_blank><i class="fas fa-tasks fa-fw"></i>Create project issue</a></div><nav id=TableOfContents><ul><li><a href=#workflow-templates>Workflow templates</a><ul><li><a href=#rna-seq-sample>RNA-Seq sample</a><ul><li><a href=#run-workflow>Run workflow</a></li></ul></li><li><a href=#chip-seq-sample>ChIP-Seq sample</a><ul><li><a href=#run-workflow-1>Run workflow</a></li></ul></li><li><a href=#var-seq-sample>VAR-Seq sample</a><ul><li><a href=#var-seq-workflow-for-the-single-machine>VAR-Seq workflow for the single machine</a></li><li><a href=#run-workflow-2>Run workflow</a></li><li><a href=#var-seq-workflow-for-computer-cluster>VAR-Seq workflow for computer cluster</a></li></ul></li><li><a href=#ribo-seq-sample>Ribo-Seq sample</a><ul><li><a href=#run-workflow-3>Run workflow</a></li></ul></li></ul></li><li><a href=#version-information>Version information</a></li><li><a href=#references>References</a></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 pl-md-5" role=main><nav aria-label=breadcrumb class="d-none d-md-block d-print-none"><ol class="breadcrumb spb-1"><li class=breadcrumb-item><a href=../../../spr/>Documentation</a></li><li class=breadcrumb-item><a href=../../../spr/systempiper/>systemPipeR</a></li><li class="breadcrumb-item active" aria-current=page><a href=../../../spr/systempiper/templates/>Workflow Templates</a></li></ol></nav><div class=td-content><h1>Workflow Templates</h1><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){document.querySelector("h1").className="title";});</script><script type=text/javascript>document.addEventListener("DOMContentLoaded",function(){var links=document.links;for(var i=0,linksLength=links.length;i<linksLength;i++)
if(links[i].hostname!=window.location.hostname)
links[i].target='_blank';});</script><h1 id=workflow-templates>Workflow templates</h1><p>The intended way of running <em><code>systemPipeR</code></em> workflows is via <em><code>*.Rmd</code></em> files, which
can be executed either line-wise in interactive mode or with a single command from
R or the command-line. This way comprehensive and reproducible analysis reports
can be generated in PDF or HTML format in a fully automated manner by making use
of the highly functional reporting utilities available for R.
The following shows how to execute a workflow (<em>e.g.</em>, systemPipeRNAseq.Rmd)
from the command-line.</p><pre><code class=language-bash>Rscript -e &quot;rmarkdown::render('systemPipeRNAseq.Rmd')&quot;
</code></pre><p>Templates for setting up custom project reports are provided as *<code>*.Rmd</code>* files by the helper package *<code>systemPipeRdata</code>* and in the vignettes subdirectory of <em><code>systemPipeR</code><em>. The corresponding HTML of these report templates are available here: </em><code>systemPipeRNAseq</code></em><a href=http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRNAseq.html></a>, <a href=http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.html><em><code>systemPipeRIBOseq</code></em></a>, <a href=http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeChIPseq.html><em><code>systemPipeChIPseq</code></em></a> and <a href=http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeVARseq.html><em><code>systemPipeVARseq</code></em></a>. To work with <em><code>*.Rnw</code>* or <em><code>*.Rmd</code>* files efficiently, basic knowledge of </em><code>Sweave</code></em><a href=https://www.stat.uni-muenchen.de/~leisch/Sweave/></a> or <a href=http://yihui.name/knitr/><em><code>knitr</code></em></a> and <a href=http://www.latex-project.org/><em><code>Latex</code></em></a> or <a href=http://rmarkdown.rstudio.com/><em><code>R Markdown v2</code></em></a> is required.</p><h2 id=rna-seq-sample>RNA-Seq sample</h2><p>Load the RNA-Seq sample workflow into your current working directory.</p><pre><code class=language-r>library(systemPipeRdata)
genWorkenvir(workflow = &quot;rnaseq&quot;)
setwd(&quot;rnaseq&quot;)
</code></pre><h3 id=run-workflow>Run workflow</h3><p>Next, run the chosen sample workflow <em><code>systemPipeRNAseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq.pdf?raw=true">PDF</a>, <a href=https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq.Rmd>Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>rnaseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p><p>The workflow includes following steps:</p><ol><li>Read preprocessing<ul><li>Quality filtering (trimming)</li><li>FASTQ quality report</li></ul></li><li>Alignments: <em><code>Tophat2</code></em> (or any other RNA-Seq aligner)</li><li>Alignment stats</li><li>Read counting</li><li>Sample-wise correlation analysis</li><li>Analysis of differentially expressed genes (DEGs)</li><li>GO term enrichment analysis</li><li>Gene-wise clustering</li></ol><h2 id=chip-seq-sample>ChIP-Seq sample</h2><p>Load the ChIP-Seq sample workflow into your current working directory.</p><pre><code class=language-r>library(systemPipeRdata)
genWorkenvir(workflow = &quot;chipseq&quot;)
setwd(&quot;chipseq&quot;)
</code></pre><h3 id=run-workflow-1>Run workflow</h3><p>Next, run the chosen sample workflow <em><code>systemPipeChIPseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq.pdf?raw=true">PDF</a>, <a href=https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq.Rmd>Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>chipseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p><p>The workflow includes the following steps:</p><ol><li>Read preprocessing<ul><li>Quality filtering (trimming)</li><li>FASTQ quality report</li></ul></li><li>Alignments: <em><code>Bowtie2</code></em> or <em><code>rsubread</code></em></li><li>Alignment stats</li><li>Peak calling: <em><code>MACS2</code></em>, <em><code>BayesPeak</code></em></li><li>Peak annotation with genomic context</li><li>Differential binding analysis</li><li>GO term enrichment analysis</li><li>Motif analysis</li></ol><h2 id=var-seq-sample>VAR-Seq sample</h2><h3 id=var-seq-workflow-for-the-single-machine>VAR-Seq workflow for the single machine</h3><p>Load the VAR-Seq sample workflow into your current working directory.</p><pre><code class=language-r>library(systemPipeRdata)
genWorkenvir(workflow = &quot;varseq&quot;)
setwd(&quot;varseq&quot;)
</code></pre><h3 id=run-workflow-2>Run workflow</h3><p>Next, run the chosen sample workflow <em><code>systemPipeVARseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.pdf?raw=true">PDF</a>, <a href=https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.Rmd>Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>varseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p><p>The workflow includes following steps:</p><ol><li>Read preprocessing<ul><li>Quality filtering (trimming)</li><li>FASTQ quality report</li></ul></li><li>Alignments: <em><code>gsnap</code></em>, <em><code>bwa</code></em></li><li>Variant calling: <em><code>VariantTools</code></em>, <em><code>GATK</code></em>, <em><code>BCFtools</code></em></li><li>Variant filtering: <em><code>VariantTools</code></em> and <em><code>VariantAnnotation</code></em></li><li>Variant annotation: <em><code>VariantAnnotation</code></em></li><li>Combine results from many samples</li><li>Summary statistics of samples</li></ol><h3 id=var-seq-workflow-for-computer-cluster>VAR-Seq workflow for computer cluster</h3><p>The workflow template provided for this step is called <em><code>systemPipeVARseq.Rmd</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.pdf?raw=true">PDF</a>, <a href=https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.Rmd>Rmd</a>).
It runs the above VAR-Seq workflow in parallel on multiple compute nodes of an HPC system using Slurm as the scheduler.</p><h2 id=ribo-seq-sample>Ribo-Seq sample</h2><p>Load the Ribo-Seq sample workflow into your current working directory.</p><pre><code class=language-r>library(systemPipeRdata)
genWorkenvir(workflow = &quot;riboseq&quot;)
setwd(&quot;riboseq&quot;)
</code></pre><h3 id=run-workflow-3>Run workflow</h3><p>Next, run the chosen sample workflow <em><code>systemPipeRIBOseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/riboseq/systemPipeRIBOseq.pdf?raw=true">PDF</a>, <a href=https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/ribseq/systemPipeRIBOseq.Rmd>Rmd</a>) by executing from the command-line <em><code>make -B</code></em> within the <em><code>ribseq</code></em> directory. Alternatively, one can run the code from the provided <em><code>*.Rmd</code></em> template file from within R interactively.</p><p>The workflow includes following steps:</p><ol><li>Read preprocessing<ul><li>Adaptor trimming and quality filtering</li><li>FASTQ quality report</li></ul></li><li>Alignments: <em><code>Tophat2</code></em> (or any other RNA-Seq aligner)</li><li>Alignment stats</li><li>Compute read distribution across genomic features</li><li>Adding custom features to the workflow (e.g. uORFs)</li><li>Genomic read coverage along with transcripts</li><li>Read counting</li><li>Sample-wise correlation analysis</li><li>Analysis of differentially expressed genes (DEGs)</li><li>GO term enrichment analysis</li><li>Gene-wise clustering</li><li>Differential ribosome binding (translational efficiency)</li></ol><h1 id=version-information>Version information</h1><p><strong>Note:</strong> the most recent version of this tutorial can be found <a href=http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html>here</a>.</p><pre><code class=language-r>sessionInfo()
</code></pre><pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-4.0.3/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] magrittr_2.0.1              batchtools_0.9.15          
##  [3] ape_5.4-1                   ggplot2_3.3.3              
##  [5] systemPipeR_1.25.6          ShortRead_1.48.0           
##  [7] GenomicAlignments_1.26.0    SummarizedExperiment_1.20.0
##  [9] Biobase_2.50.0              MatrixGenerics_1.2.1       
## [11] matrixStats_0.58.0          BiocParallel_1.24.1        
## [13] Rsamtools_2.6.0             Biostrings_2.58.0          
## [15] XVector_0.30.0              GenomicRanges_1.42.0       
## [17] GenomeInfoDb_1.26.2         IRanges_2.24.1             
## [19] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [21] BiocStyle_2.18.1           
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-0         rjson_0.2.20             hwriter_1.3.2           
##   [4] ellipsis_0.3.1           bit64_4.0.5              AnnotationDbi_1.52.0    
##   [7] xml2_1.3.2               codetools_0.2-18         splines_4.0.3           
##  [10] cachem_1.0.3             knitr_1.31               jsonlite_1.7.2          
##  [13] annotate_1.68.0          GO.db_3.12.1             dbplyr_2.1.0            
##  [16] png_0.1-7                pheatmap_1.0.12          graph_1.68.0            
##  [19] BiocManager_1.30.10      compiler_4.0.3           httr_1.4.2              
##  [22] GOstats_2.56.0           backports_1.2.1          assertthat_0.2.1        
##  [25] Matrix_1.3-2             fastmap_1.1.0            limma_3.46.0            
##  [28] formatR_1.7              htmltools_0.5.1.1        prettyunits_1.1.1       
##  [31] tools_4.0.3              gtable_0.3.0             glue_1.4.2              
##  [34] GenomeInfoDbData_1.2.4   Category_2.56.0          dplyr_1.0.4             
##  [37] rsvg_2.1                 rappdirs_0.3.3           V8_3.4.0                
##  [40] Rcpp_1.0.6               vctrs_0.3.6              nlme_3.1-152            
##  [43] blogdown_1.1.7           rtracklayer_1.50.0       xfun_0.21               
##  [46] stringr_1.4.0            lifecycle_1.0.0.9000     XML_3.99-0.5            
##  [49] edgeR_3.32.1             zlibbioc_1.36.0          scales_1.1.1            
##  [52] BSgenome_1.58.0          VariantAnnotation_1.36.0 hms_1.0.0               
##  [55] RBGL_1.66.0              RColorBrewer_1.1-2       yaml_2.2.1              
##  [58] curl_4.3                 memoise_2.0.0            biomaRt_2.46.3          
##  [61] latticeExtra_0.6-29      stringi_1.5.3            RSQLite_2.2.3           
##  [64] genefilter_1.72.1        checkmate_2.0.0          GenomicFeatures_1.42.1  
##  [67] DOT_0.1                  rlang_0.4.10             pkgconfig_2.0.3         
##  [70] bitops_1.0-6             evaluate_0.14            lattice_0.20-41         
##  [73] purrr_0.3.4              bit_4.0.4                tidyselect_1.1.0        
##  [76] GSEABase_1.52.1          AnnotationForge_1.32.0   bookdown_0.21           
##  [79] R6_2.5.0                 generics_0.1.0           base64url_1.4           
##  [82] DelayedArray_0.16.1      DBI_1.1.1                withr_2.4.1             
##  [85] pillar_1.4.7             survival_3.2-7           RCurl_1.98-1.2          
##  [88] tibble_3.0.6             crayon_1.4.1             BiocFileCache_1.14.0    
##  [91] rmarkdown_2.6            jpeg_0.1-8.1             progress_1.2.2          
##  [94] locfit_1.5-9.4           grid_4.0.3               data.table_1.13.6       
##  [97] blob_1.2.1               Rgraphviz_2.34.0         digest_0.6.27           
## [100] xtable_1.8-4             brew_1.0-6               openssl_1.4.3           
## [103] munsell_0.5.0            askpass_1.1
</code></pre><h1 id=references>References</h1><div class="text-muted mt-5 pt-3 border-top">Last modified 2021-03-04: <a href=https://github.com/systemPipeR/systemPipeR.github.io/commit/cee9f4108731fd68bd9162da8d8c2cb81b878b21>Merge pull request #3 from systemPipeR/dependabot/npm_and_yarn/postcss-8.2.7 (cee9f41)</a></div></div></main></div></div><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2021 The systemPipeR Project All Rights Reserved</small><p class=mt-2><a href=../../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src=../../../js/mathjax3.1.2.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../../js/main.min.b5fc1b29d2465835844254bd4d804738eaf24c0cec53009b4c6899989be55a47.js integrity="sha256-tfwbKdJGWDWEQlS9TYBHOOryTAzsUwCbTGiZmJvlWkc=" crossorigin=anonymous></script><script src=../../../js/prism.js></script></body></html>