<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot read distribution across genomic features — featuretypeCounts • systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot read distribution across genomic features — featuretypeCounts" />
<meta property="og:description" content="Counts how many reads in short read alignment files (BAM format) overlap with
entire annotation categories. This utility is useful for analyzing the
distribution of the read mappings across feature types, e.g. coding versus non-coding
genes. By default the read counts are reported for the sense and antisense
strand of each feature type separately. To minimize memory consumption, the
BAM files are processed in a stream using utilities from the Rsamtools
and GenomicAlignment packages.  The counts can be reported for each
read length separately or as a single value for reads of any length.
Subsequently, the counting results can be plotted with the associated
plotfeaturetypeCounts function." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot read distribution across genomic features</h1>
    
    <div class="hidden name"><code>featuretypeCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Counts how many reads in short read alignment files (BAM format) overlap with
entire annotation categories. This utility is useful for analyzing the
distribution of the read mappings across feature types, e.g. coding versus non-coding
genes. By default the read counts are reported for the sense and antisense
strand of each feature type separately. To minimize memory consumption, the
BAM files are processed in a stream using utilities from the <code>Rsamtools</code>
and <code>GenomicAlignment</code> packages.  The counts can be reported for each
read length separately or as a single value for reads of any length.
Subsequently, the counting results can be plotted with the associated
<code>plotfeaturetypeCounts</code> function.</p>
    </div>

    <pre class="usage"><span class='fu'>featuretypeCounts</span><span class='op'>(</span><span class='va'>bfl</span>, <span class='va'>grl</span>, singleEnd <span class='op'>=</span> <span class='cn'>TRUE</span>, readlength <span class='op'>=</span> <span class='cn'>NULL</span>, type <span class='op'>=</span> <span class='st'>"data.frame"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bfl</th>
      <td><p><code>BamFileList</code> object containing the paths to the target BAM files stored
on disk. Note, memory-efficient processing is achieved by streaming through
BAM files rather than reading entire files into memory at once. The maximum
number of alignments to process in each iteration is determined by the
<code>yieldSize</code> value passed on to the <code>BamFileList</code> function. For
details see <code>?BamFileList</code>.</p></td>
    </tr>
    <tr>
      <th>grl</th>
      <td><p><code>GRangesList</code> object containing in each list component the range set of
a feature type. Typically, this object is generated with the <code>genFeatures</code>
function. For details see the example section below and <code><a href='genFeatures.html'>?genFeatures</a></code>.</p></td>
    </tr>
    <tr>
      <th>singleEnd</th>
      <td><p>Specifies whether the targets BAM files contain alignments for single-end (SE) or paired-end
read data. <code>TRUE</code> is for SE and <code>FALSE</code> for PE data.</p></td>
    </tr>
    <tr>
      <th>readlength</th>
      <td><p>Integer vector specifying the read length values for which to report counts
separately. If <code>readlength=NULL</code> the length of the reads will be ignored
resulting in a single value for each feature type and strand. Note, for PE data
the two reads in a pair may differ in length. In those cases the length of the
two reads is averaged and then assigned to the corresponding length category
after rounding the mean length to the closest integer. This is not an ideal
solution but a reasonable compromise for the purpose of the summary statistics
generated by <code>featuretypeCounts</code>.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>Determines whether the results are returned as <code>data.frame</code> (<code>type="data.frame"</code>) 
or as <code>list</code> (<code>type="list"</code>). Each list component contains the counting
results for one BAM file and is named after the corresponding sample. The 
<code>data.frame</code> result contains this sample assignment information in a separate 
column.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The results are returned as <code>data.frame</code> or <code>list</code> of <code>data.frames</code>.
For details see above under <code>types</code> argument. The result <code>data.frames</code> contain
the following columns in the given order:</p>
<dt>SampleName</dt><dd><p>Sample names obtained from <code>BamFileList</code> object.</p></dd>
<dt>Strand</dt><dd><p>Sense or antisense strand of read mappings.</p></dd>  
<dt>Featuretype</dt><dd><p>Name of feature type provided by <code>GRangesList</code> object. Note, the total number
of aligned reads is reported under the special feature type 'N_total_aligned'. This value 
is useful for scaling/normalization purposes in plots, e.g. counts per million reads.</p></dd>
<dt>Featuretypelength</dt><dd><p>Total genomic length of each reduced feature type in bases. This value is 
useful to normalize the read counts by genomic length units, e.g. in plots.</p></dd>
<dt>Subsequent columns</dt><dd><p>Counts for reads of any length or for individual read lengths.</p></dd>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thomas Girke</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>plotfeaturetypeCounts</code>, <code>genFeatures</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Construct SYSargs object from param and targets files </span>
<span class='va'>param</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"tophat.param"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>targets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"targets.txt"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>args</span> <span class='op'>&lt;-</span> <span class='fu'><a href='systemArgs.html'>systemArgs</a></span><span class='op'>(</span>sysma<span class='op'>=</span><span class='va'>param</span>, mytargets<span class='op'>=</span><span class='va'>targets</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: path[1]="./data/SRR446027_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[2]="./data/SRR446028_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[3]="./data/SRR446029_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[4]="./data/SRR446030_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[5]="./data/SRR446031_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[6]="./data/SRR446032_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[7]="./data/SRR446033_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[8]="./data/SRR446034_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[9]="./data/SRR446035_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[10]="./data/SRR446036_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[11]="./data/SRR446037_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[12]="./data/SRR446038_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[13]="./data/SRR446039_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[14]="./data/SRR446040_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[15]="./data/SRR446041_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[16]="./data/SRR446042_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[17]="./data/SRR446043_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[18]="./data/SRR446044_1.fastq.gz": No such file or directory</span></div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'>## Features from sample data of systemPipeRdata package</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicFeatures'>GenomicFeatures</a></span><span class='op'>)</span>
<span class='va'>file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/annotation"</span>, <span class='st'>"tair10.gff"</span>, package<span class='op'>=</span><span class='st'>"systemPipeRdata"</span><span class='op'>)</span>
<span class='va'>txdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span>file<span class='op'>=</span><span class='va'>file</span>, format<span class='op'>=</span><span class='st'>"gff3"</span>, organism<span class='op'>=</span><span class='st'>"Arabidopsis"</span><span class='op'>)</span>
<span class='va'>feat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='genFeatures.html'>genFeatures</a></span><span class='op'>(</span><span class='va'>txdb</span>, featuretype<span class='op'>=</span><span class='st'>"all"</span>, reduce_ranges<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>1000</span>, downstream<span class='op'>=</span><span class='fl'>0</span>, verbose<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'>## Generate and plot feature counts for specific read lengths</span>
<span class='va'>fc</span> <span class='op'>&lt;-</span> <span class='fu'>featuretypeCounts</span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span>, yieldSize<span class='op'>=</span><span class='fl'>50000</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>feat</span>, singleEnd<span class='op'>=</span><span class='cn'>TRUE</span>, readlength<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>74</span><span class='op'>:</span><span class='fl'>76</span>,<span class='fl'>99</span><span class='op'>:</span><span class='fl'>102</span><span class='op'>)</span>, type<span class='op'>=</span><span class='st'>"data.frame"</span><span class='op'>)</span>
<span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='plotfeaturetypeCounts.html'>plotfeaturetypeCounts</a></span><span class='op'>(</span>x<span class='op'>=</span><span class='va'>fc</span>, graphicsfile<span class='op'>=</span><span class='st'>"featureCounts.pdf"</span>, graphicsformat<span class='op'>=</span><span class='st'>"pdf"</span>, scales<span class='op'>=</span><span class='st'>"fixed"</span>, anyreadlength<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>

<span class='co'>## Generate and plot feature counts for any read length  </span>
<span class='va'>fc2</span> <span class='op'>&lt;-</span> <span class='fu'>featuretypeCounts</span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span>, yieldSize<span class='op'>=</span><span class='fl'>50000</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>feat</span>, singleEnd<span class='op'>=</span><span class='cn'>TRUE</span>, readlength<span class='op'>=</span><span class='cn'>NULL</span>, type<span class='op'>=</span><span class='st'>"data.frame"</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='plotfeaturetypeCounts.html'>plotfeaturetypeCounts</a></span><span class='op'>(</span>x<span class='op'>=</span><span class='va'>featureCounts2</span>, graphicsfile<span class='op'>=</span><span class='st'>"featureCounts2.pdf"</span>, graphicsformat<span class='op'>=</span><span class='st'>"pdf"</span>, scales<span class='op'>=</span><span class='st'>"fixed"</span>, anyreadlength<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


