<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate feature ranges from TxDb — genFeatures • systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generate feature ranges from TxDb — genFeatures" />
<meta property="og:description" content="Function to generate a variety of feature types from TxDb objects using 
utilities provided by the GenomicFeatures package. The feature types are 
organized per gene and can be returned on that level in their non-reduced or 
reduced form.
Currently, supported features include intergenic, promoter,
intron, exon, cds, 5'/3'UTR and different
transcript types. The latter contains as many transcript types as
available in the tx_type column when extracting transcripts from
TxDb objects as follows: 
transcripts(txdb, c(&quot;tx_name&quot;, &quot;gene_id&quot;, &quot;tx_type&quot;))" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate feature ranges from TxDb</h1>
    
    <div class="hidden name"><code>genFeatures.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function to generate a variety of feature types from <code>TxDb</code> objects using 
utilities provided by the <code>GenomicFeatures</code> package. The feature types are 
organized per gene and can be returned on that level in their non-reduced or 
reduced form.</p>
<p>Currently, supported features include <code>intergenic</code>, <code>promoter</code>,
<code>intron</code>, <code>exon</code>, <code>cds</code>, <code>5'/3'UTR</code> and different
<code>transcript</code> types. The latter contains as many transcript types as
available in the <code>tx_type</code> column when extracting transcripts from
<code>TxDb</code> objects as follows: 
<code><a href='https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html'>transcripts(txdb, c("tx_name", "gene_id", "tx_type"))</a></code></p>
    </div>

    <pre class="usage"><span class='fu'>genFeatures</span><span class='op'>(</span><span class='va'>txdb</span>, featuretype <span class='op'>=</span> <span class='st'>"all"</span>, <span class='va'>reduce_ranges</span>, upstream <span class='op'>=</span> <span class='fl'>1000</span>, downstream <span class='op'>=</span> <span class='fl'>0</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>txdb</th>
      <td><p><code>TxDb</code> object</p></td>
    </tr>
    <tr>
      <th>featuretype</th>
      <td><p>Feature types can be specified by assigning a <code>character</code> vector containing 
any of the following: c("tx_type", "promoter", "intron", "exon", "cds", "fiveUTR", "threeUTR", "intergenic").
The default <code>all</code> is a shorthand to select all supported features.</p></td>
    </tr>
    <tr>
      <th>reduce_ranges</th>
      <td><p>If set to <code>TRUE</code> the feature ranges will be reduced on the gene level. As a result 
overlapping feature components of the same type and from the same gene will be
merged to a single range, e.g. two overlapping exons from the same gene are merged 
to one. Intergenic ranges are not affected by this setting. Note, all reduced feature
types are labeled with the suffix '_red'.</p></td>
    </tr>
    <tr>
      <th>upstream</th>
      <td><p>Defines for promoter features the number of bases upstream from the transcription
start site.</p></td>
    </tr>
    <tr>
      <th>downstream</th>
      <td><p>Defines for promoter features the number of bases downstream from the transcription
start site.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p><code>verbose=FALSE</code> turns off all print messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The results are returned as a <code>GRangesList</code> where each component is a
<code>GRanges</code> object containing the range set of each feature type. Intergenic
ranges are assigned unique identifiers and recorded in the <code>featuretype_id</code> 
column of the metadata block. For this the ids of their adjacent genes are concatenated with two 
underscores as separator. If the adjacent genes overlap with other genes then 
their identifiers are included in the id string as well and separated by a single
underscore.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thomas Girke</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>transcripts</code> and associated <code>TxDb</code> accessor functions from 
the <code>GenomicFeatures</code> package.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Sample from GenomicFeatures package</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicFeatures'>GenomicFeatures</a></span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: AnnotationDbi</span></div><div class='input'><span class='va'>gffFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"GFF3_files"</span>, <span class='st'>"a.gff3"</span>, package<span class='op'>=</span><span class='st'>"GenomicFeatures"</span><span class='op'>)</span>
<span class='va'>txdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span>file<span class='op'>=</span><span class='va'>gffFile</span>, format<span class='op'>=</span><span class='st'>"gff3"</span>, organism<span class='op'>=</span><span class='st'>"Solanum lycopersicum"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='warning'>Warning: The "phase" metadata column contains non-NA values for features of type</span>
#&gt; <span class='warning'>  exon. This information was ignored.</span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='input'><span class='va'>feat</span> <span class='op'>&lt;-</span> <span class='fu'>genFeatures</span><span class='op'>(</span><span class='va'>txdb</span>, featuretype<span class='op'>=</span><span class='st'>"all"</span>, reduce_ranges<span class='op'>=</span><span class='cn'>FALSE</span>, upstream<span class='op'>=</span><span class='fl'>1000</span>, downstream<span class='op'>=</span><span class='fl'>0</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Created feature ranges: mRNA 
#&gt; Created feature ranges: promoter 
#&gt; Created feature ranges: intron 
#&gt; Created feature ranges: exon 
#&gt; Created feature ranges: cds 
#&gt; Created feature ranges: fiveUTR 
#&gt; Created feature ranges: threeUTR 
#&gt; Created feature ranges: intergenic </div><div class='input'>
<span class='co'>## List extracted feature types</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>feat</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "mRNA"       "promoter"   "intron"     "exon"       "cds"       
#&gt; [6] "fiveUTR"    "threeUTR"   "intergenic"</div><div class='input'>
<span class='co'>## Obtain feature lists by genes, here for promoter</span>
<span class='fu'><a href='https://rdrr.io/r/base/split.html'>split</a></span><span class='op'>(</span><span class='va'>feat</span><span class='op'>$</span><span class='va'>promoter</span>, <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='fu'>mcols</span><span class='op'>(</span><span class='va'>feat</span><span class='op'>$</span><span class='va'>promoter</span><span class='op'>)</span><span class='op'>$</span><span class='va'>feature_by</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRangesList object of length 488:
#&gt; $Solyc00g005000.2
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;                        seqnames      ranges strand |       feature_by
#&gt;                           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;CharacterList&gt;
#&gt;   Solyc00g005000.2.1 SL2.40ch00 15437-16436      + | Solyc00g005000.2
#&gt;                          featuretype_id featuretype
#&gt;                             &lt;character&gt; &lt;character&gt;
#&gt;   Solyc00g005000.2.1 Solyc00g005000.2.1    promoter
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths
#&gt; 
#&gt; $Solyc00g005020.1
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;                        seqnames      ranges strand |       feature_by
#&gt;                           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;CharacterList&gt;
#&gt;   Solyc00g005020.1.1 SL2.40ch00 67062-68061      + | Solyc00g005020.1
#&gt;                          featuretype_id featuretype
#&gt;                             &lt;character&gt; &lt;character&gt;
#&gt;   Solyc00g005020.1.1 Solyc00g005020.1.1    promoter
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths
#&gt; 
#&gt; $Solyc00g005040.2
#&gt; GRanges object with 1 range and 3 metadata columns:
#&gt;                        seqnames        ranges strand |       feature_by
#&gt;                           &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;CharacterList&gt;
#&gt;   Solyc00g005040.2.1 SL2.40ch00 549920-550919      + | Solyc00g005040.2
#&gt;                          featuretype_id featuretype
#&gt;                             &lt;character&gt; &lt;character&gt;
#&gt;   Solyc00g005040.2.1 Solyc00g005040.2.1    promoter
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths
#&gt; 
#&gt; ...
#&gt; &lt;485 more elements&gt;</div><div class='input'>
<span class='co'>## Return all features in single GRanges object</span>
<span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='va'>feat</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 4886 ranges and 3 metadata columns:
#&gt;                    seqnames            ranges strand |       feature_by
#&gt;                       &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;CharacterList&gt;
#&gt;             mRNA SL2.40ch00       16437-18189      + | Solyc00g005000.2
#&gt;             mRNA SL2.40ch00       68062-68764      + | Solyc00g005020.1
#&gt;             mRNA SL2.40ch00     550920-551576      + | Solyc00g005040.2
#&gt;             mRNA SL2.40ch00   1115784-1117712      + | Solyc00g005100.1
#&gt;             mRNA SL2.40ch00   1204073-1205342      + | Solyc00g005150.1
#&gt;              ...        ...               ...    ... .              ...
#&gt;   intergenic.479 SL2.40ch00 13647836-13649254      * |    INTER00000479
#&gt;   intergenic.480 SL2.40ch00 13650638-13663835      * |    INTER00000480
#&gt;   intergenic.481 SL2.40ch00 13666166-13763879      * |    INTER00000481
#&gt;   intergenic.482 SL2.40ch00 13767947-13768659      * |    INTER00000482
#&gt;   intergenic.483 SL2.40ch00 13769332-13784265      * |    INTER00000483
#&gt;                          featuretype_id featuretype
#&gt;                             &lt;character&gt; &lt;character&gt;
#&gt;             mRNA     Solyc00g005000.2.1        mRNA
#&gt;             mRNA     Solyc00g005020.1.1        mRNA
#&gt;             mRNA     Solyc00g005040.2.1        mRNA
#&gt;             mRNA     Solyc00g005100.1.1        mRNA
#&gt;             mRNA     Solyc00g005150.1.1        mRNA
#&gt;              ...                    ...         ...
#&gt;   intergenic.479 Solyc00g050030.1__So..  intergenic
#&gt;   intergenic.480 Solyc00g050130.1__So..  intergenic
#&gt;   intergenic.481 Solyc00g050430.2__So..  intergenic
#&gt;   intergenic.482 Solyc00g052430.2_Sol..  intergenic
#&gt;   intergenic.483 Solyc00g052540.1__So..  intergenic
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'>## Sample from systemPipeRdata package</span>
<span class='va'>file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/annotation"</span>, <span class='st'>"tair10.gff"</span>, package<span class='op'>=</span><span class='st'>"systemPipeRdata"</span><span class='op'>)</span>
<span class='va'>txdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span>file<span class='op'>=</span><span class='va'>file</span>, format<span class='op'>=</span><span class='st'>"gff3"</span>, organism<span class='op'>=</span><span class='st'>"Arabidopsis"</span><span class='op'>)</span>
<span class='va'>feat</span> <span class='op'>&lt;-</span> <span class='fu'>genFeatures</span><span class='op'>(</span><span class='va'>txdb</span>, featuretype<span class='op'>=</span><span class='st'>"all"</span>, reduce_ranges<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>1000</span>, downstream<span class='op'>=</span><span class='fl'>0</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


