<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>SYSargs2 accessor methods — sysargs2 • systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="SYSargs2 accessor methods — sysargs2" />
<meta property="og:description" content="Methods to access information from SYSargs2 object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SYSargs2 accessor methods</h1>
    
    <div class="hidden name"><code>sysargs2-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Methods to access information from <code>SYSargs2</code> object.</p>
    </div>

    <pre class="usage"><span class='fu'>sysargs2</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>object of class <code>SYSargs2</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>various outputs</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Daniela Cassol and Thomas Girke</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Construct SYSargs2 object from CWl param, CWL input, and targets files </span>
<span class='va'>targets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"targets.txt"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>dir_path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/cwl/hisat2/hisat2-se"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>WF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='loadWorkflow.html'>loadWorkflow</a></span><span class='op'>(</span>targets<span class='op'>=</span><span class='va'>targets</span>, wf_file<span class='op'>=</span><span class='st'>"hisat2-mapping-se.cwl"</span>, 
                  input_file<span class='op'>=</span><span class='st'>"hisat2-mapping-se.yml"</span>, dir_path<span class='op'>=</span><span class='va'>dir_path</span><span class='op'>)</span>
<span class='va'>WF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='renderWF.html'>renderWF</a></span><span class='op'>(</span><span class='va'>WF</span>, inputvars<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>FileName<span class='op'>=</span><span class='st'>"_FASTQ_PATH1_"</span>, SampleName<span class='op'>=</span><span class='st'>"_SampleName_"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>WF</span>
</div><div class='output co'>#&gt; Instance of 'SYSargs2':
#&gt;    Slot names/accessors: 
#&gt;       targets: 18 (M1A...V12B), targetsheader: 4 (lines)
#&gt;       modules: 1
#&gt;       wf: 0, clt: 1, yamlinput: 7 (components)
#&gt;       input: 18, output: 18
#&gt;       cmdlist: 18
#&gt;    WF Steps:
#&gt;       1. hisat2-mapping-se (rendered: TRUE)
#&gt; 
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span>; <span class='fu'><a href='sysargs.html'>modules</a></span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span>; <span class='fu'>targets</span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>; <span class='fu'>cmdlist</span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span>; <span class='fu'>output</span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  [1] "targets"       "targetsheader" "modules"       "wf"           
#&gt;  [5] "clt"           "yamlinput"     "cmdlist"       "input"        
#&gt;  [9] "output"        "cwlfiles"      "inputvars"    </div><div class='output co'>#&gt;        module1 
#&gt; "hisat2/2.1.0" </div><div class='output co'>#&gt; $M1A
#&gt; $M1A$FileName
#&gt; [1] "./data/SRR446027_1.fastq.gz"
#&gt; 
#&gt; $M1A$SampleName
#&gt; [1] "M1A"
#&gt; 
#&gt; $M1A$Factor
#&gt; [1] "M1"
#&gt; 
#&gt; $M1A$SampleLong
#&gt; [1] "Mock.1h.A"
#&gt; 
#&gt; $M1A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $M1A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; </div><div class='output co'>#&gt; $M1A
#&gt; $M1A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $M1B
#&gt; $M1B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M1B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446028_1.fastq.gz --threads 4"
#&gt; 
#&gt; </div><div class='output co'>#&gt; $M1A
#&gt; $M1A$`hisat2-mapping-se`
#&gt; [1] "./results/M1A.sam"
#&gt; 
#&gt; 
#&gt; $M1B
#&gt; $M1B$`hisat2-mapping-se`
#&gt; [1] "./results/M1B.sam"
#&gt; 
#&gt; 
#&gt; $A1A
#&gt; $A1A$`hisat2-mapping-se`
#&gt; [1] "./results/A1A.sam"
#&gt; 
#&gt; 
#&gt; $A1B
#&gt; $A1B$`hisat2-mapping-se`
#&gt; [1] "./results/A1B.sam"
#&gt; 
#&gt; 
#&gt; $V1A
#&gt; $V1A$`hisat2-mapping-se`
#&gt; [1] "./results/V1A.sam"
#&gt; 
#&gt; 
#&gt; $V1B
#&gt; $V1B$`hisat2-mapping-se`
#&gt; [1] "./results/V1B.sam"
#&gt; 
#&gt; 
#&gt; $M6A
#&gt; $M6A$`hisat2-mapping-se`
#&gt; [1] "./results/M6A.sam"
#&gt; 
#&gt; 
#&gt; $M6B
#&gt; $M6B$`hisat2-mapping-se`
#&gt; [1] "./results/M6B.sam"
#&gt; 
#&gt; 
#&gt; $A6A
#&gt; $A6A$`hisat2-mapping-se`
#&gt; [1] "./results/A6A.sam"
#&gt; 
#&gt; 
#&gt; $A6B
#&gt; $A6B$`hisat2-mapping-se`
#&gt; [1] "./results/A6B.sam"
#&gt; 
#&gt; 
#&gt; $V6A
#&gt; $V6A$`hisat2-mapping-se`
#&gt; [1] "./results/V6A.sam"
#&gt; 
#&gt; 
#&gt; $V6B
#&gt; $V6B$`hisat2-mapping-se`
#&gt; [1] "./results/V6B.sam"
#&gt; 
#&gt; 
#&gt; $M12A
#&gt; $M12A$`hisat2-mapping-se`
#&gt; [1] "./results/M12A.sam"
#&gt; 
#&gt; 
#&gt; $M12B
#&gt; $M12B$`hisat2-mapping-se`
#&gt; [1] "./results/M12B.sam"
#&gt; 
#&gt; 
#&gt; $A12A
#&gt; $A12A$`hisat2-mapping-se`
#&gt; [1] "./results/A12A.sam"
#&gt; 
#&gt; 
#&gt; $A12B
#&gt; $A12B$`hisat2-mapping-se`
#&gt; [1] "./results/A12B.sam"
#&gt; 
#&gt; 
#&gt; $V12A
#&gt; $V12A$`hisat2-mapping-se`
#&gt; [1] "./results/V12A.sam"
#&gt; 
#&gt; 
#&gt; $V12B
#&gt; $V12B$`hisat2-mapping-se`
#&gt; [1] "./results/V12B.sam"
#&gt; 
#&gt; </div><div class='input'><span class='fu'>sysargs2</span><span class='op'>(</span><span class='va'>WF</span><span class='op'>)</span>
</div><div class='output co'>#&gt; $targets
#&gt; $targets$M1A
#&gt; $targets$M1A$FileName
#&gt; [1] "./data/SRR446027_1.fastq.gz"
#&gt; 
#&gt; $targets$M1A$SampleName
#&gt; [1] "M1A"
#&gt; 
#&gt; $targets$M1A$Factor
#&gt; [1] "M1"
#&gt; 
#&gt; $targets$M1A$SampleLong
#&gt; [1] "Mock.1h.A"
#&gt; 
#&gt; $targets$M1A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M1A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$M1B
#&gt; $targets$M1B$FileName
#&gt; [1] "./data/SRR446028_1.fastq.gz"
#&gt; 
#&gt; $targets$M1B$SampleName
#&gt; [1] "M1B"
#&gt; 
#&gt; $targets$M1B$Factor
#&gt; [1] "M1"
#&gt; 
#&gt; $targets$M1B$SampleLong
#&gt; [1] "Mock.1h.B"
#&gt; 
#&gt; $targets$M1B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M1B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A1A
#&gt; $targets$A1A$FileName
#&gt; [1] "./data/SRR446029_1.fastq.gz"
#&gt; 
#&gt; $targets$A1A$SampleName
#&gt; [1] "A1A"
#&gt; 
#&gt; $targets$A1A$Factor
#&gt; [1] "A1"
#&gt; 
#&gt; $targets$A1A$SampleLong
#&gt; [1] "Avr.1h.A"
#&gt; 
#&gt; $targets$A1A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A1A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A1B
#&gt; $targets$A1B$FileName
#&gt; [1] "./data/SRR446030_1.fastq.gz"
#&gt; 
#&gt; $targets$A1B$SampleName
#&gt; [1] "A1B"
#&gt; 
#&gt; $targets$A1B$Factor
#&gt; [1] "A1"
#&gt; 
#&gt; $targets$A1B$SampleLong
#&gt; [1] "Avr.1h.B"
#&gt; 
#&gt; $targets$A1B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A1B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V1A
#&gt; $targets$V1A$FileName
#&gt; [1] "./data/SRR446031_1.fastq.gz"
#&gt; 
#&gt; $targets$V1A$SampleName
#&gt; [1] "V1A"
#&gt; 
#&gt; $targets$V1A$Factor
#&gt; [1] "V1"
#&gt; 
#&gt; $targets$V1A$SampleLong
#&gt; [1] "Vir.1h.A"
#&gt; 
#&gt; $targets$V1A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V1A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V1B
#&gt; $targets$V1B$FileName
#&gt; [1] "./data/SRR446032_1.fastq.gz"
#&gt; 
#&gt; $targets$V1B$SampleName
#&gt; [1] "V1B"
#&gt; 
#&gt; $targets$V1B$Factor
#&gt; [1] "V1"
#&gt; 
#&gt; $targets$V1B$SampleLong
#&gt; [1] "Vir.1h.B"
#&gt; 
#&gt; $targets$V1B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V1B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$M6A
#&gt; $targets$M6A$FileName
#&gt; [1] "./data/SRR446033_1.fastq.gz"
#&gt; 
#&gt; $targets$M6A$SampleName
#&gt; [1] "M6A"
#&gt; 
#&gt; $targets$M6A$Factor
#&gt; [1] "M6"
#&gt; 
#&gt; $targets$M6A$SampleLong
#&gt; [1] "Mock.6h.A"
#&gt; 
#&gt; $targets$M6A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M6A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$M6B
#&gt; $targets$M6B$FileName
#&gt; [1] "./data/SRR446034_1.fastq.gz"
#&gt; 
#&gt; $targets$M6B$SampleName
#&gt; [1] "M6B"
#&gt; 
#&gt; $targets$M6B$Factor
#&gt; [1] "M6"
#&gt; 
#&gt; $targets$M6B$SampleLong
#&gt; [1] "Mock.6h.B"
#&gt; 
#&gt; $targets$M6B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M6B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A6A
#&gt; $targets$A6A$FileName
#&gt; [1] "./data/SRR446035_1.fastq.gz"
#&gt; 
#&gt; $targets$A6A$SampleName
#&gt; [1] "A6A"
#&gt; 
#&gt; $targets$A6A$Factor
#&gt; [1] "A6"
#&gt; 
#&gt; $targets$A6A$SampleLong
#&gt; [1] "Avr.6h.A"
#&gt; 
#&gt; $targets$A6A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A6A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A6B
#&gt; $targets$A6B$FileName
#&gt; [1] "./data/SRR446036_1.fastq.gz"
#&gt; 
#&gt; $targets$A6B$SampleName
#&gt; [1] "A6B"
#&gt; 
#&gt; $targets$A6B$Factor
#&gt; [1] "A6"
#&gt; 
#&gt; $targets$A6B$SampleLong
#&gt; [1] "Avr.6h.B"
#&gt; 
#&gt; $targets$A6B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A6B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V6A
#&gt; $targets$V6A$FileName
#&gt; [1] "./data/SRR446037_1.fastq.gz"
#&gt; 
#&gt; $targets$V6A$SampleName
#&gt; [1] "V6A"
#&gt; 
#&gt; $targets$V6A$Factor
#&gt; [1] "V6"
#&gt; 
#&gt; $targets$V6A$SampleLong
#&gt; [1] "Vir.6h.A"
#&gt; 
#&gt; $targets$V6A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V6A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V6B
#&gt; $targets$V6B$FileName
#&gt; [1] "./data/SRR446038_1.fastq.gz"
#&gt; 
#&gt; $targets$V6B$SampleName
#&gt; [1] "V6B"
#&gt; 
#&gt; $targets$V6B$Factor
#&gt; [1] "V6"
#&gt; 
#&gt; $targets$V6B$SampleLong
#&gt; [1] "Vir.6h.B"
#&gt; 
#&gt; $targets$V6B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V6B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$M12A
#&gt; $targets$M12A$FileName
#&gt; [1] "./data/SRR446039_1.fastq.gz"
#&gt; 
#&gt; $targets$M12A$SampleName
#&gt; [1] "M12A"
#&gt; 
#&gt; $targets$M12A$Factor
#&gt; [1] "M12"
#&gt; 
#&gt; $targets$M12A$SampleLong
#&gt; [1] "Mock.12h.A"
#&gt; 
#&gt; $targets$M12A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M12A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$M12B
#&gt; $targets$M12B$FileName
#&gt; [1] "./data/SRR446040_1.fastq.gz"
#&gt; 
#&gt; $targets$M12B$SampleName
#&gt; [1] "M12B"
#&gt; 
#&gt; $targets$M12B$Factor
#&gt; [1] "M12"
#&gt; 
#&gt; $targets$M12B$SampleLong
#&gt; [1] "Mock.12h.B"
#&gt; 
#&gt; $targets$M12B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$M12B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A12A
#&gt; $targets$A12A$FileName
#&gt; [1] "./data/SRR446041_1.fastq.gz"
#&gt; 
#&gt; $targets$A12A$SampleName
#&gt; [1] "A12A"
#&gt; 
#&gt; $targets$A12A$Factor
#&gt; [1] "A12"
#&gt; 
#&gt; $targets$A12A$SampleLong
#&gt; [1] "Avr.12h.A"
#&gt; 
#&gt; $targets$A12A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A12A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$A12B
#&gt; $targets$A12B$FileName
#&gt; [1] "./data/SRR446042_1.fastq.gz"
#&gt; 
#&gt; $targets$A12B$SampleName
#&gt; [1] "A12B"
#&gt; 
#&gt; $targets$A12B$Factor
#&gt; [1] "A12"
#&gt; 
#&gt; $targets$A12B$SampleLong
#&gt; [1] "Avr.12h.B"
#&gt; 
#&gt; $targets$A12B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$A12B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V12A
#&gt; $targets$V12A$FileName
#&gt; [1] "./data/SRR446043_1.fastq.gz"
#&gt; 
#&gt; $targets$V12A$SampleName
#&gt; [1] "V12A"
#&gt; 
#&gt; $targets$V12A$Factor
#&gt; [1] "V12"
#&gt; 
#&gt; $targets$V12A$SampleLong
#&gt; [1] "Vir.12h.A"
#&gt; 
#&gt; $targets$V12A$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V12A$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; $targets$V12B
#&gt; $targets$V12B$FileName
#&gt; [1] "./data/SRR446044_1.fastq.gz"
#&gt; 
#&gt; $targets$V12B$SampleName
#&gt; [1] "V12B"
#&gt; 
#&gt; $targets$V12B$Factor
#&gt; [1] "V12"
#&gt; 
#&gt; $targets$V12B$SampleLong
#&gt; [1] "Vir.12h.B"
#&gt; 
#&gt; $targets$V12B$Experiment
#&gt; [1] 1
#&gt; 
#&gt; $targets$V12B$Date
#&gt; [1] "23-Mar-2012"
#&gt; 
#&gt; 
#&gt; 
#&gt; $targetsheader
#&gt; $targetsheader$targetsheader
#&gt; [1] "# Project ID: Arabidopsis - Pseudomonas alternative splicing study (SRA: SRP010938; PMID: 24098335)"                                                                              
#&gt; [2] "# The following line(s) allow to specify the contrasts needed for comparative analyses, such as DEG identification. All possible comparisons can be specified with 'CMPset: ALL'."
#&gt; [3] "# &lt;CMP&gt; CMPset1: M1-A1, M1-V1, A1-V1, M6-A6, M6-V6, A6-V6, M12-A12, M12-V12, A12-V12"                                                                                             
#&gt; [4] "# &lt;CMP&gt; CMPset2: ALL"                                                                                                                                                             
#&gt; 
#&gt; 
#&gt; $modules
#&gt; $modules$module1
#&gt; [1] "hisat2/2.1.0"
#&gt; 
#&gt; 
#&gt; $wf
#&gt; list()
#&gt; 
#&gt; $clt
#&gt; $clt$`hisat2-mapping-se.cwl`
#&gt; $clt$`hisat2-mapping-se.cwl`$cwlVersion
#&gt; [1] "v1.0"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$class
#&gt; [1] "CommandLineTool"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$doc
#&gt; [1] "[HISAT2](https://ccb.jhu.edu/software/hisat2/index.shtml): graph-based alignment of next generation sequencing reads to a population of genomes"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$label
#&gt; [1] "Last updated 08/2019"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$hints
#&gt; $clt$`hisat2-mapping-se.cwl`$hints$SoftwareRequirement
#&gt; $clt$`hisat2-mapping-se.cwl`$hints$SoftwareRequirement$packages
#&gt; $clt$`hisat2-mapping-se.cwl`$hints$SoftwareRequirement$packages[[1]]
#&gt; $clt$`hisat2-mapping-se.cwl`$hints$SoftwareRequirement$packages[[1]]$package
#&gt; [1] "hisat2"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$hints$SoftwareRequirement$packages[[1]]$version
#&gt; [1] "2.1.0"
#&gt; 
#&gt; 
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$baseCommand
#&gt; [1] "hisat2"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$requirements
#&gt; $clt$`hisat2-mapping-se.cwl`$requirements$InitialWorkDirRequirement
#&gt; $clt$`hisat2-mapping-se.cwl`$requirements$InitialWorkDirRequirement$listing
#&gt; [1] "$(inputs.results_path)"
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[1]]
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[1]]$prefix
#&gt; [1] "-S"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[1]]$valueFrom
#&gt; [1] "$(inputs.results_path.path)/$(inputs.SampleName).sam"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[2]]
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[2]]$prefix
#&gt; [1] "-x"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[2]]$valueFrom
#&gt; [1] "$(inputs.idx_basedir.path)/$(inputs.idx_basename)"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[3]]
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[3]]$prefix
#&gt; [1] "-k"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[3]]$valueFrom
#&gt; [1] "1"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[4]]
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[4]]$prefix
#&gt; [1] "--min-intronlen"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[4]]$valueFrom
#&gt; [1] "30"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[5]]
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[5]]$prefix
#&gt; [1] "--max-intronlen"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$arguments[[5]]$valueFrom
#&gt; [1] "3000"
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basedir
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basedir$label
#&gt; [1] "Path to the directory containing the index for the reference genome"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basedir$type
#&gt; [1] "Directory"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basename
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basename$label
#&gt; [1] "Basename of the hisat2 index files"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$idx_basename$type
#&gt; [1] "string"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1$label
#&gt; [1] "Comma-separated list of files containing unpaired reads to be aligned"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1$type
#&gt; [1] "File"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1$inputBinding
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1$inputBinding$prefix
#&gt; [1] "-U"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$fq1$inputBinding$itemSeparator
#&gt; [1] ","
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$thread
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$thread$label
#&gt; [1] "Launch NTHREADS parallel search threads"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$thread$type
#&gt; [1] "int"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$thread$inputBinding
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$thread$inputBinding$prefix
#&gt; [1] "--threads"
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$SampleName
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$SampleName$label
#&gt; [1] "Filename to write output to"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$SampleName$type
#&gt; [1] "string"
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$results_path
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$results_path$label
#&gt; [1] "Path to the results directory"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$inputs$results_path$type
#&gt; [1] "Directory"
#&gt; 
#&gt; 
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$outputs
#&gt; $clt$`hisat2-mapping-se.cwl`$outputs$hisat2_sam
#&gt; $clt$`hisat2-mapping-se.cwl`$outputs$hisat2_sam$type
#&gt; [1] "File"
#&gt; 
#&gt; $clt$`hisat2-mapping-se.cwl`$outputs$hisat2_sam$outputBinding
#&gt; $clt$`hisat2-mapping-se.cwl`$outputs$hisat2_sam$outputBinding$glob
#&gt; [1] "$(inputs.results_path.path)/$(inputs.SampleName).sam"
#&gt; 
#&gt; 
#&gt; 
#&gt; 
#&gt; 
#&gt; 
#&gt; $yamlinput
#&gt; $yamlinput$idx_basedir
#&gt; $yamlinput$idx_basedir$class
#&gt; [1] "Directory"
#&gt; 
#&gt; $yamlinput$idx_basedir$path
#&gt; [1] "./data"
#&gt; 
#&gt; 
#&gt; $yamlinput$idx_basename
#&gt; [1] "tair10.fasta"
#&gt; 
#&gt; $yamlinput$fq1
#&gt; $yamlinput$fq1$class
#&gt; [1] "File"
#&gt; 
#&gt; $yamlinput$fq1$path
#&gt; [1] "_FASTQ_PATH1_"
#&gt; 
#&gt; 
#&gt; $yamlinput$SampleName
#&gt; [1] "_SampleName_"
#&gt; 
#&gt; $yamlinput$results_path
#&gt; $yamlinput$results_path$class
#&gt; [1] "Directory"
#&gt; 
#&gt; $yamlinput$results_path$path
#&gt; [1] "./results"
#&gt; 
#&gt; 
#&gt; $yamlinput$thread
#&gt; [1] 4
#&gt; 
#&gt; $yamlinput$ModulesToLoad
#&gt; $yamlinput$ModulesToLoad$module1
#&gt; [1] "hisat2/2.1.0"
#&gt; 
#&gt; 
#&gt; 
#&gt; $cmdlist
#&gt; $cmdlist$M1A
#&gt; $cmdlist$M1A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$M1B
#&gt; $cmdlist$M1B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M1B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446028_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A1A
#&gt; $cmdlist$A1A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446029_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A1B
#&gt; $cmdlist$A1B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A1B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446030_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V1A
#&gt; $cmdlist$V1A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446031_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V1B
#&gt; $cmdlist$V1B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V1B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446032_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$M6A
#&gt; $cmdlist$M6A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M6A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446033_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$M6B
#&gt; $cmdlist$M6B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M6B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446034_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A6A
#&gt; $cmdlist$A6A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A6A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446035_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A6B
#&gt; $cmdlist$A6B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A6B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446036_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V6A
#&gt; $cmdlist$V6A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V6A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446037_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V6B
#&gt; $cmdlist$V6B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V6B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446038_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$M12A
#&gt; $cmdlist$M12A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M12A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446039_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$M12B
#&gt; $cmdlist$M12B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/M12B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446040_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A12A
#&gt; $cmdlist$A12A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A12A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446041_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$A12B
#&gt; $cmdlist$A12B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/A12B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446042_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V12A
#&gt; $cmdlist$V12A$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V12A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446043_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; $cmdlist$V12B
#&gt; $cmdlist$V12B$`hisat2-mapping-se`
#&gt; [1] "hisat2 -S ./results/V12B.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446044_1.fastq.gz --threads 4"
#&gt; 
#&gt; 
#&gt; 
#&gt; $input
#&gt; $input$M1A
#&gt; $input$M1A$FileName
#&gt; [1] "./data/SRR446027_1.fastq.gz"
#&gt; 
#&gt; $input$M1A$SampleName
#&gt; [1] "M1A"
#&gt; 
#&gt; 
#&gt; $input$M1B
#&gt; $input$M1B$FileName
#&gt; [1] "./data/SRR446028_1.fastq.gz"
#&gt; 
#&gt; $input$M1B$SampleName
#&gt; [1] "M1B"
#&gt; 
#&gt; 
#&gt; $input$A1A
#&gt; $input$A1A$FileName
#&gt; [1] "./data/SRR446029_1.fastq.gz"
#&gt; 
#&gt; $input$A1A$SampleName
#&gt; [1] "A1A"
#&gt; 
#&gt; 
#&gt; $input$A1B
#&gt; $input$A1B$FileName
#&gt; [1] "./data/SRR446030_1.fastq.gz"
#&gt; 
#&gt; $input$A1B$SampleName
#&gt; [1] "A1B"
#&gt; 
#&gt; 
#&gt; $input$V1A
#&gt; $input$V1A$FileName
#&gt; [1] "./data/SRR446031_1.fastq.gz"
#&gt; 
#&gt; $input$V1A$SampleName
#&gt; [1] "V1A"
#&gt; 
#&gt; 
#&gt; $input$V1B
#&gt; $input$V1B$FileName
#&gt; [1] "./data/SRR446032_1.fastq.gz"
#&gt; 
#&gt; $input$V1B$SampleName
#&gt; [1] "V1B"
#&gt; 
#&gt; 
#&gt; $input$M6A
#&gt; $input$M6A$FileName
#&gt; [1] "./data/SRR446033_1.fastq.gz"
#&gt; 
#&gt; $input$M6A$SampleName
#&gt; [1] "M6A"
#&gt; 
#&gt; 
#&gt; $input$M6B
#&gt; $input$M6B$FileName
#&gt; [1] "./data/SRR446034_1.fastq.gz"
#&gt; 
#&gt; $input$M6B$SampleName
#&gt; [1] "M6B"
#&gt; 
#&gt; 
#&gt; $input$A6A
#&gt; $input$A6A$FileName
#&gt; [1] "./data/SRR446035_1.fastq.gz"
#&gt; 
#&gt; $input$A6A$SampleName
#&gt; [1] "A6A"
#&gt; 
#&gt; 
#&gt; $input$A6B
#&gt; $input$A6B$FileName
#&gt; [1] "./data/SRR446036_1.fastq.gz"
#&gt; 
#&gt; $input$A6B$SampleName
#&gt; [1] "A6B"
#&gt; 
#&gt; 
#&gt; $input$V6A
#&gt; $input$V6A$FileName
#&gt; [1] "./data/SRR446037_1.fastq.gz"
#&gt; 
#&gt; $input$V6A$SampleName
#&gt; [1] "V6A"
#&gt; 
#&gt; 
#&gt; $input$V6B
#&gt; $input$V6B$FileName
#&gt; [1] "./data/SRR446038_1.fastq.gz"
#&gt; 
#&gt; $input$V6B$SampleName
#&gt; [1] "V6B"
#&gt; 
#&gt; 
#&gt; $input$M12A
#&gt; $input$M12A$FileName
#&gt; [1] "./data/SRR446039_1.fastq.gz"
#&gt; 
#&gt; $input$M12A$SampleName
#&gt; [1] "M12A"
#&gt; 
#&gt; 
#&gt; $input$M12B
#&gt; $input$M12B$FileName
#&gt; [1] "./data/SRR446040_1.fastq.gz"
#&gt; 
#&gt; $input$M12B$SampleName
#&gt; [1] "M12B"
#&gt; 
#&gt; 
#&gt; $input$A12A
#&gt; $input$A12A$FileName
#&gt; [1] "./data/SRR446041_1.fastq.gz"
#&gt; 
#&gt; $input$A12A$SampleName
#&gt; [1] "A12A"
#&gt; 
#&gt; 
#&gt; $input$A12B
#&gt; $input$A12B$FileName
#&gt; [1] "./data/SRR446042_1.fastq.gz"
#&gt; 
#&gt; $input$A12B$SampleName
#&gt; [1] "A12B"
#&gt; 
#&gt; 
#&gt; $input$V12A
#&gt; $input$V12A$FileName
#&gt; [1] "./data/SRR446043_1.fastq.gz"
#&gt; 
#&gt; $input$V12A$SampleName
#&gt; [1] "V12A"
#&gt; 
#&gt; 
#&gt; $input$V12B
#&gt; $input$V12B$FileName
#&gt; [1] "./data/SRR446044_1.fastq.gz"
#&gt; 
#&gt; $input$V12B$SampleName
#&gt; [1] "V12B"
#&gt; 
#&gt; 
#&gt; 
#&gt; $output
#&gt; $output$M1A
#&gt; $output$M1A$`hisat2-mapping-se`
#&gt; [1] "./results/M1A.sam"
#&gt; 
#&gt; 
#&gt; $output$M1B
#&gt; $output$M1B$`hisat2-mapping-se`
#&gt; [1] "./results/M1B.sam"
#&gt; 
#&gt; 
#&gt; $output$A1A
#&gt; $output$A1A$`hisat2-mapping-se`
#&gt; [1] "./results/A1A.sam"
#&gt; 
#&gt; 
#&gt; $output$A1B
#&gt; $output$A1B$`hisat2-mapping-se`
#&gt; [1] "./results/A1B.sam"
#&gt; 
#&gt; 
#&gt; $output$V1A
#&gt; $output$V1A$`hisat2-mapping-se`
#&gt; [1] "./results/V1A.sam"
#&gt; 
#&gt; 
#&gt; $output$V1B
#&gt; $output$V1B$`hisat2-mapping-se`
#&gt; [1] "./results/V1B.sam"
#&gt; 
#&gt; 
#&gt; $output$M6A
#&gt; $output$M6A$`hisat2-mapping-se`
#&gt; [1] "./results/M6A.sam"
#&gt; 
#&gt; 
#&gt; $output$M6B
#&gt; $output$M6B$`hisat2-mapping-se`
#&gt; [1] "./results/M6B.sam"
#&gt; 
#&gt; 
#&gt; $output$A6A
#&gt; $output$A6A$`hisat2-mapping-se`
#&gt; [1] "./results/A6A.sam"
#&gt; 
#&gt; 
#&gt; $output$A6B
#&gt; $output$A6B$`hisat2-mapping-se`
#&gt; [1] "./results/A6B.sam"
#&gt; 
#&gt; 
#&gt; $output$V6A
#&gt; $output$V6A$`hisat2-mapping-se`
#&gt; [1] "./results/V6A.sam"
#&gt; 
#&gt; 
#&gt; $output$V6B
#&gt; $output$V6B$`hisat2-mapping-se`
#&gt; [1] "./results/V6B.sam"
#&gt; 
#&gt; 
#&gt; $output$M12A
#&gt; $output$M12A$`hisat2-mapping-se`
#&gt; [1] "./results/M12A.sam"
#&gt; 
#&gt; 
#&gt; $output$M12B
#&gt; $output$M12B$`hisat2-mapping-se`
#&gt; [1] "./results/M12B.sam"
#&gt; 
#&gt; 
#&gt; $output$A12A
#&gt; $output$A12A$`hisat2-mapping-se`
#&gt; [1] "./results/A12A.sam"
#&gt; 
#&gt; 
#&gt; $output$A12B
#&gt; $output$A12B$`hisat2-mapping-se`
#&gt; [1] "./results/A12B.sam"
#&gt; 
#&gt; 
#&gt; $output$V12A
#&gt; $output$V12A$`hisat2-mapping-se`
#&gt; [1] "./results/V12A.sam"
#&gt; 
#&gt; 
#&gt; $output$V12B
#&gt; $output$V12B$`hisat2-mapping-se`
#&gt; [1] "./results/V12B.sam"
#&gt; 
#&gt; 
#&gt; 
#&gt; $cwlfiles
#&gt; $cwlfiles$cwl
#&gt; [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e2165ead1/systemPipeR/extdata/cwl/hisat2/hisat2-se/hisat2-mapping-se.cwl"
#&gt; 
#&gt; $cwlfiles$yml
#&gt; [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e2165ead1/systemPipeR/extdata/cwl/hisat2/hisat2-se/hisat2-mapping-se.yml"
#&gt; 
#&gt; $cwlfiles$steps
#&gt; [1] "hisat2-mapping-se"
#&gt; 
#&gt; $cwlfiles$targets
#&gt; [1] "/tmp/RtmpOJ4T8t/temp_libpath3139e2165ead1/systemPipeR/extdata/targets.txt"
#&gt; 
#&gt; 
#&gt; $inputvars
#&gt; $inputvars$FileName
#&gt; [1] "_FASTQ_PATH1_"
#&gt; 
#&gt; $inputvars$SampleName
#&gt; [1] "_SampleName_"
#&gt; 
#&gt; </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


