<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipeShiny • systemPipeShiny</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipeShiny">
<meta property="og:description" content="systemPipeShiny">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeShiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://systempipe.org/sps/">
    <span class="fa fa fa fa-home"></span>
     
    SPS Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeShiny/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeShiny_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipeShiny</h1>
                        <h4 class="author">Author: Le Zhang, Daniela Cassol, Ponmathi Ramasamy, Jianhai Zhang, Gordon Mosher and Thomas Girke</h4>
            
            <h4 class="date">Last update: 12 February, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/systemPipeShiny/blob/master/vignettes/systemPipeShiny.Rmd"><code>vignettes/systemPipeShiny.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeShiny.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}

body {
    max-width: 2000px !important;
}

.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>/*Bootstrap colors*/
:root {
  --blue: #007bff;
  --indigo: #6610f2;
  --purple: #6f42c1;
  --pink: #e83e8c;
  --red: #dc3545;
  --orange: #fd7e14;
  --yellow: #ffc107;
  --green: #28a745;
  --teal: #20c997;
  --cyan: #17a2b8;
  --white: #fff;
  --gray: #6c757d;
  --gray-dark: #343a40;
  --primary: #337ab7;
  --secondary: #6c757d;
  --success: #28a745;
  --info: #17a2b8;
  --warning: #ffc107;
  --danger: #dc3545;
  --light: #f8f9fa;
  --dark: #343a40;
  --breakpoint-xs: 0;
  --breakpoint-sm: 576px;
  --breakpoint-md: 768px;
  --breakpoint-lg: 992px;
  --breakpoint-xl: 1200px;
  --font-family-sans-serif: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  --font-family-monospace: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
}


/*
systemPipeShiny CSS Styles
*/

/*buttons, default is too ugly*/
.sps .btn-default {
  color: #fff;
  background-color: var(--primary);
  border-color: #2e6da4;
}

.sps .btn-default:hover{
  background-color: #4c92cf;
}

/*widget user in shinydashboardplus*/
.widget-user {
  overflow-y:auto;
}

/*for clearableTextInput component*/
.text-input-clearable {
  width: 100%;
  border:1px solid #ccc;
  border-radius: 3px;
  padding:1px 6px 1px 1px;
  display:inline-block;
}

.text-input-clearable input {
  width: 93%;
  border:none;
  background:none;
  outline:none;
  padding:0 0;
  margin:0 0;
  font:inherit;
}

.text-input-clearable span {
  float: right;
  height: 14px;
  font-size: 14px;
  cursor: pointer;
  color: #ccc;
  visibility:hidden;
}
.text-input-focused {
  border-color: #66afe9;
  outline: 0;
  -webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  -moz-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
}

/*for gallery */
img-gallery{
  max-width:100%;
  height:auto;
  object-fit: cover;
  vertical-align: middle;
  text-align: center;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

img-gallery:hover {
  opacity: 0.7;
  transition: opacity 0.5s;
}


/*for sps hyperlink button  and tables*/
.href-button {
  background-color: var(--primary);
  border: none;
  border-radius: 10px 10px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
  color: white;
  padding: 10px 25px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 15px;
  margin: 4px 2px;
  cursor: pointer;
         }

.href-button:hover {
  opacity: 0.9;
  text-decoration: none;
  color: white;
  transition: opacity 0.5s;
}

.href-button:active {
  transform: translateY(2px);
}

.table-href td:first-child{
  width: 1px;
  vertical-align: middle;
  white-space: nowrap;
}

/* about tab background */
.widget-user-header {
    background: linear-gradient(-45deg, #428bca, #ccccff, #b3b3b3);
    background-size: 400% 400%;
    animation: gradient 10s ease infinite;
}

@keyframes gradient {
    0% {
        background-position: 0% 50%;
    }
    50% {
        background-position: 100% 50%;
    }
    100% {
        background-position: 0% 50%;
    }
}

/* sps description container*/
.desc-body.collapse {
  display: block;
  height: 150px;
  overflow: hidden;
}
.desc-body.in {
  height: auto;
}
.desc a.collapsed:after  {
    content: '+ Show More Description';
}
.desc > a:not(.collapsed):after {
    content: '- Collapse Description';
}

.desc a:not(.desc-body):link {
  text-decoration: none;
}


.desc hr {
    border-top: 1px solid #3c8dbc;
}

/* sps description link*/
.desc  a {
  margin: 0 var(--spacing, 0px);
  font-size: var(--a_size, 1em);
  -webkit-transition: margin 0.25s;
  transition: margin 0.25s;
  -webkit-transition: font-size 0.25s;
  transition: font-size 0.25s;
}

.desc  a:hover {
  --spacing: 4px;
  --a_size: 1.5em;
}

/* sps control-panel*/
.control-panel {
    opacity: 0.7;
    transition: opacity .25s ease-in-out;
   -moz-transition: opacity .25s ease-in-out;
   -webkit-transition: opacity .25s ease-in-out;
}

.control-panel:hover { opacity: 1;}


/* sps hexagon icon and hex panel*/
.hex-container {
  display: inline-block;
  text-align: center;
  width: 100%;
}

.hex {
  opacity: 0.8;
}

.hex:hover {
  opacity: 1;
}

.powerby-link {
  font: italic bold 25px serif;
  fill: var(--primary);
  opacity: 0.5;
}

.powerby-link:hover {
  opacity: 1;
}

.hex-box {
  width: 3vw;
  min-width: 50px;
  display: inline-block;
  padding-top: 0;
}

.hex-container span {
 display: inline-block;
 text-align: center;
 font-size: 0.6vw;
}

.hex-panel {
  position: auto;
}
.hex-panel > .hex-item {
  display: inline-block;
  text-align: left;
  padding-left: 0;
  padding-right: 0;
  margin-left: 0;
  margin-right:0;
}

.hex-panel h5 {
  display: inline-block;
  margin-left: 20px;
  font-weight: 900;
  font-size: 1em;
  transform: translateY(-2.2vw);
}

/* sps canvas*/
.snap-drag {
  opacity: 0;
}

.snap-drag:hover {
  opacity: 1;
}

.snap-drag button{
  position: absolute;
  outline: none;
  background-color: Transparent;
  top: 2px;
  right: 4px;
}

.snap-drag button:focus {
  outline: 0 !important;
  box-shadow: none !important;
}

.snap-drag h4{
  margin-bottom: 0;
  padding-bottom: 10px;
}

a.bttn-default {
    background: #3c8dbc !important;
}

.sps-warning h4 {
    color: var(--danger);
}
.sps-warning h4::before{
    content: "\2022 ";
}


/* background color of boxplus inside bt3 accordion*/
.panel-collapse .panel-body .box {
    background-color: #f6f6f6;
}

/* timeline */
.sps-timeline {
  list-style-type: none;
  display: flex;
  align-items: center;
  justify-content: center;
}

.sps-timestamp {
  margin-bottom: 20px;
  padding: 0px 40px;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-weight: 100;
}

.sps-timeline .timeline-i {
  width: 30px;
  height: 30px;
  padding: 7px;
  background-color: #D6DCE0;
  border-radius: 25px;
  display: inline;
  justify-content: center;
  position: relative;
  transition: all 200ms ease-in;
  left: 40%;
  z-index: 98;
}
.sps-timeline .timeline-i i {
  position: relative;
  z-index: 99;
  top: 25%;
  color: white;
}
.sps-timeline .status {
  padding: 0px 40px;
  display: flex;
  justify-content: center;
  border-top: 5px solid #D6DCE0;
  top: -10px;
  position: relative;
  transition: all 200ms ease-in;
}
.sps-timeline .status h4 {
  font-weight: 600;
}

.sps-timeline .li.complete .timeline-i {
  background-color: #5cb85c;
}
.sps-timeline .li.complete .status {
  border-top: 5px solid #5cb85c;
  top: -10px;
}
.sps-timeline .li.complete .status:before {
  background-color: #5cb85c;
  border: none;
  transition: all 200ms ease-in;
}
.sps-timeline .li.complete .status h4 {
  color: #5cb85c;
}

/*go top button*/
.sps-gotop {
  background-color: #fff;
  box-shadow: 0 1px 6px 0 rgba(0, 0, 0, 0.2);
  position: fixed;
  display: none;
  text-align: center;
  cursor: pointer;
  right: 1rem;
  bottom: 10rem;
  width: 3rem;
  height: 3rem;
  border-radius: 0.25rem;
  line-height: 2.5rem;
  z-index: 99;
  fill: #337ab7;
  opacity: 0.7;
  transition: opacity 1s;
  -webkit-transition: opacity 1s;
}

.sps-gotop:hover {
    opacity: 1;
}


/*blink class*/

.sps-blink {
  animation: blinking 1.5s linear infinite;
}

@keyframes blinking {
  50% { opacity: 0; }
}


/*disconnect*/
/*credit to daattali: https://github.com/daattali/shinydisconnect*/
#shiny-disconnected-overlay {
    background-color: #000000 !important;
    z-index: 9997 !important;
    cursor: not-allowed !important;
    pointer-events: auto !important;
}
#ss-connect-dialog label { display: none !important; }

#ss-overlay {
  background-color: #000000 !important;
  opacity: 0.6 !important;
  position: fixed !important;
  top: 0 !important;
  left: 0 !important;
  bottom: 0 !important;
  right: 0 !important;
  z-index: 9998 !important;
  overflow: hidden !important;
  cursor: not-allowed !important;
}

#ss-connect-dialog {
  background: white !important;
  color: #444444 !important;
  width: 550px !important;
  transform: translateX(-50%) translateY(-50%) !important;
  font-size: 22px !important;
  top: 50% !important;
  position: fixed !important;
  bottom: auto !important;
  left: 50% !important;
  padding: 0.8em 0.5em !important;
  text-align: center !important;
  height: auto !important;
  opacity: 1 !important;
  z-index: 9999 !important;
  border-radius: 15px !important;
  box-shadow: rgba(0, 0, 0, 0.3) 3px 3px 10px !important;
}

#ss-connect-dialog > a {
  display: block !important;
  color: #337ab7 !important;
  font-size: 0 !important;
  margin-top: 22px !important;
  font-weight: normal !important;
}

#ss-connect-dialog > a::before {
  content: 'Reconnect';
  font-size: 22px;
}

#ss-connect-dialog img{
  display: block;
  margin: auto;
}

#ss-connect-dialog span {
    font-size: 15px;
}

#ss-connect-dialog span > a {
    display: inline !important;
    color: #337ab7 !important;
}

.shiny-discon::before { content: 'SPS Disconnected: Error occurred or timeout reached' }

.shiny-discon-noserver::before { content: 'Cannot connect to the server' }


div.active:focus, :focus {
    outline: 0;
}

.plot-control-panel {
    border: 1px solid #bce8f1;
    min-height: 500px;
    margin-left: 5px;
    padding: 0 5px;
}


/* reset some values changed by shiny accessibility*/
.carousel:hover .carousel-caption, .carousel.contrast .carousel-caption {
    background-color: initial;
}

a.carousel-control:focus {
    outline: initial;
    background-image: initial;
    box-shadow: initial;
}
</style>
<p><img src="../inst/app/www/img/sps.png" width="50%" style="display: block; margin: auto;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong> (SPS) extends the widely used <a href="https://bioconductor.org/packages/systemPipeR/">systemPipeR</a> (SPR) workflow environment with a versatile graphical user interface provided by a <a href="https://shiny.rstudio.com">Shiny App</a>. This allows non-R users, such as experimentalists, to run many systemPipeR’s workflow designs, control, and visualization functionalities interactively without requiring knowledge of R. Most importantly, <code>SPS</code> has been designed as a general purpose framework for interacting with other R packages in an intuitive manner. Like most Shiny Apps, SPS can be used on both local computers as well as centralized server-based deployments that can be accessed remotely as a public web service for using SPR’s functionalities with community and/or private data. The framework can integrate many core packages from the R/Bioconductor ecosystem. Examples of SPS’ current functionalities include: (a) interactive creation of experimental designs and metadata using an easy to use tabular editor or file uploader; (b) visualization of workflow topologies combined with auto-generation of R Markdown preview for interactively designed workflows; (c) access to a wide range of data processing routines; (d) and an extendable set of visualization functionalities. Complex visual results can be managed on a ‘Canvas Workbench’ allowing users to organize and to compare plots in an efficient manner combined with a session snapshot feature to continue work at a later time. The present suite of pre-configured visualization examples include different methods to plot a count table. The modular design of SPR makes it easy to design custom functions without any knowledge of Shiny, as well as extending the environment in the future with contributions from the community.</p>
<p>An example web instance for testing <code>SPS</code> online is available <a href="https://tgirke.shinyapps.io/systemPipeShiny/" target="_blank">here</a>. At the time of writing, there are five interactive tutorials embedded in this instance that users can access from the upper-right corner. Each tutorial introduces major functionalities of SPS.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>The <code>systemPipeShiny</code> package can be installed from a user’s R console as follows.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeShiny"</span><span class="op">)</span>

<span class="co"># or use 'pak'</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"pak"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"pak"</span>, 
    repos <span class="op">=</span> <span class="st">"https://r-lib.github.io/p/pak/dev/"</span><span class="op">)</span>
<span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="http://pak.r-lib.org/reference/pkg_install.html">pkg_install</a></span><span class="op">(</span><span class="st">"systemPipeShiny"</span><span class="op">)</span></code></pre></div>
<p>If the user’s OS is Linux then one may need to install the following dependencies <strong>before</strong> installing SPS. The following example applies to Ubuntu.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libcurl4-openssl-dev</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libssl-dev</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libv8-dev</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># other dependencies</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libicu-dev</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y pandoc</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y make</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libglpk-dev</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y libgmp3-dev</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install -y zlib1g-dev</span></code></pre></div>
<p>Most likely you only need to install first 3, <em>libcurl4-openssl-dev</em> may not be required for Ubuntu &gt; 18.04. In case any other problem happens, try to install other system dependencies.</p>
<p>On other Linux distributions, the install commands may be slightly different.</p>
</div>
<div id="main-functionalities" class="section level1">
<h1 class="hasAnchor">
<a href="#main-functionalities" class="anchor"></a>Main functionalities</h1>
<p>Currently, <code>SPS</code> includes 3 main functional categories (<em>Fig 1</em>):</p>
<ol style="list-style-type: decimal">
<li>Some pre-defined modules include:
<ol style="list-style-type: lower-alpha">
<li>A workbench for designing and configuring data analysis workflows,</li>
<li>Downstream analysis and visualization tools for RNA-Seq, and</li>
<li>A space to make quick ggplots.</li>
</ol>
</li>
<li>A section with user custom tabs: users define their own shiny tabs.<br>
</li>
<li>An image editing tab “Canvas” which allows users to edit plots made from the previous two categories.</li>
</ol>
<p>Besides, SPS provides many functions to extend the default Shiny development, like more UI components, server functions. Also, SPS has some useful general R ulitlies like error catching, logging, and more.</p>
<center id="sps_structure">
<img src="img/sps_structure.jpg" title="fig:" alt="SPS_structure">
</center>
<p><strong>Figure 1.</strong> Design of <code>systemPipeShiny.</code></p>
<p>The framework provides an interactive web interface for workflow management and data visualization.</p>
<p>Within the functional categories, <code>SPS</code> functions are modularized in sub-components, here referred to as <strong>SPS tabs</strong> that are similar to menu tabs in other GUI applications that organize related and inter-connected functionalies into groups. On the backend, <strong>SPS tabs</strong> are based on <a href="https://shiny.rstudio.com/articles/modules.html">Shiny modules</a>, that are stored in separate files. This modular structure is highly extensible and greatly simplifies the design of new <code>SPS</code> tabs by both users and/or developers. Details about extending existing tabs and designing new ones are provided in <a href="https://systempipe.org/sps/">advanced sections on our website</a>.</p>
</div>
<div id="sps-example-usage" class="section level1">
<h1 class="hasAnchor">
<a href="#sps-example-usage" class="anchor"></a>SPS example usage</h1>
<p>The following instructions go through use each module step by step.</p>
<p>Please read our website: <a href="https://systempipe.org/sps/" class="uri">https://systempipe.org/sps/</a> for full instructions, animations, videos, and advanced sections.</p>
<div id="load-package" class="section level2">
<h2 class="hasAnchor">
<a href="#load-package" class="anchor"></a>Load package</h2>
<p>Load the <code>systemPipeShiny</code> package in your R session.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/systemPipeR/systemPipeShiny">systemPipeShiny</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="initialize-sps-project" class="section level2">
<h2 class="hasAnchor">
<a href="#initialize-sps-project" class="anchor"></a>Initialize <code>SPS</code> project</h2>
<p>Before launching the <code>SPS</code> application, a project environment needs to be created with the following command.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/spsInit.html">spsInit</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>For this toy example, the project directory structure is written to a temporary directory on a user’s system. For a real project, it should be written to a defined and user controlled location on a system rather than a temporary directory.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sps_tmp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/spsInit.html">spsInit</a></span><span class="op">(</span>app_path <span class="op">=</span> <span class="va">sps_tmp_dir</span>, change_wd <span class="op">=</span> <span class="cn">FALSE</span>, project_name <span class="op">=</span> <span class="st">"SPSProject"</span><span class="op">)</span></code></pre></div>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Start to create a new SPS project</code></pre>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Create project under /tmp/RtmpnUJlw2/SPSProject</code></pre>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Now copy files</code></pre>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Create SPS database</code></pre>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Created SPS database method container</code></pre>
<pre><code>## [SPS-INFO] 2021-02-12 14:36:13 Creating SPS db...</code></pre>
<pre><code>## [SPS-DANGER] 2021-02-12 14:36:13 Db created at '/tmp/RtmpnUJlw2/SPSProject/config/sps.db'. DO NOT share this file with others
## [SPS-INFO] 2021-02-12 14:36:13 Key md5 2da9140eb6a2c6a0933a6daa26ad8d3e
## [SPS-INFO] 2021-02-12 14:36:13 SPS project setup done!</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sps_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">sps_tmp_dir</span>, <span class="st">"SPSProject"</span><span class="op">)</span></code></pre></div>
<div id="sps-project-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#sps-project-structure" class="anchor"></a>SPS project structure</h3>
<p>The file and directory structure of an SPS project is organized as follows.</p>
<pre>
SPS_xx/  
├── server.R               <span class="text-success">|</span> 
├── global.R               <span class="text-success">| Most important server, UI and global files, unless special needs, `global.R` is the only file you need to edit manually</span>   
├── ui.R                   <span class="text-success">|</span>  
├── deploy.R               <span class="text-info">| Deploy helper file</span>  
├── config                 <span class="text-success">| Important app config files. Do not edit them if you don't know</span>  
│   ├── sps.db             <span class="text-gray">| SPS database</span> 
│   ├── sps_options.yaml   <span class="text-gray">| SPS default option list</span> 
│   └── tabs.csv           <span class="text-gray">| SPS tab information</span> 
├── data                   <span class="text-primary">| App example data files</span> 
│   ├── xx.csv             
├── R                      <span class="text-primary">| All SPS additional tab files and helper R function files</span> 
│   ├── tab_xx.R            
├── README.md              
├── results                <span class="text-gray">| Not in use for this current version, you can store some data been generated from the app</span> 
│   └── README.md          
└── www                    <span class="text-primary">| Internet resources</span>  
    ├── about              <span class="text-gray">| About tab information</span> 
    │   └── xx.md          
    ├── css                <span class="text-info">| CSS files</span>  
    │   └── sps.css         
    ├── img                <span class="text-info">| App image resources</span>    
    │   └── xx.png         
    ├── js                 <span class="text-info">| Javascripts</span>
    │   └── xx.js           
    ├── loading_themes     <span class="text-info">| Loading screen files</span> 
    │   └── xx.html         
    └── plot_list          <span class="text-info">| Image files for plot gallery</span>  
        └── plot_xx.jpg      
</pre>
</div>
</div>
<div id="launch-sps" class="section level2">
<h2 class="hasAnchor">
<a href="#launch-sps" class="anchor"></a>Launch <code>SPS</code>
</h2>
<p>By default, the working directory will be set inside the project folder automatically. To launch the <code>SPS</code> Shiny application, one only needs to execute the following command.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html">runApp</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>After the SPS app has been launched, clicking the “Continue to app” button on the welcome screen will open the main dashboard (Fig.2).</p>
<center id="sps_ui">
<img src="img/sps_ui_main.jpg" title="fig:" alt="sps_ui">
</center>
<p><strong>Figure 2:</strong> Snapshot of SPS’ UI.</p>
<ol style="list-style-type: decimal">
<li>Welcome screen.</li>
<li>Module tabs.</li>
<li>User defined custom tabs.</li>
<li>The Canvas tab.</li>
<li>All SPS tabs has this description on top. It is highly recommend to click here to expand and read the full the description for the first time.</li>
</ol>
<p>Alternatively, when using RStudio one can click the <img src="img/app_start.png" style="display: inline;
padding-left:0; width:10px; height:15px;"><code>Run App</code> button in the top right corner.</p>
<p>In addition, in Rstudio the <em>global.R</em> file will be automatically opened when the <code>SPS</code> project is created. Custom changes can be made inside this file before the app launches. The <a href="#extend-sps">advanced section</a> explains how to change and create new custom tabs.</p>
</div>
<div id="workflow-management" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-management" class="anchor"></a>Workflow management</h2>
<p>The workflow management module in <code>SPS</code> allows one to modify or create the configuration files required for running data analysis workflows in <a href="https://systempipe.org/docs/systemPipeR/">systemPipeR</a> (SPR). This includes three types of important files: a sample metadata (targets) file, a workflow file (in R Markdown format) defining the workflow steps, and workflow running files in <a href="https://www.commonwl.org/">Common Workflow Language (CWL)</a> format. In SPS, one can easily create these files under the “Workflow Management” module, located in navigation bar on the left of the dashboard (Fig2 - 2).</p>
<p>The current version of <code>SPS</code> allows to:</p>
<ol style="list-style-type: decimal">
<li>create a workflow environment;</li>
<li>create and/or check the format of targets / workflow / CWL files;</li>
<li>download the prepared workflow files to run elsewhere, like a cluster;</li>
<li>directly execute the workflow from SPS.</li>
</ol>
<div id="setup-a-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#setup-a-workflow" class="anchor"></a>1. setup a workflow</h3>
<center>
<img src="img/sps_wf_setup.jpg" title="fig:" alt="wf_setup">
</center>
<p><strong>Figure 3. A.</strong> Workflow Management - Targets File</p>
<ol style="list-style-type: decimal">
<li>In the workflow module, read the instructions and choose step 1. Step 1 should be automatically opened for you on start.</li>
<li>Choose a folder where you want to create the workflow environment.</li>
<li>Choose a workflow template. These are SPR workflows and mainly are next-generation sequencing workflows.</li>
<li>Click “Gen workflow” to create the workflow project.</li>
<li>You should see a pop-up saying about the project path and other information. Clicking the pop-up will jump you to the step 2. The status tracker and banner for step 1 should all turn green.</li>
</ol>
</div>
<div id="prepare-a-target-file" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-a-target-file" class="anchor"></a>2. Prepare a target file</h3>
<p>The targets file defines all input file paths and other sample information of analysis workflows. To better undertand the structure of this file, one can consult the <a href="https://systempipe.org/docs/systemPipeR/#structure-of-targets-file">“Structure of targets file”</a> section in the SPR vignette. Essentially, this is the tabular file representation of the <code>colData</code> slot in an <code>SummarizedExperiment</code> object which stores sample IDs and other meta information.</p>
<p>The following step-by-step instructions explain how to create and/or modify targets files using RNA-Seq as an example (Fig.3 A):</p>
<ol style="list-style-type: decimal">
<li>Your project targets file is loaded for you, but you can choose to upload a different one.</li>
<li>You can edit, right click to add/remove rows/columns (The first row is treated as column names).</li>
<li>SPR target file includes a header block, that can also be edited in the SPS app. Each headers needs to start with a “#.” Header is only useful for RNA-Seq workflow in current SPR. You can define sample comparison groups here. Leave it as default for other projects.</li>
<li>The section on the left provides sample statistics and information whether files exist inside the workflow project’s <code>data</code> directory. Choose any column you want from the dropdown to check and watch the statistics bar change in this section.</li>
<li>statistic status bar.</li>
<li>Clicking on “Add to task” can help you to check if your target file has any formatting problem. You should see a green success pop-up if everything is right. Now your target file is ready and you can click “save” to download it and later use in a SPR project.</li>
</ol>
<center>
<img src="img/sps_ui_target.jpg" title="fig:" alt="wf_targets">
</center>
<p><strong>Figure 3. A.</strong> Workflow Management - Targets File</p>
</div>
<div id="prepare-a-workflow-file" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-a-workflow-file" class="anchor"></a>3. Prepare a workflow file</h3>
<p>In SPR, workflows are defined in Rmarkdown files, you can read details and obtain them <a href="https://systempipe.org/pages/pipelines_area/">here</a>.</p>
<p>Now let us follow the order below to see how SPS helps you to prepare a workflow file for a RNAseq project (Fig.3 B):</p>
<ol style="list-style-type: decimal">
<li>Your project workflow file is loaded for you, but you can choose to upload a different one.</li>
<li>The workflow structure is displayed in a tree-leaf-like plot.</li>
<li>Check all steps in the workflow that you want to include. You can skip (uncheck) some steps but it may cause the workflow to fail. Read more SPR instructions before do so.</li>
<li>Clicking on the “Plot steps” will show a flow chart of what the step execution orders will be when you run the workflow in SPR.</li>
<li>Clicking “Report preview” generates a preview of what the final report will look like for current RNAseq workflow (hidden in Fig 3.B), but in the preview, no code is evaluated.</li>
<li>If you are satisfied with your workflow file, click “Add to task” to save your workflow file.</li>
</ol>
<center>
<img src="img/sps_ui_wf.jpg" title="fig:" alt="wf_wf">
</center>
<p><strong>Figure 3. B.</strong> Workflow Management - Workflow File</p>
</div>
<div id="prepare-cwl-files-optional" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-cwl-files-optional" class="anchor"></a>4. Prepare CWL files (optional)</h3>
<p>In the new version of SPR, all individual system workflow steps are called by the CWL files. Each SPR workflow has a set of CWL files and normally users do not need to make any change. In case you want to learn more about CWL and create some new CWL files, Step 4 is a good place to practice.</p>
<p>To run a CWL step in SPR, 3 files are required:</p>
<ul>
<li>targets: to determine how many samples will be run and sample names.</li>
<li>CWL running file: can be translated to bash code;</li>
<li>CWL input: variables to inject into the running file</li>
</ul>
<p>SPR is the parser between R and CWL by injecting sample information from targets to <code>CWL input</code> file and then CWL parser translates it to bash code.</p>
<ol style="list-style-type: decimal">
<li>Most people are not familiar this part, so read instructions carefully.</li>
<li>Your project targets has been loaded for you, and an example CWL running and input for <strong>hisat2</strong> is also loaded for you. Directly parse the code. See what commandline code you get.</li>
<li>Change the targets injecting column, and parse again, see what has changed.</li>
<li>You can edit the CWL running and input files</li>
<li>Try to parse the new file and see what has changed.</li>
<li>If new CWL files has been created, you can edit workflow Rmd files by adding your new steps.</li>
</ol>
<center>
<img src="img/sps_wf_cwl.jpg" title="fig:" alt="wf_cwl">
</center>
<p><strong>Figure 3. C.</strong> Workflow Management - CWL File</p>
</div>
<div id="run-or-finish-workflow-preparation" class="section level3">
<h3 class="hasAnchor">
<a href="#run-or-finish-workflow-preparation" class="anchor"></a>5. Run or finish workflow preparation</h3>
<p>Up until this step, congratulations, the workflow is prepared. You can choose to download the workflow project files as a bundle or continue to run the workflow.</p>
<center>
<img src="img/sps_wf_run.jpg" title="fig:" alt="wf_cwl">
</center>
<p><strong>Figure 4.A.B</strong> Workflow Runner</p>
<ol style="list-style-type: decimal">
<li>On step 5 you can choose to download the prepared workflow or directly run the workflow. However, if you do not have the required commandline tools, workflow will most likely fail. Make sure you system has these tools (<a href="http://bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#261_Third-party_software_tools">Read about these tools</a>).</li>
<li>Open up the runner. It is a “Rstudio-like” interface.</li>
<li>Code editor. Required workflow running code is pre-entered for you. You can simply hit “Run” to start. Of course, you can delete the default code and run random R code.</li>
<li>Output R console.</li>
<li>Workflow running log.</li>
<li>View any plot output. and send a copy of your current plot to SPS Canvas tab or download it.</li>
</ol>
</div>
</div>
<div id="rna-seq-module" class="section level2">
<h2 class="hasAnchor">
<a href="#rna-seq-module" class="anchor"></a>RNA-Seq Module</h2>
<p>This is a module which takes a <strong>raw count table</strong> to do normalization, Differential gene expression (DEG) analysis, and finally helps users to generate different plots to visualize the results.</p>
<div id="prepare-metadata-and-count-table" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-metadata-and-count-table" class="anchor"></a>Prepare metadata and count table</h3>
<p>To start, we require two files, the metadata file (targets) and a raw count table (Fig. 5).</p>
<center>
<img src="img/sps_rna_data.jpg" title="fig:" alt="rna_data">
</center>
<p><strong>Figure 5</strong> RNAseq</p>
<ol style="list-style-type: decimal">
<li>This is the RNAseq module UI when you first click it. All sub-tabs are disbled at the beginning. Other tabs will enabled as you proceed with different options.</li>
<li>First, we need a metadata file to tell SPS what samples and conditions to use. Here, we use the metadata file from SPR, which is also known as “targets” file. If you are not familiar with the targets file, we suggest to use the workflow module step 2 to practice creating and checking the format. You can also use the example to see how it looks like.</li>
<li>The loaded targets table is display here. You can use the box below each column name to filter what samples to include/exclude. Only the “SampleName” and “Factor” columns are used, other columns are ignored. <code>SampleName</code> should be a unique character string without space for each row. <code>Factor</code> is the experiment design factors, or conditions, or treatments.</li>
<li>If you want to DEG analysis, DEG comparison groups are defined in the targets file header.</li>
<li>The header will be parsed into comparison groups which contain individual comparisons. If the parsed comparison is not what you want, edit the header lines and reupload.</li>
<li>If everything is expected, confirm to use this table.</li>
<li>You should see the progress timeline of step 1 becomes green if your targets and header pass the format checking.</li>
<li>(Not on figure) Similarly, use example or upload a count table and confirm to use it.</li>
</ol>
<p>Note: For the count table, the first column will be used as gene names. Other column names will be treated as sample names, and values in these columns are treated as raw counts. Make sure columns except the first one are <strong>numeric</strong>, and replace <code>NA</code> with <code>0</code>.</p>
<p>Upon successfully confirm targets and count table, you should see the “Normalize Data” subtab is enabled. You can click on the top navigation or click the pop-up for the next step.</p>
</div>
<div id="process-raw-count" class="section level3">
<h3 class="hasAnchor">
<a href="#process-raw-count" class="anchor"></a>Process raw count</h3>
<p>If this UI is displayed, that means your targets and count table are accepted by SPS (Fig 6). On this sub-tab, you can choose:</p>
<ol style="list-style-type: decimal">
<li>Transform your count data with “raw,” “rlog” or “VST” and visualize the results in other sub-tabs.</li>
<li>Do DEG analysis.</li>
</ol>
<p>These two options are independent.</p>
<center>
<img src="img/sps_rna_norm.jpg" title="fig:" alt="rna_norm">
</center>
<p><strong>Figure 6</strong> RNAseq Normalization</p>
<ol style="list-style-type: decimal">
<li>At step 1 panel, choose how SPS can help you, count transformation or DEG analysis. The former will jump you to step 2, latter will jump to step 3.</li>
<li>There are many options. If you are not clear, hover your mouse on the option, and some tips will show up.</li>
<li>To start data transformation or DEG analysis.</li>
<li>A gallery of different plot options will show up when the data process is done.</li>
<li>When the data process is done, you can download results from the right side panel. Check all items you want and SPS will help you to zip it into one file to download.</li>
<li>If at least one item is checked, downloading is enabled.</li>
<li>Progress timeline will also change upon successful data process.</li>
<li>Different visualization options will be enabled depending on the data process options.</li>
</ol>
</div>
<div id="plot-options" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-options" class="anchor"></a>Plot options</h3>
<p>SPS RNAseq module provides 6 different plot options to cluster transformed count table.</p>
<center>
<img src="img/sps_rna_plot.jpg" title="fig:" alt="rna_plot">
</center>
<p><strong>Figure 6</strong> RNAseq plots</p>
<ol style="list-style-type: decimal">
<li>Change plot options to customize your plots.</li>
<li>Most plots are <a href="https://plotly.com">Plotly</a> plots, which means you can interact with these plots, like hiding/show groups, zoom in/out, etc.</li>
<li>All SPS plots are resizable. Dragging the bottom-right corner icon to resize your plot.</li>
<li>Click “To canvas” to take a screenshot of current plot and edit it in <code>SPS Canvas</code> tab. Or clicking the down-arrow button to directly save current plot to a png or jpg.</li>
</ol>
</div>
<div id="deg-report" class="section level3">
<h3 class="hasAnchor">
<a href="#deg-report" class="anchor"></a>DEG report</h3>
<p>This is a special sub-tab designed to filter and visualize DEG results. This sub-tab can be accessed once the DEG is calculated on the “Normalize Data” sub-tab.</p>
<center>
<img src="img/sps_rna_deg.jpg" title="fig:" alt="rna_deg">
</center>
<p><strong>Figure 7</strong> RNAseq DEG</p>
<ol style="list-style-type: decimal">
<li>DEG summary plot. You can view what are the DEG results across different comparision groups.</li>
<li>Switch to view a ggplot friendly table. Different from the table you could download from “Normalize Data” subtab, this DEG table is rearranged so you can easily make a ggplot from it.</li>
<li>You can change the filter settings here, so DEGs will be re-filtered and you do not need to go back to “Normalize Data” subtab to recalculate DEG.</li>
<li>DEG plotting options. Choose from a volcano plot, an upset plot (intersection), a MA plot or a heatmap.</li>
</ol>
</div>
<div id="interact-with-other-bioconductor-packages-" class="section level3">
<h3 class="hasAnchor">
<a href="#interact-with-other-bioconductor-packages-" class="anchor"></a>Interact with other bioconductor packages.</h3>
<div id="locally" class="section level4">
<h4 class="hasAnchor">
<a href="#locally" class="anchor"></a>Locally</h4>
<p>If you are familiar with R and want to continue other analysis after these, simple stop SPS:</p>
<ol style="list-style-type: decimal">
<li>After count transformation, there is a <code>spsRNA_trans</code> object stored in your R environment. <code>raw</code> method gives you a normalized count table. Other two methods give you a <code>DESeq2</code> class object. You can use it for other analysis.</li>
<li>After DEG analysis, SPS stores a global object called <code>spsDEG.</code> It is a <code>summerizedExperiment</code> object which has all individual tables from all DEG comparisons. You can use it for other downstream analysis.</li>
</ol>
</div>
<div id="remotely" class="section level4">
<h4 class="hasAnchor">
<a href="#remotely" class="anchor"></a>Remotely</h4>
<p>If you are using SPS from a remote server, you can choose to download results from “Normalize Data” sub-tab. Choose results in tabular format or <code>summerizedExperiment</code> format which is saved in a <code>.rds</code> file.</p>
</div>
</div>
</div>
<div id="quick-ggplot-module" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-ggplot-module" class="anchor"></a>Quick {ggplot} module</h2>
<p>This module enables you to quickly upload datasets and make a {<a href="https://ggplot2.tidyverse.org/">ggplot</a>} in a second by using some functionalities from {<a href="https://dreamrs.github.io/esquisse/index.html">Esquisse</a>}.</p>
<center>
<img src="img/sps_quickgg.jpg" title="fig:" alt="quickgg">
</center>
<p><strong>Figure 8</strong> Quick ggplot</p>
<ol style="list-style-type: decimal">
<li>Provide a tabular data table by uploading or use example.</li>
<li>Drag variables from into different ggplot aesthetic boxes to make a ggplot.</li>
<li>Change to different plot types.</li>
<li>Customize other different plotting options.</li>
</ol>
<p>For a more specific guide, read <a href="https://dreamrs.github.io/esquisse/articles/get-started.html">Esquisse official guide</a>.</p>
</div>
<div id="canvas" class="section level2">
<h2 class="hasAnchor">
<a href="#canvas" class="anchor"></a>Canvas</h2>
<p>SPS Canvas is a place to display and edit scrennshots from different plots. To start to use Canvas, you need to take some screenshots but clicking “To Canvas” buttons on different tabs/modules. After clicking, the screenshots will be automatically sent from these places to this Canvas.</p>
<center>
<div class="figure">
<img src="img/sps_canvas.jpg" alt=""><p class="caption">plot_tab</p>
</div>
</center>
<p><strong>Figure 9</strong> Canvas</p>
<ol style="list-style-type: decimal">
<li>The Canvas area.</li>
<li>Canvas drawing grids. By default, your objects are limited to these drawing grids, but you can change it from top options inside “canvas.” The grid area size is automatically calculated to fit your screen size when you start SPS.</li>
<li>Object information. When you select any object on the Canvas, a bounding box will show to display the object’s dimensions, scale, angle and other information. You can disable them in the “View” menu</li>
<li>To edit your screenshots, simply drag your screenshots from left to Canvas working area.</li>
<li>You can add text or titles, and change the font color, decorations in this panel.</li>
<li>Different Canvas options. Several menus and buttons help you to better control the Canvas. Hover your mouse on buttons will display a tooltip of their functionality.</li>
</ol>
<p>Keyboard shortcuts are also enabled with SPS Canvas. Go to “help” menu to see these options.</p>
</div>
</div>
<div id="advanced-features" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-features" class="anchor"></a>Advanced features</h1>
<p>To read the detailed advanced features, please to go our <a href="https://systempipe.org/sps/">website</a>.</p>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version Information</h1>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] systemPipeShiny_1.1.12 shinyTree_0.2.7        shiny_1.6.0.9000      
## [4] BiocStyle_2.18.1      
## 
## loaded via a namespace (and not attached):
##   [1] shinydashboard_0.7.1        R.utils_2.10.1             
##   [3] esquisse_0.3.1              tidyselect_1.1.0           
##   [5] RSQLite_2.2.3               AnnotationDbi_1.52.0       
##   [7] htmlwidgets_1.5.3           grid_4.0.3                 
##   [9] BiocParallel_1.24.1         Rtsne_0.15                 
##  [11] base64url_1.4               munsell_0.5.0              
##  [13] codetools_0.2-18            ragg_0.4.1                 
##  [15] DT_0.17                     systemPipeR_1.24.3         
##  [17] miniUI_0.1.1.1              withr_2.4.1                
##  [19] shinytoastr_2.1.1           colorspace_2.0-0           
##  [21] Biobase_2.50.0              Category_2.56.0            
##  [23] highr_0.8                   knitr_1.31                 
##  [25] rstudioapi_0.13             stats4_4.0.3               
##  [27] shinyWidgets_0.5.6          MatrixGenerics_1.2.0       
##  [29] GenomeInfoDbData_1.2.4      hwriter_1.3.2              
##  [31] bit64_4.0.5                 pheatmap_1.0.12            
##  [33] rprojroot_2.0.2             batchtools_0.9.15          
##  [35] vctrs_0.3.6                 treeio_1.14.3              
##  [37] generics_0.1.0              xfun_0.20                  
##  [39] BiocFileCache_1.14.0        R6_2.5.0                   
##  [41] markdown_1.1                GenomeInfoDb_1.26.2        
##  [43] locfit_1.5-9.4              rsvg_2.1                   
##  [45] bitops_1.0-6                cachem_1.0.1               
##  [47] shinyAce_0.4.1              DelayedArray_0.16.1        
##  [49] assertthat_0.2.1            vroom_1.3.2                
##  [51] promises_1.1.1              networkD3_0.4              
##  [53] scales_1.1.1                gtable_0.3.0               
##  [55] processx_3.4.5              systemPipeRdata_1.18.1     
##  [57] rlang_0.4.10                genefilter_1.72.1          
##  [59] systemfonts_1.0.1           splines_4.0.3              
##  [61] rtracklayer_1.50.0          lazyeval_0.2.2             
##  [63] shinyjqui_0.3.3             brew_1.0-6                 
##  [65] checkmate_2.0.0             bsplus_0.1.2               
##  [67] BiocManager_1.30.10         yaml_2.2.1                 
##  [69] GenomicFeatures_1.42.1      backports_1.2.1            
##  [71] httpuv_1.5.5                RBGL_1.66.0                
##  [73] tools_4.0.3                 bookdown_0.21.6            
##  [75] ggplot2_3.3.3               ellipsis_0.3.1             
##  [77] RColorBrewer_1.1-2          BiocGenerics_0.36.0        
##  [79] Rcpp_1.0.6                  plyr_1.8.6                 
##  [81] progress_1.2.2              zlibbioc_1.36.0            
##  [83] purrr_0.3.4                 RCurl_1.98-1.2             
##  [85] ps_1.5.0                    prettyunits_1.1.1          
##  [87] openssl_1.4.3               S4Vectors_0.28.1           
##  [89] SummarizedExperiment_1.20.0 fs_1.5.0                   
##  [91] magrittr_2.0.1              data.table_1.13.6          
##  [93] colourpicker_1.1.0          matrixStats_0.58.0         
##  [95] hms_1.0.0                   patchwork_1.1.1            
##  [97] shinyjs_2.0.0               mime_0.9                   
##  [99] evaluate_0.14               xtable_1.8-4               
## [101] XML_3.99-0.5                jpeg_0.1-8.1               
## [103] IRanges_2.24.1              gridExtra_2.3              
## [105] compiler_4.0.3              biomaRt_2.46.1             
## [107] tibble_3.0.6                V8_3.4.0                   
## [109] crayon_1.3.4                R.oo_1.24.0                
## [111] htmltools_0.5.1.1           GOstats_2.56.0             
## [113] later_1.1.0.1               tidyr_1.1.2                
## [115] geneplotter_1.68.0          aplot_0.0.6                
## [117] lubridate_1.7.9.2           DBI_1.1.1                  
## [119] formatR_1.7                 dbplyr_2.0.0               
## [121] MASS_7.3-53                 rappdirs_0.3.2             
## [123] ShortRead_1.48.0            Matrix_1.3-2               
## [125] readr_1.4.0                 glmpca_0.2.0               
## [127] R.methodsS3_1.8.1           parallel_4.0.3             
## [129] igraph_1.2.6                GenomicRanges_1.42.0       
## [131] pkgconfig_2.0.3             pkgdown_1.6.1              
## [133] rvcheck_0.1.8               GenomicAlignments_1.26.0   
## [135] plotly_4.9.3                xml2_1.3.2                 
## [137] ggtree_2.4.1                annotate_1.68.0            
## [139] XVector_0.30.0              AnnotationForge_1.32.0     
## [141] VariantAnnotation_1.36.0    stringr_1.4.0              
## [143] callr_3.5.1                 digest_0.6.27              
## [145] graph_1.68.0                Biostrings_2.58.0          
## [147] rmarkdown_2.6               tidytree_0.3.3             
## [149] edgeR_3.32.1                GSEABase_1.52.1            
## [151] curl_4.3                    Rsamtools_2.6.0            
## [153] rjson_0.2.20                lifecycle_0.2.0            
## [155] nlme_3.1-151                jsonlite_1.7.2             
## [157] BSgenome_1.58.0             desc_1.2.0                 
## [159] viridisLite_0.3.0           askpass_1.1                
## [161] limma_3.46.0                pillar_1.4.7               
## [163] lattice_0.20-41             shinyFiles_0.9.0           
## [165] fastmap_1.1.0               httr_1.4.2                 
## [167] survival_3.2-7              GO.db_3.12.1               
## [169] remotes_2.2.0               glue_1.4.2                 
## [171] zip_2.1.1                   UpSetR_1.4.0               
## [173] png_0.1-7                   rhandsontable_0.3.7        
## [175] Rgraphviz_2.34.0            bit_4.0.4                  
## [177] stringi_1.5.3               blob_1.2.1                 
## [179] textshaping_0.2.1           DESeq2_1.30.0              
## [181] latticeExtra_0.6-29         shinydashboardPlus_0.7.5   
## [183] memoise_2.0.0               DOT_0.1                    
## [185] styler_1.3.2                dplyr_1.0.3                
## [187] ape_5.4-1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://lezhang.me/">Le Zhang</a>, <a href="https://www.dcassol.com/">Daniela Cassol</a>, <a href="https://github.com/mathrj">Ponmathi Ramasamy</a>, <a href="https://github.com/jianhaizhang">Jianhai Zhang</a>, <a href="https://github.com/gdmosher">Gordon Mosher</a>, <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
