<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.27.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="EnvModules-class.html">EnvModules-class</a></code> </p>
        </td>
        <td><p>Class <code>"EnvModules"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GOHyperGAll.html">makeCATdb()</a></code> <code><a href="GOHyperGAll.html">GOHyperGAll()</a></code> <code><a href="GOHyperGAll.html">GOHyperGAll_Subset()</a></code> <code><a href="GOHyperGAll.html">GOHyperGAll_Simplify()</a></code> <code><a href="GOHyperGAll.html">GOCluster_Report()</a></code> <code><a href="GOHyperGAll.html">goBarplot()</a></code> </p>
        </td>
        <td><p>GO term enrichment analysis for large numbers of gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="INTERSECTset-class.html">INTERSECTset-class</a></code> </p>
        </td>
        <td><p>Class <code>"INTERSECTset"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="LineWise-class.html">LineWise-class</a></code> </p>
        </td>
        <td><p>Class <code>"LineWise"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="LineWise.html">LineWise()</a></code> </p>
        </td>
        <td><p>LineWise</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ParamFiles.html">printParam()</a></code> <code><a href="ParamFiles.html">subsetParam()</a></code> <code><a href="ParamFiles.html">replaceParam()</a></code> <code><a href="ParamFiles.html">renameParam()</a></code> <code><a href="ParamFiles.html">appendParam()</a></code> </p>
        </td>
        <td><p>Accessories function to modify the Command-line</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SPRdata.html">SPRdata()</a></code> </p>
        </td>
        <td><p>SPRdata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SPRproject.html">SPRproject()</a></code> </p>
        </td>
        <td><p>Workflow Project Initiation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SYSargs-class.html">SYSargs-class</a></code> </p>
        </td>
        <td><p>Class <code>"SYSargs"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SYSargs2-class.html">SYSargs2-class</a></code> </p>
        </td>
        <td><p>Class <code>"SYSargs2"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SYSargsList-class.html">SYSargsList-class</a></code> </p>
        </td>
        <td><p>Class <code>"SYSargsList"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SYSargsList.html">SYSargsList()</a></code> </p>
        </td>
        <td><p>SYSargsList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VENNset-class.html">VENNset-class</a></code> </p>
        </td>
        <td><p>Class <code>"VENNset"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add-methods.html">addAssay()</a></code> <code><a href="add-methods.html">addMetadata()</a></code> </p>
        </td>
        <td><p>Extension accessor methods for SummarizedExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="alignStats.html">alignStats()</a></code> </p>
        </td>
        <td><p>Alignment statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="catDB-class.html">catDB-class</a></code> </p>
        </td>
        <td><p>Class <code>"catDB"</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="catmap.html">catmap()</a></code> </p>
        </td>
        <td><p>catDB accessor methods</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check.output.html">check.output()</a></code> </p>
        </td>
        <td><p>Checking if the output files exist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterRun.html">clusterRun()</a></code> </p>
        </td>
        <td><p>Submit command-line tools to cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="config.param.html">config.param()</a></code> </p>
        </td>
        <td><p>Adding param file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="configWF.html">configWF()</a></code> </p>
        </td>
        <td><p>Workflow Steps Selection</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countRangeset.html">countRangeset()</a></code> </p>
        </td>
        <td><p>Read counting for several range sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createParamFiles.html">createParamFiles()</a></code> <code><a href="createParamFiles.html">writeParamFiles()</a></code> </p>
        </td>
        <td><p>createParamFiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="evalCode.html">evalCode()</a></code> </p>
        </td>
        <td><p>Toggles option <code>eval</code> on the RMarkdown files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="featureCoverage.html">featureCoverage()</a></code> </p>
        </td>
        <td><p>Genome read coverage by transcript models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="featuretypeCounts.html">featuretypeCounts()</a></code> </p>
        </td>
        <td><p>Plot read distribution across genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterDEGs.html">filterDEGs()</a></code> </p>
        </td>
        <td><p>Filter and plot DEG results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterVars.html">filterVars()</a></code> </p>
        </td>
        <td><p>Filter VCF files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genFeatures.html">genFeatures()</a></code> </p>
        </td>
        <td><p>Generate feature ranges from TxDb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getQsubargs.html">getQsubargs()</a></code> </p>
        </td>
        <td><p>Arguments for qsub</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importWF.html">importWF()</a></code> </p>
        </td>
        <td><p>Import RMarkdown file as workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadWorkflow.html">loadWorkflow()</a></code> </p>
        </td>
        <td><p>Constructs SYSargs2 object from CWL param and targets files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeBamByFactor.html">mergeBamByFactor()</a></code> </p>
        </td>
        <td><p>Merge BAM files based on factor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="moduleload.html">module()</a></code> <code><a href="moduleload.html">moduleload()</a></code> <code><a href="moduleload.html">moduleUnload()</a></code> <code><a href="moduleload.html">modulelist()</a></code> <code><a href="moduleload.html">moduleAvail()</a></code> <code><a href="moduleload.html">moduleClear()</a></code> <code><a href="moduleload.html">moduleInit()</a></code> </p>
        </td>
        <td><p>Interface to allow full use of the Environment Modules system for Unix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="olBarplot.html">olBarplot()</a></code> </p>
        </td>
        <td><p>Bar plot for intersect sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="olRanges.html">olRanges()</a></code> </p>
        </td>
        <td><p>Identify Range Overlaps for IRanges and GRanges Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="output_update.html">output_update()</a></code> </p>
        </td>
        <td><p>Updates the output files paths in the <code>SYSargs2</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overLapper.html">overLapper()</a></code> </p>
        </td>
        <td><p>Set Intersect and Venn Diagram Functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotWF.html">plotWF()</a></code> </p>
        </td>
        <td><p>Graphviz Plot Workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotfeatureCoverage.html">plotfeatureCoverage()</a></code> </p>
        </td>
        <td><p>Plot feature coverage results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotfeaturetypeCounts.html">plotfeaturetypeCounts()</a></code> </p>
        </td>
        <td><p>Plot read distribution across genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="predORF.html">predORF()</a></code> </p>
        </td>
        <td><p>Predict ORFs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocessReads.html">preprocessReads()</a></code> </p>
        </td>
        <td><p>Run custom read preprocessing functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qsubRun.html">qsubRun()</a></code> </p>
        </td>
        <td><p>Submit command-line tools to cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readComp.html">readComp()</a></code> </p>
        </td>
        <td><p>Import sample comparisons from targets file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="renderReport.html">renderReport()</a></code> </p>
        </td>
        <td><p>Render RMarkdown Report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="renderWF.html">renderWF()</a></code> </p>
        </td>
        <td><p>Populate all the command-line in an <code>SYSargs2</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="returnRPKM.html">returnRPKM()</a></code> </p>
        </td>
        <td><p>RPKM Normalization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runCommandline.html">runCommandline()</a></code> </p>
        </td>
        <td><p>Execute SYSargs and SYSargs2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDiff.html">runDiff()</a></code> </p>
        </td>
        <td><p>Differential abundance analysis for many range sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runWF.html">runWF()</a></code> </p>
        </td>
        <td><p>Execute SYSargsList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_DESeq2.html">run_DESeq2()</a></code> </p>
        </td>
        <td><p>Runs DESeq2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_edgeR.html">run_edgeR()</a></code> </p>
        </td>
        <td><p>Runs edgeR</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaleRanges.html">scaleRanges()</a></code> </p>
        </td>
        <td><p>Scale spliced ranges to genome coordinates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seeFastq.html">seeFastq()</a></code> <code><a href="seeFastq.html">seeFastqPlot()</a></code> </p>
        </td>
        <td><p>Quality reports for FASTQ files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetWF.html">subsetWF()</a></code> </p>
        </td>
        <td><p>Subsetting SYSargs2 class slots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="symLink2bam.html">symLink2bam()</a></code> </p>
        </td>
        <td><p>Symbolic links for IGV</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sysargs.html">sysargs()</a></code> </p>
        </td>
        <td><p>SYSargs accessor methods</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="systemArgs.html">systemArgs()</a></code> </p>
        </td>
        <td><p>Constructs SYSargs object from param and targets files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="systemPipeR-package.html">systemPipeR-package</a></code> </p>
        </td>
        <td><p>systemPipeR package for Workflow Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="targets.as.df.html">targets.as.df()</a></code> </p>
        </td>
        <td><p>Convert targets list to data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tryCL.html">tryCL()</a></code> </p>
        </td>
        <td><p>Collect information about the third-party software</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tryPath.html">tryPath()</a></code> </p>
        </td>
        <td><p>Validation of the files or directories</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="variantReport.html">variantReport()</a></code> <code><a href="variantReport.html">combineVarReports()</a></code> <code><a href="variantReport.html">varSummary()</a></code> </p>
        </td>
        <td><p>Generate Variant Report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vennPlot.html">vennPlot()</a></code> </p>
        </td>
        <td><p>Plot 2-5 way Venn diagrams</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeTargetsRef.html">writeTargetsRef()</a></code> </p>
        </td>
        <td><p>Generate targets file with reference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeTargetsout.html">writeTargetsout()</a></code> </p>
        </td>
        <td><p>Write updated targets out to file</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


