<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot feature coverage results — plotfeatureCoverage • systemPipeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot feature coverage results — plotfeatureCoverage" />
<meta property="og:description" content="Plots the 3 tabular data types (A-C) generated by the featureCoverage 
function. It accepts data from single or many features (e.g. CDSs) and
samples (BAM files). The coverage from multiple features will be summarized
using methods such as mean, while the data from multiple samples will 
be plotted in separate panels." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.27.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://systempipe.org/spr/">
    <span class="fa fa-home"></span>
     
    SPR Home
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot feature coverage results</h1>
    
    <div class="hidden name"><code>plotfeatureCoverage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots the 3 tabular data types (A-C) generated by the <code>featureCoverage</code> 
function. It accepts data from single or many features (e.g. CDSs) and
samples (BAM files). The coverage from multiple features will be summarized
using methods such as <code>mean</code>, while the data from multiple samples will 
be plotted in separate panels.</p>
    </div>

    <pre class="usage"><span class='fu'>plotfeatureCoverage</span><span class='op'>(</span><span class='va'>covMA</span>, method <span class='op'>=</span> <span class='va'>mean</span>, scales <span class='op'>=</span> <span class='st'>"fixed"</span>, extendylim<span class='op'>=</span><span class='fl'>2</span>, scale_count_val <span class='op'>=</span> <span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>covMA</th>
      <td><p>Object of class <code>data.frame</code> generated by <code>featureCoverage</code> function.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Defines the summary statistics to use when <code>covMA</code> contains coverage 
data from  multiple features (e.g. transcripts). The default calculates the 
mean coverage for each positon and/or bin of the corresponding coverage 
vectors.</p></td>
    </tr>
    <tr>
      <th>scales</th>
      <td><p>Scales setting passed on to the <code>facet_wrap</code> function of <code>ggplot2</code>. 
For details see <code><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>ggplot2::facet_wrap</a></code>. The default <code>fixed</code> assures a
constant scale across all bar plot panels, while <code>free</code> uses the optimum
scale within each bar plot panel. To evaluate plots in all their details, it
may be necessary to generate two graphics files one for each scaling option.</p></td>
    </tr>
    <tr>
      <th>extendylim</th>
      <td><p>Allows to extend the upper limit of the y axis when <code>scales=fixed</code>. Internally, the 
function identifies the maximum value in the data and then multiplies this maximum value 
by the value provided under <code>extendylim</code>. The default is set to <code>extendylim=2</code>.</p></td>
    </tr>
    <tr>
      <th>scale_count_val</th>
      <td><p>Scales (normalizes) the read counts to a fixed value of aligned reads
in each sample such as counts per million aligned reads (default is 10^6). 
For this calculation the <code>N_total_aligned</code> values are used that are 
reported in the input <code>data.frame</code> generated by the upstream 
<code>featureCoverage</code> function. Assign <code>NULL</code> to turn off
scaling.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Currently, the function returns ggplot2 bar plot graphics.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thomas Girke</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>featureCoverage</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Construct SYSargs object from param and targets files </span>
<span class='va'>param</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"tophat.param"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>targets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"targets.txt"</span>, package<span class='op'>=</span><span class='st'>"systemPipeR"</span><span class='op'>)</span>
<span class='va'>args</span> <span class='op'>&lt;-</span> <span class='fu'><a href='systemArgs.html'>systemArgs</a></span><span class='op'>(</span>sysma<span class='op'>=</span><span class='va'>param</span>, mytargets<span class='op'>=</span><span class='va'>targets</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: path[1]="./data/SRR446027_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[2]="./data/SRR446028_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[3]="./data/SRR446029_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[4]="./data/SRR446030_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[5]="./data/SRR446031_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[6]="./data/SRR446032_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[7]="./data/SRR446033_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[8]="./data/SRR446034_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[9]="./data/SRR446035_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[10]="./data/SRR446036_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[11]="./data/SRR446037_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[12]="./data/SRR446038_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[13]="./data/SRR446039_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[14]="./data/SRR446040_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[15]="./data/SRR446041_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[16]="./data/SRR446042_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[17]="./data/SRR446043_1.fastq.gz": No such file or directory</span></div><div class='output co'>#&gt; <span class='warning'>Warning: path[18]="./data/SRR446044_1.fastq.gz": No such file or directory</span></div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'>## Features from sample data of systemPipeRdata package</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicFeatures'>GenomicFeatures</a></span><span class='op'>)</span>
<span class='va'>file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/annotation"</span>, <span class='st'>"tair10.gff"</span>, package<span class='op'>=</span><span class='st'>"systemPipeRdata"</span><span class='op'>)</span>
<span class='va'>txdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span>file<span class='op'>=</span><span class='va'>file</span>, format<span class='op'>=</span><span class='st'>"gff3"</span>, organism<span class='op'>=</span><span class='st'>"Arabidopsis"</span><span class='op'>)</span>

<span class='co'>## (A) Generate binned coverage for two BAM files and 4 transcripts</span>
<span class='va'>grl</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html'>cdsBy</a></span><span class='op'>(</span><span class='va'>txdb</span>, <span class='st'>"tx"</span>, use.names<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>fcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='featureCoverage.html'>featureCoverage</a></span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>grl</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span>, resizereads<span class='op'>=</span><span class='cn'>NULL</span>, 
                    readlengthrange<span class='op'>=</span><span class='cn'>NULL</span>, Nbins<span class='op'>=</span><span class='fl'>20</span>, method<span class='op'>=</span><span class='va'>mean</span>, fixedmatrix<span class='op'>=</span><span class='cn'>FALSE</span>, 
                    resizefeatures<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>20</span>, downstream<span class='op'>=</span><span class='fl'>20</span><span class='op'>)</span>
<span class='va'>fcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='featureCoverage.html'>featureCoverage</a></span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>grl</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span>, resizereads<span class='op'>=</span><span class='cn'>NULL</span>, 
                    readlengthrange<span class='op'>=</span><span class='cn'>NULL</span>, Nbins<span class='op'>=</span><span class='fl'>20</span>, method<span class='op'>=</span><span class='va'>mean</span>, fixedmatrix<span class='op'>=</span><span class='cn'>TRUE</span>, 
                    resizefeatures<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>20</span>, downstream<span class='op'>=</span><span class='fl'>20</span><span class='op'>)</span>
<span class='fu'>plotfeatureCoverage</span><span class='op'>(</span>covMA<span class='op'>=</span><span class='va'>fcov</span>, method<span class='op'>=</span><span class='va'>mean</span>, scales<span class='op'>=</span><span class='st'>"fixed"</span>, scale_count_val<span class='op'>=</span><span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span><span class='op'>)</span>

<span class='co'>## (B) Coverage matrix upstream and downstream of start/stop codons</span>
<span class='va'>fcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='featureCoverage.html'>featureCoverage</a></span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>grl</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span>, resizereads<span class='op'>=</span><span class='cn'>NULL</span>, 
                    readlengthrange<span class='op'>=</span><span class='cn'>NULL</span>, Nbins<span class='op'>=</span><span class='cn'>NULL</span>, method<span class='op'>=</span><span class='va'>mean</span>, fixedmatrix<span class='op'>=</span><span class='cn'>TRUE</span>, 
                    resizefeatures<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>20</span>, downstream<span class='op'>=</span><span class='fl'>20</span><span class='op'>)</span>
<span class='fu'>plotfeatureCoverage</span><span class='op'>(</span>covMA<span class='op'>=</span><span class='va'>fcov</span>, method<span class='op'>=</span><span class='va'>mean</span>, scales<span class='op'>=</span><span class='st'>"fixed"</span>, scale_count_val<span class='op'>=</span><span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span><span class='op'>)</span>

<span class='co'>## (C) Combined matrix for both binned and start/stop codon </span>
<span class='va'>fcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='featureCoverage.html'>featureCoverage</a></span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>grl</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span>, resizereads<span class='op'>=</span><span class='cn'>NULL</span>, 
                    readlengthrange<span class='op'>=</span><span class='cn'>NULL</span>, Nbins<span class='op'>=</span><span class='fl'>20</span>, method<span class='op'>=</span><span class='va'>mean</span>, fixedmatrix<span class='op'>=</span><span class='cn'>TRUE</span>, 
                    resizefeatures<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>20</span>, downstream<span class='op'>=</span><span class='fl'>20</span>, outfile<span class='op'>=</span><span class='st'>"results/test.xls"</span><span class='op'>)</span>
<span class='fu'>plotfeatureCoverage</span><span class='op'>(</span>covMA<span class='op'>=</span><span class='va'>fcov</span>, method<span class='op'>=</span><span class='va'>mean</span>, scales<span class='op'>=</span><span class='st'>"fixed"</span>, scale_count_val<span class='op'>=</span><span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span><span class='op'>)</span>

<span class='co'>## (D) Rle coverage objects one for each query feature</span>
<span class='va'>fcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='featureCoverage.html'>featureCoverage</a></span><span class='op'>(</span>bfl<span class='op'>=</span><span class='fu'>BamFileList</span><span class='op'>(</span><span class='fu'><a href='sysargs.html'>outpaths</a></span><span class='op'>(</span><span class='va'>args</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, grl<span class='op'>=</span><span class='va'>grl</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span>, resizereads<span class='op'>=</span><span class='cn'>NULL</span>, 
                    readlengthrange<span class='op'>=</span><span class='cn'>NULL</span>, Nbins<span class='op'>=</span><span class='cn'>NULL</span>, method<span class='op'>=</span><span class='va'>mean</span>, fixedmatrix<span class='op'>=</span><span class='cn'>FALSE</span>, 
                    resizefeatures<span class='op'>=</span><span class='cn'>TRUE</span>, upstream<span class='op'>=</span><span class='fl'>20</span>, downstream<span class='op'>=</span><span class='fl'>20</span><span class='op'>)</span>

<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://girke.bioinformatics.ucr.edu/'>Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


