<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../../favicons/favicon.ico><link rel=apple-touch-icon href=../../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../../favicons/android-192x192.png sizes=192x192><title>_systemPipeR_ | sysPipe</title><meta property="og:title" content="_systemPipeR_"><meta property="og:description" content="layout: true background-image: url(https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png) background-position: 99% 1% background-size: 10%
 class: middle
Outline Introduction Design How to run a Workflow Workflows Tutorial Live Demo  class: inverse, center, middle
 Introduction  Introduction  systemPipeR provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports
 systemPipeR offers many utilities to build, control, and execute workflows entirely from R"><meta property="og:type" content="article"><meta property="og:url" content="/presentations/spr/spr_project/"><meta property="article:modified_time" content="2022-11-03T18:09:25-07:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="_systemPipeR_"><meta itemprop=description content="layout: true background-image: url(https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png) background-position: 99% 1% background-size: 10%
 class: middle
Outline Introduction Design How to run a Workflow Workflows Tutorial Live Demo  class: inverse, center, middle
 Introduction  Introduction  systemPipeR provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports
 systemPipeR offers many utilities to build, control, and execute workflows entirely from R"><meta itemprop=dateModified content="2022-11-03T18:09:25-07:00"><meta itemprop=wordCount content="2138"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="_systemPipeR_"><meta name=twitter:description content="layout: true background-image: url(https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png) background-position: 99% 1% background-size: 10%
 class: middle
Outline Introduction Design How to run a Workflow Workflows Tutorial Live Demo  class: inverse, center, middle
 Introduction  Introduction  systemPipeR provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports
 systemPipeR offers many utilities to build, control, and execute workflows entirely from R"><link rel=preload href=../../../scss/main.min.3f0edcebe8aec089d420546012b86e51389077da5d39bfa7ecd52911e400e479.css as=style><link href=../../../scss/main.min.3f0edcebe8aec089d420546012b86e51389077da5d39bfa7ecd52911e400e479.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><script src=https://unpkg.com/lunr@2.3.8/lunr.min.js integrity=sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY crossorigin=anonymous></script><link rel=stylesheet href=../../../css/prism.css><script src=../../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../../css/driver0.9.8.min.css><link rel=stylesheet href=../../../css/site.css><script src=../../../js/site.js></script></head><body class=td-page><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../../><span class=navbar-logo><svg xmlns="http://www.w3.org/2000/svg" xmlns:svg="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" id="Layer_1" width="55" height="59" enable-background="new 0 0 128 128" viewBox="0 0 55 59"><defs id="defs217"><linearGradient id="linearGradient2530"><stop style="stop-color:#a6dae8;stop-opacity:1" id="stop2526" offset="0"/><stop style="stop-color:#a6dae8;stop-opacity:0" id="stop2528" offset="1"/></linearGradient><radialGradient id="radialGradient2534" cx="48.106" cy="96.236" r="19.564" fx="48.106" fy="96.236" gradientTransform="matrix(1,0,0,1.623578,0,-60.010935)" gradientUnits="userSpaceOnUse" xlink:href="#linearGradient2530"/><meshgradient id="meshgradient2632" x="17.689" y="-106.706" gradientTransform="matrix(0.45556163,0,0,0.45556163,69.688068,2.0372999)" gradientUnits="userSpaceOnUse"><meshrow id="meshrow2634"><meshpatch id="meshpatch2636"><stop style="stop-color:#fff;stop-opacity:1" id="stop2638" path="c 23.5857,-23.5857 61.8257,-23.5857 85.4114,0"/><stop style="stop-color:#d1ecfb;stop-opacity:1" id="stop2640" path="c 23.5857,23.5857 23.5857,61.8257 0,85.4114"/><stop style="stop-color:#fff;stop-opacity:1" id="stop2642" path="c -23.5857,23.5857 -61.8257,23.5857 -85.4114,7.10543e-15"/><stop style="stop-color:#d1ecfb;stop-opacity:1" id="stop2644" path="c -23.5857,-23.5857 -23.5857,-61.8257 -1.42109e-14,-85.4114"/></meshpatch></meshrow></meshgradient></defs><g id="g212" transform="translate(0,-69)"><circle id="circle2" cx="97.202" cy="-27.119" r="27.514" transform="rotate(90)" style="opacity:.83600003;fill:url(#meshgradient2632);fill-opacity:1;stroke-width:.45556161"/><g id="g210" transform="matrix(0.45556163,0,0,0.45556163,-2.0372999,69.688068)"><path id="path52" d="m89.338 83.479-3.862.922c-.6-1.295-1.291-2.522-2.061-3.688l2.783-2.633c.666-.633.693-1.683.063-2.356l-4.57-4.83c-.635-.666-1.688-.693-2.354-.065l-2.895 2.737c-1.146-.813-2.356-1.539-3.617-2.162l1.098-3.682c.26-.883-.24-1.806-1.122-2.07l-6.374-1.901c-.88-.259-1.804.24-2.065 1.121l-1.136 3.813c-1.393-.133-2.803-.149-4.221-.06L58.114 64.888C57.9 63.994 57.005 63.44 56.108 63.654l-6.469 1.543c-.899.213-1.451 1.112-1.237 2.005l.921 3.867c-1.289.594-2.518 1.279-3.682 2.055L43.013 70.34c-.637-.671-1.688-.696-2.364-.063l-4.83 4.574c-.665.627-.695 1.681-.063 2.349l2.739 2.895c-.813 1.146-1.542 2.355-2.163 3.616l-3.679-1.098c-.885-.26-1.806.246-2.071 1.125l-1.905 6.375c-.264.877.24 1.806 1.122 2.068l3.812 1.135c-.128 1.39-.147 2.798-.058 4.215l-3.731.89c-.899.215-1.45 1.114-1.237 2.01l1.542 6.469c.213.896 1.109 1.447 2.01 1.232l3.862-.921c.592 1.292 1.28 2.528 2.054 3.688l-2.785 2.634c-.668.632-.694 1.682-.066 2.349l4.57 4.838c.629.664 1.687.694 2.354.063l2.892-2.736c1.151.812 2.361 1.541 3.622 2.165l-1.094 3.684c-.268.879.235 1.808 1.113 2.066l6.382 1.896c.878.263 1.806-.235 2.063-1.113l1.139-3.813c1.389.125 2.798.146 4.218.053l.893 3.74c.213.894 1.105 1.442 1.997 1.229l6.477-1.539c.895-.213 1.445-1.11 1.232-2.005l-.922-3.865c1.289-.597 2.521-1.281 3.688-2.055l2.633 2.779c.632.672 1.677.697 2.351.066l4.836-4.572c.666-.627.692-1.683.063-2.354l-2.739-2.891c.815-1.147 1.544-2.356 2.169-3.619l3.682 1.1c.883.26 1.804-.242 2.068-1.121l1.899-6.379c.262-.877-.243-1.805-1.116-2.066l-3.814-1.141c.131-1.384.147-2.794.063-4.21l3.734-.891c.893-.213 1.438-1.11 1.228-2.008l-1.541-6.47C91.13 83.815 90.229 83.269 89.338 83.479zm-25.89 23.716c-6.292 1.498-12.605-2.387-14.105-8.678-1.499-6.29 2.386-12.604 8.677-14.104 6.291-1.499 12.606 2.385 14.106 8.677 1.498 6.289-2.387 12.606-8.678 14.105z" style="fill:#2e4b79;fill-opacity:.91176471"/><path id="path54" d="m111.8295 28.871996-3.24301.776905c-.50038-1.091866-1.07658-2.130285-1.72291-3.113347l2.33322-2.223819c.55914-.53448.58189-1.422099.0474-1.989029l-3.83059-4.079229c-.52881-.561204-1.41396-.58602-1.9731-.05536l-2.42231 2.309717c-.96475-.68528-1.976891-1.298023-3.03831-1.824867l.919264-3.10762c.216072-.746364-.200913-1.525178-.940115-1.750423l-5.34121-1.60344c-.739202-.21952-1.510626.200429-1.732387.946793L89.930165 16.37852C88.7645 16.265897 87.583671 16.250626 86.395262 16.32889l-.746784-3.157247c-.181958-.753999-.930637-1.21976-1.683107-1.040328l-5.426502 1.301842c-.754365.181341-1.21684.941066-1.036778 1.695065l.773319 3.264147c-1.082268.500121-2.111464 1.080414-3.091381 1.735151l-2.200548-2.353621c-.53071-.566931-1.412067-.587928-1.978788-.05536l-4.050448 3.863528c-.559141.528753-.585677 1.42019-.05117 1.981394l2.295318 2.445247c-.680446.967791-1.296448 1.989029-1.810099 3.054172l-3.087592-.927705c-.741098-.21761-1.514416.208065-1.736178.950611l-1.595918 5.382979c-.221761.742546.19712 1.525177.936323 1.748513l3.197523.958246c-.109933 1.172039-.1232 2.363166-.04549 3.558111l-3.127393.752091c-.750574.18134-1.218736.939158-1.036779 1.695065l1.294552 5.463151c.180062.757816.930637 1.225486 1.686897 1.044144l3.235431-.776905c.500384 1.089958 1.074687 2.132194 1.719119 3.111438l-2.329435 2.225728c-.56293.532571-.581884 1.42019-.05307 1.985212l3.830584 4.083046c.526918.561204 1.413961.587928 1.9731.05536l2.422309-2.311626c.968544.687189 1.984474 1.30184 3.040206 1.828686l-.917368 3.109528c-.219866.742546.197119 1.527086.936322 1.746605l5.348791 1.601532c.737307.221428 1.514417-.198521 1.728596-.941067l.955277-3.222151c1.165665.106897 2.346493.124075 3.538695.04581l.746783 3.157251c.180062.753998.930636 1.217851 1.67742 1.038418l5.428398-1.301841c.750574-.177523 1.213049-.937249 1.031092-1.691248l-.771424-3.264146c1.082268-.503939 2.115255-1.082322 3.091385-1.735152l2.20813 2.345986c.5326.566931 1.40827.589838 1.97499.05726l4.05235-3.86162c.55914-.528753.58567-1.42019.0531-1.98712l-2.29911-2.441429c.68613-.967791 1.29834-1.98903 1.81768-3.056081l3.08569.927704c.7411.219519 1.51442-.202339 1.73808-.946793l1.59402-5.386796c.21607-.740633-.20471-1.523264-.93633-1.746601l-3.19941-.962063c.11182-1.168222.1213-2.35744.0512-3.552385l3.12739-.752089c.74868-.181342 1.20736-.939158 1.0292-1.696975l-1.29266-5.463148c-.18196-.755907-.93442-1.219759-1.68121-1.038418zM90.123492 48.894003c-5.276767 1.267482-10.572488-2.013844-11.827235-7.324286-1.256645-5.312351 2.003427-10.647608 7.274507-11.911271 5.274872-1.265573 10.568697 2.013844 11.823445 7.326195 1.258539 5.310442-1.997741 10.647607-7.270717 11.909362z" style="fill:#97baed;fill-opacity:1;stroke-width:1.90211201"/><path id="path56" d="m64.166 112.713c-9.08 1.85-18.058-3.824-20.221-12.906-2.165-9.083 3.288-18.198 12.227-20.642l-3.5-14.692-3.03.724c-.899.213-1.451 1.112-1.237 2.005l.921 3.867c-1.289.594-2.518 1.279-3.682 2.055L43.016 70.34c-.637-.671-1.688-.696-2.364-.063l-4.83 4.574c-.665.627-.695 1.681-.063 2.349l2.739 2.895c-.813 1.146-1.542 2.355-2.163 3.616l-3.679-1.098c-.885-.26-1.806.246-2.071 1.125l-1.905 6.375c-.264.877.24 1.806 1.122 2.068l3.812 1.135c-.128 1.39-.147 2.798-.058 4.215l-3.73.89c-.9.215-1.451 1.114-1.238 2.01l1.542 6.469c.213.896 1.109 1.447 2.01 1.232l3.862-.921c.592 1.292 1.28 2.528 2.054 3.688l-2.785 2.634c-.668.632-.693 1.682-.066 2.349l4.57 4.838c.629.664 1.687.694 2.354.063l2.892-2.736c1.151.812 2.361 1.541 3.622 2.165l-1.094 3.684c-.268.879.236 1.808 1.113 2.066l6.382 1.896c.878.263 1.806-.235 2.063-1.113l1.139-3.813c1.389.125 2.798.146 4.218.053l.893 3.74c.213.894 1.105 1.442 1.997 1.229l2.316-.549z" style="fill:url(#radialGradient2534);fill-opacity:1"/><g id="g60" transform="matrix(1.3314607,0,0,1.3314607,69.623325,-30.275292)"><path id="path58" d="m28.131 78.601-.8.19c-.185.045-.419-.051-.52-.214l-.233-.325c-.122-.146-.142-.396-.042-.56l.429-.698c.101-.162.05-.376-.113-.476l-1.079-.664c-.164-.101-.376-.049-.477.111l-.429.699c-.101.161-.332.259-.52.215L23.951 76.816C23.762 76.798 23.57 76.634 23.526 76.45l-.19-.8c-.043-.184-.232-.299-.417-.256l-1.233.295c-.185.045-.3.231-.256.416l.189.799c.044.186-.053.418-.215.519l-.323.235c-.146.123-.397.141-.561.041l-.701-.43C19.66 77.168 19.445 77.22 19.345 77.381l-.664 1.082c-.102.16-.05.375.113.476l.7.429c.159.103.257.334.216.519l-.063.396c-.017.189-.182.381-.367.425l-.798.19c-.184.043-.302.23-.257.416l.294 1.233c.043.185.234.302.418.257l.798-.19c.185-.043.417.052.52.214l.234.324c.122.146.141.397.043.561l-.432.7c-.1.161-.048.377.113.476l1.082.666c.162.098.376.047.475-.113l.43-.698c.101-.164.333-.261.52-.218l.395.063c.189.017.381.183.425.367l.189.798c.046.186.233.301.417.258l1.234-.295c.185-.045.301-.23.257-.418l-.19-.797c-.043-.188.053-.419.215-.52l.323-.235c.147-.122.397-.14.562-.04l.697.432c.164.098.377.047.477-.113l.664-1.082c.1-.161.048-.377-.113-.477l-.7-.429c-.161-.101-.258-.334-.215-.519l.063-.396c.015-.189.182-.381.367-.425l.798-.189c.184-.045.301-.233.255-.417l-.293-1.234c-.045-.186-.233-.3-.416-.257zm-4.209 3.73c-.897.213-1.8-.342-2.012-1.24-.215-.896.34-1.8 1.237-2.013.897-.215 1.8.339 2.014 1.237.215.899-.341 1.801-1.239 2.016z" style="fill:#a6dae9"/></g><path id="path72" d="m56.823 79.401 1.196 5.013c6.291-1.499 12.606 2.385 14.106 8.677 1.498 6.288-2.387 12.605-8.678 14.104l1.193 5.013C73.7 110.049 79.293 100.952 77.136 91.895 74.978 82.837 65.883 77.242 56.823 79.401z" style="fill:#3e5a85;fill-opacity:1"/><g id="g76"><path id="path74" d="m60.719 113.066c-8.016.0-14.913-5.452-16.773-13.26-2.206-9.258 3.53-18.585 12.786-20.792 1.319-.313 2.67-.474 4.015-.474 8.017.0 14.916 5.453 16.778 13.263 2.205 9.257-3.532 18.583-12.791 20.788-1.32.316-2.67.475-4.015.475zm.026-33.732c-1.282.0-2.57.151-3.831.452-8.831 2.105-14.304 11.005-12.198 19.837 1.774 7.448 8.355 12.65 16.002 12.65 1.283.0 2.572-.151 3.831-.451 8.833-2.105 14.308-11.004 12.202-19.833C74.976 84.538 68.394 79.334 60.745 79.334z" style="fill:#a6dae9;fill-opacity:1"/></g><g id="g208" transform="matrix(1.1075584,0,0,1.1075584,-60.874681,-16.511015)"><g id="g122"><path id="path102" d="m88.853 44.64c0 .26-.209.471-.471.471h-3.201c-.259.0-.47-.21-.47-.471v-6.965c0-.26.211-.471.47-.471h3.201c.262.0.471.21.471.471z" style="fill:#8fc4d7"/><path id="path104" d="m82.745 45.318c.152.21.107.505-.104.658l-2.59 1.881c-.211.152-.505.106-.656-.104L75.3 42.118c-.153-.21-.108-.504.104-.657l2.59-1.881c.21-.152.503-.106.656.104z" style="fill:#8fc4d7"/><path id="path106" d="m78.205 49.458c.247.08.382.346.302.593l-.988 3.043c-.08.247-.346.382-.594.302L70.3 51.244C70.054 51.163 69.919 50.898 69.997 50.65l.988-3.042c.082-.247.347-.382.596-.302z" style="fill:#8fc4d7"/><path id="path108" d="m76.963 55.475c.248-.079.515.055.593.303l.99 3.043c.081.247-.057.513-.302.594l-6.626 2.151c-.246.081-.512-.055-.592-.301l-.988-3.044c-.081-.248.054-.513.302-.594z" style="fill:#8fc4d7"/><path id="path110" d="m79.496 61.073c.153-.21.448-.257.657-.104l2.59 1.88c.209.152.257.447.104.657l-4.095 5.636c-.151.213-.446.258-.655.104l-2.591-1.881c-.21-.152-.256-.446-.104-.657z" style="fill:#8fc4d7"/><path id="path112" d="m84.835 64.113c.001-.26.213-.471.473-.471h3.199c.262.0.473.21.473.471v6.966c0 .26-.213.471-.473.471h-3.199c-.26.0-.472-.211-.473-.471z" style="fill:#8fc4d7"/><path id="path114" d="m90.942 63.434c-.15-.211-.105-.505.104-.658l2.59-1.88c.21-.153.505-.106.657.104l4.093 5.635c.153.212.106.507-.104.657l-2.589 1.881c-.213.152-.504.107-.657-.104z" style="fill:#8fc4d7"/><path id="path116" d="M95.483 59.295C95.237 59.213 95.1 58.949 95.18 58.702l.988-3.045c.082-.246.349-.381.595-.301l6.623 2.152c.248.08.381.347.303.593l-.989 3.044c-.08.246-.343.381-.593.302z" style="fill:#8fc4d7"/><path id="path118" d="m96.727 53.277c-.248.08-.516-.055-.595-.301l-.986-3.045c-.082-.247.053-.512.301-.592l6.622-2.151c.248-.081.514.055.595.301l.989 3.044c.08.247-.057.511-.303.593z" style="fill:#8fc4d7"/><path id="path120" d="m94.192 47.68c-.153.21-.447.258-.657.104L90.944 45.902C90.737 45.75 90.688 45.455 90.841 45.246l4.093-5.635c.152-.211.448-.257.658-.105l2.587 1.882c.213.152.258.445.104.658z" style="fill:#8fc4d7"/></g><g id="g206"><circle id="circle124" cx="86.845" cy="54.376" r="14.798" style="fill:#8fc4d7"/><circle id="circle126" cx="86.844" cy="54.376" r="13.872" style="fill:#8fc4d7"/><circle id="circle128" cx="86.844" cy="54.376" r="10.791" style="fill:#fff"/><path id="path130" d="m77.21 55.148c0-5.959 4.83-10.791 10.79-10.791 1.966.0 3.806.528 5.396 1.446-1.818-1.39-4.088-2.217-6.551-2.217-5.959.0-10.792 4.831-10.792 10.791.0 3.994 2.175 7.48 5.397 9.345-2.577-1.973-4.24-5.078-4.24-8.574z" style="fill:#d1ecfb"/><circle id="circle132" cx="86.844" cy="54.376" r="1.927" style="fill:#3e5984;fill-opacity:1"/></g></g><g id="g76-8" transform="translate(27.024379,-56.342917)" style="fill:#2e4b79;fill-opacity:1"><path style="fill:#2e4b79;fill-opacity:1" id="path74-1" d="m60.719 113.066c-8.016.0-14.913-5.452-16.773-13.26-2.206-9.258 3.53-18.585 12.786-20.792 1.319-.313 2.67-.474 4.015-.474 8.017.0 14.916 5.453 16.778 13.263 2.205 9.257-3.532 18.583-12.791 20.788-1.32.316-2.67.475-4.015.475zm.026-33.732c-1.282.0-2.57.151-3.831.452-8.831 2.105-14.304 11.005-12.198 19.837 1.774 7.448 8.355 12.65 16.002 12.65 1.283.0 2.572-.151 3.831-.451 8.833-2.105 14.308-11.004 12.202-19.833C74.976 84.538 68.394 79.334 60.745 79.334z"/></g></g></g></svg></span><span class=font-weight-bold>sysPipe</span></a><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sp/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sp/spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../../sp/spr/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../../sp/spr/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../../sp/spr/sp_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../../sp/spr/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../../sp/spr/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../../sp/spr/cwl_syntax/ style=padding-left:2rem;font-weight:calc(3/4*500)>CWL syntax</a>
<a class=dropdown-item href=../../../sp/spr/cwl_and_spr/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR and CWL</a>
<a class=dropdown-item href=../../../sp/spr/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRTools</a>
<a class=dropdown-item href=../../../sp/spt/ style=padding-left:2rem;font-weight:calc(3/4*500)>Data Visualization</a>
<a class=dropdown-item href=../../../sp/spr/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../../sp/sprdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../../sp/sprdata/systempiperdata/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../../sp/sprdata/newwf/ style=padding-left:2rem;font-weight:calc(3/4*500)>Rules</a>
<a class=dropdown-item href=../../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeTools style=padding-left:2rem;font-weight:calc(3/4*500)>SPT GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=../../../spr_wf/templates style=padding-left:1rem;font-weight:calc(3/3*500)>Templates</a>
<a class=dropdown-item href=../../../spr_wf/templates/rnaseq style=padding-left:2rem;font-weight:calc(3/4*500)>RNA-Seq</a>
<a class=dropdown-item href=../../../spr_wf/templates/varseq style=padding-left:2rem;font-weight:calc(3/4*500)>VAR-Seq</a>
<a class=dropdown-item href=../../../spr_wf/templates/chipseq style=padding-left:2rem;font-weight:calc(3/4*500)>ChIP-Seq</a>
<a class=dropdown-item href=../../../spr_wf/templates/chipseq style=padding-left:2rem;font-weight:calc(3/4*500)>RIBO-Seq</a>
<a class=dropdown-item href=../../../spr_wf/installwf/ style=padding-left:1rem;font-weight:calc(3/3*500)>How to install Workflow</a>
<a class=dropdown-item href=../../../spr_wf/newwf style=padding-left:1rem;font-weight:calc(3/3*500)>Rules to create Workflow</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../outreach/ id=navbarDropdown role=button>Outreach</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../outreach/conferences/ style=padding-left:1rem;font-weight:calc(3/3*500)>Conferences</a>
<a class=dropdown-item href=../../../outreach/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../../outreach/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../../news/release_3_13/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.13</a>
<a class=dropdown-item href=../../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../../offline-search-index.e9a41006eefa71c6788f58fd97b26a00.json data-offline-search-base-href=../../../ data-offline-search-max-results=10></div></nav></header><div class="container-fluid td-default td-outer"><main role=main class=td-main><link href=../../../rmarkdown-libs/font-awesome/css/all.css rel=stylesheet><link href=../../../rmarkdown-libs/font-awesome/css/v4-shims.css rel=stylesheet><link href=../../../rmarkdown-libs/tile-view/tile-view.css rel=stylesheet><script src=../../../rmarkdown-libs/tile-view/tile-view.js></script><link href=../../../rmarkdown-libs/animate.css/animate.xaringan.css rel=stylesheet><link href=../../../rmarkdown-libs/tachyons/tachyons.min.css rel=stylesheet><p>layout: true
background-image: url(<a href=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png>https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/systemPipeR.png</a>)
background-position: 99% 1%
background-size: 10%</p><hr><p>class: middle</p><h2 id=outline>Outline</h2><h4 id=introduction>Introduction</h4><h4 id=design>Design</h4><h4 id=how-to-run-a-workflow>How to run a Workflow</h4><h4 id=workflows-tutorial>Workflows Tutorial</h4><h4 id=live-demo>Live Demo</h4><hr><p>class: inverse, center, middle</p><h1 id=i-classfas-fa-file-codei-introduction><i class="fas fa-file-code"></i>Introduction</h1><hr><h2 id=introduction-1>Introduction</h2><p><i class="fas fa-hand-point-right" style=color:#00758a></i><a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html><em>systemPipeR</em></a> provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>systemPipeR</em> offers many utilities to build, control, and execute
workflows entirely from R</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>The environment takes advantage of central community S4 classes of the Bioconductor ecosystem</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Workflows are managed by generic workflow management containers supporting both analysis routines implemented in R code and/or command-line software</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Simple annotation system targets</p><hr><h3 id=systempipers-core-functionalities>systemPipeR’s Core Functionalities</h3><p>.center[
<img src=highLevel_color.svg height=490px class=center>
]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features><i class="fas fa-toolbox"></i>Structural Features</h2><p>&ndash;
.left-column[</p><h4 id=wf-infrastructure>WF infrastructure</h4><p>]</p><p>.right-column[
<em>systemPipeR</em> offers many utilities to build, control, and execute
workflows entirely from R. The environment takes advantage of central community S4 classes
of the Bioconductor ecosystem. Workflows are managed by generic workflow management
containers supporting both analysis routines implemented in R code and/or command-line
software. A layered monitoring infrastructure is provided to design, control and debug each
step in a workflow. The run environment allows to execute workflows entirely or with a
intuitive to use step-wise execution syntax using R’s standard subsetting syntax
(runWF(sys[1:3])) or pipes (%>%).
]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features-1><i class="fas fa-toolbox"></i>Structural Features</h2><p>.left-column[</p><h4 id=wf-infrastructure-1>WF infrastructure</h4><h4 id=command-line-support>Command-line support</h4><p>]</p><p>.right-column[
An important feature of systemPipeR is support for
running command-line software by adopting the Common Workflow Language (CWL). The
latter is a widely adopted community standard for describing analysis workflows. This design
offers several advantages such as:</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>seamless integration of most command-line software</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>support to run systemPipeR workflows from R or many other popular computer languages</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>efficient sharing of workflows across different workflow environments.
]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features-2><i class="fas fa-toolbox"></i>Structural Features</h2><p>.left-column[</p><h4 id=wf-infrastructure-2>WF infrastructure</h4><h4 id=command-line-support-1>Command-line support</h4><h4 id=parallel-evaluation>Parallel evaluation</h4><p>]</p><p>.right-column[
The processing time of workflows can be greatly reduced by making use
of parallel evaluations across several CPU cores on single machines, or multiple nodes of
computer clusters and cloud-based systems. systemPipeR simplifies these parallelization tasks
without creating any limitations for users who do not have access to high-performance
computer resources
]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features-3><i class="fas fa-toolbox"></i>Structural Features</h2><p>.left-column[</p><h4 id=wf-infrastructure-3>WF infrastructure</h4><h4 id=command-line-support-2>Command-line support</h4><h4 id=parallel-evaluation-1>Parallel evaluation</h4><h4 id=reports-infrastructure>Reports infrastructure</h4><p>]</p><p>.right-column[
<em>systemPipeR’s</em> reporting infrastructure
includes three types of interconnected reports each serving a different purpose:</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>a scientific report, based on R Markdown, contains all scientifically relevant results</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>a technical report captures all technical information important for each workflow step, including parameter
settings, software versions, and warning/error messages, etc.</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>a visual report depicts the entire workflow topology including its run status in form of a workflow graph</p><p>]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features-4><i class="fas fa-toolbox"></i>Structural Features</h2><p>.left-column[</p><h4 id=wf-infrastructure-4>WF infrastructure</h4><h4 id=command-line-support-3>Command-line support</h4><h4 id=parallel-evaluation-2>Parallel evaluation</h4><h4 id=reports-infrastructure-1>Reports infrastructure</h4><h4 id=shiny-web-interface>Shiny Web Interface</h4><p>]</p><p>.right-column[
Recently, the systemPipeShiny package has been added that allows
users to design workflows in an interactive graphical user interface (GUI). In addition to
designing workflows, this new interface allows users to run and to monitor workflows in an
intuitive manner without the need of knowing R.
]</p><hr><h2 id=i-classfas-fa-toolboxi-structural-features-5><i class="fas fa-toolbox"></i>Structural Features</h2><p>.left-column[</p><h4 id=wf-infrastructure-5>WF infrastructure</h4><h4 id=command-line-support-4>Command-line support</h4><h4 id=parallel-evaluation-3>Parallel evaluation</h4><h4 id=reports-infrastructure-2>Reports infrastructure</h4><h4 id=shiny-web-interface-1>Shiny Web Interface</h4><h4 id=workflow-templates>Workflow Templates</h4><p>]</p><p>.right-column[
A rich set of end-to-end workflow templates is provided by this
project for a wide range omics applications. In addition, users can contribute and share their
workflows with the community by submitting them to a central GitHub repository
]</p><hr><h2 id=important-functions>Important Functions</h2><p>.small[</p><table><thead><tr><th>Function Name</th><th>Description</th><th>Category</th></tr></thead><tbody><tr><td><code>genWorkenvir</code></td><td>Generates workflow templates provided by systemPipeRdata helper package / or from the individuals’ pipelines packages</td><td>Accessory</td></tr><tr><td><code>loadWorkflow</code></td><td>Constructs <code>SYSargs2</code> object from CWL param and targets files</td><td><code>SYSargs2</code></td></tr><tr><td><code>renderWF</code></td><td>Populate all the command-line in an <code>SYSargs2</code> object</td><td><code>SYSargs2</code></td></tr><tr><td><code>subsetWF</code></td><td>Subsetting <code>SYSargs2</code> class slots</td><td><code>SYSargs2</code></td></tr><tr><td><code>runCommandline</code></td><td>Executes command-line software on samples and parameters specified in <code>SYSargs2</code> object</td><td><code>SYSargs2</code></td></tr><tr><td><code>clusterRun</code></td><td>Runs command-line software in parallel mode on a computer cluster</td><td><code>SYSargs2</code></td></tr><tr><td><code>writeTargetsout</code></td><td>Write updated targets out to file/Generate targets file with reference</td><td><code>SYSargs2</code></td></tr><tr><td><code>output_update</code></td><td>Updates the output files paths in the <code>SYSargs2</code> object</td><td><code>SYSargs2</code></td></tr><tr><td><code>singleYML</code></td><td>Create automatically the <code>param.yml</code></td><td><code>SYSargs2</code></td></tr><tr><td><code>createWF</code></td><td>Create automatically <code>param.cwl</code> and the <code>param.yml</code> based on the command line</td><td><code>SYSargs2</code></td></tr><tr><td><code>config.param</code></td><td>Custom configuration of the CWL param files from R</td><td><code>SYSargs2</code></td></tr><tr><td>]</td><td></td><td></td></tr></tbody></table><hr><h2 id=important-functions-1>Important Functions</h2><p>.small[</p><table><thead><tr><th>Function Name</th><th>Description</th><th>Category</th></tr></thead><tbody><tr><td><code>initWF</code></td><td>Constructs <code>SYSargsList</code> workflow control module (S4 object) from script file</td><td><code>SYSargsList</code></td></tr><tr><td><code>configWF</code></td><td>Control of which step of the workflow will be run and the generation of the new RMarkdown</td><td><code>SYSargsList</code></td></tr><tr><td><code>runWF</code></td><td>Runs all the R chunk define in the RMarkdown file or a subset, e.g. <code>runWF[1:3]</code></td><td><code>SYSargsList</code></td></tr><tr><td><code>renderReport</code></td><td>Render Scientific Report based on RMarkdown</td><td><code>SYSargsList</code></td></tr><tr><td><code>subsetRmd</code></td><td>Write updated subset Rmarkdown of R chunk with text associate in the step selected</td><td><code>SYSargsList</code></td></tr><tr><td><code>plotWF</code></td><td>Plot visual workflow designs and topologies with different graphical layouts</td><td><code>SYSargsList</code></td></tr><tr><td><code>statusWF</code></td><td>Return the overview of the workflow steps computational status</td><td><code>SYSargsList</code></td></tr><tr><td><code>evalCode</code></td><td>Turn eval option <code>TRUE</code> or <code>FALSE</code> on RMarkdown file</td><td>Accessory</td></tr><tr><td><code>tryCL</code></td><td>Checks if third-party software or utility is installed and set in the PATH</td><td>Accessory</td></tr><tr><td>]</td><td></td><td></td></tr></tbody></table><hr><p>class: inverse, center, middle</p><h1 id=i-classfas-fa-drafting-compassi-design><i class="fas fa-drafting-compass"></i>Design</h1><hr><h3 id=workflow-management-solutions>Workflow Management Solutions</h3><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>systemPipeR</em> central concept for designing workflows is <strong>workflow
management containers</strong> (S4 class)</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>SYSargs2</em> controls workflow steps with input/output file operations</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>SYSargs2</em> requires a targets and a set of workflow definition files (here param.cwl and param.yml)</p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>SYSargsList</em> objects organize one or many <em>SYSargs2</em>
containers in a single compound object capturing all information required to run, control and
monitor complex workflows from start to finish</p><p>.center[
<img src=SPRclass.png class=center>
]</p><hr><h2 id=i-classfar-fa-folder-openi-directory-structure><i class="far fa-folder-open"></i>Directory Structure</h2><p>The workflow templates generated by <em><code>genWorkenvir</code></em> contain the following preconfigured directory structure:</p><p><img src=directory.png height=425px class=center></p><hr><h2 id=i-classfas-fa-project-diagrami-workflows-collection><i class="fas fa-project-diagram"></i>Workflows Collection</h2><p>Browse pipelines that are currently available as part of the <em>systemPipeR</em> toolkit</p><p>.small[</p><table><thead><tr><th style=text-align:center>WorkFlow</th><th style=text-align:center>Description</th><th style=text-align:center>Version</th><th style=text-align:center>GitHub</th><th style=text-align:center>CI Testing</th></tr></thead><tbody><tr><td style=text-align:center><a href=https://systempiper.github.io/systemPipeChIPseq/articles/systemPipeChIPseq.html>systemPipeChIPseq</a></td><td style=text-align:center>ChIP-Seq Workflow Template</td><td style=text-align:center>v1.0 <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png align=center style=width:20px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeChIPseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeChIPseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://systempiper.github.io/systemPipeRIBOseq/articles/systemPipeRIBOseq.html>systemPipeRIBOseq</a></td><td style=text-align:center>RIBO-Seq Workflow Template</td><td style=text-align:center>v1.0 <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png align=center style=width:20px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeRIBOseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeRIBOseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://systempiper.github.io/systemPipeRNAseq/articles/systemPipeRNAseq.html>systemPipeRNAseq</a></td><td style=text-align:center>RNA-Seq Workflow Template</td><td style=text-align:center>v1.0 <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png align=center style=width:20px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeRNAseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeRNAseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://systempiper.github.io/systemPipeVARseq/articles/systemPipeVARseq.html>systemPipeVARseq</a></td><td style=text-align:center>VAR-Seq Workflow Template</td><td style=text-align:center>v1.0 <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/check.png align=center style=width:20px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeVARseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeVARseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeMethylseq>systemPipeMethylseq</a></td><td style=text-align:center>Methyl-Seq Workflow Template</td><td style=text-align:center>devel <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png align=center style=width:25px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeMethylseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeMethylseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeDeNovo>systemPipeDeNovo</a></td><td style=text-align:center>De novo transcriptome assembly Workflow Template</td><td style=text-align:center>devel <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png align=center style=width:25px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeDeNovo><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeDeNovo/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeCLIPseq>systemPipeCLIPseq</a></td><td style=text-align:center>CLIP-Seq Workflow Template</td><td style=text-align:center>devel <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png align=center style=width:25px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeCLIPseq><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeCLIPseq/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeMetaTrans>systemPipeMetaTrans</a></td><td style=text-align:center>Metatranscriptomic Sequencing Workflow Template</td><td style=text-align:center>devel <img src=https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/images/devel.png align=center style=width:25px></td><td style=text-align:center><a href=https://github.com/systemPipeR/systemPipeMetaTrans><img src=https://raw.githubusercontent.com/systemPipeR/systemPipeR/gh-pages/images/GitHub-Mark-120px-plus.png align=center style=width:20px></td><td style=text-align:center><img src=https://github.com/systemPipeR/systemPipeMetaTrans/workflows/R-CMD-check/badge.svg alt=R-CMD-check></td></tr><tr><td style=text-align:center>]</td><td></td><td></td><td></td><td></td></tr></tbody></table><hr><p>class: inverse, center, middle</p><h1 id=cwl>CWL</h1><hr><h2 id=i-classfas-fa-box-openi-cwl><i class="fas fa-box-open"></i>CWL</h2><p><i class="fas fa-circle" style=color:#00758a></i><strong>TODO:</strong>
Add section with CWL details</p><p>.center[
<img src=cwl.png height=490px class=center>
]</p><hr><h2 id=i-classfas-fa-box-openi-cwl-and-spr><i class="fas fa-box-open"></i>CWL and SPR</h2><p><i class="fas fa-circle" style=color:#00758a></i><strong>TODO:</strong>
How to use CWL definition with <em>systemPipeR</em></p><ul><li><em>SYSargs2</em> instances are constructed from a <em><code>targets</code></em> file and two <em><code>param</code></em> file<ul><li><em><code>hisat2-mapping-se.cwl</code></em> file contains the settings for running command-line software</li><li><em><code>hisat2-mapping-se.yml</code></em> file define all the variables to be input in the specific command-line step</h4></li></ul></li></ul><pre><code class=language-r>targets &lt;- system.file(&quot;extdata&quot;, &quot;targets.txt&quot;, package=&quot;systemPipeR&quot;)
dir_path &lt;- system.file(&quot;extdata/cwl/hisat2/hisat2-se&quot;, package=&quot;systemPipeR&quot;)
align &lt;- loadWF(targets=targets, wf_file=&quot;hisat2-mapping-se.cwl&quot;,
                   input_file=&quot;hisat2-mapping-se.yml&quot;, dir_path=dir_path)
align &lt;- renderWF(align, inputvars=c(FileName=&quot;_FASTQ_PATH_&quot;, SampleName=&quot;_SampleName_&quot;))

## Instance of 'SYSargs2':
##    Slot names/accessors: 
##       targets: 18 (M1A...V12B), targetsheader: 4 (lines)
##       modules: 2
##       wf: 0, clt: 1, yamlinput: 7 (components)
##       input: 18, output: 18
##       cmdlist: 18
##    WF Steps:
##       1. hisat2-mapping-se.cwl (rendered: TRUE)
</code></pre><hr><h2 id=i-classfas-fa-box-openi-cwl-and-spr-1><i class="fas fa-box-open"></i>CWL and SPR</h2><h2 id=sysargs2-instance><em><code>SYSargs2</code></em> instance</h2><ul><li>Slots and accessor functions have the same names</li></ul><pre><code class=language-r>names(align)
#  [1] &quot;targets&quot;       &quot;targetsheader&quot; &quot;modules&quot;       &quot;wf&quot;            &quot;clt&quot;          
#  [6] &quot;yamlinput&quot;     &quot;cmdlist&quot;       &quot;input&quot;         &quot;output&quot;        &quot;cwlfiles&quot;     
# [11] &quot;inputvars&quot; 
</code></pre><ul><li><em><code>cmdlist</code></em> return command-line arguments for the specific software, here <em><code>HISAT2</code></em> for the first sample</li></ul><pre><code class=language-r>cmdlist(align)[1]
# $M1A
# $M1A$`hisat2-mapping-se.cwl`
# [1] &quot;hisat2 -S results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -U ./data/SRR446027_1.fastq.gz --threads 4&quot;
</code></pre><ul><li>The output components of <em><code>SYSargs2</code></em> define all the expected output files for each step in the workflow; some of which are the input for the next workflow step</li></ul><pre><code class=language-r>output(align)[1]
# $M1A
# $M1A$`hisat2-mapping-se.cwl`
# [1] &quot;results/M1A.sam&quot;
</code></pre><hr><p>class: inverse, center, middle</p><h1 id=i-classfas-fa-tablei-metadata><i class="fas fa-table"></i>Metadata</h1><hr><h3 id=i-classfas-fa-box-openi-targets-file-organizes-samples><i class="fas fa-box-open"></i><em>Targets</em> file organizes samples</h3><ul><li>Structure of <em><code>targets</code></em> file for single-end (SE) library</li></ul><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targets.txt&quot;, package=&quot;systemPipeR&quot;)
read.delim(targetspath, comment.char = &quot;#&quot;)[1:3,1:4]
</code></pre><pre><code>##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
</code></pre><ul><li>Structure of <em><code>targets</code></em> file for paired-end (PE) library</li></ul><pre><code class=language-r>targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE.txt&quot;, package=&quot;systemPipeR&quot;)
read.delim(targetspath, comment.char = &quot;#&quot;)[1:3,1:5]
</code></pre><pre><code>##                     FileName1                   FileName2 SampleName Factor
## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz        M1A     M1
## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz        M1B     M1
## 3 ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz        A1A     A1
##   SampleLong
## 1  Mock.1h.A
## 2  Mock.1h.B
## 3   Avr.1h.A
</code></pre><hr><h3 id=i-classfas-fa-box-openi-integration-with-summarizedexperiment><i class="fas fa-box-open"></i>Integration with <em>SummarizedExperiment</em></h3><ul><li>Integrates <em>targets</em> files and count table from <em>systemPipeR</em> to a <em>SummarizedExperiment</em> object</li></ul><pre><code class=language-r>## Create an object with targets file and comparison and count table
sprSE &lt;- SPRdata(targetspath = targetspath, cmp=TRUE)
metadata(sprSE)
# $version
# [1] ‘1.23.9’
# 
# $comparison
# $comparison$CMPset1
#       [,1]  [,2] 
#  [1,] &quot;M1&quot;  &quot;A1&quot; 
#  [2,] &quot;M1&quot;  &quot;V1&quot; 
#  [3,] &quot;A1&quot;  &quot;V1&quot; 
#  [4,] &quot;M6&quot;  &quot;A6&quot; 

colData(sprSE)
# DataFrame with 18 rows and 6 columns
#                    FileName  SampleName      Factor  SampleLong 
#                 &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; 
# M1A  ./data/SRR446027_1.f..         M1A          M1   Mock.1h.A 
# M1B  ./data/SRR446028_1.f..         M1B          M1   Mock.1h.B 
# ...                     ...         ...         ...         ... 
# M12B ./data/SRR446040_1.f..        M12B         M12  Mock.12h.B
</code></pre><hr><p>class: inverse, center, middle</p><h1 id=i-classfas-fa-codei-live-demo><i class="fas fa-code"></i>Live Demo</h1><hr><h2 id=i-classfas-fa-box-openi-install-package><i class="fas fa-box-open"></i>Install Package</h2><p>Install the <strong>systemPipeRdata</strong> package from <a href=http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html>Bioconductor</a>:</p><pre><code class=language-r>if (!requireNamespace(&quot;BiocManager&quot;, quietly=TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;systemPipeR&quot;)
</code></pre><h3 id=i-classfas-fa-booki-load-package-and-documentation><i class="fas fa-book"></i>Load Package and Documentation</h3><p><i class="fas fa-question" style=color:#00758a></i>Load package:</p><pre><code class=language-r>library(&quot;systemPipeR&quot;)
</code></pre><p><i class="fas fa-question" style=color:#00758a></i>Access help:</p><pre><code class=language-r>library(help=&quot;systemPipeR&quot;)
vignette(&quot;systemPipeR&quot;)
</code></pre><hr><h2 id=i-classfas-fa-codei-quick-start><i class="fas fa-code"></i>Quick Start</h2><h3 id=load-sample-workflow>Load Sample Workflow</h3><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>systemPipeRdata</em></p><ul><li>Helper package to generate with a single command workflow templates for <em>systemPipeR</em></li><li>Includes sample data for testing</li><li>User can create new workflows or change and extend existing ones</li><li>Template Workflows:<ul><li>Sample workflows can be loaded with the <em><code>genWorkenvir</code></em> function from <strong>systemPipeRdata</strong></li></ul></li></ul><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Generate workflow template:</p><pre><code class=language-r>library(systemPipeRdata)
genWorkenvir(workflow=&quot;rnaseq&quot;)
setwd(&quot;rnaseq&quot;)
</code></pre><p><i class="fas fa-hand-point-right" style=color:#00758a></i>More details about <em>systemPipeRdata</em> package <a href=https://systempipe.org/pages/SPRdata_intro/SPRdata_intro#1>here</a></p><hr><h2 id=i-classfas-fa-box-openi-install-workflow><i class="fas fa-box-open"></i>Install Workflow</h2><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Check the workflow template availability</p><pre><code class=language-r>availableWF(github = TRUE)

# $systemPipeRdata
# [1] &quot;chipseq&quot; &quot;new&quot;     &quot;riboseq&quot; &quot;rnaseq&quot;  &quot;varseq&quot; 
# 
# $github
#                          workflow                       branches version                                               html       description
# 1   systemPipeR/systemPipeChIPseq                         master release   https://github.com/systemPipeR/systemPipeChIPseq Workflow Template
# 2   systemPipeR/systemPipeRIBOseq                         master release   https://github.com/systemPipeR/systemPipeRIBOseq Workflow Template
# 3    systemPipeR/systemPipeRNAseq cluster, master, singleMachine release    https://github.com/systemPipeR/systemPipeRNAseq Workflow Template
# 4    systemPipeR/systemPipeVARseq                         master release    https://github.com/systemPipeR/systemPipeVARseq Workflow Template
# 5   systemPipeR/systemPipeCLIPseq                         master   devel   https://github.com/systemPipeR/systemPipeCLIPseq Workflow Template
# 6    systemPipeR/systemPipeDeNovo                         master   devel    https://github.com/systemPipeR/systemPipeDeNovo Workflow Template
# 7 systemPipeR/systemPipeMetaTrans                         master   devel https://github.com/systemPipeR/systemPipeMetaTrans Workflow Template
# 8 systemPipeR/systemPipeMethylseq                         master   devel https://github.com/systemPipeR/systemPipeMethylseq Workflow Template
</code></pre><hr><h2 id=i-classfas-fa-mouse-pointeri-dynamic-workflow-template><i class="fas fa-mouse-pointer"></i>Dynamic Workflow Template</h2><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Create dynamic Workflow Templates with <em>RStudio</em></p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><code>File -> New File -> R Markdown -> From Template</code>
.center[
<img src=rstudio.png height=425px class=center>
]</p><hr><h2 id=i-classfas-fa-toolboxi-run-a-workflow><i class="fas fa-toolbox"></i>Run a Workflow</h2><p>.left-column[</p><h3 id=setup>Setup</h3><p>]</p><p>.right-column[</p><pre><code class=language-r>library(systemPipeR)
targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targets.txt&quot;, package=&quot;systemPipeR&quot;) 
read.delim(targetspath, comment.char = &quot;#&quot;)[1:4,1:4]
</code></pre><pre><code>##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
## 4 ./data/SRR446030_1.fastq.gz        A1B     A1   Avr.1h.B
</code></pre><pre><code class=language-r>script &lt;- system.file(&quot;extdata/workflows/rnaseq&quot;, &quot;systemPipeRNAseq.Rmd&quot;, package=&quot;systemPipeRdata&quot;)
</code></pre><p>]</p><hr><h2 id=i-classfas-fa-toolboxi-run-a-workflow-1><i class="fas fa-toolbox"></i>Run a Workflow</h2><p>.left-column[</p><h3 id=setup-1>Setup</h3><h3 id=initwf><code>initWF</code></h3><p>]</p><p>.right-column[</p><pre><code class=language-r>sysargslist &lt;- initWF(script = script, targets = targetspath, overwrite = TRUE)
# Project started with success: ./SYSproject and SYSconfig.yml were created.
</code></pre><p>]</p><hr><h2 id=i-classfas-fa-toolboxi-run-a-workflow-2><i class="fas fa-toolbox"></i>Run a Workflow</h2><p>.left-column[</p><h3 id=setup-2>Setup</h3><h3 id=initwf-1><code>initWF</code></h3><h3 id=configwf><code>configWF</code></h3><p>]</p><p>.right-column[</p><pre><code class=language-r>sysargslist &lt;- configWF(sysargslist, input_steps = &quot;1:3&quot;)
sysargslist
# Instance of 'SYSargsList':
#    WF Steps:
# 1. Rmarkdown/HTML setting
# 2. Introduction
# 3. Samples and environment settings
#     3.1. Environment settings and input data
#     3.2. Required packages and resources
#     3.3. Experiment definition provided by `targets` file
</code></pre><p>]</p><hr><h2 id=i-classfas-fa-toolboxi-run-a-workflow-3><i class="fas fa-toolbox"></i>Run a Workflow</h2><p>.left-column[</p><h3 id=setup-3>Setup</h3><h3 id=initwf-2><code>initWF</code></h3><h3 id=configwf-1><code>configWF</code></h3><h3 id=runwf><code>runWF</code></h3><p>]</p><p>.right-column[</p><pre><code class=language-r>sysargslist &lt;- runWF(sysargslist, steps = &quot;1:2&quot;)
# Step: 1: Introduction --&gt; DONE 
# Step: 2: Samples and environment settings --&gt; DONE 
# Step: 2.1: Environment settings and input data --&gt; DONE 
# Step: 2.2: Required packages and resources --&gt; DONE 
# Step: 2.3: Experiment definition provided by `targets` file --&gt; DONE 
</code></pre><pre><code class=language-r>sysargslist &lt;- runWF(sysargslist, steps = &quot;ALL&quot;)
</code></pre><p>]</p><hr><h2 id=i-classfas-fa-toolboxi-run-a-workflow-4><i class="fas fa-toolbox"></i>Run a Workflow</h2><p>.left-column[</p><h3 id=setup-4>Setup</h3><h3 id=initwf-3><code>initWF</code></h3><h3 id=configwf-2><code>configWF</code></h3><h3 id=runwf-1><code>runWF</code></h3><h3 id=renderreport><code>renderReport</code></h3><p>]</p><p>.right-column[</p><pre><code class=language-r>sysargslist &lt;- renderReport(sysargslist = sysargslist)
</code></pre><p>]</p><hr><h2 id=how-to-use-pipes->How to Use Pipes <code>%>%</code></h2><p>Consider the following example, in which the steps are the initialization, configuration and running the entire workflow.</p><pre><code class=language-r>library(systemPipeR)
sysargslist &lt;- initWF(script =&quot;systemPipeRNAseq.Rmd&quot;, overwrite = TRUE) %&gt;%
    configWF(input_steps = &quot;1:6&quot;) %&gt;%
    runWF(steps = &quot;1:2&quot;)
</code></pre><hr><p>class: inverse, center, middle</p><h1 id=i-classfas-fa-tasksi-project-updates><i class="fas fa-tasks"></i>Project Updates</h1><hr><h3 id=targets-x-summarizedexperiment><em>targets</em> x <em>SummarizedExperiment</em></h3><p><i class="fas fa-check-circle" style=color:#00758a></i>Extension &ldquo;SummarizedExperiment&rdquo; methods:</p><pre><code class=language-r>sprSE &lt;- addAssay(sprSE, assay(countMatrix), xName=&quot;countMatrix&quot;)
sprSE &lt;- addMetadata(sprSE, list(targets), xName=&quot;metadata&quot;)
</code></pre><p><i class="fas fa-check-circle" style=color:#00758a></i>New Function:</p><pre><code class=language-r>## Create empty SummarizedExperiment
sprSE &lt;- SPRdata()

## Create an object with targets file and comparison and count table
sprSE &lt;- SPRdata(counts = countMatrix, cmp=TRUE, targetspath = targetspath)
metadata(sprSE)
colData(sprSE)
assays(sprSE)
</code></pre><hr><h2 id=spr-paper>SPR Paper</h2><p><i class="fas fa-check-circle" style=color:#00758a></i><a href=https://www.overleaf.com/project/5e94948a23918a000116311c>Link to draft</a></p><p><i class="fas fa-check-circle" style=color:#00758a></i>Added the main points to discuss in the draft</p><p><i class="fas fa-arrow-circle-right" style=color:#00758a></i>Writing: Results and introduction</p><p><i class="fas fa-circle" style=color:#00758a></i>Improve Graphical Abstract</p><p><i class="fas fa-circle" style=color:#00758a></i>Show case?</p><hr><h2 id=sysargslist><em>SYSargsList</em></h2><p><i class="fas fa-check-circle" style=color:#00758a></i>Explain how <em>SYSargsList</em> is implemented - Vignette</p><p>.small[</p><table><thead><tr><th>Function Name</th><th>Description</th></tr></thead><tbody><tr><td><code>initWF</code></td><td>Constructs <code>SYSargsList</code> workflow control module (S4 object) from script file</td></tr><tr><td><code>configWF</code></td><td>Control of which step of the workflow will be run and the generation of the new RMarkdown</td></tr><tr><td><code>runWF</code></td><td>Runs all the R chunk define in the RMarkdown file or a subset, e.g. <code>runWF[1:3]</code></td></tr><tr><td><code>renderReport</code></td><td>Render Scientific Report based on RMarkdown</td></tr><tr><td><code>renderLog</code></td><td>Render logs Report based on RMarkdown</td></tr><tr><td><code>updateWF</code></td><td>Recover the <code>SYSargsList</code> workflow previous ran and restarts the WF</td></tr><tr><td><code>plotWF</code></td><td>Plot visual workflow designs and topologies with different graphical layouts</td></tr><tr><td><code>statusWF</code></td><td>Return the overview of the workflow steps computational status</td></tr><tr><td><code>evalCode</code></td><td>Turn eval option <code>TRUE</code> or <code>FALSE</code> on RMarkdown file</td></tr><tr><td><code>tryCL</code></td><td>Checks if third-party software or utility is installed and set in the PATH</td></tr><tr><td>]</td><td></td></tr></tbody></table><p><i class="fas fa-circle" style=color:#00758a></i>Improve <code>statusWF()</code></p><hr><h2 id=visualization-in-systempiper>Visualization in <em>systemPipeR</em></h2><p><i class="fas fa-check-circle" style=color:#00758a></i>Add to vignette (SPR or SPS)</p><ul><li><code>exploreDDS</code>, <code>exploreDDSplot</code>, <code>GLMplot</code>, <code>MAplot</code>, <code>MDSplot</code>, <code>PCAplot</code>, <code>hclustplot</code>, <code>heatMaplot</code>, <code>tSNEplot</code>, <code>volcanoplot</code></li></ul><p><i class="fas fa-circle" style=color:#00758a></i>Enrichment analysis and visualization tool for SPR</p><ul><li>Integration with <code>FGSEA</code></li></ul><hr><h2 id=website>WebSite</h2><p><i class="fas fa-check-circle" style=color:#00758a></i>Updated the vignette</p><p><i class="fas fa-check-circle" style=color:#00758a></i>Added <em>systemPipeRdata</em> vignette and presentation: <a href=https://systempipe.org/docs/systemPipeRdata/>link</a></p><p><i class="fas fa-circle" style=color:#00758a></i>Redirect <a href=http://girke.bioinformatics.ucr.edu/systemPipeR/>http://girke.bioinformatics.ucr.edu/systemPipeR/</a> to new page</p><p><i class="fas fa-circle" style=color:#00758a></i>Add content to FAQ section</p><p><i class="fas fa-circle" style=color:#00758a></i>Add tutorials videos</p><hr><p>class: middle</p><h1 id=thanks>Thanks!</h1><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Browse source code at <a href=https://github.com/tgirke/systemPipeR/><i class="fab fa-github fa-2x"></i></a></p><p><i class="fas fa-hand-point-right" style=color:#00758a></i>Ask a question about systemPipeR at Bioconductor Support Page
<a href=https://support.bioconductor.org/t/systempiper/><i class="far fa-question-circle fa-2x"></i></a></p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><em>systemPipeRdata</em> at <a href=http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html>Bioconductor</a></p><p><i class="fas fa-hand-point-right" style=color:#00758a></i><a href=https://systempipe.org/>https://systempipe.org/</a></p></main><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2022 The systemPipe Project All Rights Reserved</small><p class=mt-2><a href=../../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script id=MathJax-script async src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js integrity="sha256-+CC/xQfD6+/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin=anonymous></script><script src=../../../js/prism.js></script></body></html>