<!doctype html><html lang=en class=no-js><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=generator content="Hugo 0.79.1"><meta name=ROBOTS content="NOINDEX, NOFOLLOW"><link rel="shortcut icon" href=../../favicons/favicon.ico><link rel=apple-touch-icon href=../../favicons/apple-touch-icon-180x180.png sizes=180x180><link rel=icon type=image/png href=../../favicons/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=../../favicons/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=../../favicons/android-36x36.png sizes=36x36><link rel=icon type=image/png href=../../favicons/android-48x48.png sizes=48x48><link rel=icon type=image/png href=../../favicons/android-72x72.png sizes=72x72><link rel=icon type=image/png href=../../favicons/android-96x96.png sizes=96x96><link rel=icon type=image/png href=../../favicons/android-144x144.png sizes=144x144><link rel=icon type=image/png href=../../favicons/android-192x192.png sizes=192x192><title>Bioconductor 3.13 | sysPipe</title><meta property="og:title" content="Bioconductor 3.13"><meta property="og:description" content="systemPipeShiny 1.0.0 is available systemPipeTools package extends the widely used systemPipeR (SPR) workflow environment with an enhanced toolkit for data visualization, including utilities to automate the data visualization for analysis of differentially expressed genes (DEGs). systemPipeTools provides data transformation and data exploration functions via scatterplots, hierarchical clustering heatMaps, principal component analysis, multidimensional scaling, generalized principal components, t-Distributed Stochastic Neighbor embedding (t-SNE), and MA and volcano plots. All these utilities can be integrated with the modular design of the systemPipeR environment that allows users to easily substitute any of these features and/or custom with alternatives."><meta property="og:type" content="article"><meta property="og:url" content="/news/release_3_13/"><meta property="article:modified_time" content="2021-11-13T02:00:58+00:00"><meta property="og:site_name" content="sysPipe"><meta itemprop=name content="Bioconductor 3.13"><meta itemprop=description content="systemPipeShiny 1.0.0 is available systemPipeTools package extends the widely used systemPipeR (SPR) workflow environment with an enhanced toolkit for data visualization, including utilities to automate the data visualization for analysis of differentially expressed genes (DEGs). systemPipeTools provides data transformation and data exploration functions via scatterplots, hierarchical clustering heatMaps, principal component analysis, multidimensional scaling, generalized principal components, t-Distributed Stochastic Neighbor embedding (t-SNE), and MA and volcano plots. All these utilities can be integrated with the modular design of the systemPipeR environment that allows users to easily substitute any of these features and/or custom with alternatives."><meta itemprop=dateModified content="2021-11-13T02:00:58+00:00"><meta itemprop=wordCount content="1888"><meta itemprop=keywords content><meta name=twitter:card content="summary"><meta name=twitter:title content="Bioconductor 3.13"><meta name=twitter:description content="systemPipeShiny 1.0.0 is available systemPipeTools package extends the widely used systemPipeR (SPR) workflow environment with an enhanced toolkit for data visualization, including utilities to automate the data visualization for analysis of differentially expressed genes (DEGs). systemPipeTools provides data transformation and data exploration functions via scatterplots, hierarchical clustering heatMaps, principal component analysis, multidimensional scaling, generalized principal components, t-Distributed Stochastic Neighbor embedding (t-SNE), and MA and volcano plots. All these utilities can be integrated with the modular design of the systemPipeR environment that allows users to easily substitute any of these features and/or custom with alternatives."><link rel=preload href=../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css as=style><link href=../../scss/main.min.9156453689c2a62860578c0dd5be0d8047f7ff28c3529bf4b8b3279efe82e52a.css rel=stylesheet integrity><script src=https://code.jquery.com/jquery-3.5.1.min.js integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin=anonymous></script><script src=https://unpkg.com/lunr@2.3.8/lunr.min.js integrity=sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY crossorigin=anonymous></script><link rel=stylesheet href=../../css/prism.css><script src=../../js/driver0.9.8.min.js></script><link rel=stylesheet href=../../css/driver0.9.8.min.css><link rel=stylesheet href=../../css/site.css><script src=../../js/site.js></script></head><body class="td-page line-numbers"><header><nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar"><a class=navbar-brand href=../../><span class=font-weight-bold>sysPipe</span></a>
<row class=header-switch><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=sidebarSwitch checked>
<label class=custom-control-label for=sidebarSwitch>sidebar</label></div><div class="custom-control custom-switch"><input type=checkbox class=custom-control-input id=tocSwitch checked>
<label class=custom-control-label for=tocSwitch>toc</label></div></row><div class="td-navbar-nav-scroll ml-md-auto" id=main_navbar><ul class="navbar-nav mt-2 mt-lg-0"><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../about/ id=navbarDropdown role=button>About</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../about/overview/ style=padding-left:1rem;font-weight:calc(3/3*500)>Overview</a>
<a class=dropdown-item href=../../about/team/ style=padding-left:1rem;font-weight:calc(3/3*500)>Our Team</a>
<a class=dropdown-item href=../../about/publications/ style=padding-left:1rem;font-weight:calc(3/3*500)>Publications</a>
<a class=dropdown-item href=../../about/acknowledgement/ style=padding-left:1rem;font-weight:calc(3/3*500)>Acknowledgement</a>
<a class=dropdown-item href=../../about/contact/ style=padding-left:1rem;font-weight:calc(3/3*500)>Contact</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../sp/ id=navbarDropdown role=button>systemPipeR</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../sp/spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeR Package</a>
<a class=dropdown-item href=../../sp/spr/introduction/ style=padding-left:2rem;font-weight:calc(3/4*500)>Introduction</a>
<a class=dropdown-item href=../../sp/spr/gettingstarted/ style=padding-left:2rem;font-weight:calc(3/4*500)>Getting Started</a>
<a class=dropdown-item href=../../sp/spr/spr_run/ style=padding-left:2rem;font-weight:calc(3/4*500)>How to run a Workflow</a>
<a class=dropdown-item href=../../sp/spr/steps/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow steps overview</a>
<a class=dropdown-item href=../../sp/spr/templates/ style=padding-left:2rem;font-weight:calc(3/4*500)>Workflow Templates</a>
<a class=dropdown-item href=../../sp/spr/cwl_syntax/ style=padding-left:2rem;font-weight:calc(3/4*500)>CWL syntax</a>
<a class=dropdown-item href=../../sp/spr/cwl_and_spr/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR and CWL</a>
<a class=dropdown-item href=../../sp/spr/sprinstall/ style=padding-left:2rem;font-weight:calc(3/4*500)>SPR detailed installation</a>
<a class=dropdown-item href=../../sp/spt/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRTools</a>
<a class=dropdown-item href=../../sp/spt/ style=padding-left:2rem;font-weight:calc(3/4*500)>Data Visualization</a>
<a class=dropdown-item href=../../sp/spr/spr_funcs/ style=padding-left:2rem;font-weight:calc(3/4*500)>Reference Manual</a>
<a class=dropdown-item href=../../sp/sprdata/ style=padding-left:1rem;font-weight:calc(3/3*500)>systemPipeRdata</a>
<a class=dropdown-item href=../../sp/sprdata/systempiperdata/ style=padding-left:2rem;font-weight:calc(3/4*500)>Vignette</a>
<a class=dropdown-item href=../../sp/sprdata/newwf/ style=padding-left:2rem;font-weight:calc(3/4*500)>Rules</a>
<a class=dropdown-item href=../../spr/ style=padding-left:1rem;font-weight:calc(3/3*500)>Links</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/bioc/html/systemPipeR.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPR Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeR style=padding-left:2rem;font-weight:calc(3/4*500)>SPR GitHub</a>
<a class=dropdown-item href=http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata Bioconductor</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeRdata style=padding-left:2rem;font-weight:calc(3/4*500)>SPRdata GitHub</a>
<a class=dropdown-item href=https://github.com/tgirke/systemPipeTools style=padding-left:2rem;font-weight:calc(3/4*500)>SPT GitHub</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../sps/ id=navbarDropdown role=button>Shiny</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../sps/ style=padding-left:1rem;font-weight:calc(3/3*500)>SPS Documentation</a>
<a class=dropdown-item href=https://github.com/systemPipeR/systemPipeShiny style=padding-left:1rem;font-weight:calc(3/3*500)>Github</a>
<a class=dropdown-item href=http://bioconductor.org/packages/3.13/bioc/html/systemPipeShiny.html style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../spr_wf id=navbarDropdown role=button>Workflows</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../spr_wf style=padding-left:1rem;font-weight:calc(3/3*500)>Collection</a>
<a class=dropdown-item href=../../spr_wf/installwf/ style=padding-left:1rem;font-weight:calc(3/3*500)>How to install Workflow</a>
<a class=dropdown-item href=../../spr_wf/newwf style=padding-left:1rem;font-weight:calc(3/3*500)>Rules to create Workflow</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../outreach/ id=navbarDropdown role=button>Outreach</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../training/conferences/ style=padding-left:1rem;font-weight:calc(3/3*500)>Conferences</a>
<a class=dropdown-item href=../../training/presentations/ style=padding-left:1rem;font-weight:calc(3/3*500)>Presentations</a>
<a class=dropdown-item href=../../training/videos/ style=padding-left:1rem;font-weight:calc(3/3*500)>Videos Tutorials</a></div></li><li class="nav-item mr-4 mb-2 mb-lg-0" style=position:relative><a class="nav-link dropdown-toggle" href=../../news/ id=navbarDropdown role=button>News</a><div class=dropdown-menu aria-labelledby=navbarDropdownMenuLink><a class=dropdown-item href=../../news/release_3_13/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.13</a>
<a class=dropdown-item href=../../news/release_3_12/ style=padding-left:1rem;font-weight:calc(3/3*500)>Bioconductor 3.12</a></div></li></ul></div><div class="navbar-nav d-none d-lg-block"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../offline-search-index.a27377dce846ec069f6d13ce96c94d0d.json data-offline-search-base-href=../../ data-offline-search-max-results=10></div></nav></header><div class="container-fluid td-outer"><div class=td-main><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none"><div id=td-sidebar-menu class=td-sidebar__inner><form class="td-sidebar__search d-flex align-items-center"><input type=search class="form-control td-search-input" placeholder="&#xf002; Search this site…" aria-label="Search this site…" autocomplete=off data-offline-search-index-json-src=../../offline-search-index.a27377dce846ec069f6d13ce96c94d0d.json data-offline-search-base-href=../../ data-offline-search-max-results=10>
<button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type=button data-toggle=collapse data-target=#td-section-nav aria-controls=td-docs-nav aria-expanded=false aria-label="Toggle section navigation"></button></form><nav class="collapse td-sidebar-nav" id=td-section-nav><ul class="td-sidebar-nav__section pr-md-3"><li class=td-sidebar-nav__section-title><a href=../../news/ class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">News</a></li><ul><li class="collapse show" id=news><a class="td-sidebar-link td-sidebar-link__page active" id=m-newsrelease_3_13 href=../../news/release_3_13/>Bioconductor 3.13</a>
<a class="td-sidebar-link td-sidebar-link__page" id=m-newsrelease_3_12 href=../../news/release_3_12/>Bioconductor 3.12</a></li></ul></ul></nav></div></div><div class="d-none d-xl-block col-xl-2 td-toc d-print-none"><div class="td-page-meta ml-2 pb-1 pt-2 mb-0"><a href=https://raw.githubusercontent.com/systemPipeR/systemPipeR.github.io/main/content/en/news/release_3_13.md target=_blank class=source-link><i class="fa fa-code fa-fw"></i>View source code</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/blob/main/content/en/news/release_3_13.md target=_blank><i class="fab fa-github fa-fw"></i>View on Github</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/edit/main/content/en/news/release_3_13.md target=_blank><i class="fa fa-edit fa-fw"></i>Edit this page</a>
<a href="https://github.com/systemPipeR/systemPipeR.github.io/new/main/content/en/news/release_3_13.md?filename=change-me.md&value=---%0Atitle%3A+%22Long+Page+Title%22%0AlinkTitle%3A+%22Short+Nav+Title%22%0Aweight%3A+100%0Adescription%3A+%3E-%0A+++++Page+description+for+heading+and+indexes.%0A---%0A%0A%23%23+Heading%0A%0AEdit+this+template+to+create+your+new+page.%0A%0A%2A+Give+it+a+good+name%2C+ending+in+%60.md%60+-+e.g.+%60getting-started.md%60%0A%2A+Edit+the+%22front+matter%22+section+at+the+top+of+the+page+%28weight+controls+how+its+ordered+amongst+other+pages+in+the+same+directory%3B+lowest+number+first%29.%0A%2A+Add+a+good+commit+message+at+the+bottom+of+the+page+%28%3C80+characters%3B+use+the+extended+description+field+for+more+detail%29.%0A%2A+Create+a+new+branch+so+you+can+preview+your+new+file+and+request+a+review+via+Pull+Request.%0A" target=_blank><i class="fa fa-edit fa-fw"></i>Create child page</a>
<a href=https://github.com/systemPipeR/systemPipeR.github.io/issues/new target=_blank><i class="fas fa-tasks fa-fw"></i>Create project issue</a></div><nav id=TableOfContents><ul><li><a href=#systempipeshiny-100-is-available><em>systemPipeShiny</em> 1.0.0 is available</a></li><li><a href=#systempipeshiny-12-is-available><em>systemPipeShiny</em> 1.2 is available</a><ul><li><a href=#systempipeshiny-1140>systemPipeShiny 1.1.40</a><ul><li><a href=#major-change>Major change</a></li><li><a href=#minor-change>Minor change</a></li><li><a href=#bug-fix>Bug fix</a></li></ul></li><li><a href=#systempipeshiny-1135>systemPipeShiny 1.1.35</a><ul><li><a href=#major-change-1>Major change</a></li><li><a href=#minor-change-1>Minor change</a></li><li><a href=#bug-fix-1>Bug fix</a></li></ul></li><li><a href=#systempipeshiny-1130>systemPipeShiny 1.1.30</a><ul><li><a href=#major-change-2>Major change</a></li><li><a href=#bug-fix-2>Bug fix</a></li></ul></li></ul></li><li><a href=#systempipeshiny-1120>systemPipeShiny 1.1.20</a><ul><li><ul><li><a href=#major-change-3>Major change</a></li><li><a href=#minor-change-2>Minor change</a></li><li><a href=#bug-fix-3>Bug fix</a></li></ul></li><li><a href=#systempipeshiny-1110>systemPipeShiny 1.1.10</a><ul><li><a href=#workflow-module-r-session>Workflow module R session</a></li><li><a href=#rnaseq-module>RNAseq module</a></li><li><a href=#general-ui>General UI</a></li><li><a href=#workflow-module-cwl-tab>Workflow module CWL tab</a></li><li><a href=#workflow-module-fully-functioning>Workflow module fully functioning</a></li><li><a href=#rework-on-the-workflow-part>Rework on the workflow part</a></li><li><a href=#other-changes>Other changes</a></li><li><a href=#change-on-visualization>Change on visualization</a></li></ul></li></ul></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 pl-md-5" role=main><nav aria-label=breadcrumb class="d-none d-md-block d-print-none"><ol class="breadcrumb spb-1"><li class=breadcrumb-item><a href=../../news/>News</a></li><li class="breadcrumb-item active" aria-current=page><a href=../../news/release_3_13/>Bioconductor 3.13</a></li></ol></nav><div class=td-content><h1>Bioconductor 3.13</h1><h1 id=systempipeshiny-100-is-available><em>systemPipeShiny</em> 1.0.0 is available</h1><p><code>systemPipeTools</code> package extends the widely used systemPipeR (SPR) workflow environment with an enhanced toolkit for data visualization, including utilities to automate the data visualization for analysis of differentially expressed genes (DEGs). <code>systemPipeTools</code> provides data transformation and data exploration functions via scatterplots, hierarchical clustering heatMaps, principal component analysis, multidimensional scaling, generalized principal components, t-Distributed Stochastic Neighbor embedding (t-SNE), and MA and volcano plots. All these utilities can be integrated with the modular design of the systemPipeR environment that allows users to easily substitute any of these features and/or custom with alternatives.</p><h1 id=systempipeshiny-12-is-available><em>systemPipeShiny</em> 1.2 is available</h1><h2 id=systempipeshiny-1140>systemPipeShiny 1.1.40</h2><h3 id=major-change>Major change</h3><ul><li><p>Add <code>is_demo</code> option: only affect workflow module right now. Lock users inside a temp folder when using the WF module and give users a new temp folder every time they refresh. This will prevent directory exist problem if many users are using a same deploy instance.</p></li><li><p>Add <code>welcome_guide</code> option: whether to enable the welcome guide which highlights the guide dropdown menu.</p></li><li><p>Rewrite welcome tab with a gallery to show all SPS features.</p></li><li><p><code>loadDF</code>, <code>dynamicFile</code> and <code>dynamicFileServer</code> added back to this mainframe work package instead of <code>spsComps</code>, because these dependencies have already been using in <code>SPS</code>. Leave these functions in <code>spsComps</code> will introduce extra dependencies, and these functions are not too frequently used outside the framework.</p></li></ul><h3 id=minor-change>Minor change</h3><ul><li><p>Option <code>warning_toast</code> now also checks if you are on &ldquo;local&rdquo; mode.</p></li><li><p>Deleted some unwanted entries in reference generating <code>yaml</code> file.</p></li><li><p>Fix some typos.</p></li><li><p>More informative error message when the config file cannot be found for <code>spsOptions</code></p></li><li><p>Add some <code>.onLoad</code> methods so users can use the <code>spsOption</code> to get default values on package load.</p></li><li><p>Updated <code>essquise</code> functions</p></li><li><p>Add more guides.</p></li><li><p>Removed the scroll to top button by <code>shinyDashboardPlus</code>, we have our own &ldquo;go top&rdquo; button.</p></li><li><p>Add assertions to <code>spsInit</code>.</p></li><li><p>Add some screenshots to readme.</p></li></ul><h3 id=bug-fix>Bug fix</h3><ul><li>Fix a bug when that loads the server twice</li><li>Fix some default option values</li><li>Fix a bug on <code>addResourcePath</code> when the working directory and app directory is not the same.</li><li>Fix links not working caused by website change</li><li>Fix code in <code>spsInit</code> overwrite all current SPS options.</li><li>Fix errors on admin page when server stats cannot be found, better text and warning messages</li><li>Fix new version of <code>essquise</code> introduced errors</li><li>Fix a warning in <code>vroom</code> due to the column type problem</li></ul><h2 id=systempipeshiny-1135>systemPipeShiny 1.1.35</h2><h3 id=major-change-1>Major change</h3><ul><li><p>Login feature added:</p><ul><li><p>Users can choose whether to enable the login or not in <code>global.R</code>SPS options.</p></li><li><p>There are also the login loading screen feature which can be turned on and off.</p></li><li><p>There are 3 different login loading screens right now and users can interact with them.</p></li></ul></li><li><p>Website updated. <a href=https://systempipe.org/sps>https://systempipe.org/sps</a></p></li><li><p>Updates on the admin panel:</p><ul><li><p>App information: added live charts for CPU, temperature, and RAM</p></li><li><p>User control: admins now can add/delete/change users directly from this tab, instead of only from command line.</p></li></ul></li></ul><h3 id=minor-change-1>Minor change</h3><ul><li>Add<code>target="_blank"</code> to all external links in the app, so when they are clicked, it will open in a new tab.</li></ul><h3 id=bug-fix-1>Bug fix</h3><ul><li><p>FIx bugs due to login page caused server not loading</p></li><li><p>Add 1s delay in javascript after login to resize the page so the dashboard can be displayed normal.</p></li><li><p>Fix a table rendering bug in workflow cwl step markdown text.</p></li></ul><h2 id=systempipeshiny-1130>systemPipeShiny 1.1.30</h2><h3 id=major-change-2>Major change</h3><ul><li><p>new <code>spsAccount</code> class. This class is associated with login management , which allows users to create/delete user/admin accounts, change password, change roles.</p></li><li><p>Deprecated the <code>spsPlotContainer</code> class since we rewrite the Canvas feature and move to a separate package {drawer}.</p></li><li><p>New <code>spsCoreTabReplace</code>, which allows users to overwrite the default core tabs.</p></li><li><p>A lot more SPS options.</p><ul><li><p>Users can now choose whether to load or not load certain tabs on start, even for default core tabs. Combining the <code>spsCoreTabReplace</code> function, now users can customize everything of the original app.</p></li><li><p>Users can change the app title, and logo image.</p></li></ul></li><li><p>Admin panel added to app. Users now can visit the admin panel by adding &ldquo;<code>?user_definded_string</code>&rdquo; to the end of the url. Default is &ldquo;<code>admin</code>&rdquo;. Login with an admin account is required. Users can use the <code>spsAccount</code> class to add/change an admin account before starting the app.</p><ul><li><p>App information: a tab displays current SPS app server information, like CPU, RAM, size, etc.</p></li><li><p>User control: a tab to see account information of current SPS app.</p></li></ul></li><li><p>Changed the way to install modules. Default modules, workflow, RNAseq and quick ggplot dependency packages are not installed by default, unless you use <code>dependency = TRUE</code> in installation command. It means all these dependencies are moved from <code>Imports</code> to the <code>Suggests</code> class. This helps to save quite some time on SPS package installation. Users install these packages based on their needs. When users loads these modules but depend packages are not fully installed, app will not crash, instead, a warning message with install instructions will be displayed on both console and app UI.</p></li><li><p>Based on the module installation change, module loading methods are also changed. Module server functions are only called if users set the option to load them. In previous versions, the server functions are still loaded, just hide the unloaded module UI. This saves a lot of time on app starting time, roughly from > 10s to &lt; 3s if none of the default modules are loaded.</p></li></ul><h3 id=bug-fix-2>Bug fix</h3><ul><li><p>update all links to our new website: <a href=https://systempipe.org/sps>https://systempipe.org/sps</a></p></li><li><p>Fix some bugs in the guide system</p></li></ul><h1 id=systempipeshiny-1120>systemPipeShiny 1.1.20</h1><h3 id=major-change-3>Major change</h3><ul><li><p>3 default modules complete: <code>workflow</code>, <code>RNAseq</code>,<code>quick ggplot</code>. Details of these modules updated in our website: <a href=https://systempipe.org/sps>&lt;https://systempipe.org/sps></a>{.uri}.</p></li><li><p>Separation of SPS smaller functions into 3 different packages. We hope these packages can help people in their own Shiny app, or other R projects.</p><ul><li><p><code>{spsComps}</code>: SPS components, all new Shiny custom components and utility functions that are used in Shiny server side.</p></li><li><p><code>{drawer}</code>: the redesign of Canvas, purely front-end image editing tool.</p></li><li><p><code>{spsUtil}</code>: SPS utilities, general useful utility functions that can be used with/without Shiny.</p></li></ul></li><li><p>Redesigned the new tab feature. Now users use <code>spsNewTab</code> function to create their new custom visualization tab. The old <code>newSpsTab</code> function is deprecated. Easier syntax and templates are used. By default it will use the &ldquo;<em>simple</em>&rdquo; template which wraps 90% of the shiny code from users so they can focus on the plotting code. There is also the &ldquo;<em>full</em>&rdquo; template which expose all the Shiny code to users.</p><ul><li>New <code>spsEzUI</code> and <code>spsEzServer</code> functions are used in the &ldquo;simple&rdquo; template to wrap complex Shiny code.</li></ul></li><li><p>New <code>spsOptDefaults</code>, which prints out all the default SPS options and current values of these options on console.</p></li><li><p>New notification system. Developers can write some notifications which stores in a remote location and when app starts, it will try to download and parse this file to notifications messages to broadcast to users. This way, developers can send messages to users often without re-deploy the app. The notification will appear on the top right corner.</p></li><li><p>The interactive guide is back. After a few versions of tests, we added the guide system back. This time, developers can customize their own guides. A <code>guide_content.R</code> file is created when a SPS project initialize. It is stored in <code>R</code> of folder relate to the project root. The guide will also be displayed on the app top right corner.</p></li></ul><h3 id=minor-change-2>Minor change</h3><ul><li>updated all unit test to testthat v3 format.</li></ul><h3 id=bug-fix-3>Bug fix</h3><ul><li><p>fix bugs due to shiny updates to 1.6.0</p></li><li><p>Fix all bugs caused by <code>{shinydashboardPlus}</code> v2.0 updates.</p></li></ul><h2 id=systempipeshiny-1110>systemPipeShiny 1.1.10</h2><p>Changes made from 1.1.0 to 1.1.05</p><h3 id=workflow-module-r-session>Workflow module R session</h3><ul><li>Now workflow module R session uses a background child R process, which runs independently to the parent R session which runs shiny.</li><li>So the shiny will be not blocked while code is running in the background (you can still click other buttons when the child session is busy) &ndash; synchronous and non-blocking. A child indicator is also placed in the UI, updates every second.</li><li>The UI design of R session is similar to Rstudio. Four panels, source code, console, log (specific to SPR), and plots.</li><li>Standard out/error and plots are captured in the workflow folder. Users can download them in the bundle on step 5 <em>Workflow Run.</em></li><li>Plots will be displayed on the plots panel. Now supports plots that opens R <code>device</code> (base and ggplot), html widget plots are not supported as this moment.</li><li>A new shiny disconnection popup for SPS. Besides the gray layer on shiny disconnection, a panel will be displayed to users to indicate the problem. Similar to what shows on a shiny server, but more informative and also works locally.</li><li>Results of this session can be downloaded by closing the session and go back to <em>step 5</em> of workflow module and there is a button to download all in a zipped file.</li></ul><h3 id=rnaseq-module>RNAseq module</h3><ul><li>redesigned the UI and server logic. Plots for DEG analysis and Canvas connections.</li><li><code>{SummarizedExperiment}</code> supports. Now it returns <code>SummarizedExperiment</code> objects to global environment once the normalization or DEG calculation is done.</li></ul><h3 id=general-ui>General UI</h3><ul><li>Added a &ldquo;Go Top&rdquo; button on the right bottom corner, clicking on this button will automatically scroll to the top of the page. This button only shows up when client has > 50px scroll height.</li></ul><h3 id=workflow-module-cwl-tab>Workflow module CWL tab</h3><ul><li>Now the CWL file and CWL input file can be be edited. The edits will be imported to CWL parser every one second. Now this is a very useful place to test or write new CWL scripts.</li><li>Now this tab has a dynamically rendered dropdown panel which allows users to choose which column for the targets table to map to the variables in CWL input file.</li></ul><h3 id=workflow-module-fully-functioning>Workflow module fully functioning</h3><ul><li><p>Now you can run a full example workflow in SPS by choosing the &ldquo;Example&rdquo; option on workflow setup step.</p></li><li><p>Other systemPipeR preconfiged workflows will cause problems because formatting issues that will cause errors in <code>systemPipeR::runWF</code> function, not introduced by SPS. Please wait the updates on systemPipeR to fix this. You can still use SPS to prepare files for all workflows. That means, step 1-4 will work, step 5 will give you errors if you choose a workflow which is not &ldquo;Example&rdquo;.</p></li></ul><h3 id=rework-on-the-workflow-part>Rework on the workflow part</h3><ul><li><p>All 3 tabs merged into the main tab</p></li><li><p>changed config tab to CWL tab</p></li><li><p>added support for the running wf as a sub tab</p></li><li><p>Now the main tab has 5 subtabs, they are all connected.</p></li><li><p>Better guidelines for users, step-like usage, can&rsquo;t reach other steps if a previous step is not completed.</p></li><li><p>Original snapshot management drop down page changed to running workflow session. This session will lock users to a unique page, they can&rsquo;t interactive other app parts on the page(working directory changed), to prevent app crash due to wd change.</p></li></ul><h3 id=other-changes>Other changes</h3><ul><li><p>A new UI component <code>spsTimeline</code> : horizontal timeline UI unit, has status, can be updated on server by <code>updateSpsTimeline</code>.</p></li><li><p>A new UI <code>bsHoverPopover</code>: enhanced high level function of <code>bsPlus::HoverPopover</code>, additional JS used to make the popover work on buttons as well.</p></li><li><p>Fixed some link problems in <code>renderDesc</code>. Better links in <code>renderDesc</code>, enlarged and spacing animation for links.</p></li><li><p>Change on about page</p><ul><li><p>The news is now rendered on <em>about</em> tab in the app</p></li><li><p>reduced developer content on about page.</p></li><li><p>changed developer emails to github links.</p></li></ul></li></ul><h3 id=change-on-visualization>Change on visualization</h3><ul><li><p>RNAseq part is now only in one tab as big module: users upload the targets file and a raw count table, and make different plots in subtabs.</p><ul><li>This introduced a lot of dependencies, will decide later if we keep as it is or separate it to <em>spsBio.</em></li></ul></li></ul><div class="text-muted mt-5 pt-3 border-top">Last modified 2021-11-13: <a href=https://github.com/systemPipeR/systemPipeR.github.io/commit/4de37549fa0e3fccff1fb665059ebd26c2012c7d>Bot build site (4de3754)</a></div></div></main></div></div><footer class="bg-dark py-5 row d-print-none"><div class="container-fluid mx-sm-5"><div class=row><div class="col-6 col-sm-4 text-xs-center order-sm-2"></div><div class="col-6 col-sm-4 text-right text-xs-center order-sm-3"><ul class="list-inline mb-0"><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=GitHub aria-label=GitHub><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/><i class="fab fa-github"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title=Bioconductor aria-label=Bioconductor><a class=text-white target=_blank rel="noopener noreferrer" href=https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html><i class="fa fa-dna"></i></a></li><li class="list-inline-item mx-2 h3" data-toggle=tooltip data-placement=top title="Find an issue?" aria-label="Find an issue?"><a class=text-white target=_blank rel="noopener noreferrer" href=https://github.com/systemPipeR/systemPipeR/issues><i class="fa fa-envelope"></i></a></li></ul></div><div class="col-12 col-sm-4 text-center py-2 order-sm-2"><small class=text-white>&copy; 2021 The systemPipe Project All Rights Reserved</small><p class=mt-2><a href=../../about/>About systemPipe Project</a></p></div></div></div></footer></div><script src=https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js integrity=sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49 crossorigin=anonymous></script><script src=https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js integrity=sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy crossorigin=anonymous></script><script>(function(){var i,text,code,codes=document.getElementsByTagName('code');for(i=0;i<codes.length;){code=codes[i];if(code.parentNode.tagName!=='PRE'&&code.childElementCount===0){text=code.textContent;if(/^\$[^$]/.test(text)&&/[^$]\$$/.test(text)){text=text.replace(/^\$/,'\\(').replace(/\$$/,'\\)');code.textContent=text;}
if(/^\\\((.|\s)+\\\)$/.test(text)||/^\\\[(.|\s)+\\\]$/.test(text)||/^\$(.|\s)+\$$/.test(text)||/^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)){code.outerHTML=code.innerHTML;continue;}}
i++;}})();</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script id=MathJax-script async src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script><script type=text/x-mathjax-config>
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script><script src=../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js integrity="sha256-+CC/xQfD6+/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin=anonymous></script><script src=../../js/prism.js></script></body></html>